% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/utils.R
\name{pcaSCoPE2}
\alias{pcaSCoPE2}
\title{PCA plot suggested by SCoPE2}
\usage{
pcaSCoPE2(object, scale = FALSE, center = FALSE)
}
\arguments{
\item{object}{A \code{SingleCellExperiment} object}

\item{scale}{A \code{logical(1)} indicating whether the columns of the
weighted input matrix should be scaled to unit variance.}

\item{center}{A \code{logical(1)} indicating whether the columns of the
weighted input matrix should be mean centered.}
}
\value{
An object of class \code{eigen} containing the computed
eigenvector and eigenvalues.
}
\description{
The function performs a weighted principal component analysis (PCA)
as suggested by Specht et al. The PCA is performed on the
correlation matrix where the rows (features) are weighted according
to the sum of the correlation with the other rows.
}
\examples{
data("feat2")
sce <- as(feat2[[3]], "SingleCellExperiment")
pcaSCoPE2(sce)

}
\references{
Specht, Harrison, Edward Emmott, Aleksandra A. Petelski, R. Gray
Huffman, David H. Perlman, Marco Serra, Peter Kharchenko, Antonius
Koller, and Nikolai Slavov. 2021. "Single-Cell Proteomic and
Transcriptomic Analysis of Macrophage Heterogeneity Using SCoPE2.â€
Genome Biology 22 (1): 50.
\href{http://dx.doi.org/10.1186/s13059-021-02267-5}{link to article},
\href{https://www.biorxiv.org/content/10.1101/665307v5}{link to preprint}
}
