<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SCP Replication Vignettes • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="SCP Replication Vignettes">
<meta property="og:description" content="Repository of vignettes about the replication of
    single-cell analyses.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="articles/brunner2022.html">Replication of the cell cycle state study (Brunner et al. 2021)</a>
    </li>
    <li>
      <a href="articles/derks2022.html">Replication of the plexDIA analysis (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="articles/leduc2022.html">Replication of the nPOP analysis (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="articles/schoof2021.html">Replication of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="articles/scplainer_derks2022.html">scplainer: reanalysis of the plexDIA dataset (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="articles/scplainer_leduc2022.html">scplainer: reanalysis of the nPOP dataset (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="articles/scplainer_woo2022.html">scplainer: reanalysis of the macrophage activation dataset (Woo et al. 2022)</a>
    </li>
    <li>
      <a href="articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="single-cell-replication-package">Single cell replication package<a class="anchor" aria-label="anchor" href="#single-cell-replication-package"></a>
</h1></div>
<p>This package contains a series of vignettes that reproduce published SCP data analyses. Emphasis is put on standardization of the workflows.</p>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>All vignettes are pre-compiled and can be found on the <a href="https://uclouvain-cbio.github.io/SCP.replication/index.html" class="external-link">website</a> under the <code>Articles</code> tab.</p>
<p>The replication package can be installed with R &gt;= 4.1 by running:</p>
<pre><code><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"UCLouvain-CBIO/SCP.replication"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="reproduction-of-scp-data-analyses">Reproduction of SCP data analyses<a class="anchor" aria-label="anchor" href="#reproduction-of-scp-data-analyses"></a>
</h2>
<p>The currently available reproduction vignettes are listed below.</p>
<div class="section level3">
<h3 id="reproduction-of-the-scope2-analysis-specht-et-al-2021">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/SCoPE2.html" class="external-link">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a><a class="anchor" aria-label="anchor" href="#reproduction-of-the-scope2-analysis-specht-et-al-2021"></a>
</h3>
<p>Project tag: <code>SCoPE2</code> Docker image: <code>cvanderaa/scp_replication_docker:v1</code></p>
<blockquote>
<p>Specht H, Emmott E, Petelski AA, Huffman RG, Perlman DH, Serra M, et al. Single-cell proteomic and transcriptomic analysis of macrophage heterogeneity using SCoPE2. Genome Biol. <a href="http://dx.doi.org/10.1186/s13059-021-02267-5" class="external-link">2021;22: 50</a>.</p>
</blockquote>
<p>See also</p>
<ul>
<li>The SCoPE2 <a href="https://scope2.slavovlab.net/" class="external-link">website</a>
</li>
<li>The SCoPE2 Github <a href="https://github.com/SlavovLab/SCoPE2" class="external-link">repository</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="exploring-the-autopots-data-liang-et-al-2020">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/liang2020.html" class="external-link">Exploring the autoPOTS data (Liang et al. 2020)</a><a class="anchor" aria-label="anchor" href="#exploring-the-autopots-data-liang-et-al-2020"></a>
</h3>
<p>Project tag: <code>liang2020</code> Docker image: <code>cvanderaa/scp_replication_docker:v1</code></p>
<blockquote>
<p>Liang, Yiran, Hayden Acor, Michaela A McCown, Andikan J Nwosu, Hannah Boekweg, Nathaniel B Axtell, Thy Truong, Yongzheng Cong, Samuel H Payne, and Ryan T Kelly. 2020. “Fully Automated Sample Processing and Analysis Workflow for Low-Input Proteome Profiling.” Anal. Chem., December.2021, <a href="https://pubs.acs.org/doi/10.1021/acs.analchem.0c04240" class="external-link">93, 3, 1658–1666</a>.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="reproduction-of-the-hair-cell-development-analysis-zhu-et-al-2019-elife">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/zhu2019EL.html" class="external-link">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a><a class="anchor" aria-label="anchor" href="#reproduction-of-the-hair-cell-development-analysis-zhu-et-al-2019-elife"></a>
</h3>
<p>Project tag: <code>zhu2019EL</code> Docker image: <code>cvanderaa/scp_replication_docker:v1</code></p>
<blockquote>
<p>Zhu, Ying, Mirko Scheibinger, Daniel Christian Ellwanger, Jocelyn F Krey, Dongseok Choi, Ryan T Kelly, Stefan Heller, and Peter G Barr-Gillespie. 2019. “Single-Cell Proteomics Reveals Changes in Expression During Hair-Cell Development.” <a href="https://elifesciences.org/articles/50777" class="external-link">Elife 8 (November)</a>.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="reproduction-of-the-aml-model-analysis-schoof-et-al-2021">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/schoof2021.html" class="external-link">Reproduction of the AML model analysis (Schoof et al. 2021)</a><a class="anchor" aria-label="anchor" href="#reproduction-of-the-aml-model-analysis-schoof-et-al-2021"></a>
</h3>
<p>Project tag: <code>schoof2021</code> Docker image: <code>cvanderaa/scp_replication_docker:v1</code></p>
<blockquote>
<p>Schoof, Erwin M., Benjamin Furtwängler, Nil Üresin, Nicolas Rapin, Simonas Savickas, Coline Gentil, Eric Lechman, Ulrich auf Dem Keller, John E. Dick, and Bo T. Porse. 2021. “Quantitative Single-Cell Proteomics as a Tool to Characterize Cellular Hierarchies.” Nature Communications 12 (1): 745679.</p>
</blockquote>
<p>See also</p>
<ul>
<li>The SCeptre Github <a href="https://github.com/bfurtwa/SCeptre" class="external-link">repository</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="reproducing-the-multiplexed-scp-analysis-by-williams-et-al-2020">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/williams2020_tmt.html" class="external-link">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a><a class="anchor" aria-label="anchor" href="#reproducing-the-multiplexed-scp-analysis-by-williams-et-al-2020"></a>
</h3>
<p>Project tag: <code>williams2020_tmt</code> Docker image: <code>cvanderaa/scp_replication_docker:v1</code></p>
<blockquote>
<p>Williams, Sarah M., Andrey V. Liyu, Chia-Feng Tsai, Ronald J. Moore, Daniel J. Orton, William B. Chrisler, Matthew J. Gaffrey, et al. 2020. “Automated Coupling of Nanodroplet Sample Preparation with Liquid Chromatography-Mass Spectrometry for High-Throughput Single-Cell Proteomics.” Analytical Chemistry 92 (15): 10588–96.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="replication-of-the-npop-analysis-leduc-et-al-2022">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/leduc2022.html" class="external-link">Replication of the nPOP analysis (Leduc et al. 2022)</a><a class="anchor" aria-label="anchor" href="#replication-of-the-npop-analysis-leduc-et-al-2022"></a>
</h3>
<p>Project tag: <code>leduc2022</code> Docker image: <code>cvanderaa/scp_replication_docker:v1</code></p>
<blockquote>
<p>Leduc, Andrew, R. Gray Huffman, Joshua Cantlon, Saad Khan, and Nikolai Slavov. 2022. “Exploring Functional Protein Covariation across Single Cells Using nPOP.” bioRxiv. <a href="https://doi.org/10.1101/2021.04.24.441211" class="external-link uri">https://doi.org/10.1101/2021.04.24.441211</a>.</p>
</blockquote>
<p>See also</p>
<ul>
<li>The nPOP <a href="https://scp.slavovlab.net/nPOP" class="external-link">website</a>
</li>
<li>The nPOP Github <a href="https://github.com/SlavovLab/nPOP" class="external-link">repository</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="replication-of-the-plexdia-analysis-derks-et-al-2022">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/derks2022.html" class="external-link">Replication of the plexDIA analysis (Derks et al. 2022)</a><a class="anchor" aria-label="anchor" href="#replication-of-the-plexdia-analysis-derks-et-al-2022"></a>
</h3>
<p>Project tag: <code>derks2022</code> Docker image: <code>cvanderaa/scp_replication_docker:v1</code></p>
<blockquote>
<p>Derks, Jason, Andrew Leduc, Georg Wallmann, R. Gray Huffman, Matthew Willetts, Saad Khan, Harrison Specht, Markus Ralser, Vadim Demichev, and Nikolai Slavov. 2022. “Increasing the Throughput of Sensitive Proteomics by plexDIA.” Nature Biotechnology, July, 2021.11.03.467007.</p>
</blockquote>
<p>See also</p>
<ul>
<li>The plexDIA <a href="https://scp.slavovlab.net/plexDIA" class="external-link">website</a>
</li>
<li>The plexDIA Github <a href="https://github.com/SlavovLab/plexDIA" class="external-link">repository</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="replication-of-the-cell-cycle-state-study-brunner-et-al-2021">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/brunner2022.html" class="external-link">Replication of the cell cycle state study (Brunner et al. 2021)</a><a class="anchor" aria-label="anchor" href="#replication-of-the-cell-cycle-state-study-brunner-et-al-2021"></a>
</h3>
<p>Project tag: <code>brunner2022</code> Docker image: <code>cvanderaa/scp_replication_docker:v1</code></p>
<blockquote>
<p>Brunner, Andreas-David, Marvin Thielert, Catherine Vasilopoulou, Constantin Ammar, Fabian Coscia, Andreas Mund, Ole B. Hoerning, et al. 2022. “Ultra-High Sensitivity Mass Spectrometry Quantifies Single-Cell Proteome Changes upon Perturbation.” Molecular Systems Biology 18 (3): e10798.</p>
</blockquote>
<p>See also</p>
<ul>
<li>The author’s Github <a href="https://github.com/theislab/singlecell_proteomics" class="external-link">repository</a>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="replicate-the-analyses-locally">Replicate the analyses locally<a class="anchor" aria-label="anchor" href="#replicate-the-analyses-locally"></a>
</h2>
<p>You can reproduce the analysis vignettes on your local machine using <code>Docker</code>. You must first install Docker. Then, pull the image from the <a href="https://hub.docker.com/repository/docker/cvanderaa/scp_replication_docker" class="external-link">DockerHub repository</a>.</p>
<pre><code>docker pull cvanderaa/scp_replication_docker:vX</code></pre>
<p>As we’ll release more vignettes, we will release new image tags. Make sure to pull the correct version <code>vX</code> for the vignette you want to reproduce (cf sections above).</p>
<p>Then, you can start an Rstudio session within a Docker container using:</p>
<pre><code>docker run \
    -e PASSWORD=bioc \
    -p 8787:8787 \
    -v `pwd`:/home/rstudio/SCP.replication/ \ 
    cvanderaa/scp_replication_docker:vX</code></pre>
<p>Note you should use <code>%CD%</code> instead of <code>pwd</code> when using Windows.</p>
<p>Open your browser and go to <a href="http://localhost:8787" class="external-link uri">http://localhost:8787</a>. The USER is <code>rstudio</code> and the password is <code>bioc</code>. See the <a href="https://hub.docker.com/repository/docker/cvanderaa/scp_replication_docker" class="external-link">DockerHub repository</a> for more detailed information on getting started with <code>Docker</code>.</p>
</div>
<div class="section level2">
<h2 id="note-to-future-self">Note to future self<a class="anchor" aria-label="anchor" href="#note-to-future-self"></a>
</h2>
<div class="section level3">
<h3 id="build-the-docker-image">Build the docker image<a class="anchor" aria-label="anchor" href="#build-the-docker-image"></a>
</h3>
<p>Commit all recent changes to <code>SCP.replication</code> and push to the GitHub repo. The docker image is built from the master branch of <code>UCLouvain-CBIO/SCP.replication</code>, so make sure all changes needed for building the new vignette are live on GitHub. This is also true for:</p>
<ul>
<li><code>rformassspectrometry/QFeatures</code></li>
<li><code>UCLouvain-CBIO/scpdata</code></li>
<li><code>UCLouvain-CBIO/scp</code></li>
</ul>
<p>A new docker image should be built every time we add a new vignette. If new dependencies are required, update the <code>Dockerfile</code> accordingly.</p>
<p>Then build the image (make sure to <code>cd</code> in the <code>SCP.replication</code> repo):</p>
<pre><code>docker build -t cvanderaa/scp_replication_docker:vX .</code></pre>
<p>Make sure to <strong>increment vX</strong> with the latest version. See the <a href="https://hub.docker.com/repository/docker/cvanderaa/scp_replication_docker" class="external-link">DockerHub repo</a> for the latest tag. When complete, push the new image to DockerHub:</p>
<pre><code>docker push cvanderaa/scp_replication_docker:vX </code></pre>
</div>
<div class="section level3">
<h3 id="compile-the-vignette">Compile the vignette<a class="anchor" aria-label="anchor" href="#compile-the-vignette"></a>
</h3>
<p>You can compile a new vignette or update an existing one using:</p>
<pre><code>docker run -e PASSWORD=scp \
        -v `pwd`:/home/rstudio/SCP.replication/ \
        -it cvanderaa/scp_replication_docker:vX \
        R --quiet -e "setwd('SCP.replication');pkgdown::build_article('VIGNETTE_NAME')"</code></pre>
<p>Make sure to adapt <code>vX</code> and <code>VIGNETTE_NAME</code> with the correct tag and the name of the vignette to add or update.</p>
</div>
<div class="section level3">
<h3 id="build-the-website">Build the website<a class="anchor" aria-label="anchor" href="#build-the-website"></a>
</h3>
<p>When you’re done compiling the vignette(s), rebuild the website shell using:</p>
<pre><code>make website_shell</code></pre>
<p>Push the changes in <code>docs/</code> to Github.</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h1>
<p>To cite the SCP.replication website in publications please use:</p>
<blockquote>
<p>Vanderaa Christophe and Laurent Gatto. The current state of single-cell proteomics data analysis. arXiv:2210.01020; doi: <a href="https://doi.org/10.48550/arXiv.2210.01020" class="external-link uri">https://doi.org/10.48550/arXiv.2210.01020</a> (2022).</p>
</blockquote>
</div>
<div class="section level1">
<h1 id="licence">Licence<a class="anchor" aria-label="anchor" href="#licence"></a>
</h1>
<p>All vignettes are distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA licence</a> licence.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 3)</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SCP.replication</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Christophe Vanderaa <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-7443-5427" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Laurent Gatto <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-1520-2268" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
