<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SCP Replication Vignettes • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="SCP Replication Vignettes">
<meta property="og:description" content="Repository of vignettes about the replication of
    single-cell analyses.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="articles/schoof2021.html">Reproduction of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="single-cell-replication-package">Single cell replication package<a class="anchor" aria-label="anchor" href="#single-cell-replication-package"></a>
</h1></div>
<p>This package contains a series of vignettes that reproduce published SCP data analyses. Emphasis is put on standardization of the workflows.</p>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>All vignettes are pre-compiled and can be found on the <a href="https://uclouvain-cbio.github.io/SCP.replication/index.html" class="external-link">website</a> under the <code>Articles</code> tab.</p>
<p>The replication package can be installed with R &gt;= 4.1 by running:</p>
<pre><code><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"UCLouvain/SCP.replication"</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="reproduction-of-scp-data-analyses">Reproduction of SCP data analyses<a class="anchor" aria-label="anchor" href="#reproduction-of-scp-data-analyses"></a>
</h2>
<p>The currently available reproduction vignettes are listed below.</p>
<div class="section level4">
<h4 id="reproduction-of-the-scope2-analysis-specht-et-al-2021">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/SCoPE2.html" class="external-link">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a><a class="anchor" aria-label="anchor" href="#reproduction-of-the-scope2-analysis-specht-et-al-2021"></a>
</h4>
<p>Project tag: <code>SCoPE2</code></p>
<blockquote>
<p>Specht H, Emmott E, Petelski AA, Huffman RG, Perlman DH, Serra M, et al. Single-cell proteomic and transcriptomic analysis of macrophage heterogeneity using SCoPE2. Genome Biol. <a href="http://dx.doi.org/10.1186/s13059-021-02267-5" class="external-link">2021;22: 50</a>.</p>
</blockquote>
<p>See also</p>
<ul>
<li>The SCoPE2 <a href="https://scope2.slavovlab.net/" class="external-link">website</a>
</li>
<li>The SCoPE2 Github <a href="https://github.com/SlavovLab/SCoPE2" class="external-link">repository</a>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="exploring-the-autopots-data-liang-et-al-2020">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/liang2020.html" class="external-link">Exploring the autoPOTS data (Liang et al. 2020)</a><a class="anchor" aria-label="anchor" href="#exploring-the-autopots-data-liang-et-al-2020"></a>
</h4>
<p>Project tag: <code>liang2020</code></p>
<blockquote>
<p>Liang, Yiran, Hayden Acor, Michaela A McCown, Andikan J Nwosu, Hannah Boekweg, Nathaniel B Axtell, Thy Truong, Yongzheng Cong, Samuel H Payne, and Ryan T Kelly. 2020. “Fully Automated Sample Processing and Analysis Workflow for Low-Input Proteome Profiling.” Anal. Chem., December.2021, <a href="https://pubs.acs.org/doi/10.1021/acs.analchem.0c04240" class="external-link">93, 3, 1658–1666</a>.</p>
</blockquote>
</div>
<div class="section level4">
<h4 id="reproduction-of-the-hair-cell-development-analysis-zhu-et-al-2019-elife">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/zhu2019EL.html" class="external-link">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a><a class="anchor" aria-label="anchor" href="#reproduction-of-the-hair-cell-development-analysis-zhu-et-al-2019-elife"></a>
</h4>
<p>Project tag: <code>zhu2019EL</code></p>
<blockquote>
<p>Zhu, Ying, Mirko Scheibinger, Daniel Christian Ellwanger, Jocelyn F Krey, Dongseok Choi, Ryan T Kelly, Stefan Heller, and Peter G Barr-Gillespie. 2019. “Single-Cell Proteomics Reveals Changes in Expression During Hair-Cell Development.” <a href="https://elifesciences.org/articles/50777" class="external-link">Elife 8 (November)</a>.</p>
</blockquote>
</div>
<div class="section level4">
<h4 id="reproduction-of-the-aml-model-analysis-schoof-et-al-2021">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/schoof2021.html" class="external-link">Reproduction of the AML model analysis (Schoof et al. 2021)</a><a class="anchor" aria-label="anchor" href="#reproduction-of-the-aml-model-analysis-schoof-et-al-2021"></a>
</h4>
<p>Project tag: <code>schoof2021</code></p>
<blockquote>
<p>Schoof, Erwin M., Benjamin Furtwängler, Nil Üresin, Nicolas Rapin, Simonas Savickas, Coline Gentil, Eric Lechman, Ulrich auf Dem Keller, John E. Dick, and Bo T. Porse. 2021. “Quantitative Single-Cell Proteomics as a Tool to Characterize Cellular Hierarchies.” Nature Communications 12 (1): 745679.</p>
</blockquote>
<p>See also</p>
<ul>
<li>The SCeptre Github <a href="https://github.com/bfurtwa/SCeptre" class="external-link">repository</a>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="reproducing-the-multiplexed-scp-analysis-by-williams-et-al-2020">
<a href="https://uclouvain-cbio.github.io/SCP.replication/articles/williams2020_tmt.html" class="external-link">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a><a class="anchor" aria-label="anchor" href="#reproducing-the-multiplexed-scp-analysis-by-williams-et-al-2020"></a>
</h4>
<p>Project tag: <code>williams2020_tmt</code></p>
<blockquote>
<p>Williams, Sarah M., Andrey V. Liyu, Chia-Feng Tsai, Ronald J. Moore, Daniel J. Orton, William B. Chrisler, Matthew J. Gaffrey, et al. 2020. “Automated Coupling of Nanodroplet Sample Preparation with Liquid Chromatography-Mass Spectrometry for High-Throughput Single-Cell Proteomics.” Analytical Chemistry 92 (15): 10588–96.</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="replicate-the-analyses-locally">Replicate the analyses locally<a class="anchor" aria-label="anchor" href="#replicate-the-analyses-locally"></a>
</h2>
<p>You can reproduce the analysis vignettes on your local machine. First, clone this repository and <code>cd</code> into it. The Rmarkdown files that created the articles are available in the <code>vignettes</code> directory. Then, you need to activate and restore the corresponding <code>renv</code> project (see tags above). To do so, run the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"renv"</span><span class="op">)</span>
<span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="st">"inst/renvs/PROJECT_TAG"</span><span class="op">)</span>
<span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/restore.html" class="external-link">restore</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Make sure that <code><a href="https://rstudio.github.io/renv/reference/restore.html" class="external-link">renv::restore</a></code> completes successfully. This procedure installs all the required packages (in R) and modules (in Python) that are required to run the vignettes. <strong>Warning</strong>: <code>renv</code> makes sure to restore the correct dependency versions, but it does not consider the R or Python software version. You may run into trouble if your R and/or Python versions do not match with the versions indicated at the end of each analysis (section <em>Requirements</em>).</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 3)</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SCP.replication</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Christophe Vanderaa <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-7443-5427" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Laurent Gatto <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-1520-2268" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
