<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Replication of the cell cycle state study (Brunner et al. 2021) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Replication of the cell cycle state study (Brunner et al. 2021)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/brunner2022.html">Replication of the cell cycle state study (Brunner et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/derks2022.html">Replication of the plexDIA analysis (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/leduc2022.html">Replication of the nPOP analysis (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Replication of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Replication of the cell cycle state study
(Brunner et al. 2021)</h1>
                        <h4 data-toc-skip class="author">Christophe
Vanderaa, Computational Biology, UCLouvain</h4>
                        <h4 data-toc-skip class="author">Laurent Gatto,
Computational Biology, UCLouvain</h4>
            
            <h4 data-toc-skip class="date">5 October 2022</h4>
      
      
      <div class="hidden name"><code>brunner2022.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The study by Brunner et al. on the cell cycle is the first mass
spectrometry-based single-cell proteomics (SCP) experiment that acquired
hundreds of single cells using a label-free protocol (<span class="citation">Brunner et al. (2022)</span>). The data was acquired
using a newly designed mass spectrometer that will later be released by
Bruker as the timsTOF-SCP device. This instrument can be used for data
dependent acquisition (DDA) or data independent acquisition (DIA).
Single-cell samples were acquired using the latter and the DIA-NN
software (<span class="citation">Demichev et al. (2020)</span>) was used
to identify and quantify the MS data.</p>
<p>Let’s first load the replication package to make use of some helper
functions. Those functions are only meant for this replication vignette
and are not designed for general use.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SCP.replication"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="scp-and-the-plexdia-data-analysis-workflow">
<code>scp</code> and the plexDIA data analysis workflow<a class="anchor" aria-label="anchor" href="#scp-and-the-plexdia-data-analysis-workflow"></a>
</h3>
<p>The code provided along with the article can be retrieved from <a href="https://github.com/theislab/singlecell_proteomics" class="external-link">this GitHub
repository</a>. It provides the python code used to reproduce the
article’s analysis. The objective of this vignette is to replicate the
analysis script using R code. We focus on standardized, easy-to-read,
and well documented code. Therefore, our first contribution is to
formalize the data processing into a conceptual flow chart.</p>
<div class="figure" style="text-align: center">
<img src="figs/brunner2022-workflow.png" alt="Overview of the processing workflow by Derks et al." width="20%"><p class="caption">
Overview of the processing workflow by Derks et al.
</p>
</div>
<p>This replication vignette relies on a data framework dedicated to SCP
data analysis that combines two Bioconductor classes (<span class="citation">Vanderaa and Gatto (2021)</span>):</p>
<ul>
<li>The <code>SingleCellExperiment</code> class provides an interface to
many cutting edge methods for single-cell analysis</li>
<li>The <code>QFeatures</code> class facilitates manipulation and
processing of MS-based quantitative data.</li>
</ul>
<p>The <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/scp/inst/doc/scp.html" class="external-link"><code>scp</code>
vignette</a> provides detailed information about the data structure. The
<code>scp</code> package extends the functionality of
<code>QFeatures</code> for single-cell application. <code>scp</code>
offers a standardized implementation for single-cell processing
methods.</p>
<p>The required packages for running this workflow are listed below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Core packages of this workflow</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">"scp"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"scpdata"</span><span class="op">)</span></span>
<span><span class="co">## External package for SingleCellExperiment objects</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/scater/" class="external-link">"scater"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"scuttle"</span><span class="op">)</span></span>
<span><span class="co">## Utility packages for data manipulation and visualization</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://tidyverse.tidyverse.org" class="external-link">"tidyverse"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="scpdata-and-the-brunner2022-dataset">
<code>scpdata</code> and the <code>brunner2022</code> dataset<a class="anchor" aria-label="anchor" href="#scpdata-and-the-brunner2022-dataset"></a>
</h3>
<p>We also implemented a data package called <code>scpdata</code> (<span class="citation">Vanderaa and Gatto (2022)</span>). It distributes
published SCP datasets, such as the <code>brunner2022</code> dataset.
The datasets were downloaded from the data source provided in the
publication and formatted to a <code>QFeatures</code> object so that it
is compatible with our software. The underlying data storage is based on
the <code>ExperimentHub</code> package that provides a cloud-based
storage infrastructure.</p>
<p>The <code>brunner2022</code> dataset is provided at different levels
of processing:</p>
<ul>
<li>The <strong>.d</strong> files that were generated by the
mass-spectrometer. This data is not included in
<code>scpdata</code>.</li>
<li>The <strong>DIA-NN main output report table</strong> that contains
the results of the spectrum identification and quantification.</li>
<li>The <strong>DIA-NN protein group matrix</strong>: normalised
quantities for protein groups, filtered at 1% FDR.</li>
<li>Other <strong>DIA-NN matrix</strong>: normalised quantities for
precursors, gene groups or unique genes, filtered at 1% FDR. These files
were not added to <code>scpdata</code>.</li>
</ul>
<p>More information about the DIA-NN output tables can be found in the
corresponding <a href="https://github.com/vdemichev/DiaNN#output" class="external-link">documentation</a>.</p>
<p>The workflow starts with the protein group matrix and will replicate
Figure 4 B, C and D from the original article. Note that the script did
not mention which input file was used and we assumed from the methods
section that the protein group matrix was used.</p>
<p>We formatted the <code>brunner2022</code> dataset following the
<code>scp</code> data framework. The formatted data can be retrieved
from the <code>scpdata</code> package using the
<code><a href="https://rdrr.io/pkg/scpdata/man/brunner2022.html" class="external-link">brunner2022()</a></code> function. All datasets in
<code>scpdata</code> are called after the first author and the date of
publication. More information about the data and how it was processed
can be find in <code><a href="https://rdrr.io/pkg/scpdata/man/brunner2022.html" class="external-link">?brunner2022</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">brunner</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/brunner2022.html" class="external-link">brunner2022</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## An instance of class QFeatures containing 435 assays:</span></span>
<span><span class="co">##  [1] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201009_TIMS04_Evo07_AnBr_1Cell_HeLa_NB_01_S3-G2_1_3873.d: SingleCellExperiment with 3772 rows and 1 columns </span></span>
<span><span class="co">##  [2] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201009_TIMS04_Evo07_AnBr_1Cell_HeLa_NB_02_S3-G3_1_3874.d: SingleCellExperiment with 7759 rows and 1 columns </span></span>
<span><span class="co">##  [3] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201009_TIMS04_Evo07_AnBr_1Cell_HeLa_NB_03_S3-G4_1_3875.d: SingleCellExperiment with 4481 rows and 1 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [433] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201126_TIMS04_Evo07_SA_ADB_1cell_HeLa_UB_98_S2-C2_1_5109.d: SingleCellExperiment with 2569 rows and 1 columns </span></span>
<span><span class="co">##  [434] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201126_TIMS04_Evo07_SA_ADB_1cell_HeLa_UB_99_S2-C3_1_5110.d: SingleCellExperiment with 2327 rows and 1 columns </span></span>
<span><span class="co">##  [435] proteins: SingleCellExperiment with 2476 rows and 434 columns</span></span></code></pre></div>
<p>The data are stored in a <code>QFeatures</code> object. In total, it
contains 435 different <code>SingleCellExperiment</code> objects that we
refer to as <strong>assays</strong>. Each assay contains expression data
along with feature metadata. Each row in an assay represents a
<strong>feature</strong>, in this case a precursor or a protein
depending on the assay. Each column in an assay represents a
<strong>sample</strong>. Each acquisition contains 1 cell, hence the
precursor data contains 1 column per acquisition.</p>
<p>Using <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code>, we can have a quick overview of the
assays.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">)</span></span>
<span><span class="co">## Warning in plot.QFeatures(brunner): The QFeatures object contains many assays.</span></span>
<span><span class="co">## You may want to consider creating an interactive plot (set 'interactive = TRUE')</span></span></code></pre></div>
<p><img src="brunner2022_files/figure-html/unnamed-chunk-4-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>This figure is crowded, you can use
<code>plot(brunner, interactive = TRUE)</code> to interactively explore
this map. However, we can see that all precursor assays in the top of
the figure are linked to the protein data.</p>
<p>Note that we inferred sample annotations from the file names. These
are accessible from the <code>colData</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] "File.Name"        "Run"              "Date"             "MsInstrument"    </span></span>
<span><span class="co">##  [5] "Purification"     "User"             "SampleAnnotation" "SampleType"      </span></span>
<span><span class="co">##  [9] "CellCycleStage"   "..undetermined.." "PlatePosition"    "CellNumber"      </span></span>
<span><span class="co">## [13] "RunID"            "Label"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-cells">Filter cells<a class="anchor" aria-label="anchor" href="#filter-cells"></a>
</h3>
<p>The authors first perform a sample quality control to make sure all
cells have at least 600 quantified proteins. To perform this, we rely on
<code><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics()</a></code> from the <code>scuttle</code> package.
To use this function we must first extract the assay we want to apply
the function on, using double bracket subsetting (<code>[[</code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">brunner</span><span class="op">[[</span><span class="st">"proteins"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>Next the function counts the number of features that are not zero,
but missing data is encoded as <code>NA</code>. We therefore impote
missing data with zero using <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute()</a></code> from the
<code>QFeatures</code> package.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute</a></span><span class="op">(</span><span class="va">sce</span>, method <span class="op">=</span> <span class="st">"zero"</span><span class="op">)</span></span></code></pre></div>
<p>We can now apply the QC function that will compute 3 metrics: the
total abundance recorded per cell (<code>sum</code> and
<code>total</code>, they are the same) and the number detected proteins
per cell (<code>detected</code>).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellqc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span>, assay.type <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cellqc</span><span class="op">)</span></span>
<span><span class="co">## [1] "sum"      "detected" "total"</span></span></code></pre></div>
<p>Next, we insert the computed metrics in the <code>colData</code>.
This allows to use <code>subsetByColData()</code> and enable to subset
single cells with 600 or more detected proteins.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cellqc</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">cellqc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">brunner</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">brunner</span>, <span class="va">brunner</span><span class="op">$</span><span class="va">detected</span> <span class="op">&gt;=</span> <span class="fl">600</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-proteins">Filter proteins<a class="anchor" aria-label="anchor" href="#filter-proteins"></a>
</h3>
<p>The authors then filter for proteins that are detected in at least
15% of the cells. We apply the same procedure, but using
<code><a href="https://rdrr.io/pkg/scuttle/man/perFeatureQCMetrics.html" class="external-link">perFeatureQCMetrics()</a></code></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">brunner</span><span class="op">[[</span><span class="st">"proteins"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute</a></span><span class="op">(</span><span class="va">sce</span>, method <span class="op">=</span> <span class="st">"zero"</span><span class="op">)</span></span>
<span><span class="va">protqc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perFeatureQCMetrics.html" class="external-link">perFeatureQCMetrics</a></span><span class="op">(</span><span class="va">sce</span>, assay.type <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">protqc</span><span class="op">)</span></span>
<span><span class="co">## [1] "mean"     "detected"</span></span></code></pre></div>
<p><code><a href="https://rdrr.io/pkg/scuttle/man/perFeatureQCMetrics.html" class="external-link">perFeatureQCMetrics()</a></code> computes the mean abundance for
each protein (<code>mean</code>) and the detection rate across all cells
(<code>detected</code>). Since this information is related to features,
we must insert it in the <code>rowData</code>. This is a bit more
complicated than for the <code>colData</code> since each assay has its
own specific <code>rowData</code>. The QC metric we computed are related
to the protein level and should only be included for the
<code>proteins</code> assay. This can be done by providing a list. The
name of the elements will indicate in which assay the
<code>rowData</code> must be added. Columns of <code>protqc</code> are
automatically added based on the protein names (see details in
?<code>QFeatures-class</code> for more details).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">List</span><span class="op">(</span>proteins <span class="op">=</span> <span class="va">protqc</span><span class="op">)</span></span></code></pre></div>
<p>Since the metrics are now in the <code>rowData</code>, we cannot use
<code>subsetByColData</code>. Instead, we use
<code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures()</a></code> from the <code>QFeatures</code> package
that will automactilly apply the filter on the <code>rowData</code> of
each assay. Since <code>detected</code> is only available for proteins,
we provide <code>keep = TRUE</code> to ignore the filter and hence keep
the features for the other assays.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">brunner</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">brunner</span>, <span class="op">~</span> <span class="va">detected</span> <span class="op">&gt;</span> <span class="fl">0.15</span>, </span>
<span>                          keep <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="performance-assessment">Performance assessment<a class="anchor" aria-label="anchor" href="#performance-assessment"></a>
</h3>
<p>In Figure 4 B and C of the original paper, the authors assessed the
analytical performance of their technology for single-cell samples. The
cells were analysed at different cell cycle stages thanks to thymidine
and nocodazole block. We recode the annotation to match the
corresponding cell cycle stage.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">brunner</span><span class="op">$</span><span class="va">CellCycleStage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">recode</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">$</span><span class="va">CellCycleStage</span>,</span>
<span>                                 TB <span class="op">=</span> <span class="st">"G1-S"</span>,</span>
<span>                                 NB <span class="op">=</span> <span class="st">"G2-M"</span>,</span>
<span>                                 UB <span class="op">=</span> <span class="st">"untreated"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">$</span><span class="va">CellCycleStage</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##        G1      G1-S        G2      G2-M untreated </span></span>
<span><span class="co">##        87        41        52        45       203</span></span></code></pre></div>
<p>The counts above are very close to those reported in the legend of
Figure 4C. We next plot the number of quantified proteins per cell. This
is already available from the <code>colData</code>. Using
<code>ggplot2</code>, we replicate Figure 4B.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">"id"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">detected</span>, </span>
<span>        fill <span class="op">=</span> <span class="va">CellCycleStage</span>,</span>
<span>        color <span class="op">=</span> <span class="va">CellCycleStage</span>,</span>
<span>        x <span class="op">=</span> <span class="va">id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="op">~</span> <span class="va">CellCycleStage</span>, scales <span class="op">=</span> <span class="st">"free_x"</span>, space <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Quantified proteins"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="brunner2022_files/figure-html/unnamed-chunk-14-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>This plot is very similar to the original figure, except there here
is the additional `“untreated”’ category.</p>
<p>We also plot the total signal per cell that replicated Figure 4C.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">brunner</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sum</span>, </span>
<span>        x <span class="op">=</span> <span class="va">CellCycleStage</span>,</span>
<span>        fill <span class="op">=</span> <span class="va">CellCycleStage</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="brunner2022_files/figure-html/unnamed-chunk-15-1.png" width="70%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="log-transformation">Log-Transformation<a class="anchor" aria-label="anchor" href="#log-transformation"></a>
</h3>
<p>The protein data is then log2-transformed using
<code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform()</a></code>. This will add a new assay that we call
<code>proteins_log</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">brunner</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">brunner</span>, i <span class="op">=</span> <span class="st">"proteins"</span>, </span>
<span>                        name <span class="op">=</span> <span class="st">"proteins_log"</span>, </span>
<span>                        base <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, pc <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## An instance of class QFeatures containing 436 assays:</span></span>
<span><span class="co">##  [1] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201009_TIMS04_Evo07_AnBr_1Cell_HeLa_NB_01_S3-G2_1_3873.d: SingleCellExperiment with 3772 rows and 1 columns </span></span>
<span><span class="co">##  [2] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201009_TIMS04_Evo07_AnBr_1Cell_HeLa_NB_02_S3-G3_1_3874.d: SingleCellExperiment with 7759 rows and 1 columns </span></span>
<span><span class="co">##  [3] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201009_TIMS04_Evo07_AnBr_1Cell_HeLa_NB_03_S3-G4_1_3875.d: SingleCellExperiment with 4481 rows and 1 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [434] D:\Andreas_Brunner\Projects\SingleCellProteomics\202109_WorkHere\SingleCells\SingleShots_RawData\20201126_TIMS04_Evo07_SA_ADB_1cell_HeLa_UB_99_S2-C3_1_5110.d: SingleCellExperiment with 2327 rows and 1 columns </span></span>
<span><span class="co">##  [435] proteins: SingleCellExperiment with 2476 rows and 428 columns </span></span>
<span><span class="co">##  [436] proteins_log: SingleCellExperiment with 2476 rows and 428 columns</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="imputation-by-downshifted-norm">Imputation by downshifted norm<a class="anchor" aria-label="anchor" href="#imputation-by-downshifted-norm"></a>
</h3>
<p>The final step is missing data imputation. As mentioned above, the
<code>QFeatures</code> package has a function to perform imputation. The
authors used imputation by sampling from a down-shifted normal
distribution. Unfortunately, this method is not available, but we can
provide it as a custom function.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imputeByDownShiftedNormal</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">scale</span> <span class="op">=</span> <span class="fl">0.3</span>, <span class="va">shift</span> <span class="op">=</span> <span class="fl">1.8</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">std</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">nmis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">repl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">nmis</span>, mean <span class="op">=</span> <span class="va">m</span> <span class="op">-</span> <span class="va">shift</span> <span class="op">*</span> <span class="va">std</span>, sd <span class="op">=</span> <span class="va">scale</span> <span class="op">*</span> <span class="va">std</span><span class="op">)</span></span>
<span>    <span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">repl</span></span>
<span>    <span class="va">x</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We supply the custom function to <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute()</a></code></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">brunner</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute</a></span><span class="op">(</span><span class="va">brunner</span>, i <span class="op">=</span> <span class="st">"proteins_log"</span>, </span>
<span>                  name <span class="op">=</span> <span class="st">"proteins_dsnImpd"</span>,</span>
<span>                  FUN <span class="op">=</span> <span class="va">imputeByDownShiftedNormal</span><span class="op">)</span></span></code></pre></div>
<p>The last assay, <code>proteins_dsnImpd</code>, represents the final
step of the data processing workflow.</p>
</div>
<div class="section level3">
<h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>We finally perform dimension reduction using PCA to replicate Figure
4D. This is performed by <code><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA()</a></code> and
<code><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA()</a></code> from the <code>scater</code> package. They work
on <code>SingleCEllExperiment</code> objects and to include the sample
annotation, we extract the last assay using
<code>getWithColData()</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">brunner</span>, <span class="st">"proteins_dsnImpd"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">CellCycleStage</span> <span class="op">!=</span> <span class="st">"untreated"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">50</span>,</span>
<span>              ntop <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>              scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              exprs_values <span class="op">=</span> <span class="fl">1</span>,</span>
<span>              name <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"CellCycleStage"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkseagreen"</span>, <span class="st">"cornflowerblue"</span>, </span>
<span>                                   <span class="st">"grey"</span>, <span class="st">"darkorange2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="brunner2022_files/figure-html/unnamed-chunk-20-1.png" width="70%" style="display: block; margin: auto;">
This plot shows similar trends to the original figure although the
percent variability explained is lower in our replication (PC1: 10 vs
14% and PC2: 3 vs 4 %).</p>
<p>An important aspect of the data analysis that the authors overlooked
is the assessment of batch effect. We extracted the (presumably) date of
acquisition from the file names and plot these on the PCA.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Date"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gold"</span>, <span class="st">"gold3"</span>, <span class="st">"orange2"</span>, <span class="st">"orange4"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="brunner2022_files/figure-html/unnamed-chunk-21-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>Surprisingly, we observe the different cell cycle stages where
acquired in distinct batches. For instance, all G1 cells where acquired
on November 21, 2020 and all G2 cells where acquired on November 13,
2020. This is an issue because batch effects can dramatically influence
quantifications in proteomics experiments, especially SCP (<span class="citation">Vanderaa and Gatto (2021)</span>). This experimental
design does not allow for batch correction and hence we cannot
distinguish the proportion of variance that is explained by technical
and biological factors…</p>
<p>In the article, the authors argue:</p>
<blockquote>
<p>In addition to these drug-perturbed cells, we measured more than 200
untreated ones from two independent cell culture batches. The proteomes
of these asynchronous cells distributed well across the cell cycle
states, while different passage batches were enriched in the G1 and G2
phase</p>
</blockquote>
<p>They refer to supplementary Figure EV2E. Let’s replicate this
figure:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">brunner</span>, <span class="st">"proteins_dsnImpd"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">50</span>,</span>
<span>              ntop <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>              scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              exprs_values <span class="op">=</span> <span class="fl">1</span>,</span>
<span>              name <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"CellCycleStage"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkseagreen"</span>, <span class="st">"cornflowerblue"</span>, </span>
<span>                                   <span class="st">"grey"</span>, <span class="st">"darkorange2"</span>, <span class="st">"firebrick"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="brunner2022_files/figure-html/unnamed-chunk-22-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>Indeed, we also find that some untreated cells are close to G1 cells
in PCA space and other are closer to the G2-M phase. But let’s look at
the batch effect.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Date"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gold"</span>, <span class="st">"gold3"</span>, <span class="st">"orange2"</span>, <span class="st">"orange4"</span>,<span class="st">"red4"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="brunner2022_files/figure-html/unnamed-chunk-23-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The PCA better explains the acquisition date than the cells cycle
stage, refuting the argument stated above. A better design would have
been to randomize the cell cycle stages across several acquisition runs.
Obviously, effect of cell culture batch or FACS sorting may be
responsible as well and are practically inevitably. Spreading sample
preparation over several culture and FACS sorting batches would have
allowed to disentangle the biological and technical variability.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>This vignette present a standardized R solution to replicate the
python analysis by <span class="citation">Brunner et al. (2022)</span>.
We successfully reproduced their main results for the SCP cell cycle
experiment. By re-analyzing the data, we discovered an important flaw in
the experimental design. Technical and biological factors cannot be
distinguished hindering batch correction, leading to potentially biased
outcome.</p>
</div>
<div class="section level2">
<h2 id="reproduce-this-vignette">Reproduce this vignette<a class="anchor" aria-label="anchor" href="#reproduce-this-vignette"></a>
</h2>
<p>You can reproduce this vignette using <code>Docker</code>:</p>
<pre><code>docker pull cvanderaa/scp_replication_docker:v1
docker run \
    -e PASSWORD=bioc \
    -p 8787:8787 \
    cvanderaa/scp_replication_docker:v1</code></pre>
<p>Open your browser and go to <a href="http://localhost:8787" class="external-link uri">http://localhost:8787</a>. The USER is <code>rstudio</code>
and the password is <code>bioc</code>. You can find the vignette in the
<code>vignettes</code> folder.</p>
<p>See the <a href="https://uclouvain-cbio.github.io/SCP.replication/index.html" class="external-link">website
home page</a> for more information.</p>
</div>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<div class="section level3">
<h3 id="hardware-and-software">Hardware and software<a class="anchor" aria-label="anchor" href="#hardware-and-software"></a>
</h3>
<p>The system details of the machine that built the vignette are:</p>
<pre><code><span><span class="co">## Machine: Linux (5.15.0-48-generic)</span></span>
<span><span class="co">## R version: R.4.2.1 (svn: 82513)</span></span>
<span><span class="co">## RAM: 16.5 GB</span></span>
<span><span class="co">## CPU: 16 core(s) - 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="timing">Timing<a class="anchor" aria-label="anchor" href="#timing"></a>
</h3>
<p>The total time required to compile this vignette is:</p>
<pre><code><span><span class="co">## 7.945226 mins</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="memory">Memory<a class="anchor" aria-label="anchor" href="#memory"></a>
</h3>
<p>The final <code>leduc</code> object size is:</p>
<pre><code><span><span class="co">## [1] "2.3 Gb"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] benchmarkme_1.0.8           patchwork_1.1.2            </span></span>
<span><span class="co">##  [3] forcats_0.5.2               stringr_1.4.1              </span></span>
<span><span class="co">##  [5] dplyr_1.0.10                purrr_0.3.4                </span></span>
<span><span class="co">##  [7] readr_2.1.3                 tidyr_1.2.1                </span></span>
<span><span class="co">##  [9] tibble_3.1.8                tidyverse_1.3.2            </span></span>
<span><span class="co">## [11] scater_1.25.7               ggplot2_3.3.6              </span></span>
<span><span class="co">## [13] scuttle_1.7.4               SCP.replication_0.2.1      </span></span>
<span><span class="co">## [15] SingleCellExperiment_1.19.1 scpdata_1.5.4              </span></span>
<span><span class="co">## [17] ExperimentHub_2.5.0         AnnotationHub_3.5.2        </span></span>
<span><span class="co">## [19] BiocFileCache_2.5.0         dbplyr_2.2.1               </span></span>
<span><span class="co">## [21] scp_1.7.4                   QFeatures_1.7.3            </span></span>
<span><span class="co">## [23] MultiAssayExperiment_1.23.9 SummarizedExperiment_1.27.3</span></span>
<span><span class="co">## [25] Biobase_2.57.1              GenomicRanges_1.49.1       </span></span>
<span><span class="co">## [27] GenomeInfoDb_1.33.7         IRanges_2.31.2             </span></span>
<span><span class="co">## [29] S4Vectors_0.35.4            BiocGenerics_0.43.4        </span></span>
<span><span class="co">## [31] MatrixGenerics_1.9.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## [33] BiocStyle_2.25.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] utf8_1.2.2                    reticulate_1.26              </span></span>
<span><span class="co">##   [3] tidyselect_1.1.2              RSQLite_2.2.17               </span></span>
<span><span class="co">##   [5] AnnotationDbi_1.59.1          grid_4.2.1                   </span></span>
<span><span class="co">##   [7] BiocParallel_1.31.12          munsell_0.5.0                </span></span>
<span><span class="co">##   [9] ScaledMatrix_1.5.1            codetools_0.2-18             </span></span>
<span><span class="co">##  [11] ragg_1.2.3                    withr_2.5.0                  </span></span>
<span><span class="co">##  [13] colorspace_2.0-3              filelock_1.0.2               </span></span>
<span><span class="co">##  [15] highr_0.9                     knitr_1.40                   </span></span>
<span><span class="co">##  [17] labeling_0.4.2                GenomeInfoDbData_1.2.9       </span></span>
<span><span class="co">##  [19] bit64_4.0.5                   farver_2.1.1                 </span></span>
<span><span class="co">##  [21] rprojroot_2.0.3               vctrs_0.4.2                  </span></span>
<span><span class="co">##  [23] generics_0.1.3                xfun_0.33                    </span></span>
<span><span class="co">##  [25] doParallel_1.0.17             R6_2.5.1                     </span></span>
<span><span class="co">##  [27] ggbeeswarm_0.6.0              clue_0.3-61                  </span></span>
<span><span class="co">##  [29] rsvd_1.0.5                    locfit_1.5-9.6               </span></span>
<span><span class="co">##  [31] MsCoreUtils_1.9.1             AnnotationFilter_1.21.0      </span></span>
<span><span class="co">##  [33] bitops_1.0-7                  cachem_1.0.6                 </span></span>
<span><span class="co">##  [35] DelayedArray_0.23.2           assertthat_0.2.1             </span></span>
<span><span class="co">##  [37] promises_1.2.0.1              scales_1.2.1                 </span></span>
<span><span class="co">##  [39] googlesheets4_1.0.1           beeswarm_0.4.0               </span></span>
<span><span class="co">##  [41] gtable_0.3.1                  beachmat_2.13.4              </span></span>
<span><span class="co">##  [43] OrgMassSpecR_0.5-3            benchmarkmeData_1.0.4        </span></span>
<span><span class="co">##  [45] sva_3.45.0                    rlang_1.0.6                  </span></span>
<span><span class="co">##  [47] genefilter_1.79.0             systemfonts_1.0.4            </span></span>
<span><span class="co">##  [49] splines_4.2.1                 lazyeval_0.2.2               </span></span>
<span><span class="co">##  [51] gargle_1.2.1                  broom_1.0.1                  </span></span>
<span><span class="co">##  [53] BiocManager_1.30.18           yaml_2.3.5                   </span></span>
<span><span class="co">##  [55] modelr_0.1.9                  backports_1.4.1              </span></span>
<span><span class="co">##  [57] httpuv_1.6.6                  tools_4.2.1                  </span></span>
<span><span class="co">##  [59] bookdown_0.29                 ellipsis_0.3.2               </span></span>
<span><span class="co">##  [61] jquerylib_0.1.4               Rcpp_1.0.9                   </span></span>
<span><span class="co">##  [63] sparseMatrixStats_1.9.0       zlibbioc_1.43.0              </span></span>
<span><span class="co">##  [65] RCurl_1.98-1.9                viridis_0.6.2                </span></span>
<span><span class="co">##  [67] cowplot_1.1.1                 haven_2.5.1                  </span></span>
<span><span class="co">##  [69] ggrepel_0.9.1                 cluster_2.1.4                </span></span>
<span><span class="co">##  [71] fs_1.5.2                      magrittr_2.0.3               </span></span>
<span><span class="co">##  [73] reprex_2.0.2                  googledrive_2.0.0            </span></span>
<span><span class="co">##  [75] ProtGenerics_1.29.0           hms_1.1.2                    </span></span>
<span><span class="co">##  [77] mime_0.12                     evaluate_0.16                </span></span>
<span><span class="co">##  [79] xtable_1.8-4                  XML_3.99-0.11                </span></span>
<span><span class="co">##  [81] readxl_1.4.1                  gridExtra_2.3                </span></span>
<span><span class="co">##  [83] compiler_4.2.1                crayon_1.5.2                 </span></span>
<span><span class="co">##  [85] htmltools_0.5.3               mgcv_1.8-40                  </span></span>
<span><span class="co">##  [87] later_1.3.0                   tzdb_0.3.0                   </span></span>
<span><span class="co">##  [89] lubridate_1.8.0               DBI_1.1.3                    </span></span>
<span><span class="co">##  [91] MASS_7.3-58.1                 rappdirs_0.3.3               </span></span>
<span><span class="co">##  [93] Matrix_1.5-1                  cli_3.4.1                    </span></span>
<span><span class="co">##  [95] parallel_4.2.1                igraph_1.3.5                 </span></span>
<span><span class="co">##  [97] pkgconfig_2.0.3               pkgdown_2.0.6                </span></span>
<span><span class="co">##  [99] foreach_1.5.2                 xml2_1.3.3                   </span></span>
<span><span class="co">## [101] annotate_1.75.0               vipor_0.4.5                  </span></span>
<span><span class="co">## [103] bslib_0.4.0                   XVector_0.37.1               </span></span>
<span><span class="co">## [105] rvest_1.0.3                   digest_0.6.29                </span></span>
<span><span class="co">## [107] Biostrings_2.65.6             rmarkdown_2.16               </span></span>
<span><span class="co">## [109] cellranger_1.1.0              edgeR_3.39.6                 </span></span>
<span><span class="co">## [111] DelayedMatrixStats_1.19.1     curl_4.3.2                   </span></span>
<span><span class="co">## [113] shiny_1.7.2                   lifecycle_1.0.2              </span></span>
<span><span class="co">## [115] nlme_3.1-159                  jsonlite_1.8.2               </span></span>
<span><span class="co">## [117] BiocNeighbors_1.15.1          desc_1.4.2                   </span></span>
<span><span class="co">## [119] viridisLite_0.4.1             limma_3.53.10                </span></span>
<span><span class="co">## [121] fansi_1.0.3                   pillar_1.8.1                 </span></span>
<span><span class="co">## [123] lattice_0.20-45               KEGGREST_1.37.3              </span></span>
<span><span class="co">## [125] fastmap_1.1.0                 httr_1.4.4                   </span></span>
<span><span class="co">## [127] survival_3.4-0                interactiveDisplayBase_1.35.0</span></span>
<span><span class="co">## [129] glue_1.6.2                    iterators_1.0.14             </span></span>
<span><span class="co">## [131] png_0.1-7                     BiocVersion_3.16.0           </span></span>
<span><span class="co">## [133] bit_4.0.4                     stringi_1.7.8                </span></span>
<span><span class="co">## [135] sass_0.4.2                    BiocBaseUtils_0.99.12        </span></span>
<span><span class="co">## [137] blob_1.2.3                    textshaping_0.3.6            </span></span>
<span><span class="co">## [139] BiocSingular_1.13.1           memoise_2.0.1                </span></span>
<span><span class="co">## [141] irlba_2.3.5.1</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="licence">Licence<a class="anchor" aria-label="anchor" href="#licence"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
licence</a> licence.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Brunner2022-rd" class="csl-entry">
Brunner, Andreas-David, Marvin Thielert, Catherine Vasilopoulou,
Constantin Ammar, Fabian Coscia, Andreas Mund, Ole B Hoerning, et al.
2022. <span>“Ultra-High Sensitivity Mass Spectrometry Quantifies
Single-Cell Proteome Changes Upon Perturbation.”</span> <em>Mol. Syst.
Biol.</em> 18 (3): e10798.
</div>
<div id="ref-Demichev2020-rb" class="csl-entry">
Demichev, Vadim, Christoph B Messner, Spyros I Vernardis, Kathryn S
Lilley, and Markus Ralser. 2020. <span>“<span>DIA-NN</span>: Neural
Networks and Interference Correction Enable Deep Proteome Coverage in
High Throughput.”</span> <em>Nat. Methods</em> 17 (1): 41–44.
</div>
<div id="ref-Vanderaa2021-ue" class="csl-entry">
Vanderaa, Christophe, and Laurent Gatto. 2021. <span>“Replication of
Single-Cell Proteomics Data Reveals Important Computational
Challenges.”</span> <em>Expert Rev. Proteomics</em>, October.
</div>
<div id="ref-Vanderaa2022-qv" class="csl-entry">
———. 2022. <span>“The Current State of Single-Cell Proteomics Data
Analysis,”</span> October. <a href="https://arxiv.org/abs/2210.01020" class="external-link">https://arxiv.org/abs/2210.01020</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
