<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Replication of the plexDIA analysis (Derks et al. 2022) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Replication of the plexDIA analysis (Derks et al. 2022)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/derks2022.html">Replication of the plexDIA analysis (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Reproduction of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Replication of the plexDIA analysis (Derks et al. 2022)</h1>
                        <h4 data-toc-skip class="author">Christophe Vanderaa<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, Computational Biology, UCLouvain</h4>
                        <h4 data-toc-skip class="author">Laurent Gatto, Computational Biology, UCLouvain</h4>
            
            <h4 data-toc-skip class="date">18 August 2022</h4>
      
      
      <div class="hidden name"><code>derks2022.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>plexDIA (<a href="http://dx.doi.org/10.1038/s41587-022-01389-w" class="external-link">Derks et al. 2022</a>) enables the profiling of the proteome of single cells using a multiplexed DIA data acquisition strategy. The pipeline includes the nPOP sample processing protocole (<a href="http://dx.doi.org/10.1101/2021.04.24.441211" class="external-link">Leduc et al. 2022</a>), with mTRAQ labeling of the samples. The authors used the DIA-NN software (<a href="http://dx.doi.org/10.1038/s41592-019-0638-x" class="external-link">Demichev et al. 2022</a>) to identify and quantify the MS data.</p>
<p>Let’s first load the replication package to make use of some helper functions. Those functions are only meant for this replication vignette and are not designed for general use.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SCP.replication"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="scp-and-the-plexdia-data-analysis-workflow">
<code>scp</code> and the plexDIA data analysis workflow<a class="anchor" aria-label="anchor" href="#scp-and-the-plexdia-data-analysis-workflow"></a>
</h4>
<p>The code provided along with the article can be retrieved from <a href="https://github.com/SlavovLab/plexDIA" class="external-link">this GitHub repository</a>. The objective of this vignette is to replicate the analysis script while providing standardized, easy-to-read, and well documented code. Therefore, our first contribution is to formalize the data processing into a conceptual flow chart.</p>
<div class="figure" style="text-align: center">
<img src="figs/derks2022-workflow.png" alt="Overview of the processing workflow by Derks et al." width="100%"><p class="caption">
Overview of the processing workflow by Derks et al.
</p>
</div>
<p>This replication vignette relies on a data framework dedicated to SCP data analysis that combines two Bioconductor classes (<a href="http://dx.doi.org/10.1080/14789450.2021.1988571" class="external-link">Vanderaa et al. 2021</a>):</p>
<ul>
<li>The <code>SingleCellExperiment</code> class provides an interface to many cutting edge methods for single-cell analysis</li>
<li>The <code>QFeatures</code> class facilitates manipulation and processing of MS-based quantitative data.</li>
</ul>
<p>The <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/scp/inst/doc/scp.html" class="external-link"><code>scp</code> vignette</a> provides detailed information about the data structure. The <code>scp</code> package extends the functionality of <code>QFeatures</code> for single-cell application. <code>scp</code> offers a standardized implementation for single-cell processing methods.</p>
<p>The required packages for running this workflow are listed below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Core packages of this workflow</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">"scp"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"scpdata"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"sva"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Utility packages for data manipulation and visualization</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://tidyverse.tidyverse.org" class="external-link">"tidyverse"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/eclarke/ggbeeswarm" class="external-link">"ggbeeswarm"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/slowkow/ggrepel" class="external-link">"ggrepel"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/hadley/reshape" class="external-link">"reshape2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/LKremer/ggpointdensity" class="external-link">"ggpointdensity"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="scpdata-and-the-leduc2022-dataset">
<code>scpdata</code> and the <code>leduc2022</code> dataset<a class="anchor" aria-label="anchor" href="#scpdata-and-the-leduc2022-dataset"></a>
</h4>
<p>We also implemented a data package called <code>scpdata</code>. It distributes published SCP datasets, such as the <code>derks2022</code> dataset. The datasets were downloaded from the data source provided in the publication and formatted to a <code>QFeatures</code> object so that it is compatible with our software. The underlying data storage is based on the <code>ExperimentHub</code> package that provides a cloud-based storage infrastructure.</p>
<p>The <code>derks2022</code> dataset is provided at different levels of processing:</p>
<ul>
<li>The <strong>.raw and .d</strong> files that were generated by the mass-spectrometers. This data is not included in <code>scpdata</code>.</li>
<li>The <strong>DIA-NN main output report table</strong> that contains the results of the spectrum identification and quantification.</li>
<li>The <strong>DIA-NN MS1 extracted signal table</strong>, if a precursor was identified in at least one of the mTRAQ channels in the main report, signal (if any) will be extracted from the other channels regardless of whether there is sufficient evidence in those channels at 1% FDR.</li>
<li>A <strong>processed protein data table</strong> is provided as the final output of the data processing. This workflow will try to replicate this table.</li>
</ul>
<p>The workflow starts with precursor tables and will generate the processed protein data. The authors provided the DIA-NN output tables and the sample annotation table through a <a href="https://drive.google.com/drive/folders/1pUC2zgXKtKYn22mlor0lmUDK0frgwL_-?usp=sharing" class="external-link">Google Drive repo</a>. Protein data are shared a TXT files. We highly value the effort the authors have made to publicly share all the data generated in their project, from raw files to final expression tables (see the Slavov Lab <a href="https://scp.slavovlab.net/Derks_et_al_2022" class="external-link">website</a>).</p>
</div>
<div class="section level4">
<h4 id="to-include-in-scpdata">To include in <code>scpdata</code><a class="anchor" aria-label="anchor" href="#to-include-in-scpdata"></a>
</h4>
<p>Actually, <code>derks2022()</code> is still in development and is not yet available from <code>scpdata</code>. The code below will soon be published in <code>scpdata</code>. We here include it so that we can still run the vignette.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datadir</span> <span class="op">&lt;-</span> <span class="st">"~/PhD/.localdata/SCP/derks2022/"</span></span>
<span></span>
<span><span class="co">## Sample annotations</span></span>
<span></span>
<span><span class="va">sampleAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"Meta_SingleCell_updated_1.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Add which dataset each sample is part of</span></span>
<span><span class="va">sampleAnnot</span><span class="op">$</span><span class="va">dataset</span> <span class="op">&lt;-</span> <span class="va">sampleAnnot</span><span class="op">$</span><span class="va">Instrument</span></span>
<span><span class="va">sampleAnnot</span><span class="op">$</span><span class="va">dataset</span><span class="op">[</span><span class="op">!</span><span class="va">sampleAnnot</span><span class="op">$</span><span class="va">Real_single_cell</span> <span class="op">&amp;</span> </span>
<span>                        <span class="va">sampleAnnot</span><span class="op">$</span><span class="va">Instrument</span> <span class="op">==</span> <span class="st">"Q-Exactive"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"bulk"</span></span>
<span><span class="co">## Adapt variable to better match the DIANN output data</span></span>
<span><span class="va">sampleAnnot</span><span class="op">$</span><span class="va">Label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">sampleAnnot</span><span class="op">$</span><span class="va">Label</span><span class="op">)</span></span>
<span><span class="va">timsPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.names.html" class="external-link">make.names</a></span><span class="op">(</span><span class="st">"F:\\JD\\plexDIA\\Bruker\\OneDrive_1_3-9-2022\\"</span><span class="op">)</span></span>
<span><span class="va">qePath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.names.html" class="external-link">make.names</a></span><span class="op">(</span><span class="st">"F:\\JD\\plexDIA\\nPOP\\"</span><span class="op">)</span></span>
<span><span class="va">sampleAnnot</span><span class="op">$</span><span class="va">File.Name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">sampleAnnot</span><span class="op">$</span><span class="va">Instrument</span> <span class="op">==</span> <span class="st">"timsTOFSCP"</span>,</span>
<span>                                <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">timsPath</span>, <span class="va">sampleAnnot</span><span class="op">$</span><span class="va">Raw</span>, <span class="st">".d"</span><span class="op">)</span>,</span>
<span>                                <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">qePath</span>, <span class="va">sampleAnnot</span><span class="op">$</span><span class="va">Raw</span>, <span class="st">".raw"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Bulk data</span></span>
<span></span>
<span><span class="co"># We load the DIA-NN main output table and the MS1 extracted report table.</span></span>
<span><span class="co"># These are read and combined in a `QFeatures`object. </span></span>
<span><span class="va">extractedDataBulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"qe_bulk/Report.pr_matrix_channels_ms1_extracted.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">reportDataBulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"qe_bulk/Report.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSCPfromDIANN.html">readSCPfromDIANN</a></span><span class="op">(</span>colData <span class="op">=</span> <span class="va">sampleAnnot</span>, </span>
<span>                         reportData <span class="op">=</span> <span class="va">reportDataBulk</span>,</span>
<span>                         extractedData <span class="op">=</span> <span class="va">extractedDataBulk</span>,</span>
<span>                         multiplexing <span class="op">=</span> <span class="st">"mTRAQ"</span><span class="op">)</span></span>
<span><span class="co">## Rename the MS1Extracted assay</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"bulk_prec_extracted"</span></span>
<span></span>
<span><span class="co">## Load timsTOF-SCP data</span></span>
<span></span>
<span><span class="co"># We load the DIA-NN main output table and the MS1 extracted report table.</span></span>
<span><span class="co"># These are read and combined in a `QFeatures`object. </span></span>
<span><span class="va">extractedDataTims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"tims_sc/Report.pr_matrix_channels_ms1_extracted.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">reportDataTims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"tims_sc/Report.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># We modify the `Run` variable to match the `Run` variables in the other tables</span></span>
<span><span class="va">reportDataTims</span><span class="op">$</span><span class="va">Run</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.names.html" class="external-link">make.names</a></span><span class="op">(</span><span class="va">reportDataTims</span><span class="op">$</span><span class="va">Run</span><span class="op">)</span></span>
<span><span class="va">tims</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSCPfromDIANN.html">readSCPfromDIANN</a></span><span class="op">(</span>colData <span class="op">=</span> <span class="va">sampleAnnot</span>, </span>
<span>                         reportData <span class="op">=</span> <span class="va">reportDataTims</span>,</span>
<span>                         extractedData <span class="op">=</span> <span class="va">extractedDataTims</span>,</span>
<span>                         multiplexing <span class="op">=</span> <span class="st">"mTRAQ"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tims</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">tims</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"tims_prec_extracted"</span></span>
<span></span>
<span><span class="co">## Load Q-Exactive data</span></span>
<span></span>
<span><span class="co"># We load the DIA-NN main output table and the MS1 extracted report table.</span></span>
<span><span class="co"># These are read and combined in a `QFeatures`object. </span></span>
<span><span class="va">extractedDataQE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"qe_sc/Report.pr_matrix_channels_ms1_extracted.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">reportDataQE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"qe_sc/Report.tsv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSCPfromDIANN.html">readSCPfromDIANN</a></span><span class="op">(</span>colData <span class="op">=</span> <span class="va">sampleAnnot</span>, </span>
<span>                       reportData <span class="op">=</span> <span class="va">reportDataQE</span>,</span>
<span>                       extractedData <span class="op">=</span> <span class="va">extractedDataQE</span>,</span>
<span>                       multiplexing <span class="op">=</span> <span class="st">"mTRAQ"</span><span class="op">)</span></span>
<span><span class="co">## Warning in (function (featureData, colData, batchCol, channelCol,</span></span>
<span><span class="co">## suffix = NULL, : Missing metadata. The features are removed for</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1200.raw, F..JD.plexDIA.nPOP.wJD1205.raw,</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1203.raw, F..JD.plexDIA.nPOP.wJD1201.raw,</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1204.raw, F..JD.plexDIA.nPOP.wJD1202.raw</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">qe</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"qe_prec_extracted"</span></span>
<span></span>
<span><span class="co">## Load protein data</span></span>
<span></span>
<span><span class="va">prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="st">"Proteins_SC_IDs.txt"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/readSingleCellExperiment.html" class="external-link">readSingleCellExperiment</a></span><span class="op">(</span><span class="va">prots</span>, fname <span class="op">=</span> <span class="st">"prot"</span>,</span>
<span>                                  ecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prots</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">prots</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span><span class="va">sampleAnnot</span><span class="op">[</span><span class="va">sampleAnnot</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prots</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prots</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">prots</span><span class="op">$</span><span class="va">File.Name</span>, <span class="st">"."</span>, <span class="va">prots</span><span class="op">$</span><span class="va">Label</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Combine all datasets</span></span>
<span><span class="co">## TODO implement a c() method to simplify this process...</span></span>
<span><span class="va">derks2022</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">QFeatures</a></span><span class="op">(</span>experiments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">experiments</span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span>, <span class="fu">experiments</span><span class="op">(</span><span class="va">tims</span><span class="op">)</span>,</span>
<span>                                       <span class="fu">experiments</span><span class="op">(</span><span class="va">qe</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">tims</span><span class="op">)</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">qe</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       assayLinks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">append</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">@</span><span class="va">assayLinks</span>, <span class="va">tims</span><span class="op">@</span><span class="va">assayLinks</span>, </span>
<span>                                                        <span class="va">qe</span><span class="op">@</span><span class="va">assayLinks</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">derks2022</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">derks2022</span>, <span class="va">prots</span>, name <span class="op">=</span> <span class="st">"proteins"</span>, dropColData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">derks2022</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLink</a></span><span class="op">(</span><span class="va">derks2022</span>, </span>
<span>                          from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"extracted$"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks2022</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                          to <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                          varFrom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Protein.Group"</span>, <span class="fl">3</span><span class="op">)</span>, </span>
<span>                          varTo <span class="op">=</span> <span class="st">"prot"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Save data</span></span>
<span><span class="co">## ...</span></span></code></pre></div>
<p><strong>Stop “To include in <code>scpdata</code>”</strong></p>
<p>We formatted the <code>derks2022</code> dataset following our data framework. The formatted data can be retrieved from the <code>scpdata</code> package using the <code>derks2022()</code> function. All datasets in <code>scpdata</code> are called after the first author and the date of publication.</p>
<p>The <code>derks2022</code> data combines 3 main datasets: the 100-cell equivalent samples acquired with a Q-Exactive instrument, the single-cell samples acquired with a Q-Exactive instrument (we call this dataset <code>qe</code>), and the the single-cell samples acquired with a timsTOF-SCP instrument (we call this dataset <code>tims</code>). Each dataset contains the DIA-NN main report data split over each run and the combined MS1 extracted data (see above for more details). The data also contains the protein data for all samples as processed by the authors.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (scp &lt;- derks2022())</span></span>
<span><span class="op">(</span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="va">derks2022</span><span class="op">)</span></span>
<span><span class="co">## An instance of class QFeatures containing 66 assays:</span></span>
<span><span class="co">##  [1] F..JD.plexDIA.nPOP.wJD1194.raw: SingleCellExperiment with 19618 rows and 3 columns </span></span>
<span><span class="co">##  [2] F..JD.plexDIA.nPOP.wJD1195.raw: SingleCellExperiment with 19587 rows and 3 columns </span></span>
<span><span class="co">##  [3] F..JD.plexDIA.nPOP.wJD1196.raw: SingleCellExperiment with 19176 rows and 3 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [64] F..JD.plexDIA.nPOP.wJD1193.raw: SingleCellExperiment with 2643 rows and 3 columns </span></span>
<span><span class="co">##  [65] qe_prec_extracted: SingleCellExperiment with 8590 rows and 144 columns </span></span>
<span><span class="co">##  [66] proteins: SingleCellExperiment with 1475 rows and 164 columns</span></span></code></pre></div>
<p>The datasets are stored in a <code>QFeatures</code> object. In total, it contains 66 different <code>SingleCellExperiment</code> objects that we refer to as <strong>assays</strong>. Each assay contains expression data along with feature metadata. Each row in an assay represents a <strong>feature</strong>, in this case a precursor or a protein depending on the assay. Each column in an assay represents a <strong>sample</strong>. During sample preparation, 3 samples are pooled using mTRAQ labeling, hence some assays contain 3 columns corresponding to each measured channel.</p>
<p>Using <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code>, we can have a quick overview of the assays.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span></span>
<span><span class="co">## Warning in plot.QFeatures(derks): The QFeatures object contains many assays. You</span></span>
<span><span class="co">## may want to consider creating an interactive plot (set 'interactive = TRUE')</span></span></code></pre></div>
<p><img src="derks2022_files/figure-html/unnamed-chunk-5-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>This figure is crowded, you can use <code>plot(derks, interactive = TRUE)</code> to interactively explore this map. However, we can see that the different runs in the top of the figure converge to 3 assays that contain the extracted MS1 precursor signal for the three datasets: <code>bulk</code>, <code>tims</code> and <code>qe</code>. Finally, the protein table contains data derived from all three datasets.</p>
<p>The objective of this vignette is to replicate the combined protein data table from the precursor assays following the same data processing workflow as the original study by Derks et al. but using standardized functionality.</p>
<p>We extract the <code>proteins</code> assay and keep it for later benchmarking. <code>getWithColData()</code> extract an assay of interest along with the associated sample annotations. We then remove the assay from the <code>S</code> object for the remainder of the processing. We apply this using <code>removeAssays</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proteins_derks</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">derks</span>, <span class="st">"proteins"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">removeAssay</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"proteins"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="remove-unwanted-samples">Remove unwanted samples<a class="anchor" aria-label="anchor" href="#remove-unwanted-samples"></a>
</h2>
<p>Before starting the data processing, we remove the samples that are not considered by the authors to generate the protein data. The timsTOF-SCP dataset was acquired with a few bulk samples diluted to single-cell equivalent. These samples are ending with <code>"_t_DB"</code>. The Q-Exactive dataset contains single-cells, but also a dozen negative controls (encoded as <code>"Neg"</code>).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">derks</span><span class="op">$</span><span class="va">Celltype</span>, <span class="va">derks</span><span class="op">$</span><span class="va">dataset</span><span class="op">)</span></span>
<span><span class="co">##                </span></span>
<span><span class="co">##                 bulk Q-Exactive timsTOFSCP</span></span>
<span><span class="co">##   Melanoma         0         45          0</span></span>
<span><span class="co">##   Melanoma_DB      3          0          0</span></span>
<span><span class="co">##   Melanoma_t       0          0         10</span></span>
<span><span class="co">##   Melanoma_t_DB    0          0          1</span></span>
<span><span class="co">##   Neg              0         12          0</span></span>
<span><span class="co">##   PDAC             0         45          0</span></span>
<span><span class="co">##   PDAC_DB          3          0          0</span></span>
<span><span class="co">##   PDAC_t           0          0         10</span></span>
<span><span class="co">##   PDAC_t_DB        0          0          1</span></span>
<span><span class="co">##   U-937            0         42          0</span></span>
<span><span class="co">##   U-937_DB         3          0          0</span></span>
<span><span class="co">##   U-937_t          0          0         10</span></span>
<span><span class="co">##   U-937_t_DB       0          0          1</span></span></code></pre></div>
<p>We remove those unwanted samples by accessing the information from the <code>colData</code> using <code>subsetByColData()</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">derks</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Neg|_t_DB"</span>, <span class="va">derks</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="isotopic-carry-over-correction">Isotopic carry over correction<a class="anchor" aria-label="anchor" href="#isotopic-carry-over-correction"></a>
</h2>
<p>The first data processing step is to correct for isotopic carry over from one mTRAQ channel to another. This is performed using <code><a href="../reference/correctIsotopicCarryover.html">correctIsotopicCarryover()</a></code>. The function takes an assay and returns the same assay with corrected quantifications. Since there are 3 datasets, we will loop over the 3 assays containing the extracted MS1 signal. These corrected assays are added to the <code>QFeatures</code> object using <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay()</a></code> and the links between the precursors are added using</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bulk"</span>, <span class="st">"qe"</span>, <span class="st">"tims"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">inputAssay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">i</span> ,<span class="st">"_prec_extracted"</span><span class="op">)</span></span>
<span>    <span class="va">outputAssay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">i</span> ,<span class="st">"_prec_corrected"</span><span class="op">)</span></span>
<span>    <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/correctIsotopicCarryover.html">correctIsotopicCarryover</a></span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="va">inputAssay</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">derks</span>, <span class="va">x</span>, name <span class="op">=</span> <span class="va">outputAssay</span><span class="op">)</span></span>
<span>    <span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLinkOneToOne</a></span><span class="op">(</span><span class="va">derks</span>, from <span class="op">=</span> <span class="va">inputAssay</span>, </span>
<span>                                  to <span class="op">=</span> <span class="va">outputAssay</span><span class="op">)</span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Finished correction for the"</span>, <span class="va">i</span>, <span class="st">"dataset\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">## Finished correction for the bulk dataset</span></span>
<span><span class="co">## Finished correction for the qe dataset</span></span>
<span><span class="co">## Finished correction for the tims dataset</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="precursor-fdr-control">Precursor FDR control<a class="anchor" aria-label="anchor" href="#precursor-fdr-control"></a>
</h2>
<p><strong>!discussion</strong> this is too complicated. Refactoring of filterFeatures in QFeatures would make this much easier.</p>
<p><strong>!discussion</strong> There was a copy-paste issue in the original code: they used the scQE dataset instead of the bulkQE dataset…</p>
<p><code>PGs_0.01FDR &lt;- SC[which(SC$Lib.PG.Q.Value&lt;0.01),] %&gt;% distinct(Protein.Group)</code></p>
<p>Should have been</p>
<p><code>PGs_0.01FDR &lt;- B[which(B$Lib.PG.Q.Value&lt;0.01),] %&gt;% distinct(Protein.Group)</code></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bulk"</span>, <span class="st">"qe"</span>, <span class="st">"tims"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">repAssays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">assayLink</a></span><span class="op">(</span><span class="va">derks</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">i</span> ,<span class="st">"_prec_extracted"</span><span class="op">)</span><span class="op">)</span><span class="op">@</span><span class="va">from</span></span>
<span>    <span class="va">rd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">derks</span>, <span class="va">repAssays</span><span class="op">)</span></span>
<span>    <span class="va">PGs_0.01FDR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">rd</span><span class="op">$</span><span class="va">Protein.Group</span><span class="op">[</span><span class="va">rd</span><span class="op">$</span><span class="va">Lib.PG.Q.Value</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">derks</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">i</span> ,<span class="st">"_prec_corrected"</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">Protein.Group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">PGs_0.01FDR</span></span>
<span>    <span class="va">derks</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">i</span> ,<span class="st">"_prec_corrected"</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">ind</span>, <span class="op">]</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">ind</span><span class="op">)</span><span class="op">)</span> </span>
<span>        <span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLink</a></span><span class="op">(</span><span class="va">derks</span>, from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">i</span> ,<span class="st">"_prec_extracted"</span><span class="op">)</span>, </span>
<span>                              to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">i</span> ,<span class="st">"_prec_corrected"</span><span class="op">)</span>,</span>
<span>                              varFrom <span class="op">=</span> <span class="st">"Precursor.Id"</span>,</span>
<span>                              varTo <span class="op">=</span> <span class="st">"Precursor.Id"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Finished FDR control for the"</span>, <span class="va">i</span>, <span class="st">"dataset\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">## Finished FDR control for the bulk dataset</span></span>
<span><span class="co">## Warning in replaceAssay(x = x, y = value, i = i): Links between assays were</span></span>
<span><span class="co">## lost/removed during replacement. See '?addAssayLink' to restore them manually.</span></span>
<span><span class="co">## Finished FDR control for the qe dataset</span></span>
<span><span class="co">## Finished FDR control for the tims dataset</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="combine-the-datasets">Combine the datasets<a class="anchor" aria-label="anchor" href="#combine-the-datasets"></a>
</h2>
<p>Up to now, we kept the data belonging to each MS run in separate assays. We now combine all batches into a single assay. This can easily be done using the <code><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays()</a></code> function from the <code>QFeatures</code> package.</p>
<div class="section level3">
<h3 id="consensus-mapping-of-precursors-to-proteins">Consensus mapping of precursors to proteins<a class="anchor" aria-label="anchor" href="#consensus-mapping-of-precursors-to-proteins"></a>
</h3>
<p>We need to account for an issue in the data. <code><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays()</a></code> will only keep the metadata variables that have the same value between matching rows. However, some precursors map to one protein group in one run and to another protein group in another run. Hence, the protein group is not constant for all precursors and is removed during joining. It is important we keep the protein group information in the <code>rowData</code> since we will later need it to aggregate precursors to proteins. To avoid this issue, we replace the problematic precursor to protein group mappings through a majority vote.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Generate a list of DataFrames with the information to modify</span></span>
<span><span class="va">precAssays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"corrected$"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ppMap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="va">precAssays</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Precursor.Id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">## The majority vote happens here</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Protein.Group <span class="op">=</span></span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">Protein.Group</span><span class="op">)</span>,</span>
<span>                          decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Precursor.Id</span>, <span class="va">Protein.Group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">Precursor.Id</span>, <span class="va">Protein.Group</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">precAssays</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">Precursor.Id</span>, </span>
<span>                 <span class="va">ppMap</span><span class="op">$</span><span class="va">Precursor.Id</span><span class="op">)</span></span>
<span>    <span class="fu">DataFrame</span><span class="op">(</span>Protein.Group <span class="op">=</span> <span class="va">ppMap</span><span class="op">$</span><span class="va">Protein.Group</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">## Name the list</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">consensus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">precAssays</span></span>
<span><span class="co">## Modify the rowData</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">consensus</span></span></code></pre></div>
<p>We now join the three datasets in a single assay. This is performed using <code><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"corrected$"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     name <span class="op">=</span> <span class="st">"prec_corrected"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## An instance of class QFeatures containing 69 assays:</span></span>
<span><span class="co">##  [1] F..JD.plexDIA.nPOP.wJD1194.raw: SingleCellExperiment with 19618 rows and 3 columns </span></span>
<span><span class="co">##  [2] F..JD.plexDIA.nPOP.wJD1195.raw: SingleCellExperiment with 19587 rows and 3 columns </span></span>
<span><span class="co">##  [3] F..JD.plexDIA.nPOP.wJD1196.raw: SingleCellExperiment with 19176 rows and 3 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [67] qe_prec_corrected: SingleCellExperiment with 8416 rows and 132 columns </span></span>
<span><span class="co">##  [68] tims_prec_corrected: SingleCellExperiment with 8722 rows and 30 columns </span></span>
<span><span class="co">##  [69] prec_corrected: SingleCellExperiment with 23585 rows and 171 columns</span></span></code></pre></div>
<p>The last assay, <code>prec_corrected</code> contains the data for all 3 datasets.</p>
</div>
</div>
<div class="section level2">
<h2 id="filter-missing-data">Filter missing data<a class="anchor" aria-label="anchor" href="#filter-missing-data"></a>
</h2>
<p>Next, the authors require less than 60% missing data per cell. We compute the percent missing data for each column. We store it in the <code>colData</code> using <code>derks$ &lt;-</code> for later use.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span><span class="op">$</span><span class="va">pNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prec_corrected"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><strong>!discussion</strong>: I do not agree with how the missing data is computed. Missing data is defined by the authors as the the number of zero values and ignore the presence of NA’s in the data… I don’t understand why ignoring the NA’s, it dramatically underestimates the missingness. Instead, I would have done as depicted in the chunk below although it considers the union of all proteins found in all 3 datasets and will .</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prec_corrected"</span><span class="op">)</span></span>
<span><span class="va">derks</span><span class="op">$</span><span class="va">pNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">nNA</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prec_corrected"</span><span class="op">)</span><span class="op">$</span><span class="va">nNAcols</span><span class="op">$</span><span class="va">pNA</span></span></code></pre></div>
<p>We plot the percent missingness for each data set.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">pNA</span>, </span>
<span>        x <span class="op">=</span> <span class="va">dataset</span>, </span>
<span>        color <span class="op">=</span> <span class="va">Celltype</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html" class="external-link">geom_beeswarm</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="derks2022_files/figure-html/unnamed-chunk-14-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>Most samples comply to the threshold. We remove the few single-cells with too many missing data using <code>subsetByColdata()</code> since we have stored <code>pNA</code> in the <code>colData</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">derks</span>, <span class="va">derks</span><span class="op">$</span><span class="va">pNA</span> <span class="op">&lt;</span> <span class="fl">0.6</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="clean-missing-data">Clean missing data<a class="anchor" aria-label="anchor" href="#clean-missing-data"></a>
</h2>
<p>Log-transformation has generated infinite values because zeros were present in the data. We clean these infinite values and residual zero by replacing them with NA’s using <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">infIsNA()</a></code> and <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA()</a></code>, respectively.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">infIsNA</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prec_corrected"</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prec_corrected"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h2>
<p>The authors then apply a first batch correction by aligning the precursor means across the different acquisition runs. First, we extract the combined precursor assay along with the colData using <code>getWithColData()</code> and extract the quantification matrix from it using <code>assay()</code></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prec_corrected"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p>For each acquisition run, we divided each row by its mean.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">batch</span> <span class="kw">in</span> <span class="va">sce</span><span class="op">$</span><span class="va">File.Name</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">File.Name</span> <span class="op">==</span> <span class="va">batch</span><span class="op">)</span> </span>
<span>    <span class="va">m</span><span class="op">[</span>, <span class="va">ind</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">ind</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>, FUN <span class="op">=</span> <span class="st">"/"</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">ind</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m</span></span></code></pre></div>
<p>We add the batch corrected values as a new assay using <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay()</a></code> and keep the links with the previous assay using <code><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLinkOneToOne()</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">derks</span>, <span class="va">sce</span>, name <span class="op">=</span> <span class="st">"prec_runnorm"</span>, dropColData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLinkOneToOne</a></span><span class="op">(</span><span class="va">derks</span>, from <span class="op">=</span> <span class="st">"prec_corrected"</span>, to <span class="op">=</span> <span class="st">"prec_runnorm"</span><span class="op">)</span></span></code></pre></div>
<p><strong>!discussion</strong>: this type of batch correction is rather common and should be implemented in <code>scp</code> or maybe even in <code>QFeatures</code>.</p>
</div>
<div class="section level2">
<h2 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h2>
<p>The next step is to apply normalization. The authors perform it at the sample level and the precursor level.</p>
<div class="section level3">
<h3 id="median-normalize-per-sample">Median normalize per sample<a class="anchor" aria-label="anchor" href="#median-normalize-per-sample"></a>
</h3>
<p>Median normalization per sample is available among the methods from <code><a href="https://UCLouvain-CBIO.github.io/scp/reference/normalizeSCP.html" class="external-link">normalizeSCP()</a></code>. It will automatically add a new assay that we call <code>prec_sampnorm</code></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/normalizeSCP.html" class="external-link">normalizeSCP</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prec_runnorm"</span>, name <span class="op">=</span> <span class="st">"prec_sampnorm"</span>,</span>
<span>                      method <span class="op">=</span> <span class="st">"div.median"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="mean-normalize-per-precursor">Mean normalize per precursor<a class="anchor" aria-label="anchor" href="#mean-normalize-per-precursor"></a>
</h3>
<p>Mean normalization per sample is not available among the methods from <code><a href="https://UCLouvain-CBIO.github.io/scp/reference/normalizeSCP.html" class="external-link">normalizeSCP()</a></code>, but can be applied using <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep()</a></code> offering more flexibility.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rowm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prec_sampnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="st">"prec_sampnorm"</span>, name <span class="op">=</span> <span class="st">"prec_featnorm"</span>, </span>
<span>               MARGIN <span class="op">=</span> <span class="fl">1</span>, FUN <span class="op">=</span> <span class="st">"/"</span>, STATS <span class="op">=</span> <span class="va">rowm</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="protein-aggregation">Protein aggregation<a class="anchor" aria-label="anchor" href="#protein-aggregation"></a>
</h2>
<p>Up to now, the data was processed at the precursor level. Several precursors belong to the same protein. One protein group is represented by over 150 precursors.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prec_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">Protein.Group</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Number precursors per protein group"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></code></pre></div>
<p><img src="derks2022_files/figure-html/unnamed-chunk-22-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>We aggregate all precursors belonging to the same proteins. Median aggregation is performed using <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures()</a></code> and providing <code>fun = colMedians</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="st">"prec_featnorm"</span>, name <span class="op">=</span> <span class="st">"prot"</span>,</span>
<span>                           fcol <span class="op">=</span> <span class="st">"Protein.Group"</span>, fun <span class="op">=</span> <span class="va">colMedians</span>, </span>
<span>                           na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="batch-correction-1">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction-1"></a>
</h2>
<p>After protein agregation, the authors perform again a batch correction round. We use the same approach as above.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prot"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">batch</span> <span class="kw">in</span> <span class="va">sce</span><span class="op">$</span><span class="va">File.Name</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">File.Name</span> <span class="op">==</span> <span class="va">batch</span><span class="op">)</span> </span>
<span>    <span class="va">m</span><span class="op">[</span>, <span class="va">ind</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">ind</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>, FUN <span class="op">=</span> <span class="st">"/"</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">ind</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">derks</span>, <span class="va">sce</span>, name <span class="op">=</span> <span class="st">"prot_runnorm"</span>, dropColData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLinkOneToOne</a></span><span class="op">(</span><span class="va">derks</span>, from <span class="op">=</span> <span class="st">"prot"</span>, to <span class="op">=</span> <span class="st">"prot_runnorm"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalization-1">Normalization<a class="anchor" aria-label="anchor" href="#normalization-1"></a>
</h2>
<p>Again, the batch correction is followed by sample and protein normalization. We use the same approach as above.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/normalizeSCP.html" class="external-link">normalizeSCP</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prot_runnorm"</span>, name <span class="op">=</span> <span class="st">"prot_sampnorm"</span>,</span>
<span>                      method <span class="op">=</span> <span class="st">"div.median"</span><span class="op">)</span></span>
<span><span class="va">rowm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_sampnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="st">"prot_sampnorm"</span>, name <span class="op">=</span> <span class="st">"prot_featnorm"</span>, </span>
<span>               MARGIN <span class="op">=</span> <span class="fl">1</span>, FUN <span class="op">=</span> <span class="st">"/"</span>, STATS <span class="op">=</span> <span class="va">rowm</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="log-transform">Log-transform<a class="anchor" aria-label="anchor" href="#log-transform"></a>
</h2>
<p>The protein data is log2-transformed using <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform()</a></code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="st">"prot_featnorm"</span>, name <span class="op">=</span> <span class="st">"prot_log"</span>,</span>
<span>                      base <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="clean-missing-data-1">Clean missing data<a class="anchor" aria-label="anchor" href="#clean-missing-data-1"></a>
</h2>
<p>Log-transformation has generated infinite values because zeros were present in the data. We clean these infinite values and residual zero by replacing them with NA’s using <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">infIsNA()</a></code> and <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA()</a></code>, respectively.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">infIsNA</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prot_log"</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prot_log"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="filter-missing-data-1">Filter missing data<a class="anchor" aria-label="anchor" href="#filter-missing-data-1"></a>
</h2>
<p>Although we already performed a first filtering based on data missingness, we further filter on protein and sample missingness.</p>
<div class="section level3">
<h3 id="filter-on-protein-missingness">Filter on protein missingness<a class="anchor" aria-label="anchor" href="#filter-on-protein-missingness"></a>
</h3>
<p>We filter out proteins with more than 95% missingness. This is performed using <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA()</a></code> that removes rows (proteins) with a missingness higher than the given threshold.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="st">"prot_log"</span>, pNA <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-on-single-cell-missingness">Filter on single-cell missingness<a class="anchor" aria-label="anchor" href="#filter-on-single-cell-missingness"></a>
</h3>
<p>We then filter out samples with more than 95% missingness. This is performed using the same procedure as above.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span><span class="op">$</span><span class="va">pNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_log"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">derks</span>, <span class="va">derks</span><span class="op">$</span><span class="va">pNA</span> <span class="op">&lt;</span> <span class="fl">0.95</span><span class="op">)</span></span></code></pre></div>
<p><strong>!discussion</strong>: it would be nice to have a <code>byRow = TRUE</code> argument in filterNA to filter based on row or columns.</p>
</div>
</div>
<div class="section level2">
<h2 id="imputation">Imputation<a class="anchor" aria-label="anchor" href="#imputation"></a>
</h2>
<p>Although we filtered on missing data, the protein data is majorly composed of missing values. The graph below shows the distribution of the proportion missingness in cells. Cells contain on average 65 % missing values.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">pNA</span>, </span>
<span>        x <span class="op">=</span> <span class="va">dataset</span>, </span>
<span>        color <span class="op">=</span> <span class="va">Celltype</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html" class="external-link">geom_beeswarm</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="derks2022_files/figure-html/unnamed-chunk-30-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The datasets containing single-cells have more than 50% missing data that is imputed by the authors using a custom hierarchical clustering function. This function is available from <code>SCP.replication</code> as <code><a href="../reference/imputeKnnSCoPE2.html">imputeKnnSCoPE2()</a></code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/imputeKnnSCoPE2.html">imputeKnnSCoPE2</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="st">"prot_log"</span>, name <span class="op">=</span> <span class="st">"prot_imp"</span>, k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="batch-correction-2">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction-2"></a>
</h2>
<p>A final batch correction is applied, this time using the <code>ComBat</code> algorithm from the <code>sva</code> package.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prot_imp"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">Label</span></span>
<span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Celltype</span>, data <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat.html" class="external-link">ComBat</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, batch <span class="op">=</span> <span class="va">batch</span>, mod <span class="op">=</span> <span class="va">mod</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">derks</span>, <span class="va">sce</span>, name <span class="op">=</span> <span class="st">"prot_bc"</span>, dropColData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLinkOneToOne</a></span><span class="op">(</span><span class="va">derks</span>, from <span class="op">=</span> <span class="st">"prot_imp"</span>, to <span class="op">=</span> <span class="st">"prot_bc"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalization-2">Normalization<a class="anchor" aria-label="anchor" href="#normalization-2"></a>
</h2>
<p>Again, the batch correction is followed by sample and protein normalization. We use the same approach as above.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/normalizeSCP.html" class="external-link">normalizeSCP</a></span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prot_bc"</span>, name <span class="op">=</span> <span class="st">"prot_bc_sampnorm"</span>,</span>
<span>                      method <span class="op">=</span> <span class="st">"center.median"</span><span class="op">)</span></span>
<span><span class="va">rowm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_sampnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="st">"prot_bc_sampnorm"</span>, name <span class="op">=</span> <span class="st">"prot_bc_featnorm"</span>, </span>
<span>               MARGIN <span class="op">=</span> <span class="fl">1</span>, FUN <span class="op">=</span> <span class="st">"-"</span>, STATS <span class="op">=</span> <span class="va">rowm</span><span class="op">)</span></span></code></pre></div>
<p>This last normalization step leads to the final data as processed by Derks et al. We will now compare our results with the results published by the authors.</p>
</div>
<div class="section level2">
<h2 id="benchmark-final-data">Benchmark final data<a class="anchor" aria-label="anchor" href="#benchmark-final-data"></a>
</h2>
<p>Let’s first compare the filtered cells.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allElements</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">union</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">proteins_derks</span><span class="op">)</span>, </span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>derks2022 <span class="op">=</span> <span class="va">allElements</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">proteins_derks</span><span class="op">)</span>,</span>
<span>      scp <span class="op">=</span> <span class="va">allElements</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## derks2022 FALSE TRUE</span></span>
<span><span class="co">##     FALSE     0    7</span></span>
<span><span class="co">##     TRUE      7  157</span></span></code></pre></div>
<p>There is a good agreement between the set of filtered cells after the data processing by the authors and the data processing in this vignette.</p>
<p>Let’s now compare the filtered protein groups.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allElements</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">union</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">proteins_derks</span><span class="op">)</span>, </span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>derks2022 <span class="op">=</span> <span class="va">allElements</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">proteins_derks</span><span class="op">)</span>,</span>
<span>      scp <span class="op">=</span> <span class="va">allElements</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## derks2022 FALSE TRUE</span></span>
<span><span class="co">##     FALSE     0  561</span></span>
<span><span class="co">##     TRUE    124 1351</span></span></code></pre></div>
<p>Most protein groups are found in both data processing workflows, but the agreement is weak.</p>
<p>Let’s now compare the quantitative data between the two matrices. To do so, we need to intersect the column names and the rownames.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">proteins_derks</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">proteins_derks</span><span class="op">)</span>,</span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">err</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">proteins_derks</span><span class="op">)</span><span class="op">[</span><span class="va">rows</span>, <span class="va">cols</span><span class="op">]</span> <span class="op">-</span> </span>
<span>    <span class="fu">assay</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">rows</span>, <span class="va">cols</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>difference <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">err</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">err</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">difference</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"plexDIA - scp"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/scientific_format.html" class="external-link">scientific</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="derks2022_files/figure-html/unnamed-chunk-36-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>There are very large differences between the two workflows. Note that the differences are computed on a log2 scale. However, the mode of the distribution peaks around 0.</p>
</div>
<div class="section level2">
<h2 id="number-precursors-per-cell">Number precursors per cell<a class="anchor" aria-label="anchor" href="#number-precursors-per-cell"></a>
</h2>
</div>
<div class="section level2">
<h2 id="bulk-vs-single-cell">Bulk vs single-cell<a class="anchor" aria-label="anchor" href="#bulk-vs-single-cell"></a>
</h2>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">derks</span>, <span class="st">"prot_imp"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">logFC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>PvsU_tims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">Celltype</span> <span class="op">==</span> <span class="st">"PDAC_t"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-</span> </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">Celltype</span> <span class="op">==</span> <span class="st">"U-937_t"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                    PvsU_qe <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">Celltype</span> <span class="op">==</span> <span class="st">"PDAC"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-</span> </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">Celltype</span> <span class="op">==</span> <span class="st">"U-937"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                    PvsU_bulk <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">Celltype</span> <span class="op">==</span> <span class="st">"PDAC_DB"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-</span></span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">Celltype</span> <span class="op">==</span> <span class="st">"U-937_DB"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                    Protein.Group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">logFC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">logFC</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Protein.Group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17096;P17096-3"</span>, <span class="st">"P07437"</span>, <span class="st">"P08729"</span><span class="op">)</span>,</span>
<span>                                     Protein.Name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HMGA1"</span>, <span class="st">"TUBB"</span>, <span class="st">"KRT7"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                   by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Protein.Group"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PvsU_bulk</span>, </span>
<span>        y <span class="op">=</span> <span class="va">PvsU_qe</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Q-Exactive vs Bulk"</span>,</span>
<span>            subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rho = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">$</span><span class="va">PvsU_qe</span>, </span>
<span>                                                  <span class="va">logFC</span><span class="op">$</span><span class="va">PvsU_bulk</span>, </span>
<span>                                                  use <span class="op">=</span> <span class="st">"pairwise.complete.obs"</span>,</span>
<span>                                                  method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">Log</span><span class="op">[</span><span class="st">"2"</span><span class="op">]</span>,<span class="st">", PDAC/U-937 100 cells"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>         y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">Log</span><span class="op">[</span><span class="st">"2"</span><span class="op">]</span>,<span class="st">", PDAC/U-937 1 cell (Q-Exactive)"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PvsU_bulk</span>, </span>
<span>        y <span class="op">=</span> <span class="va">PvsU_tims</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"timsTOF-SCP vs Bulk"</span>,</span>
<span>            subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rho = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">$</span><span class="va">PvsU_tims</span>, </span>
<span>                                                  <span class="va">logFC</span><span class="op">$</span><span class="va">PvsU_bulk</span>, </span>
<span>                                                  use <span class="op">=</span> <span class="st">"pairwise.complete.obs"</span>,</span>
<span>                                                  method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">Log</span><span class="op">[</span><span class="st">"2"</span><span class="op">]</span>,<span class="st">", PDAC/U-937 100 cells"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>         y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">Log</span><span class="op">[</span><span class="st">"2"</span><span class="op">]</span>,<span class="st">", PDAC/U-937 1 cell (timTOF-SCP)"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggpointdensity/man/geom_pointdensity.html" class="external-link">geom_pointdensity</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_color_continuous</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Protein.Name</span><span class="op">)</span>,</span>
<span>                     min.segment.length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">'lines'</span><span class="op">)</span>,</span>
<span>                     box.padding <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                     size <span class="op">=</span> <span class="fl">2.75</span><span class="op">)</span> </span>
<span><span class="co">## Warning: Removed 1909 rows containing missing values (geom_label_repel).</span></span>
<span><span class="co">## Removed 1909 rows containing missing values (geom_label_repel).</span></span></code></pre></div>
<p><img src="derks2022_files/figure-html/unnamed-chunk-40-1.png" width="70%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<p>We here plot the PCA to reproduce Figure 6p from the author’s paper. Although the we could not accurately reproduce the data provided by the authors, we can clearly see the same global pattern as in the original work.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">derks</span><span class="op">[[</span><span class="st">"prot_bc_featnorm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">pcaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pcaSCoPE2.html">pcaSCoPE2</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="co">## Percent of variance explained by each principle component</span></span>
<span><span class="va">pca_var</span> <span class="op">&lt;-</span> <span class="va">pcaRes</span><span class="op">$</span><span class="va">values</span></span>
<span><span class="va">percent_var</span><span class="op">&lt;-</span> <span class="va">pca_var</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pca_var</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span></span>
<span><span class="co">## PCA scores data</span></span>
<span><span class="va">pcaDf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>PC <span class="op">=</span> <span class="va">pcaRes</span><span class="op">$</span><span class="va">vectors</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>                    <span class="fu">colData</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Type</span> <span class="op">&lt;-</span> <span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span></span>
<span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"DB"</span>, <span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Bulk"</span></span>
<span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"t$"</span>, <span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"timsTOF-SCP"</span></span>
<span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Type</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"[_]"</span>, <span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Q-Exactive"</span></span>
<span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"PDAC"</span>, <span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"PDAC"</span></span>
<span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Mel"</span>, <span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Melanoma"</span></span>
<span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^U"</span>, <span class="va">pcaDf</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Monocyte"</span></span>
<span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"mTRAQ"</span>, <span class="va">pcaDf</span><span class="op">$</span><span class="va">Label</span><span class="op">)</span></span>
<span><span class="co">## Plot    </span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pcaDf</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Label</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"orange2"</span>,<span class="st">"purple"</span>,<span class="st">"turquoise"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pcaDf</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Celltype</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#52B788"</span>,<span class="st">"#457B9D"</span>,<span class="st">"#E07A5F"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PCA of the processed data using 'scp'"</span>,</span>
<span>                    subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pcaDf</span><span class="op">$</span><span class="va">Type</span> <span class="op">!=</span> <span class="st">"Bulk"</span><span class="op">)</span>, <span class="st">" cells\n"</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>, <span class="st">" protein groups"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC.1</span>, </span>
<span>        y <span class="op">=</span> <span class="va">PC.2</span>, </span>
<span>        size <span class="op">=</span> <span class="va">Type</span>,</span>
<span>        shape <span class="op">=</span> <span class="va">Type</span>, </span>
<span>        alpha <span class="op">=</span> <span class="va">Type</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.66</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC1 ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">percent_var</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="fl">0</span><span class="op">)</span>,<span class="st">"%)"</span><span class="op">)</span>, </span>
<span>         y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC2 ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">percent_var</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,<span class="fl">0</span><span class="op">)</span>,<span class="st">"%)"</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="derks2022_files/figure-html/unnamed-chunk-41-1.png" width="70%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<div class="section level4">
<h4 id="hardware-and-software">Hardware and software<a class="anchor" aria-label="anchor" href="#hardware-and-software"></a>
</h4>
<p>The system details of the machine that built the vignette are:</p>
<pre><code><span><span class="co">## Machine: Linux (5.15.0-46-generic)</span></span>
<span><span class="co">## R version: R.4.2.0 (svn: 81131)</span></span>
<span><span class="co">## RAM: 16.5 GB</span></span>
<span><span class="co">## CPU: 16 core(s) - 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="timing">Timing<a class="anchor" aria-label="anchor" href="#timing"></a>
</h4>
<p>The total time required to compile this vignette is:</p>
<pre><code><span><span class="co">## 4.933047 mins</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="memory">Memory<a class="anchor" aria-label="anchor" href="#memory"></a>
</h4>
<p>The final <code>leduc</code> object size is:</p>
<pre><code><span><span class="co">## [1] "0.5 Gb"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h4>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## R Under development (unstable) (2021-11-02 r81131)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] benchmarkme_1.0.8           ggpointdensity_0.1.0       </span></span>
<span><span class="co">##  [3] patchwork_1.1.1             reshape2_1.4.4             </span></span>
<span><span class="co">##  [5] ggrepel_0.9.1               ggbeeswarm_0.6.0           </span></span>
<span><span class="co">##  [7] forcats_0.5.1               stringr_1.4.0              </span></span>
<span><span class="co">##  [9] dplyr_1.0.9                 purrr_0.3.4                </span></span>
<span><span class="co">## [11] readr_2.1.2                 tidyr_1.2.0                </span></span>
<span><span class="co">## [13] tibble_3.1.8                ggplot2_3.3.6              </span></span>
<span><span class="co">## [15] tidyverse_1.3.2             sva_3.45.0                 </span></span>
<span><span class="co">## [17] BiocParallel_1.31.12        genefilter_1.79.0          </span></span>
<span><span class="co">## [19] mgcv_1.8-40                 nlme_3.1-158               </span></span>
<span><span class="co">## [21] SCP.replication_0.2.1       scpdata_1.5.0              </span></span>
<span><span class="co">## [23] ExperimentHub_2.5.0         AnnotationHub_3.5.0        </span></span>
<span><span class="co">## [25] BiocFileCache_2.5.0         dbplyr_2.2.1               </span></span>
<span><span class="co">## [27] scp_1.7.1                   QFeatures_1.7.1            </span></span>
<span><span class="co">## [29] MultiAssayExperiment_1.23.5 SingleCellExperiment_1.19.0</span></span>
<span><span class="co">## [31] SummarizedExperiment_1.27.1 Biobase_2.57.1             </span></span>
<span><span class="co">## [33] GenomicRanges_1.49.0        GenomeInfoDb_1.33.3        </span></span>
<span><span class="co">## [35] IRanges_2.31.0              S4Vectors_0.35.1           </span></span>
<span><span class="co">## [37] BiocGenerics_0.43.1         MatrixGenerics_1.9.1       </span></span>
<span><span class="co">## [39] matrixStats_0.62.0          BiocStyle_2.25.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] readxl_1.4.0                  backports_1.4.1              </span></span>
<span><span class="co">##   [3] systemfonts_1.0.4             plyr_1.8.7                   </span></span>
<span><span class="co">##   [5] igraph_1.3.4                  lazyeval_0.2.2               </span></span>
<span><span class="co">##   [7] splines_4.2.0                 digest_0.6.29                </span></span>
<span><span class="co">##   [9] foreach_1.5.2                 htmltools_0.5.3              </span></span>
<span><span class="co">##  [11] fansi_1.0.3                   magrittr_2.0.3               </span></span>
<span><span class="co">##  [13] memoise_2.0.1                 doParallel_1.0.17            </span></span>
<span><span class="co">##  [15] googlesheets4_1.0.0           cluster_2.1.3                </span></span>
<span><span class="co">##  [17] tzdb_0.3.0                    limma_3.53.5                 </span></span>
<span><span class="co">##  [19] Biostrings_2.65.1             annotate_1.75.0              </span></span>
<span><span class="co">##  [21] modelr_0.1.8                  pkgdown_2.0.6                </span></span>
<span><span class="co">##  [23] colorspace_2.0-3              rvest_1.0.2                  </span></span>
<span><span class="co">##  [25] blob_1.2.3                    rappdirs_0.3.3               </span></span>
<span><span class="co">##  [27] textshaping_0.3.6             haven_2.5.0                  </span></span>
<span><span class="co">##  [29] xfun_0.31                     crayon_1.5.1                 </span></span>
<span><span class="co">##  [31] RCurl_1.98-1.8                jsonlite_1.8.0               </span></span>
<span><span class="co">##  [33] iterators_1.0.14              survival_3.3-1               </span></span>
<span><span class="co">##  [35] glue_1.6.2                    gtable_0.3.0                 </span></span>
<span><span class="co">##  [37] gargle_1.2.0                  zlibbioc_1.43.0              </span></span>
<span><span class="co">##  [39] XVector_0.37.0                OrgMassSpecR_0.5-3           </span></span>
<span><span class="co">##  [41] DelayedArray_0.23.1           scales_1.2.0                 </span></span>
<span><span class="co">##  [43] DBI_1.1.3                     edgeR_3.39.4                 </span></span>
<span><span class="co">##  [45] Rcpp_1.0.9                    viridisLite_0.4.0            </span></span>
<span><span class="co">##  [47] xtable_1.8-4                  clue_0.3-61                  </span></span>
<span><span class="co">##  [49] reticulate_1.25               bit_4.0.4                    </span></span>
<span><span class="co">##  [51] MsCoreUtils_1.9.1             httr_1.4.3                   </span></span>
<span><span class="co">##  [53] ellipsis_0.3.2                farver_2.1.1                 </span></span>
<span><span class="co">##  [55] pkgconfig_2.0.3               XML_3.99-0.10                </span></span>
<span><span class="co">##  [57] sass_0.4.2                    locfit_1.5-9.6               </span></span>
<span><span class="co">##  [59] utf8_1.2.2                    labeling_0.4.2               </span></span>
<span><span class="co">##  [61] tidyselect_1.1.2              rlang_1.0.4                  </span></span>
<span><span class="co">##  [63] later_1.3.0                   AnnotationDbi_1.59.1         </span></span>
<span><span class="co">##  [65] munsell_0.5.0                 BiocVersion_3.16.0           </span></span>
<span><span class="co">##  [67] cellranger_1.1.0              tools_4.2.0                  </span></span>
<span><span class="co">##  [69] cachem_1.0.6                  cli_3.3.0                    </span></span>
<span><span class="co">##  [71] generics_0.1.3                RSQLite_2.2.15               </span></span>
<span><span class="co">##  [73] broom_1.0.0                   evaluate_0.15                </span></span>
<span><span class="co">##  [75] fastmap_1.1.0                 yaml_2.3.5                   </span></span>
<span><span class="co">##  [77] ragg_1.2.2                    knitr_1.39                   </span></span>
<span><span class="co">##  [79] bit64_4.0.5                   fs_1.5.2                     </span></span>
<span><span class="co">##  [81] KEGGREST_1.37.3               AnnotationFilter_1.21.0      </span></span>
<span><span class="co">##  [83] mime_0.12                     xml2_1.3.3                   </span></span>
<span><span class="co">##  [85] compiler_4.2.0                rstudioapi_0.13              </span></span>
<span><span class="co">##  [87] beeswarm_0.4.0                filelock_1.0.2               </span></span>
<span><span class="co">##  [89] curl_4.3.2                    png_0.1-7                    </span></span>
<span><span class="co">##  [91] interactiveDisplayBase_1.35.0 reprex_2.0.1                 </span></span>
<span><span class="co">##  [93] bslib_0.4.0                   stringi_1.7.8                </span></span>
<span><span class="co">##  [95] highr_0.9                     desc_1.4.1                   </span></span>
<span><span class="co">##  [97] lattice_0.20-45               ProtGenerics_1.29.0          </span></span>
<span><span class="co">##  [99] Matrix_1.4-1                  vctrs_0.4.1                  </span></span>
<span><span class="co">## [101] pillar_1.8.0                  lifecycle_1.0.1              </span></span>
<span><span class="co">## [103] BiocManager_1.30.18           jquerylib_0.1.4              </span></span>
<span><span class="co">## [105] bitops_1.0-7                  httpuv_1.6.5                 </span></span>
<span><span class="co">## [107] R6_2.5.1                      bookdown_0.27                </span></span>
<span><span class="co">## [109] promises_1.2.0.1              vipor_0.4.5                  </span></span>
<span><span class="co">## [111] codetools_0.2-18              benchmarkmeData_1.0.4        </span></span>
<span><span class="co">## [113] MASS_7.3-58.1                 assertthat_0.2.1             </span></span>
<span><span class="co">## [115] rprojroot_2.0.3               withr_2.5.0                  </span></span>
<span><span class="co">## [117] GenomeInfoDbData_1.2.8        parallel_4.2.0               </span></span>
<span><span class="co">## [119] hms_1.1.1                     grid_4.2.0                   </span></span>
<span><span class="co">## [121] rmarkdown_2.14                googledrive_2.0.0            </span></span>
<span><span class="co">## [123] lubridate_1.8.0               shiny_1.7.2</span></span></code></pre></div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:christophe.vanderaa@uclouvain.be" class="email">christophe.vanderaa@uclouvain.be</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
