<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Replication of the nPOP analysis (Leduc et al. 2022) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Replication of the nPOP analysis (Leduc et al. 2022)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/brunner2022.html">Replication of the cell cycle state study (Brunner et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/derks2022.html">Replication of the plexDIA analysis (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/leduc2022.html">Replication of the nPOP analysis (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Replication of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Replication of the nPOP analysis (Leduc et
al. 2022)</h1>
                        <h4 data-toc-skip class="author">Christophe
Vanderaa<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, Computational Biology, UCLouvain</h4>
                        <h4 data-toc-skip class="author">Laurent Gatto,
Computational Biology, UCLouvain</h4>
            
            <h4 data-toc-skip class="date">5 October 2022</h4>
      
      
      <div class="hidden name"><code>leduc2022.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>nPOP (<a href="http://dx.doi.org/10.1101/2021.04.24.441211" class="external-link">Leduc et
al. 2022</a>) is an upgrade of the SCoPE2 protocole (<a href="http://dx.doi.org/10.1186/s13059-021-02267-5" class="external-link">Specht et
al. 2021</a> and <a href="http://dx.doi.org/10.1038/s41596-021-00616-z" class="external-link">Petelski et
al. 2021</a>), where the mPOP sample preparation method is replaced by
the nPOP method. nPOP processes samples using the Cellenion dispensing
device and uses DMSO as lysis reagent instead of a freeze-thaw
procedure. They also include the prioritized data acquisition mode as
described by <a href="http://dx.doi.org/10.1101/2022.03.16.484655" class="external-link">Huffman et
al. 2022</a>.</p>
<p>Let’s first load the replication package to make use of some helper
functions. Those functions are only meant for this replication vignette
and are not designed for general use.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SCP.replication"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="scp-and-the-scope2-workflow">
<code>scp</code> and the SCoPE2 workflow<a class="anchor" aria-label="anchor" href="#scp-and-the-scope2-workflow"></a>
</h4>
<p>The code provided along with the article can be retrieved from <a href="https://github.com/SlavovLab/nPOP" class="external-link">this GitHub repository</a>. The
objective of this vignette is to replicate the analysis script while
providing standardized, easy-to-read, and well documented code.
Therefore, our first contribution is to formalize the data processing
into a conceptual flow chart.</p>
<div class="figure" style="text-align: center">
<img src="figs/leduc2022-workflow.png" alt="Overview of the processing workflow by Leduc et al." width="100%"><p class="caption">
Overview of the processing workflow by Leduc et al.
</p>
</div>
<p>This replication vignette relies on a data framework dedicated to SCP
data analysis that combines two Bioconductor classes (<a href="http://dx.doi.org/10.1080/14789450.2021.1988571" class="external-link">Vanderaa et
al. 2021</a>):</p>
<ul>
<li>The <code>SingleCellExperiment</code> class provides an interface to
many cutting edge methods for single-cell analysis</li>
<li>The <code>QFeatures</code> class facilitates manipulation and
processing of MS-based quantitative data.</li>
</ul>
<p>The <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/scp/inst/doc/scp.html" class="external-link"><code>scp</code>
vignette</a> provides detailed information about the data structure. The
<code>scp</code> package extends the functionality of
<code>QFeatures</code> for single-cell application. <code>scp</code>
offers a standardized implementation for single-cell processing
methods.</p>
<p>The required packages for running this workflow are listed below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Core packages of this workflow</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">scp</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scpdata</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="co">## Utility packages for data manipulation and visualization</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="scpdata-and-the-leduc2022-dataset">
<code>scpdata</code> and the <code>leduc2022</code> dataset<a class="anchor" aria-label="anchor" href="#scpdata-and-the-leduc2022-dataset"></a>
</h4>
<p>We also implemented a data package called <code>scpdata</code> (<span class="citation">@Vanderaa2022-qv</span>). It distributes published SCP
datasets, such as the <code>leduc2022</code> dataset. The datasets were
downloaded from the data source provided in the publication and
formatted to a <code>QFeatures</code> object so that it is compatible
with our software. The underlying data storage is based on the
<code>ExperimentHub</code> package that provides a cloud-based storage
infrastructure.</p>
<p>The <code>leduc2022</code> dataset is provided at different levels of
processing:</p>
<ul>
<li>The <strong>raw data</strong> files that were generated by the
mass-spectrometer software. This data is not included in
<code>scpdata</code>.</li>
<li>A <strong>PSM data</strong> table obtained from the MaxQuant
software that performs spectrum identification and quantification. PSM
stands for <em>peptide to spectrum match</em> where MS spectra could
successfully be assigned to peptide sequences. The provided PSM table
was further processed using DART-ID (@Chen2019-uc) to improve the
identification rate.</li>
<li>
<strong>Peptide data</strong> are provided as intermediate output
tables from the data processing. There are 2 tables: peptides normalized
by reference and peptides further normalized (median centering of
columns and rows) and log-transformed.</li>
<li>A <strong>protein data</strong> are provided as intermediate and
final output tables from the data processing. There are 2 tables:
proteins column and row centered to median and log2 transformed and
proteins after imputation, batch correction and final column and row
centering.</li>
</ul>
<p>The workflow starts with the PSM table and will generate the peptide
and the protein data. The authors provided the PSM dataset as a tabular
text file called <code>ev_updated.txt</code>. Peptide and protein data
are shared as CSV files. We highly value the effort the authors have
made to publicly share all the data generated in their project, from raw
files to final expression tables (see the Slavov Lab <a href="https://scp.slavovlab.net/Leduc_et_al_2022" class="external-link">website</a>).</p>
<p>We formatted the <code>leduc2022</code> dataset following our data
framework. The formatted data can be retrieved from the
<code>scpdata</code> package using the <code><a href="https://rdrr.io/pkg/scpdata/man/leduc2022.html" class="external-link">leduc2022()</a></code>
function. All datasets in <code>scpdata</code> are called after the
first author and the date of publication.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/leduc2022.html" class="external-link">leduc2022</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The data contain 138 different <code>SingleCellExperiment</code>
objects that we refer to as <strong>assays</strong>. Each assay contains
expression data along with feature metadata. Each row in an assay
represents a <strong>feature</strong> that can either be a PSM, a
peptide or a protein depending on the assay. Each column in an assay
represents a <strong>sample</strong>. In the <code>leduc</code> object,
samples are pooled using TMT-pro18 labeling, hence each assay contains
18 columns. Most samples are single-cells, but some samples are negative
controls, references, carriers,… Below, we show the overview of the
<code>leduc</code> object</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span></span>
<span><span class="co">## An instance of class QFeatures containing 138 assays:</span></span>
<span><span class="co">##  [1] eAL00219: SingleCellExperiment with 6269 rows and 18 columns </span></span>
<span><span class="co">##  [2] eAL00220: SingleCellExperiment with 6603 rows and 18 columns </span></span>
<span><span class="co">##  [3] eAL00221: SingleCellExperiment with 6511 rows and 18 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [136] peptides_log: SingleCellExperiment with 12284 rows and 1543 columns </span></span>
<span><span class="co">##  [137] proteins_norm2: SingleCellExperiment with 2844 rows and 1543 columns </span></span>
<span><span class="co">##  [138] proteins_processed: SingleCellExperiment with 2844 rows and 1543 columns</span></span></code></pre></div>
<p>134 out of the 138 assays are PSM data, each assay corresponding to a
separate MS run. Notice that the assays were acquired in 2 sample
preparation and chromatographic batches.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>LcBatch <span class="op">=</span> <span class="va">leduc</span><span class="op">$</span><span class="va">lcbatch</span>,</span>
<span>      SamplePrepBatch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"AL.*"</span>, <span class="st">""</span>, <span class="va">leduc</span><span class="op">$</span><span class="va">Set</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##        SamplePrepBatch</span></span>
<span><span class="co">## LcBatch    e    w</span></span>
<span><span class="co">##       A    0 1548</span></span>
<span><span class="co">##       C  864    0</span></span></code></pre></div>
<p>The dataset also contains a <code>peptides</code>,
<code>peptides_log</code>, <code>proteins_norm</code> and
<code>proteins_processed</code> assay. Those were provided by the
authors. The objective of this vignette is to replicate these assays
from the 134 PSM assays following the same procedure as the original
script but using standardized functionality.</p>
<p>We extract these latter assays and keep them for later benchmarking.
Using double brackets <code>[[...]]</code> extracts the desired assay as
a <code>SingleCellExperiment</code> object. On the other hand, using
simple brackets <code>[row, col, assay]</code> subsets the desired
elements/assays but preserves the <code>QFeatures</code> data
structure.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peptides_leduc</span> <span class="op">&lt;-</span> <span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">peptides_log_leduc</span> <span class="op">&lt;-</span> <span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides_log"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">proteins_norm_leduc</span> <span class="op">&lt;-</span> <span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_norm2"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">proteins_processed_leduc</span> <span class="op">&lt;-</span> <span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_processed"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="va">leduc</span><span class="op">[</span>, , <span class="op">-</span><span class="op">(</span><span class="fl">135</span><span class="op">:</span><span class="fl">138</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span></code></pre></div>
<p>We will compare the replications by comparing the set of filtered
features (peptides or proteins) and samples. This is performed using
this function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compareSets</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">setleduc</span>, <span class="va">setscp</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">allElements</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">setleduc</span>, <span class="va">setscp</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>leduc2022 <span class="op">=</span> <span class="va">allElements</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">setleduc</span>,</span>
<span>          scp <span class="op">=</span> <span class="va">allElements</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">setscp</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We will also compare the replication based on the quantitative data.
We again create a dedicated function to perform this.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compareQuantitativeData</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sceleduc</span>, <span class="va">scescp</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sceleduc</span><span class="op">)</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scescp</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sceleduc</span><span class="op">)</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scescp</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">err</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sceleduc</span><span class="op">)</span><span class="op">[</span><span class="va">rows</span>, <span class="va">cols</span><span class="op">]</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">scescp</span><span class="op">)</span><span class="op">[</span><span class="va">rows</span>, <span class="va">cols</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>difference <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">err</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">err</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">difference</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"nPOP - scp"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/scientific_format.html" class="external-link">scientific</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="psm-filtering">PSM filtering<a class="anchor" aria-label="anchor" href="#psm-filtering"></a>
</h2>
<p>After importing the data, Leduc et al. filter low-confidence PSMs.
Each PSM assay contains feature meta-information that are stored in the
assay <code>rowData</code>. The <code>QFeatures</code> package allows to
quickly filter the rows of an assay by using these information. The
available variables in the <code>rowData</code> are listed below for
each assay.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rowDataNames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span></span>
<span><span class="co">## CharacterList of length 134</span></span>
<span><span class="co">## [["eAL00219"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00220"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00221"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00222"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00223"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00224"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00225"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00226"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00227"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## [["eAL00228"]] Sequence Length ... Leading.razor.protein.symbol</span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## &lt;124 more elements&gt;</span></span></code></pre></div>
<div class="section level3">
<h3 id="remove-contaminant-noisy-and-low-confidence-spectra">Remove contaminant, noisy and low-confidence spectra<a class="anchor" aria-label="anchor" href="#remove-contaminant-noisy-and-low-confidence-spectra"></a>
</h3>
<p>We first remove spectra that are matched to contaminant proteins and
reverse hits. We also remove PSMs that have been matched from impure
spectra, that are spectra containing co-eluting peptides. These are
identified based on the parental ion fraction (PIF), computed by
MaxQuant. Finally, we also want to remove PSM with poor matching
confidence, as defined by the false discovery rate (FDR) computed by
DART-ID.</p>
<p>We can extract the information from the <code>rowData</code> of
several assays using the <code>rbindRowData</code> function. It takes
the <code>rowData</code> of interest and returns a single
<code>DataFrame</code> table with variables of interest. We extract such
a table for the different variables listed above to create a quality
control plot.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">rd</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dart_PEP</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">rd</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PIF</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span><span class="co">## Warning: Removed 181437 rows containing non-finite values (stat_bin).</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/qvalue_plot-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>We next remove the PSMs that are matched to potential contaminants
(<code>Potential.contaminant</code> is <code>+</code> and
<code>Proteins</code> starts with <code>CON</code>), reverse hits
(<code>Reverse</code> is <code>+</code> and
<code>Leading.razor.protein</code> starts with <code>REV</code>), noisy
spectra (<code>PIF</code> is missing or greater than 0.6) and
low-confidence spectra with at 1% FDR threshold (<code>dart_qval</code>
smaller than 0.01). We can perform this on our <code>QFeatures</code>
object using the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures()</a></code> function. The different
pieces of information are directly accessed from the
<code>rowData</code> of each assay.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">leduc</span>, <span class="op">~</span> <span class="va">Potential.contaminant</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span></span>
<span>                            <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"CON"</span>, <span class="va">Proteins</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                            <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span></span>
<span>                            <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"REV"</span>, <span class="va">Leading.razor.protein</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                            <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">PIF</span><span class="op">)</span> <span class="op">|</span> <span class="va">PIF</span> <span class="op">&gt;</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                            <span class="va">dart_qval</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sample-to-carrier-filter">Sample to carrier filter<a class="anchor" aria-label="anchor" href="#sample-to-carrier-filter"></a>
</h3>
<p>The PSMs are next filtered based on the sample to carrier ratio
(SCR), that is the TMT ion intensity of a single-cell sample divided by
the TMT ion intensity of the carrier (200 cell equivalent) acquired
during the same run as the sample. It is expected that the carrier
intensities are much higher than the single-cell intensities. We
implemented the <code><a href="https://UCLouvain-CBIO.github.io/scp/reference/computeSCR.html" class="external-link">computeSCR()</a></code> function that computes the SCR
for each PSM averaged over all samples of interest in a given assay. A
PSM is removed when the mean SCR exceeds 10 %. To perform this, we need
to tell the function which columns are the samples of interest and which
column is the carrier. The <code>colData</code> of the
<code>QFeatures</code> object is used to define this.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##       Carrier Melanoma cell      Monocyte    NegControl     Reference </span></span>
<span><span class="co">##           134           878           877           120           134 </span></span>
<span><span class="co">##        Unused </span></span>
<span><span class="co">##           269</span></span></code></pre></div>
<p>In this dataset, <code>SampleType</code> gives the type of sample
that is present in each TMT channel. There 5 types of samples:</p>
<ul>
<li>The carrier channels (<code>Carrier</code>) contain 200 cell
equivalents and are meant to boost the peptide identification rate.</li>
<li>The normalization channels (<code>Reference</code>) are used to
partially correct for between-run variation.</li>
<li>The unused channels (<code>Unused</code>) are channels that are left
empty due to isotopic cross-contamination.</li>
<li>The negative controls (<code>NegControl</code>) contain samples that
do not contain any cell but are processed as single-cell samples.</li>
<li>The single-cell sample channels contain the single-cell samples of
interest (<code>Melanoma cell</code> or <code>Monocyte</code>).</li>
</ul>
<p>The <code>computeSCR</code> function expects the user to provide a
pattern (following regular expression syntax) that uniquely identifies a
carrier channel in each run and the samples or blanks. The function will
store the mean SCR of each feature in the <code>rowData</code> of each
assay.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/computeSCR.html" class="external-link">computeSCR</a></span><span class="op">(</span><span class="va">leduc</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>,</span>
<span>                    colvar <span class="op">=</span> <span class="st">"SampleType"</span>, </span>
<span>                    samplePattern <span class="op">=</span> <span class="st">"Mel|Macro"</span>,</span>
<span>                    carrierPattern <span class="op">=</span> <span class="st">"Carrier"</span>,</span>
<span>                    sampleFUN <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>                    rowDataName <span class="op">=</span> <span class="st">"MeanSCR"</span><span class="op">)</span></span></code></pre></div>
<p>Before applying the filter, we plot the distribution of the mean
SCR.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">MeanSCR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-9-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>A great majority of the PSMs have a mean SCR that is lower than 10%,
as expected. Since the mean SCR is stored in the <code>rowData</code>,
we can apply <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures()</a></code> on the object to remove PSMs
with high average SCR.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">leduc</span>, <span class="op">~</span> </span>
<span>                            <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">MeanSCR</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html" class="external-link">is.infinite</a></span><span class="op">(</span><span class="va">MeanSCR</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                            <span class="va">MeanSCR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-on-summed-single-cell-signal">Filter on summed single-cell signal<a class="anchor" aria-label="anchor" href="#filter-on-summed-single-cell-signal"></a>
</h3>
<p>Finally, we remove PSM that have no signal in single-cell samples.
This is not explicitely implemented in <code>scp</code>. To add custom
information to <code>rowData</code>, you need to provide a list of
<code>DataFrame</code>s. The name of the elements in the list should
correspond to the names of the assays where the <code>rowData</code> is
modified. The column names of the <code>DataFrame</code> indicate which
variable should be modified or added (if they do not exist yet). So, for
each assay, we compute the summed signal in single-cells (and negative
controls) and store the results in a <code>DataFrame</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sums</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">leduc</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Mel|Macro|Neg"</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="st">"SampleType"</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span>, <span class="va">sel</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span>    <span class="va">rs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="fu">DataFrame</span><span class="op">(</span>ScSums <span class="op">=</span> <span class="va">rs</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>The list of <code>DataFrame</code> is named after the corresponding
assays and the <code>rowData</code> of the <code>leduc</code> object is
modified.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sums</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sums</span></span></code></pre></div>
<p>To verify this new piece of data was correctly added, we plot the
summed signal for each PSM.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ScSums</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-13-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>We apply the final filter using <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures()</a></code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">leduc</span>, <span class="op">~</span> <span class="va">ScSums</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="normalize-to-reference">Normalize to reference<a class="anchor" aria-label="anchor" href="#normalize-to-reference"></a>
</h2>
<p>In order to partially correct for between-run variation, Leduc et al.
compute relative reporter ion intensities. This means that intensities
measured for single-cells are divided by the reference channel. We use
the <code><a href="https://UCLouvain-CBIO.github.io/scp/reference/divideByReference.html" class="external-link">divideByReference()</a></code> function that divides channels of
interest by the reference channel. Similarly to <code>computeSCR</code>,
we can point to the samples and the reference columns in each assay
using the annotation contained in the <code>colData</code>. We will here
divide all columns (using the regular expression wildcard
<code>.</code>) by the reference channel (<code>Reference</code>).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/divideByReference.html" class="external-link">divideByReference</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>,</span>
<span>                           colvar <span class="op">=</span> <span class="st">"SampleType"</span>, </span>
<span>                           samplePattern <span class="op">=</span> <span class="st">"."</span>,</span>
<span>                           refPattern <span class="op">=</span> <span class="st">"Reference"</span><span class="op">)</span></span></code></pre></div>
<p>Notice that when taking all samples we also include the reference
channel itself. Hence, from now on, the reference channels will contain
only ones.</p>
</div>
<div class="section level2">
<h2 id="aggregate-psm-data-to-peptide-data">Aggregate PSM data to peptide data<a class="anchor" aria-label="anchor" href="#aggregate-psm-data-to-peptide-data"></a>
</h2>
<p>Now that the PSM assays are processed, we can aggregate them to
peptides. This is performed using the
<code><a href="https://UCLouvain-CBIO.github.io/scp/reference/aggregateFeaturesOverAssays.html" class="external-link">aggregateFeaturesOverAssays()</a></code> function. This is a wrapper
function in <code>scp</code> that sequentially calls the
<code>aggregateFeatures</code> from the <code>QFeatures</code> package
over the different assays. For each assay, the function aggregates
several PSMs into a unique peptide given an aggregating variable in the
<code>rowData</code> (peptide sequence) and a user-supplied aggregating
function (the median for instance). Regarding the aggregating function,
the original analysis removes duplicated peptide sequences per run by
taking the first non-missing value. While better alternatives are
documented in <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">QFeatures::aggregateFeatures</a></code>, we still use
this approach for the sake of replication and for illustrating that
custom functions can be applied.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">remove.duplicates</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">xx</span><span class="op">)</span> <span class="va">xx</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">xx</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<p>The aggregated peptide assays must be given a name. We here used the
original names with <code>peptides_</code> at the start.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peptideAssays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"peptides_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We now have all the required information to aggregate the PSMs in the
different batches to peptides.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/aggregateFeaturesOverAssays.html" class="external-link">aggregateFeaturesOverAssays</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                                     i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>,</span>
<span>                                     fcol <span class="op">=</span> <span class="st">"modseq"</span>,</span>
<span>                                     name <span class="op">=</span> <span class="va">peptideAssays</span>,</span>
<span>                                     fun <span class="op">=</span> <span class="va">remove.duplicates</span><span class="op">)</span></span></code></pre></div>
<p>Under the hood, the <code>QFeatures</code> architecture preserves the
relationship between the aggregated assays. See <code><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">?AssayLinks</a></code>
for more information on relationships between assays. Notice that
<code>aggregateFeaturesOverAssays</code> created as many new assays as
the number of supplied assays.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span></span>
<span><span class="co">## An instance of class QFeatures containing 268 assays:</span></span>
<span><span class="co">##  [1] eAL00219: SingleCellExperiment with 3555 rows and 18 columns </span></span>
<span><span class="co">##  [2] eAL00220: SingleCellExperiment with 3981 rows and 18 columns </span></span>
<span><span class="co">##  [3] eAL00221: SingleCellExperiment with 3785 rows and 18 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [266] peptides_wAL00284: SingleCellExperiment with 3211 rows and 18 columns </span></span>
<span><span class="co">##  [267] peptides_wAL00285: SingleCellExperiment with 3277 rows and 18 columns </span></span>
<span><span class="co">##  [268] peptides_wAL00286: SingleCellExperiment with 3411 rows and 18 columns</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="join-assays">Join assays<a class="anchor" aria-label="anchor" href="#join-assays"></a>
</h2>
<p>Up to now, we kept the data belonging to each MS run in separate
assays. We now combine all batches into a single assay. This can easily
be done using the <code><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays()</a></code> function from the
<code>QFeatures</code> package.</p>
<div class="section level3">
<h3 id="consensus-mapping-of-peptides-to-proteins">Consensus mapping of peptides to proteins<a class="anchor" aria-label="anchor" href="#consensus-mapping-of-peptides-to-proteins"></a>
</h3>
<p>We need to account for an issue in the data.
<code><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays()</a></code> will only keep the metadata variables that
have the same value between matching rows. However, some peptide
sequences map to one protein in one run and to another protein in
another run. Hence, the protein sequence is not constant for all
peptides and is removed during joining. It is important we keep the
protein sequence in the <code>rowData</code> since we will later need it
to aggregate peptides to proteins. To avoid this issue, we replace the
problematic peptides to protein mappings through a majority vote.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Generate a list of DataFrames with the information to modify</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^pep"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">modseq</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">## The majority vote happens here</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Leading.razor.protein.symbol <span class="op">=</span></span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">Leading.razor.protein</span><span class="op">)</span>,</span>
<span>                          decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">modseq</span>, <span class="va">Leading.razor.protein.symbol</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">modseq</span>, <span class="va">Leading.razor.protein.symbol</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span></span>
<span>    <span class="va">ppMap</span></span>
<span><span class="va">consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">peptideAssays</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">modseq</span>, <span class="va">ppMap</span><span class="op">$</span><span class="va">modseq</span><span class="op">)</span></span>
<span>    <span class="fu">DataFrame</span><span class="op">(</span>Leading.razor.protein.symbol <span class="op">=</span></span>
<span>                  <span class="va">ppMap</span><span class="op">$</span><span class="va">Leading.razor.protein.symbol</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">## Name the list</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">consensus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">peptideAssays</span></span>
<span><span class="co">## Modify the rowData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">consensus</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cleaning-missing-data">Cleaning missing data<a class="anchor" aria-label="anchor" href="#cleaning-missing-data"></a>
</h3>
<p>Another important step before we join the assays is to replace zero
and infinite values by <code>NA</code>s. The zeros can be biological
zeros or technical zeros and differentiating between the two types is a
difficult task, they are therefore better considered as missing. The
infinite values arose during the normalization by the reference because
the channel values are divide by a zero from the reference channel. This
artefact could easily be avoided if we had replace the zeros by
<code>NA</code>s at the beginning of the workflow, what we strongly
recommend for future analyses.</p>
<p>The <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">infIsNA()</a></code> and the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA()</a></code> functions
automatically detect infinite and zero values, respectively, and replace
them with <code>NA</code>s. Those two functions are provided by the
<code>QFeatures</code> package.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">infIsNA</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="va">peptideAssays</span><span class="op">)</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="va">peptideAssays</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="join-assays-1">Join assays<a class="anchor" aria-label="anchor" href="#join-assays-1"></a>
</h3>
<p>Now that the peptides are correctly matched to proteins and missing
values are correctly formatted, we can join the assays.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                    i <span class="op">=</span> <span class="va">peptideAssays</span>,</span>
<span>                    name <span class="op">=</span> <span class="st">"peptides"</span><span class="op">)</span></span></code></pre></div>
<p><code>joinAssays</code> has created a new assay called
<code>peptides</code> that combines the previously aggregated peptide
assays.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span></span>
<span><span class="co">## An instance of class QFeatures containing 269 assays:</span></span>
<span><span class="co">##  [1] eAL00219: SingleCellExperiment with 3555 rows and 18 columns </span></span>
<span><span class="co">##  [2] eAL00220: SingleCellExperiment with 3981 rows and 18 columns </span></span>
<span><span class="co">##  [3] eAL00221: SingleCellExperiment with 3785 rows and 18 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [267] peptides_wAL00285: SingleCellExperiment with 3277 rows and 18 columns </span></span>
<span><span class="co">##  [268] peptides_wAL00286: SingleCellExperiment with 3411 rows and 18 columns </span></span>
<span><span class="co">##  [269] peptides: SingleCellExperiment with 20480 rows and 2412 columns</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="filter-single-cells-based-on-median-cv">Filter single-cells based on median CV<a class="anchor" aria-label="anchor" href="#filter-single-cells-based-on-median-cv"></a>
</h2>
<p>Leduc et al. proceed with filtering the single-cells. The filtering
is mainly based on the median coefficient of variation (CV) per cell.
The median CV measures the consistency of quantification for a group of
peptides that belong to a protein. We remove cells that exhibit high
median CV over the different proteins. We compute the median CV per cell
using the <code><a href="https://UCLouvain-CBIO.github.io/scp/reference/medianCVperCell.html" class="external-link">medianCVperCell()</a></code> function from the
<code>scp</code> package. The function takes the protein information
from the <code>rowData</code> of the assays that will tell how to group
the features (peptides) when computing the CV. Note that we supply the
peptide assays before joining in a single assays
(<code>i = peptideAssays</code>). This is because SCoPE2 performs a
custom normalization (<code>norm = "SCoPE2"</code>). Each row in an
assay is normalized by a scaling factor. This scaling factor is the row
mean after dividing the columns by the median. The authors retained CVs
that are computed using at least 3 peptides (<code>nobs = 3</code>).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/medianCVperCell.html" class="external-link">medianCVperCell</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                         i <span class="op">=</span> <span class="va">peptideAssays</span>,</span>
<span>                         groupBy <span class="op">=</span> <span class="st">"Leading.razor.protein.symbol"</span>,</span>
<span>                         nobs <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                         na.rm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         colDataName <span class="op">=</span> <span class="st">"MedianCV"</span>,</span>
<span>                         norm <span class="op">=</span> <span class="st">"SCoPE2"</span><span class="op">)</span></span>
<span><span class="co">## Warning in medianCVperCell(leduc, i = peptideAssays, groupBy = "Leading.razor.protein.symbol", : The median CV could not be computed for one or more samples. You may want to try a smaller value for 'nobs'.</span></span></code></pre></div>
<p>The computed CVs are stored in the <code>colData</code>. We can now
filter cells that have reliable quantifications. The negative controls
are not expected to have reliable quantifications and hence can be used
to estimate a null distribution of the CV. This distribution helps
defining a threshold that filters out single-cells that contain noisy
quantification.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Mono|Mel|Neg"</span>, <span class="va">SampleType</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Neg"</span>, <span class="va">SampleType</span><span class="op">)</span>, <span class="st">"ctl"</span>, <span class="st">"sc"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">MedianCV</span>,</span>
<span>        fill <span class="op">=</span> <span class="va">control</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span>, adjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.42</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"black"</span>, <span class="st">"purple2"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Quantification variability"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Fraction of cells"</span><span class="op">)</span></span>
<span><span class="co">## Warning: Removed 31 rows containing non-finite values (stat_density).</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-21-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>We can see that the protein quantification for single-cells are much
more consistent within single-cell channels than within blank channels.
A threshold of 0.42 best separates single-cells from empty channels.</p>
<p>We keep the cells that pass the median CV threshold. Furthermore, we
keep melanoma cells and monocytes as those represent the samples of
interest. We can extract the sample names that pass the CV and sample
type filters using the <code>subsetByColData()</code> function.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                    <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">MedianCV</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                        <span class="va">leduc</span><span class="op">$</span><span class="va">MedianCV</span> <span class="op">&lt;</span> <span class="fl">0.42</span> <span class="op">&amp;</span></span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Mono|Mel"</span>, <span class="va">leduc</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="compare-intermediate-results">Compare intermediate results<a class="anchor" aria-label="anchor" href="#compare-intermediate-results"></a>
</h2>
<p>At this stage of the processing, the last assay should be similar to
the <code>peptides_leduc</code> data provided by the authors. Let’s
compare the filtered cells.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareSets</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">peptides_leduc</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## leduc2022 FALSE TRUE</span></span>
<span><span class="co">##     FALSE     0   27</span></span>
<span><span class="co">##     TRUE      7 1549</span></span></code></pre></div>
<p>There is an excellent agreement between the the original and the
replicated vignette. Let’s do the same for the filtered peptides.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareSets</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">peptides_leduc</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## leduc2022 FALSE  TRUE</span></span>
<span><span class="co">##     FALSE     0    27</span></span>
<span><span class="co">##     TRUE    351 20453</span></span></code></pre></div>
<p>Finally let’s compare the quantitative data</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareQuantitativeData</span><span class="op">(</span><span class="va">peptides_leduc</span>, <span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-25-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The replication is close to perfect. Note however that this vignette
is more stringent with respect to the number of selected peptides. We
cannot explain this difference.</p>
</div>
<div class="section level2">
<h2 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h2>
<p>The columns (samples) then the rows (peptides) are normalized by
dividing the relative intensities by the median relative intensities.
The column normalization is implemented as the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize()</a></code>
function with the argument <code>method = div.median</code>. The row
normalization is not available from <code>normalizet()</code>, but is
easily performed using the <code>sweep</code> function from the
<code>QFeatures</code> package that is inspired from the
<code><a href="https://rdrr.io/r/base/sweep.html" class="external-link">base::sweep</a></code> function.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Scale column with median</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                   i <span class="op">=</span> <span class="st">"peptides"</span>,</span>
<span>                   method <span class="op">=</span> <span class="st">"div.median"</span>,</span>
<span>                   name <span class="op">=</span> <span class="st">"peptides_norm1"</span><span class="op">)</span></span>
<span><span class="co">## Scale rows with median</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>               i <span class="op">=</span> <span class="st">"peptides_norm1"</span>,</span>
<span>               name <span class="op">=</span> <span class="st">"peptides_norm2"</span>,</span>
<span>               MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>               FUN <span class="op">=</span> <span class="st">"/"</span>,</span>
<span>               STATS <span class="op">=</span> <span class="fu">rowMedians</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                                  na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Each normalization step is stored in a separate assay. An important
aspect to note here is that</p>
</div>
<div class="section level2">
<h2 id="missing-data-filtering">Missing data filtering<a class="anchor" aria-label="anchor" href="#missing-data-filtering"></a>
</h2>
<p>Peptides that contain many missing values are not informative.
Therefore, the authors remove those with more than 99 % missing data.
This is done using the <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA()</a></code> function from
<code>QFeatures</code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                  i <span class="op">=</span> <span class="st">"peptides_norm2"</span>,</span>
<span>                  pNA <span class="op">=</span> <span class="fl">0.99</span><span class="op">)</span></span></code></pre></div>
<p>They also remove cells with more than 99 % missing data. This is
performed by first computing the amount of missing data in the assay
using <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">nNA()</a></code>. We then subset the cells that meet the
criterion.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nnaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">nNA</a></span><span class="op">(</span><span class="va">leduc</span>, <span class="st">"peptides_norm2"</span><span class="op">)</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="va">nnaRes</span><span class="op">$</span><span class="va">nNAcols</span><span class="op">$</span><span class="va">pNA</span> <span class="op">&lt;</span> <span class="fl">99</span></span>
<span><span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides_norm2"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides_norm2"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="va">sel</span><span class="op">]</span></span>
<span><span class="co">## Warning in replaceAssay(x = x, y = value, i = i): Links between assays were</span></span>
<span><span class="co">## lost/removed during replacement. See '?addAssayLink' to restore them manually.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="log-transformation">Log-transformation<a class="anchor" aria-label="anchor" href="#log-transformation"></a>
</h2>
<p>Peptide data is log2-transformed before aggregating to proteins. This
is performed by the <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform()</a></code> function from
<code>QFeatures</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                      base <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                      i <span class="op">=</span> <span class="st">"peptides_norm2"</span>,</span>
<span>                      name <span class="op">=</span> <span class="st">"peptides_log"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="compare-intermediate-results-1">Compare intermediate results<a class="anchor" aria-label="anchor" href="#compare-intermediate-results-1"></a>
</h2>
<p>At this stage of the processing, the last assay should be similar to
the <code>peptides_log_leduc</code> data provided by the authors. Let’s
compare the filtered cells.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareSets</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">peptides_log_leduc</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides_log"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## leduc2022 FALSE TRUE</span></span>
<span><span class="co">##     FALSE     0   27</span></span>
<span><span class="co">##     TRUE      2 1541</span></span></code></pre></div>
<p>There is an excellent agreement between the the original and the
replicated vignette. Let’s do the same for the filtered peptides.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareSets</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">peptides_log_leduc</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides_log"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## leduc2022 FALSE  TRUE</span></span>
<span><span class="co">##     FALSE     0    18</span></span>
<span><span class="co">##     TRUE     51 12233</span></span></code></pre></div>
<p>Notice here that most peptides that this vignette removed earlier are
now also removed by the original analysis. There is an excellent
agreement as well between selected peptides. Finally let’s compare the
quantitative data.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareQuantitativeData</span><span class="op">(</span><span class="va">peptides_log_leduc</span>, <span class="va">leduc</span><span class="op">[[</span><span class="st">"peptides_log"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-32-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The agreement is still very good, with a sharp peak around 0.
However, we can see that the range of differences starts to increase,
probably because numerical differences propagate as we progress through
the data processing.</p>
</div>
<div class="section level2">
<h2 id="aggregate-peptide-data-to-protein-data">Aggregate peptide data to protein data<a class="anchor" aria-label="anchor" href="#aggregate-peptide-data-to-protein-data"></a>
</h2>
<p>Similarly to aggregating PSM data to peptide data, we can aggregate
peptide data to protein data using the <code>aggregateFeatures</code>
function. Note that we here use the median as a summarizing
function.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                           i <span class="op">=</span> <span class="st">"peptides_log"</span>,</span>
<span>                           name <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                           fcol <span class="op">=</span> <span class="st">"Leading.razor.protein.symbol"</span>,</span>
<span>                           fun <span class="op">=</span> <span class="fu">matrixStats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html" class="external-link">colMedians</a></span>, </span>
<span>                           na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalization-1">Normalization<a class="anchor" aria-label="anchor" href="#normalization-1"></a>
</h2>
<p>Normalization is performed similarly to peptide normalization. We use
the same functions, but since the data were log-transformed at the
peptide level, we subtract by the median instead of dividing.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Center columns with median</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                   i <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                   method <span class="op">=</span> <span class="st">"center.median"</span>,</span>
<span>                   name <span class="op">=</span> <span class="st">"proteins_norm1"</span><span class="op">)</span></span>
<span><span class="co">## Scale rows with median</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>               i <span class="op">=</span> <span class="st">"proteins_norm1"</span>,</span>
<span>               name <span class="op">=</span> <span class="st">"proteins_norm2"</span>,</span>
<span>               MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>               FUN <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>               STATS <span class="op">=</span> <span class="fu">rowMedians</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                                  na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="compare-intermediate-results-2">Compare intermediate results<a class="anchor" aria-label="anchor" href="#compare-intermediate-results-2"></a>
</h2>
<p>At this stage of the processing, the last assay should be similar to
the <code>proteins_norm_leduc</code> data provided by the authors. Let’s
compare the filtered cells.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareSets</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">proteins_norm_leduc</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_norm2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## leduc2022 FALSE TRUE</span></span>
<span><span class="co">##     FALSE     0   27</span></span>
<span><span class="co">##     TRUE      2 1541</span></span></code></pre></div>
<p>There is an excellent agreement between the the original and the
replicated vignette. Let’s do the same for the filtered proteins.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareSets</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">proteins_norm_leduc</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_norm2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## leduc2022 FALSE TRUE</span></span>
<span><span class="co">##     FALSE     0 2837</span></span>
<span><span class="co">##     TRUE   2844    0</span></span></code></pre></div>
<p>There is an almost perfect agreement between the selected proteins.
selected Finally let’s compare the quantitative data.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareQuantitativeData</span><span class="op">(</span><span class="va">proteins_norm_leduc</span>, <span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_norm2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-37-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>Again, there is a very sharp peak around 0.</p>
</div>
<div class="section level2">
<h2 id="imputation">Imputation<a class="anchor" aria-label="anchor" href="#imputation"></a>
</h2>
<p>The protein data is majorily composed of missing values. The graph
below shows the distribution of the proportion missingness in cells.
Cells contain on average 65 % missing values.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>pNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">nNA</a></span><span class="op">(</span><span class="va">leduc</span>, <span class="st">"proteins_norm2"</span><span class="op">)</span><span class="op">$</span><span class="va">nNAcols</span><span class="op">$</span><span class="va">pNA</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pNA</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Percentage missingnes per cell"</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/missing_plot-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The missing data is imputed using K nearest neighbors. The authors
run KNN with k = 3. We made a wrapper around the author’s code to apply
imputation to our <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/imputeKnnSCoPE2.html">imputeKnnSCoPE2</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                         i <span class="op">=</span> <span class="st">"proteins_norm2"</span>,</span>
<span>                         name <span class="op">=</span> <span class="st">"proteins_impd"</span>,</span>
<span>                         k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><code>QFeatures</code> provides the <code>impute</code> function that
serves as an interface to different imputation algorithms among which
the KNN algorithm from <code><a href="https://rdrr.io/pkg/impute/man/impute.knn.html" class="external-link">impute::impute.knn</a></code>. However, the KNN
implementation in the oringal analysis and in <code>impute.knn</code>
are different. Leduc et al. perform KNN imputation in the sample space,
meaning that data from neighbouring cells are used to impute the central
cell, whereas <code><a href="https://rdrr.io/pkg/impute/man/impute.knn.html" class="external-link">impute::impute.knn</a></code> performs KNN imputation in
the feature space, meaning that data from neighbouring features are used
to impute the missing values from the central features. We provide the
code for KNN imputation with <code>QFeatures</code> but do not run in
order to replicate the original analysis.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                i <span class="op">=</span> <span class="st">"proteins_norm2"</span>,</span>
<span>                method <span class="op">=</span> <span class="st">"knn"</span>,</span>
<span>                k <span class="op">=</span> <span class="fl">3</span>, rowmax <span class="op">=</span> <span class="fl">1</span>, colmax<span class="op">=</span> <span class="fl">1</span>,</span>
<span>                maxp <span class="op">=</span> <span class="cn">Inf</span>, rng.seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h2>
<p>The next step is to correct for the remaining batch effects. The data
were acquired as a series of MS runs. Recall we had 134 assays at the
beginning of the workflow. Each MS run can be subjected to technical
perturbations that lead to differences in the data. Furthermore, TMT
labeling can also influence the quantification. These effects must be
accounted for to avoid attributing biological effects to technical
effects. The <code>limma</code> algorithm (CITE-Ritchie) is used by
Leduc et al. to correct for batch effects. It can take up to 2 batch
variables, in this case the MS acquisition batch and the TMT channel,
while protecting for variables of interest, the sample type in this
case. All the information is contained in the <code>colData</code> of
the <code>QFeatures</code> object. We first extract the assays with the
associated <code>colData</code>.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">leduc</span>, <span class="st">"proteins_impd"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span></code></pre></div>
<p>We next create the design matrix. We then perform the batch
correction and overwrite the data matrix. Recall the data matrix can be
accessed using the <code>assay</code> function.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">SampleType</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/removeBatchEffect.html" class="external-link">removeBatchEffect</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span>                                batch <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">lcbatch</span>,</span>
<span>                                batch2 <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">Channel</span>,</span>
<span>                                design <span class="op">=</span> <span class="va">model</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we add the batch corrected assay to the
<code>QFeatures</code> object and create the feature links.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">leduc</span>, y <span class="op">=</span> <span class="va">sce</span>, name <span class="op">=</span> <span class="st">"proteins_batchC"</span><span class="op">)</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLinkOneToOne</a></span><span class="op">(</span><span class="va">leduc</span>, from <span class="op">=</span> <span class="st">"proteins_impd"</span>,</span>
<span>                              to <span class="op">=</span> <span class="st">"proteins_batchC"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalization-2">Normalization<a class="anchor" aria-label="anchor" href="#normalization-2"></a>
</h2>
<p>The very last step of the data processing workflow is a new round of
normalization.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Center columns with median</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                   i <span class="op">=</span> <span class="st">"proteins_batchC"</span>,</span>
<span>                   method <span class="op">=</span> <span class="st">"center.median"</span>,</span>
<span>                   name <span class="op">=</span> <span class="st">"proteins_batchC_norm1"</span><span class="op">)</span></span>
<span><span class="co">## Scale rows with median</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>               i <span class="op">=</span> <span class="st">"proteins_batchC_norm1"</span>,</span>
<span>               name <span class="op">=</span> <span class="st">"proteins_processed"</span>,</span>
<span>               MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>               FUN <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>               STATS <span class="op">=</span> <span class="fu">rowMedians</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_batchC_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                                  na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="compare-the-final-results">Compare the final results<a class="anchor" aria-label="anchor" href="#compare-the-final-results"></a>
</h2>
<p>At the end of the processing, the last assay should be similar to the
<code>proteins_processed_leduc</code> data provided by the authors.
Let’s compare the filtered cells.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareSets</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">proteins_processed_leduc</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_processed"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## leduc2022 FALSE TRUE</span></span>
<span><span class="co">##     FALSE     0   27</span></span>
<span><span class="co">##     TRUE      2 1541</span></span></code></pre></div>
<p>There is an excellent agreement between the selected cells from the
original and this vignette. Let’s do the same for the filtered
proteins.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareSets</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">proteins_processed_leduc</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_processed"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##          scp</span></span>
<span><span class="co">## leduc2022 FALSE TRUE</span></span>
<span><span class="co">##     FALSE     0 2837</span></span>
<span><span class="co">##     TRUE   2844    0</span></span></code></pre></div>
<p>The agreement is also excellent between filtered proteins. Finally
let’s compare the quantitative data</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">compareQuantitativeData</span><span class="op">(</span><span class="va">proteins_processed_leduc</span>, </span>
<span>                        <span class="va">leduc</span><span class="op">[[</span><span class="st">"proteins_processed"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-46-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The differences are still sharply peaked around 0. However the
differences are more spread and the range is larger compared to the
previous steps.</p>
<p>Overall, we can see good replication of the data processing, although
early differences seem to get amplified as we progress through the
different processing steps. We next compare the dimension reduction
results to get a more qualitative assessment.</p>
</div>
<div class="section level2">
<h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<p>We run the same PCA procedure as performed by the authors, that is a
weighted PCA where the weight for a protein is defined as the summed
correlation with the other proteins.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">leduc</span>, <span class="st">"proteins_processed"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="co">## Warning: Ignoring redundant column names in 'colData(x)':</span></span>
<span><span class="va">pcaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pcaSCoPE2.html">pcaSCoPE2</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">## Compute percent explained variance</span></span>
<span><span class="va">pcaPercentVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">pcaRes</span><span class="op">$</span><span class="va">values</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pcaRes</span><span class="op">$</span><span class="va">values</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">## Plot PCA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>PC <span class="op">=</span> <span class="va">pcaRes</span><span class="op">$</span><span class="va">vectors</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>           <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC.1</span>, </span>
<span>        y <span class="op">=</span> <span class="va">PC.2</span>, </span>
<span>        colour <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC1 ("</span>, <span class="va">pcaPercentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"%)"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC2 ("</span>, <span class="va">pcaPercentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"%)"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"PCA on scp processed protein data"</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-47-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The PCA plot is very similar to the published PCA plot.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcaResLeduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pcaSCoPE2.html">pcaSCoPE2</a></span><span class="op">(</span><span class="va">proteins_processed_leduc</span><span class="op">)</span></span>
<span><span class="co">## Compute percent explained variance</span></span>
<span><span class="va">pcaPercentVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">pcaResLeduc</span><span class="op">$</span><span class="va">values</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pcaResLeduc</span><span class="op">$</span><span class="va">values</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">## Plot PCA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>PC <span class="op">=</span> <span class="va">pcaResLeduc</span><span class="op">$</span><span class="va">vectors</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>           <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">proteins_processed_leduc</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC.1</span>, </span>
<span>        y <span class="op">=</span> <span class="va">PC.2</span>, </span>
<span>        colour <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC1 ("</span>, <span class="va">pcaPercentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"%)"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC2 ("</span>, <span class="va">pcaPercentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"%)"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"PCA on processed protein data by Leduc et al."</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-48-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>Here again we can see the replicated PCA from this vignette is very
similar to the PCA published by the authors.</p>
<p>Using standard PCA, we obtain the same cell patterns although the
explained variance differs.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">## Perform PCA, see ?runPCA for more info about arguments</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">50</span>,</span>
<span>       ntop <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>       scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>       exprs_values <span class="op">=</span> <span class="fl">1</span>,</span>
<span>       name <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">## Plotting is performed in a single line of code</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span>colour_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span></span></code></pre></div>
<p><img src="leduc2022_files/figure-html/unnamed-chunk-49-1.png" width="70%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this vignette, we have demonstrated that the <code>scp</code>
package is able to accurately reproduce the analysis published by Leduc
et al. We not only support the reliability of the published work, but we
also offer a formalization and standardization of the pipeline by means
of easy-to-read and highly documented code. This workflow can serve as a
starting ground to improve upon the current methods and to design new
modelling tools dedicated to single-cell proteomics.</p>
</div>
<div class="section level2">
<h2 id="reproduce-this-vignette">Reproduce this vignette<a class="anchor" aria-label="anchor" href="#reproduce-this-vignette"></a>
</h2>
<p>You can reproduce this vignette using <code>Docker</code>:</p>
<pre><code>docker pull cvanderaa/scp_replication_docker:v1
docker run \
    -e PASSWORD=bioc \
    -p 8787:8787 \
    cvanderaa/scp_replication_docker:v1</code></pre>
<p>Open your browser and go to <a href="http://localhost:8787" class="external-link uri">http://localhost:8787</a>. The USER is <code>rstudio</code>
and the password is <code>bioc</code>. You can find the vignette in the
<code>vignettes</code> folder.</p>
<p>See the <a href="https://uclouvain-cbio.github.io/SCP.replication/index.html" class="external-link">website
home page</a> for more information.</p>
</div>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<div class="section level4">
<h4 id="hardware-and-software">Hardware and software<a class="anchor" aria-label="anchor" href="#hardware-and-software"></a>
</h4>
<p>The system details of the machine that built the vignette are:</p>
<pre><code><span><span class="co">## Machine: Linux (5.15.0-48-generic)</span></span>
<span><span class="co">## R version: R.4.2.1 (svn: 82513)</span></span>
<span><span class="co">## RAM: 16.5 GB</span></span>
<span><span class="co">## CPU: 16 core(s) - 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="timing">Timing<a class="anchor" aria-label="anchor" href="#timing"></a>
</h4>
<p>The total time required to compile this vignette is:</p>
<pre><code><span><span class="co">## 21.28293 mins</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="memory">Memory<a class="anchor" aria-label="anchor" href="#memory"></a>
</h4>
<p>The final <code>leduc</code> object size is:</p>
<pre><code><span><span class="co">## [1] "2.6 Gb"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h4>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] benchmarkme_1.0.8           scater_1.25.7              </span></span>
<span><span class="co">##  [3] scuttle_1.7.4               patchwork_1.1.2            </span></span>
<span><span class="co">##  [5] forcats_0.5.2               stringr_1.4.1              </span></span>
<span><span class="co">##  [7] dplyr_1.0.10                purrr_0.3.4                </span></span>
<span><span class="co">##  [9] readr_2.1.3                 tidyr_1.2.1                </span></span>
<span><span class="co">## [11] tibble_3.1.8                ggplot2_3.3.6              </span></span>
<span><span class="co">## [13] tidyverse_1.3.2             limma_3.53.10              </span></span>
<span><span class="co">## [15] SCP.replication_0.2.1       SingleCellExperiment_1.19.1</span></span>
<span><span class="co">## [17] scpdata_1.5.4               ExperimentHub_2.5.0        </span></span>
<span><span class="co">## [19] AnnotationHub_3.5.2         BiocFileCache_2.5.0        </span></span>
<span><span class="co">## [21] dbplyr_2.2.1                scp_1.7.4                  </span></span>
<span><span class="co">## [23] QFeatures_1.7.3             MultiAssayExperiment_1.23.9</span></span>
<span><span class="co">## [25] SummarizedExperiment_1.27.3 Biobase_2.57.1             </span></span>
<span><span class="co">## [27] GenomicRanges_1.49.1        GenomeInfoDb_1.33.7        </span></span>
<span><span class="co">## [29] IRanges_2.31.2              S4Vectors_0.35.4           </span></span>
<span><span class="co">## [31] BiocGenerics_0.43.4         MatrixGenerics_1.9.1       </span></span>
<span><span class="co">## [33] matrixStats_0.62.0          BiocStyle_2.25.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] utf8_1.2.2                    reticulate_1.26              </span></span>
<span><span class="co">##   [3] tidyselect_1.1.2              RSQLite_2.2.17               </span></span>
<span><span class="co">##   [5] AnnotationDbi_1.59.1          grid_4.2.1                   </span></span>
<span><span class="co">##   [7] BiocParallel_1.31.12          munsell_0.5.0                </span></span>
<span><span class="co">##   [9] ScaledMatrix_1.5.1            codetools_0.2-18             </span></span>
<span><span class="co">##  [11] ragg_1.2.3                    withr_2.5.0                  </span></span>
<span><span class="co">##  [13] colorspace_2.0-3              filelock_1.0.2               </span></span>
<span><span class="co">##  [15] highr_0.9                     knitr_1.40                   </span></span>
<span><span class="co">##  [17] labeling_0.4.2                GenomeInfoDbData_1.2.9       </span></span>
<span><span class="co">##  [19] bit64_4.0.5                   farver_2.1.1                 </span></span>
<span><span class="co">##  [21] rprojroot_2.0.3               vctrs_0.4.2                  </span></span>
<span><span class="co">##  [23] generics_0.1.3                xfun_0.33                    </span></span>
<span><span class="co">##  [25] doParallel_1.0.17             R6_2.5.1                     </span></span>
<span><span class="co">##  [27] ggbeeswarm_0.6.0              clue_0.3-61                  </span></span>
<span><span class="co">##  [29] rsvd_1.0.5                    locfit_1.5-9.6               </span></span>
<span><span class="co">##  [31] MsCoreUtils_1.9.1             AnnotationFilter_1.21.0      </span></span>
<span><span class="co">##  [33] bitops_1.0-7                  cachem_1.0.6                 </span></span>
<span><span class="co">##  [35] DelayedArray_0.23.2           assertthat_0.2.1             </span></span>
<span><span class="co">##  [37] promises_1.2.0.1              scales_1.2.1                 </span></span>
<span><span class="co">##  [39] googlesheets4_1.0.1           beeswarm_0.4.0               </span></span>
<span><span class="co">##  [41] gtable_0.3.1                  beachmat_2.13.4              </span></span>
<span><span class="co">##  [43] OrgMassSpecR_0.5-3            benchmarkmeData_1.0.4        </span></span>
<span><span class="co">##  [45] sva_3.45.0                    rlang_1.0.6                  </span></span>
<span><span class="co">##  [47] genefilter_1.79.0             systemfonts_1.0.4            </span></span>
<span><span class="co">##  [49] splines_4.2.1                 lazyeval_0.2.2               </span></span>
<span><span class="co">##  [51] gargle_1.2.1                  broom_1.0.1                  </span></span>
<span><span class="co">##  [53] BiocManager_1.30.18           yaml_2.3.5                   </span></span>
<span><span class="co">##  [55] modelr_0.1.9                  backports_1.4.1              </span></span>
<span><span class="co">##  [57] httpuv_1.6.6                  tools_4.2.1                  </span></span>
<span><span class="co">##  [59] bookdown_0.29                 ellipsis_0.3.2               </span></span>
<span><span class="co">##  [61] jquerylib_0.1.4               Rcpp_1.0.9                   </span></span>
<span><span class="co">##  [63] sparseMatrixStats_1.9.0       zlibbioc_1.43.0              </span></span>
<span><span class="co">##  [65] RCurl_1.98-1.9                viridis_0.6.2                </span></span>
<span><span class="co">##  [67] cowplot_1.1.1                 haven_2.5.1                  </span></span>
<span><span class="co">##  [69] ggrepel_0.9.1                 cluster_2.1.4                </span></span>
<span><span class="co">##  [71] fs_1.5.2                      magrittr_2.0.3               </span></span>
<span><span class="co">##  [73] reprex_2.0.2                  googledrive_2.0.0            </span></span>
<span><span class="co">##  [75] ProtGenerics_1.29.0           hms_1.1.2                    </span></span>
<span><span class="co">##  [77] mime_0.12                     evaluate_0.16                </span></span>
<span><span class="co">##  [79] xtable_1.8-4                  XML_3.99-0.11                </span></span>
<span><span class="co">##  [81] readxl_1.4.1                  gridExtra_2.3                </span></span>
<span><span class="co">##  [83] compiler_4.2.1                crayon_1.5.2                 </span></span>
<span><span class="co">##  [85] htmltools_0.5.3               mgcv_1.8-40                  </span></span>
<span><span class="co">##  [87] later_1.3.0                   tzdb_0.3.0                   </span></span>
<span><span class="co">##  [89] lubridate_1.8.0               DBI_1.1.3                    </span></span>
<span><span class="co">##  [91] MASS_7.3-58.1                 rappdirs_0.3.3               </span></span>
<span><span class="co">##  [93] Matrix_1.5-1                  cli_3.4.1                    </span></span>
<span><span class="co">##  [95] parallel_4.2.1                igraph_1.3.5                 </span></span>
<span><span class="co">##  [97] pkgconfig_2.0.3               pkgdown_2.0.6                </span></span>
<span><span class="co">##  [99] foreach_1.5.2                 xml2_1.3.3                   </span></span>
<span><span class="co">## [101] annotate_1.75.0               vipor_0.4.5                  </span></span>
<span><span class="co">## [103] bslib_0.4.0                   XVector_0.37.1               </span></span>
<span><span class="co">## [105] rvest_1.0.3                   digest_0.6.29                </span></span>
<span><span class="co">## [107] Biostrings_2.65.6             rmarkdown_2.16               </span></span>
<span><span class="co">## [109] cellranger_1.1.0              edgeR_3.39.6                 </span></span>
<span><span class="co">## [111] DelayedMatrixStats_1.19.1     curl_4.3.2                   </span></span>
<span><span class="co">## [113] shiny_1.7.2                   lifecycle_1.0.2              </span></span>
<span><span class="co">## [115] nlme_3.1-159                  jsonlite_1.8.2               </span></span>
<span><span class="co">## [117] BiocNeighbors_1.15.1          desc_1.4.2                   </span></span>
<span><span class="co">## [119] viridisLite_0.4.1             fansi_1.0.3                  </span></span>
<span><span class="co">## [121] pillar_1.8.1                  lattice_0.20-45              </span></span>
<span><span class="co">## [123] KEGGREST_1.37.3               fastmap_1.1.0                </span></span>
<span><span class="co">## [125] httr_1.4.4                    survival_3.4-0               </span></span>
<span><span class="co">## [127] interactiveDisplayBase_1.35.0 glue_1.6.2                   </span></span>
<span><span class="co">## [129] iterators_1.0.14              png_0.1-7                    </span></span>
<span><span class="co">## [131] BiocVersion_3.16.0            bit_4.0.4                    </span></span>
<span><span class="co">## [133] stringi_1.7.8                 sass_0.4.2                   </span></span>
<span><span class="co">## [135] BiocBaseUtils_0.99.12         blob_1.2.3                   </span></span>
<span><span class="co">## [137] textshaping_0.3.6             BiocSingular_1.13.1          </span></span>
<span><span class="co">## [139] memoise_2.0.1                 irlba_2.3.5.1</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="licence">Licence<a class="anchor" aria-label="anchor" href="#licence"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
licence</a> licence.</p>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:christophe.vanderaa@uclouvain.be" class="email">christophe.vanderaa@uclouvain.be</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
