<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Exploring the autoPOTS data (Liang et al. 2020) â€¢ SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Exploring the autoPOTS data (Liang et al. 2020)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Reproduction of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/schoof2021_R.html">Reproduction of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/schoof2021_python.html">Reproduction of the AML model analysis with python (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Exploring the autoPOTS data (Liang et al. 2020)</h1>
                        <h4 class="author">Christophe Vanderaa, Computational Biology, UCLouvain</h4>
                        <h4 class="author">Laurent Gatto, Computational Biology, UCLouvain</h4>
            
            <h4 class="date">1 Oktober 2021</h4>
      
      
      <div class="hidden name"><code>liang2020.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>autoPOTS (<span class="citation">Liang et al. (2020)</span>) is an acquisition protocol for label-free mass spectrometry (MS)-based single-cell proteomics (SCP). It relies on commercial material and is meant to be widely adopted and applied. Its development derives from the nanoPOTS technology (<span class="citation">Zhu et al. (2018)</span>) that uses custom-built instruments that are hardly transferable to other laboratories. Therefore, autoPOTS is the first attempt to provide a reproducible label-free SCP acquisition workflow.</p>
<p>While many efforts are put on a reproducible technology, the authors do not address the reproducibility of the downstream data analysis. We here tackle this issue by showing that the <code>scp</code> package can be used to manipulate and process the data. We improve upon the results shown in the article and demonstrate that <code>scp</code> is a keystone for principled and rigorous data exploration.</p>
<p>Running this vignette requires the <code>SCP.replication</code> package:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"SCP.replication"</span><span class="op">)</span></code></pre></div>
<div id="scp-workflow-and-data-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#scp-workflow-and-data-structure" class="anchor"></a><code>scp</code> workflow and data structure</h3>
<p>The data processing workflow we carry out in this vignette is inferred from the methods section of the paper. Since no reproducible data analysis script was provided, we are not able to accurately evaluate the accuracy of the replication of the results. We will rely on Figure S5 to assess the replication. This vignette will also extend the results to better assess the quality of the provided data.</p>
<p>To run the vignette you will need the following packages:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">clusterProfiler</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures">QFeatures</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp">scp</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scpdata</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/hms-dbmi/UpSetR">UpSetR</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="scpdata-and-the-autopots-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#scpdata-and-the-autopots-dataset" class="anchor"></a><code>scpdata</code> and the autoPOTS dataset</h3>
<p>The authors provided 4 data sets related to the article:</p>
<ol style="list-style-type: decimal">
<li>HeLa samples containing 0, 1, 10, 150 and 500 cells. Peptides and proteins were identified and quantified using MaxQuant.</li>
<li>HeLa samples containing 0, 1, 10, 150 and 500 cells. Peptides and proteins were identified using FragPipe.</li>
<li>HeLa samples containing 0, 10, 100 and 500 cells. Peptides and proteins were identified using FragPipe.</li>
<li>Lymphocyte samples containing 10 cells. Peptides and proteins were identified using FragPipe.</li>
</ol>
<p>Since no quantification data is available for the FragPipe data sets, we here only use the MaxQuant output. We retrieved the data from the PXD021882 repository. We formatted the data to a <code>QFeatures</code> object following the <code>scp</code> data format (more information can be found in the <a href="https://uclouvain-cbio.github.io/scp/articles/scp.html"><code>scp</code> vignette</a>). This formatted data is available directly from the <code>scpdata</code> package:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">autoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/liang2020_hela.html">liang2020_hela</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The data contain 17 different <code>SingleCellExperiment</code> objects that we refer to as <strong>assays</strong>. Each assay contains expression data along with feature metadata. Each row in an assay represents a <strong>feature</strong> that can either be a PSM, a peptide or a protein depending on the assay. Each column in an assay represents a <strong>sample</strong>. Below, we show the overview of the <code>autoPOTS</code> dataset.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">autoPOTS</span></code></pre></div>
<pre><code>## An instance of class QFeatures containing 17 assays:
##  [1] HeLa_0cell_B16: SingleCellExperiment with 1153 rows and 1 columns 
##  [2] HeLa_0cell_B18: SingleCellExperiment with 1214 rows and 1 columns 
##  [3] HeLa_0cell_B4: SingleCellExperiment with 1366 rows and 1 columns 
##  ...
##  [15] HeLa_500cell_K8: SingleCellExperiment with 35054 rows and 1 columns 
##  [16] peptides: SingleCellExperiment with 48705 rows and 15 columns 
##  [17] proteins: SingleCellExperiment with 3970 rows and 15 columns</code></pre>
<p>15 out of the 17 assays are PSM data, each assay corresponding to a separate MS run. Since LFQ acquisitions contain a single sample per MS run, those 15 assays contain a single column. The dataset also contains a <code>peptides</code> assay and a <code>proteins</code> assay that hold peptide and protein level information,respectively. Those assays were produced by the authors after running MaxQuant.</p>
<p>The data set also contains samples annotations that can be retrieved from the <code>colData</code> of the <code>QFeatures</code> object. The annotation contains the exact number of HeLa cells in each sample, but also additional information about cell type, sample preparation method, or MS2 settings.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 15 rows and 9 columns
##                              Raw.file   nbCells  SampleType IsolationMethod
##                           &lt;character&gt; &lt;integer&gt; &lt;character&gt;     &lt;character&gt;
## HeLa_0cell_B16_sample   HeLa_0cell...         0        HeLa      micromanip
## HeLa_0cell_B18_sample   HeLa_0cell...         0        HeLa      micromanip
## HeLa_0cell_B4_sample    HeLa_0cell...         0        HeLa      micromanip
## HeLa_10cell_G4_sample   HeLa_10cel...        10        HeLa      micromanip
## HeLa_10cell_G8_sample   HeLa_10cel...        10        HeLa      micromanip
## ...                               ...       ...         ...             ...
## HeLa_1cell_E14_sample   HeLa_1cell...         1        HeLa      micromanip
## HeLa_1cell_F8_sample    HeLa_1cell...         1        HeLa      micromanip
## HeLa_500cell_J5_sample  HeLa_500ce...       466        HeLa        dilution
## HeLa_500cell_K16_sample HeLa_500ce...       520        HeLa        dilution
## HeLa_500cell_K8_sample  HeLa_500ce...       540        HeLa        dilution
##                         MS2exclusionTime_s MS2accumulationTime_ms MS2resolution
##                                  &lt;integer&gt;              &lt;integer&gt;     &lt;integer&gt;
## HeLa_0cell_B16_sample                   90                    500         60000
## HeLa_0cell_B18_sample                   90                    500         60000
## HeLa_0cell_B4_sample                    90                    500         60000
## HeLa_10cell_G4_sample                   90                    250         60000
## HeLa_10cell_G8_sample                   90                    250         60000
## ...                                    ...                    ...           ...
## HeLa_1cell_E14_sample                   90                    500         60000
## HeLa_1cell_F8_sample                    90                    500         60000
## HeLa_500cell_J5_sample                  60                     50         30000
## HeLa_500cell_K16_sample                 60                     50         30000
## HeLa_500cell_K8_sample                  60                     50         30000
##                            MS2AGC QuantificationColumn
##                         &lt;integer&gt;          &lt;character&gt;
## HeLa_0cell_B16_sample        5000            Intensity
## HeLa_0cell_B18_sample        5000            Intensity
## HeLa_0cell_B4_sample         5000            Intensity
## HeLa_10cell_G4_sample       10000            Intensity
## HeLa_10cell_G8_sample       10000            Intensity
## ...                           ...                  ...
## HeLa_1cell_E14_sample        5000            Intensity
## HeLa_1cell_F8_sample         5000            Intensity
## HeLa_500cell_J5_sample      10000            Intensity
## HeLa_500cell_K16_sample     10000            Intensity
## HeLa_500cell_K8_sample      10000            Intensity</code></pre>
<p>To facilitate later analysis, we will group the cell numbers in bins of 0, 1, 10, 150 and 500 cells. Therefore, we add a new variable in the <code>colData</code> containing this grouping.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">$</span><span class="va">nbCells</span>
<span class="va">group</span><span class="op">[</span><span class="va">group</span> <span class="op">&gt;</span> <span class="fl">130</span> <span class="op">&amp;</span> <span class="va">group</span> <span class="op">&lt;</span> <span class="fl">170</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">150</span>
<span class="va">group</span><span class="op">[</span><span class="va">group</span> <span class="op">&gt;</span> <span class="fl">450</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">500</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##   0   1  10 150 500 
##   3   3   3   3   3</code></pre>
<p>We remove the <code>peptides</code> and <code>proteins</code> to keep only the PSM data. We can subset a <code>QFeatures</code> object using the simple bracket operation. The desired elements/assays can be selected following the <code>[row, col, assay]</code> syntax.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">autoPOTS</span> <span class="op">&lt;-</span> <span class="va">autoPOTS</span><span class="op">[</span>, , <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span></code></pre></div>
</div>
</div>
<div id="process-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#process-the-data" class="anchor"></a>Process the data</h1>
<p>In this section, we perform data processing as described in the methods section. This consist of two steps: i) removing reverse hits, peptides flagged as contaminants and keep only unmodified peptides, ii) replace missing data by NA. We will also load another data set containing data generated by nanoPOTS to compare the efficiency of autoPOTS against its parent technology. The nanoPOTS data will be processed in a similar way.</p>
<div id="autopots" class="section level2">
<h2 class="hasAnchor">
<a href="#autopots" class="anchor"></a>autoPOTS</h2>
<p>We start by filtering out undesired PSMs. Each PSM assay contains feature meta-information that are stored in the <code>rowData</code> of the assay. The <code>QFeatures</code> package allows to quickly filter the rows of an assay by using these information. The available variables in the <code>rowData</code> are listed below for each assay.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">rowDataNames</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span></code></pre></div>
<pre><code>## CharacterList of length 15
## [["HeLa_0cell_B16"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_0cell_B18"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_0cell_B4"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_10cell_G4"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_10cell_G8"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_10cell_H2"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_150cell_I12"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_150cell_I16"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_150cell_I18"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## [["HeLa_1cell_E12"]] Sequence Length ... Best.MS.MS Oxidation..M..site.IDs
## ...
## &lt;5 more elements&gt;</code></pre>
<p>Among those feature variable are <code>Reverse</code>, <code>Potential.contaminant</code> and <code>Modifications</code> that allow to filter for peptides that are not matched to a reversed sequence, that are not matched to a contaminant peptide and that are unmodified, respectively. We can perform this on the <code>QFeatures</code> object by using the <code>filterFeatures</code> function. The filters can be directly applied to the <code>rowData</code> variables of each assay.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">autoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">autoPOTS</span>, 
                           <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span> 
                               <span class="va">Potential.contaminant</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span>
                               <span class="va">Modifications</span> <span class="op">==</span> <span class="st">"Unmodified"</span><span class="op">)</span></code></pre></div>
<p>Then, we replace zero values by <code>NA</code> to explicitly encode the missing data. This is performed using the <code>zeroisNA</code> function.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">autoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">zeroIsNA</a></span><span class="op">(</span><span class="va">autoPOTS</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="nanopots" class="section level2">
<h2 class="hasAnchor">
<a href="#nanopots" class="anchor"></a>nanoPOTS</h2>
<p>In the article, the authors compare their recent results to the nanoPOTS results published in 2018 (<span class="citation">Zhu et al. (2018)</span>). This data is readily available in <code>scpdata</code>. The data set contains several types of samples: blanks, HeLa lysates, HeLa cells, MCF7 cells and THP1 cells. We will only keep the HeLa cells and the blanks.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nanoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/zhu2018NC_hela.html">zhu2018NC_hela</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">nanoPOTS</span> <span class="op">&lt;-</span> <span class="va">nanoPOTS</span><span class="op">[</span>, <span class="va">nanoPOTS</span><span class="op">$</span><span class="va">SampleType</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Hela"</span>, <span class="st">"Blank"</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">nanoPOTS</span></code></pre></div>
<pre><code>## An instance of class QFeatures containing 2 assays:
##  [1] peptides: SingleCellExperiment with 37795 rows and 12 columns 
##  [2] proteins: SingleCellExperiment with 3984 rows and 12 columns</code></pre>
<p>Note that the PSM quantification data is not available for the <code>autoPOTS</code>, but this is not an issue since we will mainly focus on peptide or protein information. We perform the same data cleaning as for autoPOTS.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nanoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">nanoPOTS</span>, 
                           <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span> 
                               <span class="va">Potential.contaminant</span> <span class="op">!=</span> <span class="st">"+"</span><span class="op">)</span>
<span class="va">nanoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">zeroIsNA</a></span><span class="op">(</span><span class="va">nanoPOTS</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>To further make the two data sets comparable, we remove observations that were identified using match between run, because this option was disabled for autoPOTS.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">matchTypeCols</span> <span class="op">&lt;-</span> 
    <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"Identification.type.Hela"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">isMBR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>, <span class="va">matchTypeCols</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">"By matching"</span>
<span class="fu">assay</span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="va">isMBR</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></code></pre></div>
<p>Finally, just like for the autoPOTS dataset, we bin the cell numbers into groups.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">)</span><span class="op">$</span><span class="va">nbCells</span>
<span class="va">group</span><span class="op">[</span><span class="va">group</span> <span class="op">&gt;</span> <span class="fl">130</span> <span class="op">&amp;</span> <span class="va">group</span> <span class="op">&lt;</span> <span class="fl">145</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">140</span>
<span class="va">group</span><span class="op">[</span><span class="va">group</span> <span class="op">&gt;</span> <span class="fl">30</span> <span class="op">&amp;</span> <span class="va">group</span> <span class="op">&lt;</span> <span class="fl">50</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">40</span>
<span class="va">group</span><span class="op">[</span><span class="va">group</span> <span class="op">&gt;</span> <span class="fl">9</span> <span class="op">&amp;</span> <span class="va">group</span> <span class="op">&lt;</span> <span class="fl">15</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">10</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">)</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">)</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##   0  10  40 140 
##   3   3   3   3</code></pre>
<p>Unfortunately, no true single-cell is available from this dataset, but sample still contain very low amounts of cells.</p>
</div>
</div>
<div id="feature-counts" class="section level1">
<h1 class="hasAnchor">
<a href="#feature-counts" class="anchor"></a>Feature counts</h1>
<p>The feature counts are the only results that are shown for the HeLa data set. Since we are using the quantifications generated by MaxQuant, we can relate to Figure S5.</p>
<div class="figure" style="text-align: center">
<img src="figs/liang2020-figureS5.png" alt="Figure S5 from Liang et al. 2020." width="70%"><p class="caption">
Figure S5 from Liang et al. 2020.
</p>
</div>
<p>We count the unique number of peptides and proteins for each data set using the <code>countUniqueFeatures</code> from the <code>Qfeatures</code> package. To perform this, we need to supply the grouping variables. Peptides are identified by <code>Sequence</code> (<em>i.e</em> the peptide sequence) and proteins are identified by <code>Leading.razor.protein</code>. Note that the features are counted from the PSM data for the autoPOTS dataset, we therefore supply all the assay names to the function.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Count peptides</span>
<span class="va">autoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/countUniqueFeatures.html">countUniqueFeatures</a></span><span class="op">(</span><span class="va">autoPOTS</span>, 
                                i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span>, 
                                groupBy <span class="op">=</span> <span class="st">"Sequence"</span>,
                                colDataName <span class="op">=</span> <span class="st">"Peptide_counts"</span><span class="op">)</span>
<span class="va">nanoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/countUniqueFeatures.html">countUniqueFeatures</a></span><span class="op">(</span><span class="va">nanoPOTS</span>, 
                                i <span class="op">=</span> <span class="st">"peptides"</span>, 
                                groupBy <span class="op">=</span> <span class="st">"Sequence"</span>,
                                colDataName <span class="op">=</span> <span class="st">"Peptide_counts"</span><span class="op">)</span>
<span class="co">## Count proteins</span>
<span class="va">autoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/countUniqueFeatures.html">countUniqueFeatures</a></span><span class="op">(</span><span class="va">autoPOTS</span>, 
                                i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span>, 
                                groupBy <span class="op">=</span> <span class="st">"Leading.razor.protein"</span>,
                                colDataName <span class="op">=</span> <span class="st">"Protein_counts"</span><span class="op">)</span>
<span class="va">nanoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/countUniqueFeatures.html">countUniqueFeatures</a></span><span class="op">(</span><span class="va">nanoPOTS</span>, 
                                i <span class="op">=</span> <span class="st">"peptides"</span>, 
                                groupBy <span class="op">=</span> <span class="st">"Leading.razor.protein"</span>,
                                colDataName <span class="op">=</span> <span class="st">"Protein_counts"</span><span class="op">)</span></code></pre></div>
</div>
<div id="compare-autopots-and-nanopots" class="section level1">
<h1 class="hasAnchor">
<a href="#compare-autopots-and-nanopots" class="anchor"></a>Compare autoPOTs and nanoPOTS</h1>
<p>We combine the sample data (<code>colData</code>) from the two data sets. We then use <code>ggplot2</code> to plot the peptides and protein counts for each data set.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                <span class="va">data.frame</span> <span class="op">%&gt;%</span> 
                <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Protein_counts</span>, <span class="va">Peptide_counts</span>, 
                              <span class="va">nbCells</span>, <span class="va">group</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"autoPOTS"</span><span class="op">)</span>,
            <span class="fu">colData</span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                <span class="va">data.frame</span> <span class="op">%&gt;%</span> 
                <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Protein_counts</span>, <span class="va">Peptide_counts</span>, 
                              <span class="va">nbCells</span>, <span class="va">group</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"nanoPOTS"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="fu">ends_with</span><span class="op">(</span><span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Plot</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">value</span>, 
        x <span class="op">=</span> <span class="va">group</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_text</span><span class="op">(</span>data <span class="op">=</span> <span class="va">.</span> <span class="op">%&gt;%</span>
                  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">group</span>, <span class="va">dataset</span>, <span class="va">name</span><span class="op">)</span> <span class="op">%&gt;%</span>
                  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span>,
                            group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span>,
                            y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1000</span><span class="op">)</span>,
              <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">y</span>,
                  label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_wrap</span><span class="op">(</span><span class="va">name</span> <span class="op">~</span> <span class="va">dataset</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p><strong>Watch out</strong>: this plot as well as the figure in the paper give the false impression of a linear increase. This is because the number of cells is treated as a factor and all categories are separated using an equal width. Let's redo the bottom right plot, but using the cell as a numeric variable, hence using a numeric scale.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">%&gt;%</span> 
    <span class="va">data.frame</span> <span class="op">%&gt;%</span> 
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">nbCells</span>,
        y <span class="op">=</span> <span class="va">value</span>,
        colour <span class="op">=</span> <span class="va">dataset</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span>, formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/poly.html">poly</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">name</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-7-1.png" width="50%"></p>
<p>The relationship between number of cells and detected features is highly non-linear and the decreased performance of <code>autoPOTS</code> compared to <code>nanoPOTS</code> is easier to assess.</p>
</div>
<div id="protein-detection" class="section level1">
<h1 class="hasAnchor">
<a href="#protein-detection" class="anchor"></a>Protein detection</h1>
<div id="in-the-autopots-data-set" class="section level2">
<h2 class="hasAnchor">
<a href="#in-the-autopots-data-set" class="anchor"></a>In the autoPOTS data set</h2>
<p>In Figure S5, the authors use a Venn diagram to show the overlap between the sets of proteins identified at different cell concentrations. While very intuitive, we dig a little bit deeper by providing an upset plot. Although upset plots are harder to read, they contain more information and make comparison between sets much easier.</p>
<p>First, we format the <code>QFeatures</code> object to a long <code>DataFrame</code> using the <code>longFormat</code> function. This will take the quantifications matrix and turn it in a single vector, and add the desired feature and sample metadata. We also add a dataset identifier for later use.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lfauto</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">[</span>, ,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"HeLa"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,
                     colvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nbCells"</span>, <span class="st">"group"</span><span class="op">)</span>,
                     rowvars <span class="op">=</span> <span class="st">"Leading.razor.protein"</span><span class="op">)</span>
<span class="va">lfauto</span><span class="op">$</span><span class="va">dataset</span> <span class="op">&lt;-</span> <span class="st">"autoPOTS"</span></code></pre></div>
<p>We then build the upset plot using the functionality from <code>upsetR</code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html">fromList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">lfauto</span><span class="op">$</span><span class="va">Leading.razor.protein</span>, 
                     <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">lfauto</span><span class="op">$</span><span class="va">group</span>, <span class="st">"cells - autoPOTS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html">upset</a></span><span class="op">(</span><span class="va">lt</span>, 
      nsets <span class="op">=</span> <span class="fl">5</span>, 
      order.by <span class="op">=</span> <span class="st">"freq"</span>, 
      nintersects <span class="op">=</span> <span class="fl">15</span>,
      keep.order <span class="op">=</span> <span class="cn">TRUE</span>,
      mainbar.y.label <span class="op">=</span> <span class="st">"Proteins in intersection"</span>, 
      sets.x.label <span class="op">=</span> <span class="st">"Proteins"</span>,
      mb.ratio <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Several observations can be drawn from this graph:</p>
<ul>
<li>The largest common set is found between 10, 150 and 500 cells, but it does not include single-cells</li>
<li>A great majority of the proteins are found in the 500 cell samples this indicates that working with smaller sample sizes cover a fraction of the bulk proteome.</li>
<li>There are 269 proteins found in all sample excluding blanks, and there are 177 proteins found in all sample including blanks. This indicates that about 40 % of the proteins found in single-cells are potentially caused by background signal (contamination by culture media or during sample processing).</li>
</ul>
</div>
<div id="autopots-against-nanopots" class="section level2">
<h2 class="hasAnchor">
<a href="#autopots-against-nanopots" class="anchor"></a>autoPOTS against nanoPOTS</h2>
<p>We will now compare to what extend the results obtained using the autoPOTS technology are consistent with the previous data acquired using the nanoPOTS technology. Similarly to the <code>autoPOTS</code> data set, we format the <code>nanoPOTS</code> in a long table.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lfnano</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat</a></span><span class="op">(</span><span class="va">nanoPOTS</span><span class="op">[</span>, ,<span class="st">"peptides"</span><span class="op">]</span>,
                     colvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nbCells"</span>, <span class="st">"group"</span><span class="op">)</span>,
                     rowvars <span class="op">=</span> <span class="st">"Leading.razor.protein"</span><span class="op">)</span>
<span class="va">lfnano</span><span class="op">$</span><span class="va">dataset</span> <span class="op">&lt;-</span> <span class="st">"nanoPOTS"</span></code></pre></div>
<p>We combine the long tables for the two data sets in a single table.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">lfnano</span>, <span class="va">lfauto</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="va">data.frame</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="va">DataFrame</span> <span class="op">-&gt;</span>
    <span class="va">lf</span></code></pre></div>
<p>In order to compare the coverage between the two datasets, we keep only samples between 10 and 150 cells (common in both experimental designs). We then create the upset plot.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lfsub</span> <span class="op">&lt;-</span> <span class="va">lf</span><span class="op">[</span><span class="va">lf</span><span class="op">$</span><span class="va">nbCells</span> <span class="op">&gt;=</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">lf</span><span class="op">$</span><span class="va">nbCells</span> <span class="op">&lt;=</span> <span class="fl">150</span>, <span class="op">]</span>
<span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html">fromList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">lfsub</span><span class="op">$</span><span class="va">Leading.razor.protein</span>, <span class="va">lfsub</span><span class="op">$</span><span class="va">dataset</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html">upset</a></span><span class="op">(</span><span class="va">lt</span>, 
      sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">lt</span><span class="op">)</span>,
      order.by <span class="op">=</span> <span class="st">"freq"</span>, 
      mainbar.y.label <span class="op">=</span> <span class="st">"Proteins in intersection"</span>, 
      sets.x.label <span class="op">=</span> <span class="st">"Proteins in sample"</span>,
      mb.ratio <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>The majority of the proteins (2300 out of 3721) are consistently found by <code>autoPOTS</code> and <code>nanoPOTS</code>. A significant number of proteins are found only by the nanoPOTS protocol. This is probably due to its increase sensitivity compared to autoPOTS. Finally, a few proteins are only found using the nanoPOTS technology.</p>
</div>
<div id="robustness-of-the-detection" class="section level2">
<h2 class="hasAnchor">
<a href="#robustness-of-the-detection" class="anchor"></a>Robustness of the detection</h2>
<p>In the previous sections, we have seen that more proteins are identified in samples with more cells. In this section, we explore how robust the identification are across samples with the same number of cells. We focus on the autoPOTs data set only.</p>
<div id="in-500-cell-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#in-500-cell-samples" class="anchor"></a>In 500-cell samples</h3>
<p>We take again the autoPOTS data formatted as a long table, and subset only data points that were acquired in 500-cell samples. We make the upset plot to assess the consistency of the protein identification across those samples.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lt</span> <span class="op">&lt;-</span> <span class="va">lfauto</span><span class="op">[</span><span class="va">lfauto</span><span class="op">$</span><span class="va">nbCells</span> <span class="op">%in%</span> <span class="fl">450</span><span class="op">:</span><span class="fl">550</span>, <span class="op">]</span>
<span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html">fromList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">lt</span><span class="op">$</span><span class="va">Leading.razor.protein</span>, <span class="va">lt</span><span class="op">$</span><span class="va">colname</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html">upset</a></span><span class="op">(</span><span class="va">lt</span>, 
      nsets <span class="op">=</span> <span class="fl">17</span>, 
      order.by <span class="op">=</span> <span class="st">"freq"</span>, 
      nintersects <span class="op">=</span> <span class="fl">30</span>,
      keep.order <span class="op">=</span> <span class="cn">TRUE</span>,
      mainbar.y.label <span class="op">=</span> <span class="st">"Proteins in intersection"</span>, 
      sets.x.label <span class="op">=</span> <span class="st">"Proteins"</span>,
      mb.ratio <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>Almost 90% of the proteins are found among all 3 samples. This is shows that the protein detection and identification is highly consistent across samples.</p>
</div>
<div id="in-single-cells" class="section level3">
<h3 class="hasAnchor">
<a href="#in-single-cells" class="anchor"></a>In single-cells</h3>
<p>We apply the same procedure for single-cell samples.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lt</span> <span class="op">&lt;-</span> <span class="va">lfauto</span><span class="op">[</span><span class="va">lfauto</span><span class="op">$</span><span class="va">nbCells</span> <span class="op">==</span> <span class="fl">1</span>, <span class="op">]</span>
<span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html">fromList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">lt</span><span class="op">$</span><span class="va">Leading.razor.protein</span>, <span class="va">lt</span><span class="op">$</span><span class="va">colname</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html">upset</a></span><span class="op">(</span><span class="va">lt</span>, 
      nsets <span class="op">=</span> <span class="fl">17</span>, 
      order.by <span class="op">=</span> <span class="st">"freq"</span>, 
      nintersects <span class="op">=</span> <span class="fl">30</span>,
      keep.order <span class="op">=</span> <span class="cn">TRUE</span>,
      mainbar.y.label <span class="op">=</span> <span class="st">"Proteins in intersection"</span>, 
      sets.x.label <span class="op">=</span> <span class="st">"Proteins"</span>,
      mb.ratio <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>The results for single-cell samples are very different from the results shown above for 500-cell samples. For each single-cell sample, about 50 % of the identified proteins were also found in the two other samples. However, about a third of the proteins are only found in one of the single-cell samples, indicating a high variability in detection rates between single-cell samples. There are two possible explanations:</p>
<ul>
<li>Technical missingness: protein concentrations are too low in single-cell samples so that only a part of the most expressed proteins are detectable. This is subjected to stochastic effects (MS selection, heterogeneity of expression,...) that lead to different sets of identified proteins.</li>
<li>Biological missingness: the inherent heterogeneity in single cells leads to the expression of different, but biologically meaningful, sets of proteins. The more cells are contained in a sample, the more diluted this heterogeneity is.</li>
</ul>
<p>The most probable scenario is that there is a combination of the two types of missingness. However, it is not trivial to decouple technical and biological missingness. Synthetic samples are required to better assess which component of the missingness are more prominent.</p>
</div>
<div id="in-blanks" class="section level3">
<h3 class="hasAnchor">
<a href="#in-blanks" class="anchor"></a>In blanks</h3>
<p>Applying the same exploration to blank samples allows to assess the background detection. Be it either due to technical contamination, computational artefacts (during peptide identification) or contamination from the culture environment.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lt</span> <span class="op">&lt;-</span> <span class="va">lfauto</span><span class="op">[</span><span class="va">lfauto</span><span class="op">$</span><span class="va">nbCells</span> <span class="op">==</span> <span class="fl">0</span>, <span class="op">]</span>
<span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html">fromList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">lt</span><span class="op">$</span><span class="va">Leading.razor.protein</span>, <span class="va">lt</span><span class="op">$</span><span class="va">colname</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html">upset</a></span><span class="op">(</span><span class="va">lt</span>, 
      nsets <span class="op">=</span> <span class="fl">17</span>, 
      order.by <span class="op">=</span> <span class="st">"freq"</span>, 
      nintersects <span class="op">=</span> <span class="fl">30</span>,
      keep.order <span class="op">=</span> <span class="cn">TRUE</span>,
      mainbar.y.label <span class="op">=</span> <span class="st">"Proteins in intersection"</span>, 
      sets.x.label <span class="op">=</span> <span class="st">"Proteins"</span>,
      mb.ratio <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>As expected, the most frequent protein sets are specific to one of the blank samples indicating inconsistent signal in the data. However, 49 proteins (out of 302) are consistently found across all blanks indicating background contamination that probably also occurs in the single-cell samples.</p>
</div>
</div>
</div>
<div id="quantification-levels" class="section level1">
<h1 class="hasAnchor">
<a href="#quantification-levels" class="anchor"></a>Quantification levels</h1>
<div id="intensity-distribution" class="section level2">
<h2 class="hasAnchor">
<a href="#intensity-distribution" class="anchor"></a>Intensity distribution</h2>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">lfauto</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">group</span>, scales <span class="op">=</span> <span class="st">"free_y"</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"autoPOTS"</span><span class="op">)</span> </code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
</div>
<div id="intensity-dependent-identification" class="section level2">
<h2 class="hasAnchor">
<a href="#intensity-dependent-identification" class="anchor"></a>Intensity dependent identification</h2>
<p>We expect that proteins that are expressed in all single-cells should be more expressed than proteins found in only in one or two cells or not found in single-cells. To assess this, we <strong>assume</strong> that the 500-cell samples approximate the population protein expression. We look at the median protein expression in those concentrated samples but separate the proteins based on the number of single-cells it was found in.</p>
<p>To plot this, we first need a table containing proteins found in single-cell samples and the number of single-cell samples in which each protein was found.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">lfauto</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span>,
           <span class="va">nbCells</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Leading.razor.protein</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">colname</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">scProts</span></code></pre></div>
<p>Another interesting information to plot is whether a protein is also found in a blank sample. We therefore create a table containing only information about blank samples.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">lfauto</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span>,
           <span class="va">nbCells</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">blankProts</span></code></pre></div>
<p>Finally we create a table containing only information about the 500-cell samples, that we consider as bulk samples. We compute the median expression for each proteins across all 500-cell samples. We then join the single-cell table to include the number single-cell samples in which each protein was found. Finally, we also include the information about whether each protein was found in a blank samples or not.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">lfauto</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span>,
           <span class="va">nbCells</span> <span class="op">%in%</span> <span class="fl">450</span><span class="op">:</span><span class="fl">550</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">colname</span>, <span class="va">Leading.razor.protein</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">left_join</span><span class="op">(</span><span class="va">scProts</span>, by <span class="op">=</span> <span class="st">"Leading.razor.protein"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span>, <span class="fl">0</span>, <span class="va">n</span><span class="op">)</span>,
           foundInBlank <span class="op">=</span> <span class="va">Leading.razor.protein</span> <span class="op">%in%</span> <span class="va">blankProts</span><span class="op">$</span><span class="va">Leading.razor.protein</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">df</span></code></pre></div>
<p>After this advanced data wrangling, we can plot the relationship between the median protein intensity in bulk samples and the number of single-cell samples that a protein is found in.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">value</span>,
        x <span class="op">=</span> <span class="va">n</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_jitter</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="va">foundInBlank</span><span class="op">)</span>,
                alpha <span class="op">=</span> <span class="fl">0.5</span>, width <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_violin</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">n</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"transparent"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_smooth</span><span class="op">(</span>formula <span class="op">=</span> <span class="st">"y ~ x"</span>, method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Protein expression in 500-cell samples"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ylab</span><span class="op">(</span><span class="st">"log10 median protein expression"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"found in # single-cell samples"</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>As expected from technical missingness, proteins that are found in more single-cell samples are proteins that have a higher expression distribution in bulk samples. However, the top 100 most expressed proteins in bulk samples are found between no and 3 single-cell samples, meaning that the protein detection probability not solely depends on protein expression. A disappointing observation is that a protein that is found in more single-cell samples has a higher chance to be also found in blanks samples. A great majority of the proteins found in all 3 single-cell samples were also found in blank samples, indicating that the background signal consist of highly expressed proteins that might have been released in the medium during the cell culture (blanks consist of culture media without cells) or by carry over of sample material between successive MS runs.</p>
</div>
</div>
<div id="quantification-variability" class="section level1">
<h1 class="hasAnchor">
<a href="#quantification-variability" class="anchor"></a>Quantification variability</h1>
<p>Because bottom-up proteomics reconstruct protein data from several peptide fragments, it is possible to assess the quantification accuracy for each protein by assessing the variability among its constituting peptides.</p>
<div id="number-of-peptides-per-proteins" class="section level2">
<h2 class="hasAnchor">
<a href="#number-of-peptides-per-proteins" class="anchor"></a>Number of peptides per proteins</h2>
<p>First, we have a look at the distribution of the number of peptides that are associated to a protein and assess the impact of the amount of sample material.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">lfauto</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">colname</span>, <span class="va">Leading.razor.protein</span>, <span class="va">group</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">n</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span> <span class="va">group</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">75</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"# peptides per protein"</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-21-1.png" width="768"></p>
<p>Irrespective of the amount of sample material, the mode of the distribution is one, meaning that many proteins are identified based on a single peptide or a low number of peptides. As more sample material is analysed, the distribution tails towards more peptides per proteins.</p>
</div>
<div id="median-cvs" class="section level2">
<h2 class="hasAnchor">
<a href="#median-cvs" class="anchor"></a>Median CVs</h2>
<p>The coefficient of variation (CV) is a measure of the variability of the peptide quantification for a protein. Using the <code>medianCVperCell</code> function, we can compute this for all proteins in all samples. The median CV in each sample is computed and added to the <code>colData</code>, here we name the new varaible <code>medianCV</code>. We only assess CVs for proteins with at least 5 peptides. Note that this means that the CV is computed for only for a few proteins for blank and single-cell proteins (see plot above).</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">autoPOTS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/medianCVperCell.html">medianCVperCell</a></span><span class="op">(</span><span class="va">autoPOTS</span>, 
                            i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"HeLa"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">)</span>,
                            groupBy <span class="op">=</span> <span class="st">"Leading.razor.protein"</span>,
                            nobs <span class="op">=</span> <span class="fl">5</span>, 
                            colDataName <span class="op">=</span> <span class="st">"MedianCV"</span><span class="op">)</span></code></pre></div>
<p>We plot the computed median CVs per sample against the number of cells per sample. The required data are immediately available in the <code>colData</code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>  
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">group</span>, 
        y <span class="op">=</span> <span class="va">MedianCV</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="liang2020_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>Interestingly, the sample amount does not seem to influence the quantification variability.</p>
</div>
</div>
<div id="assess-the-nature-of-the-proteins" class="section level1">
<h1 class="hasAnchor">
<a href="#assess-the-nature-of-the-proteins" class="anchor"></a>Assess the nature of the proteins</h1>
<p><strong>Sections below are work in progress</strong></p>
<p>Up to now, we saw that single-cells and</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">rbindRowData</a></span><span class="op">(</span><span class="va">autoPOTS</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"HeLa"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">universe</span><span class="op">$</span><span class="va">Leading.razor.protein</span><span class="op">)</span></code></pre></div>
<div id="in-many-cells-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#in-many-cells-samples" class="anchor"></a>In many cells samples</h2>
<div id="biological-process" class="section level3">
<h3 class="hasAnchor">
<a href="#biological-process" class="anchor"></a>Biological process</h3>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">common500</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">rbindRowData</a></span><span class="op">(</span><span class="va">autoPOTS</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"500"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">common500</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">common500</span><span class="op">$</span><span class="va">Leading.razor.protein</span><span class="op">)</span>
<span class="va">go_500</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">common500</span>,
                   OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,
                   keyType <span class="op">=</span> <span class="st">"UNIPROT"</span>,
                   ont <span class="op">=</span> <span class="st">"BP"</span>,
                   pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                   pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,
                   universe <span class="op">=</span> <span class="fu">keys</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="st">"UNIPROT"</span><span class="op">)</span>,
                   qvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                   minGSSize <span class="op">=</span> <span class="fl">50</span>,
                   maxGSSize <span class="op">=</span> <span class="fl">200</span>,
                   readable <span class="op">=</span> <span class="cn">FALSE</span>,
                   pool <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">go_500</span>, showCategory <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></code></pre></div>
</div>
<div id="cellular-compartment" class="section level3">
<h3 class="hasAnchor">
<a href="#cellular-compartment" class="anchor"></a>Cellular compartment</h3>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">go_blank_CC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">common500</span>,
                        OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,
                        keyType <span class="op">=</span> <span class="st">"UNIPROT"</span>,
                        ont <span class="op">=</span> <span class="st">"CC"</span>,
                        pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                        pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,
                        <span class="co"># universe = keys(org.Hs.eg.db, "UNIPROT"),</span>
                        universe <span class="op">=</span> <span class="va">universe</span>,
                        qvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                        minGSSize <span class="op">=</span> <span class="fl">200</span>,
                        maxGSSize <span class="op">=</span> <span class="fl">1000</span>,
                        readable <span class="op">=</span> <span class="cn">FALSE</span>,
                        pool <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">go_blank_CC</span>, showCategory <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="in-blanks-1" class="section level2">
<h2 class="hasAnchor">
<a href="#in-blanks-1" class="anchor"></a>In blanks</h2>
<p>This indicates contamination by skin cells? Let</p>
<div id="biological-process-1" class="section level3">
<h3 class="hasAnchor">
<a href="#biological-process-1" class="anchor"></a>Biological process</h3>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">common0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">rbindRowData</a></span><span class="op">(</span><span class="va">autoPOTS</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"_0cell"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">common0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">common0</span><span class="op">$</span><span class="va">Leading.razor.protein</span><span class="op">)</span>
<span class="va">go_blank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">common0</span>,
                     OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,
                     keyType <span class="op">=</span> <span class="st">"UNIPROT"</span>,
                     ont <span class="op">=</span> <span class="st">"BP"</span>,
                     pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                     pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,
                     universe <span class="op">=</span> <span class="fu">keys</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="st">"UNIPROT"</span><span class="op">)</span>,
                     qvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                     minGSSize <span class="op">=</span> <span class="fl">50</span>,
                     maxGSSize <span class="op">=</span> <span class="fl">200</span>,
                     readable <span class="op">=</span> <span class="cn">FALSE</span>,
                     pool <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">go_blank</span>, showCategory <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></code></pre></div>
</div>
<div id="cellular-component" class="section level3">
<h3 class="hasAnchor">
<a href="#cellular-component" class="anchor"></a>Cellular component</h3>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">go_blank_CC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">common0</span>,
                        OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,
                        keyType <span class="op">=</span> <span class="st">"UNIPROT"</span>,
                        ont <span class="op">=</span> <span class="st">"CC"</span>,
                        pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                        pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,
                        <span class="co"># universe = keys(org.Hs.eg.db, "UNIPROT"),</span>
                        universe <span class="op">=</span> <span class="va">universe</span>,
                        qvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                        minGSSize <span class="op">=</span> <span class="fl">200</span>,
                        maxGSSize <span class="op">=</span> <span class="fl">1000</span>,
                        readable <span class="op">=</span> <span class="cn">FALSE</span>,
                        pool <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">go_blank_CC</span>, showCategory <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="in-single-cells-1" class="section level2">
<h2 class="hasAnchor">
<a href="#in-single-cells-1" class="anchor"></a>In single-cells</h2>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">commonSC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">rbindRowData</a></span><span class="op">(</span><span class="va">autoPOTS</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"_1cell"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">autoPOTS</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">commonSC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">commonSC</span><span class="op">$</span><span class="va">Leading.razor.protein</span><span class="op">)</span>
<span class="va">go_sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">commonSC</span>,
                  OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,
                  keyType <span class="op">=</span> <span class="st">"UNIPROT"</span>,
                  ont <span class="op">=</span> <span class="st">"BP"</span>,
                  pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                  pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,
                  universe <span class="op">=</span> <span class="va">universe</span>,
                  qvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,
                  minGSSize <span class="op">=</span> <span class="fl">50</span>,
                  maxGSSize <span class="op">=</span> <span class="fl">200</span>,
                  readable <span class="op">=</span> <span class="cn">FALSE</span>,
                  pool <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html">dotplot</a></span><span class="op">(</span><span class="va">go_sc</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>This vignette reproduces the observations provided for Figure S5 in <span class="citation">Liang et al. (2020)</span>. However, we have pushed the data exploration further and showed that looking at the number of detected features is not sufficient to demonstrate the quality of the data. We here argue that looking at the overlap of identified protein within and between experimental groups (number of cells per sample) provide a better understanding of the data structure, namely an increase of the protein detection variability when the number of cells decreases.</p>
<p>The increase in detection variability is the consequence of technical and biological processes. The development of synthetic ground truth data sets are required to decouple the two processes and assess the importance of one process with respect to the other. This will allow to highlight the room for improvement for new MS-SCP technologies.</p>
<p>TODO add paragraph: single-cell signal ~ blank signal -&gt; list arguments =&gt; is autoPOTS really working for sc samples?</p>
<p>Finally, this vignette also provides a demonstration that using the <code>scp</code> data framework is ideal f or the manipulation and visualization of MS-SCP data. The <code>longFormat</code> output that is used throughout this vignette can be plugged in to powerful visualization tools such as <code>ggplot2</code> or <code>upsetR</code>. This allowed an in-depth exploration of the data and hence we could highlight some aspect of the data that were overlook in the original paper.</p>
</div>
<div id="requirements" class="section level1">
<h1 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h1>
<div id="hardware-and-software" class="section level3">
<h3 class="hasAnchor">
<a href="#hardware-and-software" class="anchor"></a>Hardware and software</h3>
<p>The system details of the machine that built the vignette are:</p>
</div>
<div id="timing" class="section level3">
<h3 class="hasAnchor">
<a href="#timing" class="anchor"></a>Timing</h3>
<p>The total time required to compile this vignette is:</p>
</div>
<div id="memory" class="section level3">
<h3 class="hasAnchor">
<a href="#memory" class="anchor"></a>Memory</h3>
<p>The final <code>scp</code> memory size is:</p>
</div>
<div id="session-info" class="section level3">
<h3 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h3>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Liang2020-cr">
<p>Liang, Yiran, Hayden Acor, Michaela A McCown, Andikan J Nwosu, Hannah Boekweg, Nathaniel B Axtell, Thy Truong, Yongzheng Cong, Samuel H Payne, and Ryan T Kelly. 2020. â€œFully Automated Sample Processing and Analysis Workflow for Low-Input Proteome Profiling.â€ <em>Anal. Chem.</em>, December.</p>
</div>
<div id="ref-Zhu2018-bf">
<p>Zhu, Ying, Paul D Piehowski, Rui Zhao, Jing Chen, Yufeng Shen, Ronald J Moore, Anil K Shukla, et al. 2018. â€œNanodroplet Processing Platform for Deep and Quantitative Proteome Profiling of 10-100 Mammalian Cells.â€ <em>Nat. Commun.</em> 9 (1): 882.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
