<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reproduction of the AML model analysis (Schoof et al. 2021) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Reproduction of the AML model analysis (Schoof et al. 2021)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Reproduction of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="schoof2021_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Reproduction of the AML model analysis (Schoof et al. 2021)</h1>
                        <h4 class="author">Christophe Vanderaa, Computational Biology, UCLouvain</h4>
                        <h4 class="author">Laurent Gatto, Computational Biology, UCLouvain</h4>
            
      
      
      <div class="hidden name"><code>schoof2021.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The acute myeloid leukemia (AML) dataset acquired by <span class="citation">Schoof et al. (2021)</span> is an excellent dataset to validate the application of mass spectrometry (MS)-based single-cell proteomics (SCP). This model is well characterized with expected cell types that exhibit known biological heterogeneity for some markers (namely CD34 and CD38). This dataset is also an ideal playground for method development and benchmarking.</p>
<p>The research question here is whether SCP is able to retrieve the different population that occur in the AML model. In this culture, we expect to find a small population of leukemic stem cells (LSC) that are CD34+ and CD38-. This population has the ability of self renewal or to differentiate into progenitor cells (PROG) that loose the self renewal ability and start to express CD38+. The PROG progressively differentiate into blast cells (BLAST) that loose the expression of CD38. The CD34/CD38 expression profiles in this model is already well characterized thanks to flow cytometry. A great advantage of the data set provided in <span class="citation">Schoof et al. (2021)</span> is that the flow cytometry data is also available. This allows for a characterization of the cell types before starting the data analysis on the SCP data.</p>
<p>The authors performed the analysis using Python and made the effort to provide the code needed to reproduce the results from the article. Furthermore, they developed a library called <code>SCeptre</code> that extends the <code>scanpy</code> library, a python library to process scRNA-Seq data.</p>
<p>In this vignette, we will replicate one of the analysis of the so called <code>bulk</code> dataset, but here using the R programming language. The <code>bulk</code> dataset contains single cells isolated from a bulk cell culture without further selection or cell type purification. This is a very interesting dataset since it emulates real-life research applications of SCP.</p>
<div id="scp-workflow-and-data-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#scp-workflow-and-data-structure" class="anchor"></a><code>scp</code> workflow and data structure</h3>
<p>The data processing workflow we carry out in this vignette is converted or takes the <code>R</code> equivalent of the <code>Python</code> scripts and libraries provided in the <a href="https://github.com/bfurtwa/SCeptre">SCeptre GitHub repository</a>. This repository comes with multiple notebooks, but we here reproduce the <code>bulk.ipynb</code> notebook.</p>
<p>We have recently developed a data framework to perform SCP data analysis in R (<span class="citation">Vanderaa and Gatto (2021)</span>) and built the <code>scp</code> package from this framework. The data framework combines two existing Bioconductor classes. The <code>SingleCellExperiment</code> class provides an interface to many cutting edge methods for single-cell analysis and the <code>QFeatures</code> class facilitates manipulation and processing of MS-based quantitative data. The <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/scp/inst/doc/scp.html"><code>scp</code> vignette</a> provides detailed information about the data structure. The <code>scp</code> package extends the functionality of <code>QFeatures</code> for single-cell application.</p>
<p>Below is a general overview of the workflow of the <code>bulk</code> data analysis.</p>
<div class="figure" style="text-align: center">
<img src="figs/SCeptre-workflow.png" alt="Overview of the SCoPE2 workflow." width="100%"><p class="caption">
Overview of the SCoPE2 workflow.
</p>
</div>
<p>We coloured each step depending on the underlying package that performs that step. The blue steps rely on <code>QFeatures</code>, the purple steps rely on the <code>SingleCellExperiment</code> infrastructure, the green steps rely on Python code from SCeptre (called from R) and the orange step relies on <code>scp</code> functionality. As you can see, only few steps are specific to SCP (orange and green).</p>
<p>The required packages for running this vignette are listed below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Core packages of this workflow</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/RforMassSpectrometry/QFeatures">"QFeatures"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"scpdata"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://UCLouvain-CBIO.github.io/scp">"scp"</a></span><span class="op">)</span>
<span class="co">## Utility packages for data manipulation and visualization</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://tidyverse.tidyverse.org">"tidyverse"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com">"patchwork"</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="scpdata-and-the-aml-differentiation-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#scpdata-and-the-aml-differentiation-dataset" class="anchor"></a><code>scpdata</code> and the AML differentiation dataset</h3>
<p>We also implemented a data package called <code>scpdata</code>. It distributes published MS-SCP datasets, such as the dataset that will be analysed in this vignette. The data were downloaded from the data source provided in the publication and formatted to a <code>QFeatures</code> object so that it is compatible with our software. The underlying data storage is based on the <code>ExperimentHub</code> package that provides a cloud-based storage infrastructure.</p>
<p>The dataset we uploaded to <code>scpdata</code> is the so called “bulk” dataset that contains 3072 single-cells. The authors refer to it as the bulk dataset because the cell were not sorted by flow cytometry but rather processed using an unbiased cell isolation. The dataset was retrieved from the PRIDE repository (accession ID: PXD020586). It contains a zip file with all the quantitative data and the sample annotations. The dataset contains two types of quantitative data: quantified peptide to spectrum match (PSM) data and protein data. They were generated after running ProteomeDiscoverer on the raw MS files. The quantification data are given as signal to noise ratios (S/N). See <code>?schoof2021</code> for more info about data acquisition and data formatting.</p>
<p>The formatted data can be retrieved from the <code>scpdata</code> package using the <code>schoof2021()</code> function.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu">schoof2021</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The dataset contains PSM and protein data generated by Proteome Discoverer. It also contains the protein data after processing by the authors using the SCeptre workflow. We extract the processed protein data in a separate variable that we will use later to benchmark the replication of the data processing using <code>scp</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce_sceptre</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"logNormProteins"</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span>, , <span class="fl">1</span><span class="op">:</span><span class="fl">193</span><span class="op">]</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></code></pre></div>
<p>The data now contain 194 different <code>SingleCellExperiment</code> objects that we refer to as <strong>assays</strong>. Each assay contains expression data along with feature annotations. Each row in an assay represents a <strong>feature</strong> that can either be a PSM or a protein depending on the assay. Each column in an assay represents a <strong>sample</strong>. Below, we show the overview of the <code>scp</code> dataset.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 193 assays:</span>
<span class="co">##  [1] F1: SingleCellExperiment with 4455 rows and 16 columns </span>
<span class="co">##  [2] F10: SingleCellExperiment with 4604 rows and 16 columns </span>
<span class="co">##  [3] F100: SingleCellExperiment with 5056 rows and 16 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [191] F98: SingleCellExperiment with 4025 rows and 16 columns </span>
<span class="co">##  [192] F99: SingleCellExperiment with 3898 rows and 16 columns </span>
<span class="co">##  [193] proteins: SingleCellExperiment with 2898 rows and 3072 columns</span></code></pre></div>
<p>The 192 first assays contain the PSM quantifications for each run. The samples were acquired using a TMT-16 labelling protocol meaning that every run contains 16 samples and therefore every assay contains 16 columns. The <code>proteins</code> assay contains the protein quantifications.</p>
<p>The sample annotation can be found in the <code>colData</code> of the dataset. The most informative annotation fields are:</p>
<ul>
<li>
<code>File.ID</code>: the data were acquired in different batches, each batch being associated to a unique file ID.</li>
<li>
<code>Channel</code>: the TMT used to label the sample</li>
<li>
<code>Population</code>: the cell type defined by the flow cytometry data</li>
<li>
<code>SampleType</code>: samples are either boosters (200-cell equivalents), empty wells, normalization channels (10-cell equivalents), or single</li>
<li>
<code>FSC.A</code>, <code>FSC.H</code>, …, <code>APC.Cy7.A</code>, <code>PE.A</code>: the flow cytometry data acquired during the single-cell isolation.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span>
<span class="co">## DataFrame with 6 rows and 32 columns</span>
<span class="co">##         File.ID  Channel Input.Files.Workflow.ID Input.Files.Workflow.Level</span>
<span class="co">##        &lt;factor&gt; &lt;factor&gt;               &lt;integer&gt;                  &lt;integer&gt;</span>
<span class="co">## F1126        F1     126                    -3327                          1</span>
<span class="co">## F1127N       F1     127N                   -3327                          1</span>
<span class="co">## F1127C       F1     127C                   -3327                          1</span>
<span class="co">## F1128N       F1     128N                   -3327                          1</span>
<span class="co">## F1128C       F1     128C                   -3327                          1</span>
<span class="co">## F1129N       F1     129N                   -3327                          1</span>
<span class="co">##        Input.Files.     File.Name Creation.Date RT.Range.in.min Instrument.Name</span>
<span class="co">##           &lt;integer&gt;   &lt;character&gt;   &lt;character&gt;     &lt;character&gt;     &lt;character&gt;</span>
<span class="co">## F1126          3573 20200829_M... 08/28/2020...   0.07 - 160...   Orbitrap E...</span>
<span class="co">## F1127N         3573 20200829_M... 08/28/2020...   0.07 - 160...   Orbitrap E...</span>
<span class="co">## F1127C         3573 20200829_M... 08/28/2020...   0.07 - 160...   Orbitrap E...</span>
<span class="co">## F1128N         3573 20200829_M... 08/28/2020...   0.07 - 160...   Orbitrap E...</span>
<span class="co">## F1128C         3573 20200829_M... 08/28/2020...   0.07 - 160...   Orbitrap E...</span>
<span class="co">## F1129N         3573 20200829_M... 08/28/2020...   0.07 - 160...   Orbitrap E...</span>
<span class="co">##        Software.Revision       Plate    Sample        Well     Facs.Data</span>
<span class="co">##              &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;character&gt;   &lt;character&gt;</span>
<span class="co">## F1126      2.0-2.0.12...    bulk_b_1         4          NA            NA</span>
<span class="co">## F1127N     2.0-2.0.12...    bulk_b_1         4          NA            NA</span>
<span class="co">## F1127C     2.0-2.0.12...    bulk_b_1         4          NA            NA</span>
<span class="co">## F1128N     2.0-2.0.12...    bulk_b_1         4          D4 facs_data_...</span>
<span class="co">## F1128C     2.0-2.0.12...    bulk_b_1         4          D7 facs_data_...</span>
<span class="co">## F1129N     2.0-2.0.12...    bulk_b_1         4          D5 facs_data_...</span>
<span class="co">##                Row      Col  Label.Layout   Sort.Layout Sample.Layout</span>
<span class="co">##        &lt;character&gt; &lt;factor&gt;   &lt;character&gt;   &lt;character&gt;   &lt;character&gt;</span>
<span class="co">## F1126           NA       NA            NA            NA            NA</span>
<span class="co">## F1127N          NA       NA            NA            NA            NA</span>
<span class="co">## F1127C          NA       NA            NA            NA            NA</span>
<span class="co">## F1128N           D       4  label_layo... sort_layou... sample_lay...</span>
<span class="co">## F1128C           D       7  label_layo... sort_layou... sample_lay...</span>
<span class="co">## F1129N           D       5  label_layo... sort_layou... sample_lay...</span>
<span class="co">##        SortedPopulation     FSC.A     FSC.H     FSC.W     SSC.A     SSC.H</span>
<span class="co">##             &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">## F1126                NA        NA        NA        NA        NA        NA</span>
<span class="co">## F1127N               NA        NA        NA        NA        NA        NA</span>
<span class="co">## F1127C               NA        NA        NA        NA        NA        NA</span>
<span class="co">## F1128N             BULK       247       144       439       159       100</span>
<span class="co">## F1128C             BULK       249       138       463       199       125</span>
<span class="co">## F1129N             BULK       242       139       445       199       124</span>
<span class="co">##            SSC.W CD34_APC.Cy7.A CD38_PE.A      Time    Population    SampleType</span>
<span class="co">##        &lt;integer&gt;      &lt;integer&gt; &lt;integer&gt; &lt;numeric&gt;   &lt;character&gt;   &lt;character&gt;</span>
<span class="co">## F1126         NA             NA        NA        NA            NA booster_bu...</span>
<span class="co">## F1127N        NA             NA        NA        NA            NA          norm</span>
<span class="co">## F1127C        NA             NA        NA        NA            NA         empty</span>
<span class="co">## F1128N       407            233       415   2.47494 BLAST CD38...            sc</span>
<span class="co">## F1128C       406            186       439   2.34823 BLAST CD38...            sc</span>
<span class="co">## F1129N       410            289       398   2.42698 BLAST CD38...            sc</span>
<span class="co">##          nbCells</span>
<span class="co">##        &lt;numeric&gt;</span>
<span class="co">## F1126          0</span>
<span class="co">## F1127N        10</span>
<span class="co">## F1127C         0</span>
<span class="co">## F1128N         1</span>
<span class="co">## F1128C         1</span>
<span class="co">## F1129N         1</span></code></pre></div>
<p>Throughout the vignette, we will use <code>ggplot2</code> from the <code>tidyverse</code> project (<span class="citation">Wickham et al. (2019)</span>) to visualize the data. As an example, we plot how the different cell types are distributed across batches.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">Population</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">Population</span>, 
        x <span class="op">=</span> <span class="va">File.ID</span>,
        colour <span class="op">=</span> <span class="va">Population</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_jitter</span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
          legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>As denoted in the article, this data set shows unbalance between the different cell types. This reflects biological reality as some cell types, such as blast cells, are more prevalent than others, such as progenitor cells. An important point to highlight is that all MS batches do not contain PROG or LSC, which complicates the data integration across batches.</p>
</div>
<div id="useful-links" class="section level2">
<h2 class="hasAnchor">
<a href="#useful-links" class="anchor"></a>Useful links</h2>
<ul>
<li>Link to the <a href="http://dx.doi.org/10.1038/s41467-021-23667-y">paper</a>
</li>
<li>Link to the <a href="https://www.biorxiv.org/content/10.1101/745679v2">preprint</a>
</li>
<li>Link to the <code>sceptre</code> <a href="https://github.com/bfurtwa/SCeptre/blob/master/sceptre/sceptre.py">source code</a>
</li>
<li>Link to the SCeptre <a href="https://github.com/bfurtwa/SCeptre/tree/master/Schoof_et_al/code">notebooks</a>, our report replicates the <code>bulk.ipynb</code> notebook</li>
</ul>
</div>
</div>
<div id="ms-run-qc" class="section level1">
<h1 class="hasAnchor">
<a href="#ms-run-qc" class="anchor"></a>MS run QC</h1>
<p>The SCeptre workflow starts with a MS run quality control (QC). In this section, we will check for each run the distribution of the number of PSMs, the average S/N distribution, the channel intensity distribution and the isolation interference distribution.</p>
<div id="psms-per-run" class="section level2">
<h2 class="hasAnchor">
<a href="#psms-per-run" class="anchor"></a>PSMs per run</h2>
<p>The number of features and samples per assay in a <code>QFeatures</code> object is retrieved using <code>dims</code>. After removing the last assay (<code>proteins</code>), we plot the histogram with the number of features for the remaining assays (<em>i.e.</em> PSMs).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">npsms</span> <span class="op">&lt;-</span> <span class="fu">dims</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">193</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>npsms <span class="op">=</span> <span class="va">npsms</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">npsms</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-4-1.png" width="384"></p>
<p>The plot indicates that a few runs failed with low number of PSMs.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">npsms</span><span class="op">)</span><span class="op">[</span><span class="va">npsms</span> <span class="op">&lt;</span> <span class="fl">3000</span><span class="op">]</span>
<span class="co">## [1] "F117" "F44"  "F5"   "F58"</span></code></pre></div>
<p>4 assays have less than 3000 PSMs and were flagged by the authors as failedd MS runs.</p>
</div>
<div id="intensity-distributions" class="section level2">
<h2 class="hasAnchor">
<a href="#intensity-distributions" class="anchor"></a>Intensity distributions</h2>
<p>We expect similar average reporter signal to noise (S/N) distributions across MS runs. The averages are already computed for each PSM by ProteomeDiscover and correspond to the reporter intensities averaged over the 16 TMT channels divided by the noise estimated from the corresponding MS spectrum. The higher the S/N, the higher the signal quality.</p>
<p>The average S/N is available from the <code>rowData</code> of the PSM assays. We here collect and combine the <code>rowData</code> from different assays using the <code>rbindRowData</code> function that will return a single <code>DataFrame</code> table with all the feature annotation columns common to all assays. Note that we don’t plot all assays to avoid a crowded plot. We selected 4 representative and successful MS runs and the 4 MS runs that have failed (we store this information in the table).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## A few selected runs</span>
<span class="va">succeeded</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"F26"</span>, <span class="st">"F72"</span>, <span class="st">"F108"</span>, <span class="st">"F133"</span><span class="op">)</span>
<span class="va">failed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"F5"</span>, <span class="st">"F44"</span>, <span class="st">"F58"</span>, <span class="st">"F117"</span><span class="op">)</span>
<span class="co">## Collect the PSM annotations (including the average S/N)</span>
<span class="va">rd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">rbindRowData</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">succeeded</span>, <span class="va">failed</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Add whether a run is flagged as failed or succeeded</span>
<span class="va">rd</span><span class="op">$</span><span class="va">failed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">rd</span><span class="op">$</span><span class="va">assay</span> <span class="op">%in%</span> <span class="va">failed</span>, <span class="st">"failed"</span>, <span class="st">"succeeded"</span><span class="op">)</span></code></pre></div>
<p>We plot the average S/N (stored under <code>Average.Reporter.SN</code>) as separate boxplots for each MS runs. The figure is split in 2 panels depending on whether the MS runs succeeded of failed.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">rd</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">Average.Reporter.SN</span>,
        x <span class="op">=</span> <span class="va">assay</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span> <span class="va">failed</span>, scales <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span>
<span class="co">## Warning: Removed 7 rows containing non-finite values (stat_boxplot).</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-7-1.png" width="576"></p>
<p>There are no clear differences between succeeded and failed runs.</p>
</div>
<div id="isolation-interference" class="section level2">
<h2 class="hasAnchor">
<a href="#isolation-interference" class="anchor"></a>Isolation interference</h2>
<p>Similarly to the average S/N, we can also explore the isolation interference that gives the percentage of signal that is attributed to co-isolated peptides. We plot the percent isolation interference (stored under <code>Isolation.Interference.in.Percent</code>) using the same layout as above.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">rd</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">Isolation.Interference.in.Percent</span>,
        x <span class="op">=</span> <span class="va">assay</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span> <span class="va">failed</span>, scales <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-8-1.png" width="576"></p>
<p>There are no differences between succeeded and failed runs. However, the isolation interference can be used to filter low-quality PSMs (not performed in this workflow).</p>
</div>
<div id="median-sn-per-channel" class="section level2">
<h2 class="hasAnchor">
<a href="#median-sn-per-channel" class="anchor"></a>Median S/N per channel</h2>
<p>The authors next explored the median S/N per channel for all MS runs. This time, the median S/N has to be computed. First, we focus on the 4 succeeded and the 4 failed assays that were selected above. Note that a <code>QFeatures</code> object can be seen as a three-order array: <span class="math inline">\(features \times samples \times assay\)</span>. Hence, <code>QFeatures</code> supports three-order subsetting <code>x[rows, columns, assays]</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">sub</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span>, , <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">succeeded</span>, <span class="va">failed</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span>
<span class="co">## An instance of class QFeatures containing 8 assays:</span>
<span class="co">##  [1] F26: SingleCellExperiment with 4636 rows and 16 columns </span>
<span class="co">##  [2] F72: SingleCellExperiment with 4492 rows and 16 columns </span>
<span class="co">##  [3] F108: SingleCellExperiment with 4695 rows and 16 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [6] F44: SingleCellExperiment with 167 rows and 16 columns </span>
<span class="co">##  [7] F58: SingleCellExperiment with 2727 rows and 16 columns </span>
<span class="co">##  [8] F117: SingleCellExperiment with 202 rows and 16 columns</span></code></pre></div>
<p>Then, we retrieve the S/N quantifications along with the sample annotations (<code>Channel</code> and <code>Population</code>) using the <code>longFormat</code> function. This function will create a long table from the <code>QFeatures</code> object where each row corresponds to a single quantification.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat</a></span><span class="op">(</span><span class="va">sub</span>, colvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Channel"</span>, <span class="st">"Population"</span><span class="op">)</span><span class="op">)</span>
<span class="va">lf</span><span class="op">$</span><span class="va">failed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">lf</span><span class="op">$</span><span class="va">assay</span> <span class="op">%in%</span> <span class="va">failed</span>, <span class="st">"failed"</span>, <span class="st">"succeeded"</span><span class="op">)</span>
<span class="va">lf</span>
<span class="co">## DataFrame with 348432 rows and 8 columns</span>
<span class="co">##              assay     primary     rowname     colname     value  Channel</span>
<span class="co">##        &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">## 1              F26      F26126    PSM20316      F26126    1288.1      126</span>
<span class="co">## 2              F26      F26126    PSM20324      F26126     358.9      126</span>
<span class="co">## 3              F26      F26126    PSM20326      F26126     458.6      126</span>
<span class="co">## 4              F26      F26126    PSM20327      F26126    1475.4      126</span>
<span class="co">## 5              F26      F26126    PSM20328      F26126     878.4      126</span>
<span class="co">## ...            ...         ...         ...         ...       ...      ...</span>
<span class="co">## 348428        F117    F117134N   PSM811358    F117134N        NA     134N</span>
<span class="co">## 348429        F117    F117134N   PSM815013    F117134N       2.4     134N</span>
<span class="co">## 348430        F117    F117134N   PSM816804    F117134N        NA     134N</span>
<span class="co">## 348431        F117    F117134N   PSM819328    F117134N        NA     134N</span>
<span class="co">## 348432        F117    F117134N   PSM820788    F117134N       5.4     134N</span>
<span class="co">##           Population      failed</span>
<span class="co">##          &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1                 NA   succeeded</span>
<span class="co">## 2                 NA   succeeded</span>
<span class="co">## 3                 NA   succeeded</span>
<span class="co">## 4                 NA   succeeded</span>
<span class="co">## 5                 NA   succeeded</span>
<span class="co">## ...              ...         ...</span>
<span class="co">## 348428 BLAST CD38...      failed</span>
<span class="co">## 348429 BLAST CD38...      failed</span>
<span class="co">## 348430 BLAST CD38...      failed</span>
<span class="co">## 348431 BLAST CD38...      failed</span>
<span class="co">## 348432 BLAST CD38...      failed</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># channelOrder &lt;- c("126", "127N", "127C", "128N", "128C", "129N", "129C",</span>
<span class="co">#                   "130N", "130C", "131N", "131C", "132N", "132C", "133N",</span>
<span class="co">#                   "133C", "134N")</span>
<span class="co"># lf$Channel &lt;- factor(lf$Channel, levels = channelOrder)</span></code></pre></div>
<p>We then need to compute the median S/N grouped by assay and channel. This performed using <code>dplyr</code> functionality.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lf</span> <span class="op">&lt;-</span> <span class="fu">group_by</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">lf</span><span class="op">)</span>,
               <span class="va">assay</span>, <span class="va">Channel</span>, <span class="va">failed</span>, <span class="va">Population</span><span class="op">)</span>
<span class="va">lfsumm</span> <span class="op">&lt;-</span> <span class="fu">summarize</span><span class="op">(</span><span class="va">lf</span>, medianSN <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">value</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We can now plot the median S/N for each channel in each assay. We also add a colouring scheme depending on the type of cell that was acquired per channel. Note the y-axis is displayed on a log10 scale.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">lfsumm</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Channel</span>, 
        y <span class="op">=</span> <span class="va">medianSN</span>, 
        fill <span class="op">=</span> <span class="va">Population</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">failed</span> <span class="op">+</span> <span class="va">assay</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
          legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-13-1.png" width="576"></p>
<p>Again, there are no difference in the median intensity distribution over the different channels between failed and succeeded runs and so this is not a good proxy for MS run quality. However, it is a good way to identify technical biases. For instance, and as expected, the <code>126</code> channel consistently shows higher S/N because this channel was allocated the booster sample containing more cells. We can also see that the cell type does not seem to influence the median S/N.</p>
</div>
</div>
<div id="filter-data" class="section level1">
<h1 class="hasAnchor">
<a href="#filter-data" class="anchor"></a>Filter data</h1>
<p>In this section, we proceed the processing by removing the contaminant proteins, remove the failed runs and removes the channels that are not single-cell samples.</p>
<div id="remove-contaminants" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-contaminants" class="anchor"></a>Remove contaminants</h2>
<p>Contaminant proteins were flagged when create the <code>QFeatures</code> object in the <code>scpdata</code> package. This was performed by matching the protein accession from the dataset from a list of known contaminants (provided by the authors). Whether a protein is a contaminant or not is contained in the feature annotation as the <code>isContaminant</code> field. We can directly filter this variable using the <code>filterFeatures</code> function that will retrieve the required data from the <code>rowData</code> of each assay.</p>
<p>We keep only features that are not matched to a contaminant protein.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>, <span class="op">~</span> <span class="op">!</span><span class="va">isContaminant</span><span class="op">)</span></code></pre></div>
</div>
<div id="remove-failed-runs" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-failed-runs" class="anchor"></a>Remove failed runs</h2>
<p>Next, we remove the failed runs that we flagged using the second index subsetting to remove the failed samples from the PSM as well as the protein data.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">scp</span>, <span class="op">!</span><span class="va">scp</span><span class="op">$</span><span class="va">File.ID</span> <span class="op">%in%</span> <span class="va">failed</span><span class="op">)</span>
<span class="co">## Takes 90 sec, but could probably be faster -&gt; open issue</span></code></pre></div>
</div>
<div id="keep-only-single-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#keep-only-single-cells" class="anchor"></a>Keep only single cells</h2>
<p>Finally, we keep only single-cell samples. The MS acquisitions contain different sample types. There is the booster channel that was mentioned above containing the equivalent of 200 cells. The 127N channel contains a reference sample with 10-cell equivalents and the 127C channel is left empty to due to cross-contamination of the 126 booster channel. Finally, a few negative controls (samples prepared as single-cell samples but lack a cell) were randomly included in the different MS runs.</p>
<p>Before performing the filter, we show that negative controls can be discriminated from single-cells based on the number of detected proteins and the summed quantifications in each sample. This is a reassuring indicator of data quality. To perform this, we run sample quality control using the <code>scuttle</code> package that offers utility functions for scRNA-Seq analysis. Those function can be directly applied to our assays since they are stored as <code>SingleCellExperiment</code> objects. We therefore extract the protein assay along with the sample annotation using the <code>getWithColData</code>. Then, we replace missing values by zero as expected by the <code>scuttle</code> package.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins"</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/impute.html">impute</a></span><span class="op">(</span><span class="va">sce</span>, method <span class="op">=</span> <span class="st">"zero"</span><span class="op">)</span></code></pre></div>
<p>We then apply the <code>perCellQCMetrics</code> function that computes the number of detected features and the summed abundance per sample.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"scuttle"</span><span class="op">)</span>
<span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span>, assay.type <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>We then combine the computed QC with the sample annotation and plot the sample QC data.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">qc</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">qc</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">qc</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">sum</span><span class="op">)</span>,
        y <span class="op">=</span> <span class="va">detected</span>,
        col <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.9</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>First, we can clearly see that booster and reference channels show higher total signal. This is expected since those channels contains more material than the single-cell samples. However, those channels cannot be discriminated based on the number of quantified proteins. Next, the <code>emtpy</code> empty channels (no TMT label added) are clearly separated from the single-cell channels both based on the number of quantified proteins and the summed signal. Finally, the negative control samples overlap with the lower tail of the single-cell samples both for the number of proteins and the summed signal. The overlapping single-cells should be removed, but we will see in a later section that those cells are removed during cell QC.</p>
<p>We now remove all samples but the single-cell samples.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">scp</span>, <span class="va">scp</span><span class="op">$</span><span class="va">SampleType</span> <span class="op">==</span> <span class="st">"sc"</span><span class="op">)</span>
<span class="va">scp</span><span class="op">@</span><span class="va">assayLinks</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">@</span><span class="va">assayLinks</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">]</span></code></pre></div>
</div>
</div>
<div id="batch-correction" class="section level1">
<h1 class="hasAnchor">
<a href="#batch-correction" class="anchor"></a>Batch correction</h1>
<p>The dataset needs to be processed before running downstream analyses. The first step the authors take is normalizing the data. To perform this, they implemented a new normalization method and included it in their <code>sceptre</code> Python module. Their normalization method aims to remove channel and MS run biases. In our opinion, this step should be referred to as batch correction rather than normalization.</p>
<p>We extract the assay to batch correct using <code>getWithColData</code>. The batch correction method in <code>sceptre</code> expects that the sample annotation <code>File ID</code> is present. We therefore adapt the corresponding column name as requested.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins"</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"File ID"</span>
<span class="va">sce</span>
<span class="co">## class: SingleCellExperiment </span>
<span class="co">## dim: 2870 2396 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(1): ''</span>
<span class="co">## rownames(2870): MYH9 MPO ... METTL3 MRPL48</span>
<span class="co">## rowData names(34): ProteinsUnique.Sequence.ID Checked ... rowname</span>
<span class="co">##   isContaminant</span>
<span class="co">## colnames(2396): F1128N F1128C ... F99133C F99134N</span>
<span class="co">## colData names(32): File ID Channel ... SampleType nbCells</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span></code></pre></div>
<p>We are now faced with an issue. The algorithm we want to apply is only available in Python whereas this script is written in R. Luckily, we can make use of the <code>reticulate</code> R package (<span class="citation">Ushey, Allaire, and Tang (2021)</span>) that provides an interface between an R and Python code. We setup the python environment that contains the required modules to run <code>scpetre</code> (for more info, see the <em>Requirements</em> section at the end of the vignette).</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rstudio/reticulate">"reticulate"</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_condaenv</a></span><span class="op">(</span><span class="st">"sceptre"</span><span class="op">)</span>
<span class="va">spt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/import.html">import</a></span><span class="op">(</span><span class="st">"sceptre"</span><span class="op">)</span>
<span class="co">## Warning: Python '/home/christophe/miniconda3/envs/sceptre/bin/python' was</span>
<span class="co">## requested but '/home/christophe/PhD/SCP.replication/inst/renvs/schoof2021/</span>
<span class="co">## renv/python/virtualenvs/renv-python-3.8/bin/python' was loaded instead (see</span>
<span class="co">## reticulate::py_config() for more information)</span></code></pre></div>
<p>Now that we can interface R with Python, we need to convert the <code>SingleCellExperiment</code> object (R) into an <code>AnnData</code> object (Python). This can easily be done using the <code>SCE2AnnData</code> from the <code>zellkonverter</code> package.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/theislab/zellkonverter">"zellkonverter"</a></span><span class="op">)</span>
<span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/AnnData-Conversion.html">SCE2AnnData</a></span><span class="op">(</span><span class="va">sce</span>, X_name <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>We can now apply the <code>normalize</code> function from <code>sceptre</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spt</span><span class="op">$</span><span class="fu">normalize</span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></code></pre></div>
<p>The <code>AnnData</code> object stored as <code>adata</code> now contains the normalized data. We can reinsert the object as a new assay. First, we convert the data back to a <code>SingleCellExperiment</code> object using <code>AnnData2SCE</code>. Then, we add the normalized assay into the <code>QFeatures</code> object using <code>addAssay</code>. Finally, <code>addAssayLink</code> will retrieve the relationship between the features of the old and new assay based on the protein accession.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/AnnData-Conversion.html">AnnData2SCE</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">addAssay</a></span><span class="op">(</span><span class="va">scp</span>, <span class="va">sce</span>, name <span class="op">=</span> <span class="st">"proteins_norm"</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html">addAssayLink</a></span><span class="op">(</span><span class="va">scp</span>, from <span class="op">=</span> <span class="st">"proteins"</span>, to <span class="op">=</span> <span class="st">"proteins_norm"</span>,
                    varFrom <span class="op">=</span> <span class="st">"Accession"</span>, varTo <span class="op">=</span> <span class="st">"Accession"</span><span class="op">)</span>
<span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 190 assays:</span>
<span class="co">##  [1] F1: SingleCellExperiment with 4425 rows and 13 columns </span>
<span class="co">##  [2] F10: SingleCellExperiment with 4562 rows and 13 columns </span>
<span class="co">##  [3] F100: SingleCellExperiment with 4997 rows and 13 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [188] F99: SingleCellExperiment with 3817 rows and 13 columns </span>
<span class="co">##  [189] proteins: SingleCellExperiment with 2870 rows and 2396 columns </span>
<span class="co">##  [190] proteins_norm: SingleCellExperiment with 2778 rows and 2396 columns</span></code></pre></div>
</div>
<div id="cell-qc" class="section level1">
<h1 class="hasAnchor">
<a href="#cell-qc" class="anchor"></a>Cell QC</h1>
<p>In this section, we proceed with the single-cell QC. Failed and low- quality cells must be removed to avoid biasing the downstream analyses.</p>
<div id="compute-qc" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-qc" class="anchor"></a>Compute QC</h2>
<p>We already showed how to perform cell QC using <code>perCellQCMetrics</code> in a previous section and showed that booster, reference, empty and single-cell samples can easily be discriminated. This time, we will apply the same procedure on the normalized data and show how to remove low-quality cells.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html">perCellQCMetrics</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_norm"</span><span class="op">]</span><span class="op">]</span>, assay.type <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Once the QC metrics are computed, we need to define thresholds. In the SCeptre workflow, this is performed using the median absolute deviation (MAD). We can use the <code>isOutlier</code> function that computes the MADs and return whether a cell is an outlier (low-quality cell) or not. This information is immediately stored in the QC table and we extract the thresholds for later plotting.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qc</span><span class="op">$</span><span class="va">outlier</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html">isOutlier</a></span><span class="op">(</span><span class="va">qc</span><span class="op">$</span><span class="va">sum</span>, nmads <span class="op">=</span> <span class="fl">2</span>, log <span class="op">=</span> <span class="cn">TRUE</span>, type <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span>
<span class="va">mads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">qc</span><span class="op">$</span><span class="va">outlier</span>,<span class="st">"thresholds"</span><span class="op">)</span></code></pre></div>
<p>The computed QCs are added to the <code>colData</code> of the <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">qc</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">qc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></code></pre></div>
</div>
<div id="note-about-qc-in-sceptre" class="section level2">
<h2 class="hasAnchor">
<a href="#note-about-qc-in-sceptre" class="anchor"></a>Note about QC in SCeptre</h2>
<p>The QC on single-cell is generally performed before batch correction (here called normalization in <code>scpetre</code>) since the cell quality may vary depending on the batch and hence bias the correction process. A better approach is to compute the cell QC for each batch separately. This is easily performed with <code>isOutlier</code> by providing the <code>batch</code> argument.</p>
</div>
<div id="plot-qc" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-qc" class="anchor"></a>Plot QC</h2>
<p>We plot the cell QC metrics and colour by channel.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">sum</span><span class="op">)</span>,
        y <span class="op">=</span> <span class="va">detected</span>,
        col <span class="op">=</span> <span class="va">Channel</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.9</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">mads</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">700</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_alpha_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>As we already observed previously, the single-cell sample form a compact cloud. The MAD and number of proteins thresholds remove a few outlying cells, but a large majority is retained. You may notice that the first MAD threshold (<span class="math inline">\(log_2(sum) \approx 14\)</span>) filters out all the negative controls that were discussed in a previous section. The channel does not seem to affect the QC, but remember that difference between channels were already corrected during the normalization step.</p>
</div>
<div id="filter-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-cells" class="anchor"></a>Filter cells</h2>
<p>Now that we have defined filtering thresholds, we can apply the filter. Recall that <code>outlier</code> defines whether a cell passes the MAD threshold or not.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">scp</span>, <span class="op">!</span><span class="va">scp</span><span class="op">$</span><span class="va">outlier</span> <span class="op">&amp;</span> <span class="va">scp</span><span class="op">$</span><span class="va">detected</span> <span class="op">&gt;</span> <span class="fl">700</span><span class="op">)</span></code></pre></div>
</div>
<div id="further-qc" class="section level2">
<h2 class="hasAnchor">
<a href="#further-qc" class="anchor"></a>Further QC</h2>
<p>In their workflow, the authors displayed the summed intensities per plate to identify biases linked to the plate, the rows of the plate and the column of the plate. The data required to reproduce this plot is contained in the <code>colData</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Col</span>,
        y <span class="op">=</span> <span class="va">Row</span>, 
        fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">sum</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_tile</span><span class="op">(</span>col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">Plate</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme</span><span class="op">(</span>panel.grid.major <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_x_discrete</span><span class="op">(</span>drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-29-1.png" width="1440"></p>
<p>The position on the plate does not seem to influence the quality of the cells.</p>
<p>Another QC the authors performed is to check the correlation between FACS parameters (that define the cell type) and the QC metrics. The information is again available from the <code>colData</code>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span>
<span class="co">## First plot is coloured by number of detected proteins</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD34_APC.Cy7.A</span>,
        y <span class="op">=</span> <span class="va">CD38_PE.A</span>,
        colour <span class="op">=</span> <span class="va">detected</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_colour_continuous</span><span class="op">(</span>type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span> 
<span class="co">## Second plot is coloured by log2 summed intensities</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="va">p1</span> <span class="op">+</span> <span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">sum</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Show plots</span>
<span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-30-1.png" width="384"></p>
<p>FACS signal is correlated neither with the number of detect proteins nor with the summed protein intensities.</p>
</div>
</div>
<div id="feature-qc" class="section level1">
<h1 class="hasAnchor">
<a href="#feature-qc" class="anchor"></a>Feature QC</h1>
<p>Next to cell QC, the authors also filter the data based on protein QC metrics. More specifically, they keep a proteins if it was detect in at least 3 cells.</p>
<p>We can quickly compute this using the <code>perFeatureQCMetrics</code> from the <code>scuttle</code> package. We convert the proportion of sample a protein is detected in to the number of sample.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_norm"</span><span class="op">]</span><span class="op">]</span>
<span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perFeatureQCMetrics.html">perFeatureQCMetrics</a></span><span class="op">(</span><span class="va">sce</span>, assay.type <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">qc</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&lt;-</span> <span class="va">qc</span><span class="op">$</span><span class="va">detected</span> <span class="op">/</span> <span class="fl">100</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<p>We store the resulting QC in the <code>rowData</code> of the normalized protein assay.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">rowData</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_norm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">qc</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">qc</span></code></pre></div>
<p>Next, we keep the proteins that are detected in at least 3 cells. Note that this time we don’t use the <code>filterFeatures</code> function but the 3 index subsetting. Since <code>ncells</code> is only available for the normalized protein assay, using <code>filterFeatures</code> would remove all the other assays. Instead, subsetting on the feature names will take advantage of the fact that the normalized protein assay is linked to the other assays thanks to the <code>QFeatures</code> data structure.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sel</span> <span class="op">&lt;-</span> <span class="va">qc</span><span class="op">[</span><span class="va">qc</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&gt;=</span> <span class="fl">3</span>, <span class="op">]</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sel</span><span class="op">)</span>, , <span class="op">]</span></code></pre></div>
</div>
<div id="normalization-ii" class="section level1">
<h1 class="hasAnchor">
<a href="#normalization-ii" class="anchor"></a>Normalization II</h1>
<p>The SCeptre workflow performs a second normalization step to prepare the data before embedding an dimension reduction. They apply a median shift of the total intensity across cells, as implemented in the <code>scanpy</code> python library. We perform the same normalization using R code. First, we compute median shifted total intensities that are used as scaling factor. Then, we run the <code>sweep</code> function on the <code>QFeatures</code> object. It acts as the base <code>sweep</code> for matrices: it takes the quantitative data matrix and applies a scaling vector and function to the matrix along the desired dimension. The results are stored in a new assay that we here call <code>proteins_medshift</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_norm"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sf</span> <span class="op">&lt;-</span> <span class="va">sf</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">sf</span><span class="op">)</span>
<span class="op">(</span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">sweep</a></span><span class="op">(</span><span class="va">scp</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, STATS <span class="op">=</span> <span class="va">sf</span>, FUN <span class="op">=</span> <span class="st">"/"</span>,
              i <span class="op">=</span> <span class="st">"proteins_norm"</span>, name <span class="op">=</span> <span class="st">"proteins_medshift"</span><span class="op">)</span><span class="op">)</span>
<span class="co">## An instance of class QFeatures containing 191 assays:</span>
<span class="co">##  [1] F1: SingleCellExperiment with 0 rows and 13 columns </span>
<span class="co">##  [2] F10: SingleCellExperiment with 0 rows and 8 columns </span>
<span class="co">##  [3] F100: SingleCellExperiment with 0 rows and 8 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [189] proteins: SingleCellExperiment with 2727 rows and 2094 columns </span>
<span class="co">##  [190] proteins_norm: SingleCellExperiment with 2727 rows and 2094 columns </span>
<span class="co">##  [191] proteins_medshift: SingleCellExperiment with 2727 rows and 2094 columns</span></code></pre></div>
</div>
<div id="log-transformation" class="section level1">
<h1 class="hasAnchor">
<a href="#log-transformation" class="anchor"></a>Log-transformation</h1>
<p>The next step in the SCeptre workflow is to <span class="math inline">\(log_2\)</span> transform the data. This is performed in our workflow using the <code>logTransform</code> function that adds a new assay containing the transformed data to the <code>QFeatures</code> object. Note that the authors used a pseudo-count of 1.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">logTransform</a></span><span class="op">(</span><span class="va">scp</span>, base <span class="op">=</span> <span class="fl">2</span>, pc <span class="op">=</span> <span class="fl">1</span>,
                    i <span class="op">=</span> <span class="st">"proteins_medshift"</span>, name <span class="op">=</span> <span class="st">"proteins_log"</span><span class="op">)</span> </code></pre></div>
</div>
<div id="imputation-of-missing-data" class="section level1">
<h1 class="hasAnchor">
<a href="#imputation-of-missing-data" class="anchor"></a>Imputation of missing data</h1>
<p>The SCeptre workflow then proceeds with imputation. The imputation is performed by the <code>impute</code> function implemented in <code>sceptre</code> library. Under the hood, the function relies on the <code>KNNImputer</code> from the <code>scikit-learn</code> Python library.</p>
<p>Imputation can be performed on a <code>QFeatures</code> object using the <code>impute</code> function that relies on the <code>MsCoreUtils</code> package. However, the specific implementation differs from <code>KNNImputer</code>. Therefore, for sake of replication, we rather use the latter using the R-to-Python interface by <code>reticulate</code>. Again, we need to convert the target assay to an <code>AnnData</code> object.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_log"</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span>
<span class="co">## Warning: Ignoring redundant column names in 'colData(x)': Channel, Col</span>
<span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/AnnData-Conversion.html">SCE2AnnData</a></span><span class="op">(</span><span class="va">sce</span>, X_name <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>We then perform the imputation.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">spt</span><span class="op">$</span><span class="fu">impute</span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></code></pre></div>
<p>Finally, we convert the data back to a <code>SingleCellExperiment</code> object and add it to the <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/AnnData-Conversion.html">AnnData2SCE</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">addAssay</a></span><span class="op">(</span><span class="va">scp</span>, <span class="va">sce</span>, name <span class="op">=</span> <span class="st">"proteins_imput"</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html">addAssayLink</a></span><span class="op">(</span><span class="va">scp</span>, from <span class="op">=</span> <span class="st">"proteins_log"</span>, to <span class="op">=</span> <span class="st">"proteins_imput"</span>,
                    varFrom <span class="op">=</span> <span class="st">"Accession"</span>, varTo <span class="op">=</span> <span class="st">"Accession"</span><span class="op">)</span>
<span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 193 assays:</span>
<span class="co">##  [1] F1: SingleCellExperiment with 0 rows and 13 columns </span>
<span class="co">##  [2] F10: SingleCellExperiment with 0 rows and 8 columns </span>
<span class="co">##  [3] F100: SingleCellExperiment with 0 rows and 8 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [191] proteins_medshift: SingleCellExperiment with 2727 rows and 2094 columns </span>
<span class="co">##  [192] proteins_log: SingleCellExperiment with 2727 rows and 2094 columns </span>
<span class="co">##  [193] proteins_imput: SingleCellExperiment with 2727 rows and 2094 columns</span></code></pre></div>
</div>
<div id="normalization-iii" class="section level1">
<h1 class="hasAnchor">
<a href="#normalization-iii" class="anchor"></a>Normalization III</h1>
<p>The last step of the SCeptre workflow before running the downstream analysis is to perform a last normalization, known as centering and scaling to unit variance. While this is performed using <code>scanpy</code> in the SCeptre workflow, we here apply the <code>sweep</code> function as shown previously.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_imput"</span><span class="op">)</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">sweep</a></span><span class="op">(</span><span class="va">scp</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>, STATS <span class="op">=</span> <span class="va">means</span>, FUN <span class="op">=</span> <span class="st">"-"</span>,
             i <span class="op">=</span> <span class="st">"proteins_imput"</span>, name <span class="op">=</span> <span class="st">"proteins_centered"</span><span class="op">)</span>
<span class="va">stds</span> <span class="op">&lt;-</span> <span class="fu">rowSds</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_centered"</span><span class="op">)</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">sweep</a></span><span class="op">(</span><span class="va">scp</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>, STATS <span class="op">=</span> <span class="va">stds</span>, FUN <span class="op">=</span> <span class="st">"/"</span>,
             i <span class="op">=</span> <span class="st">"proteins_centered"</span>, name <span class="op">=</span> <span class="st">"proteins_scaled"</span><span class="op">)</span></code></pre></div>
</div>
<div id="downstream-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#downstream-analysis" class="anchor"></a>Downstream analysis</h1>
<p>After this extensive processing of the data, the authors perform the downstream analysis to answer the research question of interest, that is, is SCP able to retrieve the hierarchical structure of the AML model. To evaluate this, they first performed various dimension reductions, followed by pseudo-time analysis. Cell clustering is performed as well as gene clustering to define novel gene signature.</p>
<p>So, now that the processing is complete, we can extract the last assay on which we will perform the downstream analysis.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_scaled"</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span>
<span class="co">## Warning: Ignoring redundant column names in 'colData(x)': Channel, Col, outlier</span>
<span class="co"># load("~/Desktop/SCE_temp.RData")</span></code></pre></div>
<div id="dimension-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#dimension-reduction" class="anchor"></a>Dimension reduction</h2>
<p>We perform 3 types of dimension reductions: PCA, UMAP and diffusion map.</p>
<div id="pca" class="section level3">
<h3 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h3>
<p>PCA can be directly computed from <code>SingleCellExperiment</code> objects using the <code>runPCA</code> function from the <code>scater</code> package. The resulting dimension reduction is stored in the <code>reducedDim</code> slot of the object.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/scater/">"scater"</a></span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, 
              ncomponents <span class="op">=</span> <span class="fl">50</span>,
              ntop <span class="op">=</span> <span class="cn">Inf</span>,
              scale <span class="op">=</span> <span class="cn">FALSE</span>,
              exprs_values <span class="op">=</span> <span class="fl">1</span>,
              name <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<p>The package also provides the <code>plotPCA</code> function to directly plot the PCA from the <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Population"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-42-1.png" width="700"></p>
<p>We can already see the LSC and PROG populations are close together in the latent space whereas the blast are spread over the latent space.</p>
</div>
<div id="umap" class="section level3">
<h3 class="hasAnchor">
<a href="#umap" class="anchor"></a>UMAP</h3>
<p>Similarly to PCA, we can use <code>runUMAP</code> and <code>plotUMAP</code> to generate and plot UMAP space. We modify the tuning arguments to match the parameters used in the SCeptre notebook.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, 
               ncomponents <span class="op">=</span> <span class="fl">2</span>,
               ntop <span class="op">=</span> <span class="cn">Inf</span>,
               scale <span class="op">=</span> <span class="cn">FALSE</span>,
               n_neighbors <span class="op">=</span> <span class="fl">30</span>,
               min_dist <span class="op">=</span> <span class="fl">0.5</span>,
               exprs_values <span class="op">=</span> <span class="fl">1</span>,
               dimred <span class="op">=</span> <span class="st">"PCA"</span>,
               name <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Population"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-43-1.png" width="700"></p>
<p>We can draw the same conclusion for the UMAP results than for the PCA result. LSC and PROG are also close in UMAP space while the BLAST are spread evenly. UMAP is used to visually assess the presence of clusters. In this case, we cannot see clusters but rather an homogeneous cloud. One reason could be that the present SCP data does not provide sufficient resolution (noise is too high) to discriminate between the different cell types. Another explanation is that the proteome between the different cell types is continuously shifting from one cell type to another. It is important to remember that the cell types are defined based on the CD34/CD38 flow-cytometry data using hard thresholds. To investigate whether this transition might be a smooth transition, we plot the score of the first UMAP vector along the FACS data.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, UMAP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"UMAP"</span><span class="op">)</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD34_APC.Cy7.A</span>, 
        y <span class="op">=</span> <span class="va">CD38_PE.A</span>,
        colour <span class="op">=</span> <span class="va">UMAP.1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">470</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">annotate</span><span class="op">(</span><span class="st">"label"</span>, 
             label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BLAST CD38+"</span>, <span class="st">"PROG"</span>, <span class="st">"BLAST CD38-"</span>, <span class="st">"LSC"</span><span class="op">)</span>,
             x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">300</span>, <span class="fl">700</span>, <span class="fl">300</span>, <span class="fl">700</span><span class="op">)</span>, 
             y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">750</span>, <span class="fl">750</span>, <span class="fl">200</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-44-1.png" width="700"></p>
<p>This plot indeed indicates a smooth transition between the different cell types and this transition seems to be captured by the first UMAP vector. The higher the combination of the CD34 and or CD38 signal, the higher the UMAP score.</p>
<p>While reassuring, the first scenario (too much noise to resolve cluster) might still hold as the LSC and PROG that are well separated by the FACS data and display clear phenotypical differences (LSC can self-renew, whereas PROG cannot) cannot be distinguished in UMAP space.</p>
</div>
<div id="diffusion-map" class="section level3">
<h3 class="hasAnchor">
<a href="#diffusion-map" class="anchor"></a>Diffusion map</h3>
<p>Another dimension reduction technique is diffusion maps that generates a manifold where cells are spatially organized so that it reflects their transition probabilities. This is a more suitable of representation differentiation processes, like those happening in the AML model.</p>
<p>Diffusion maps can be computed using <code>runDiffusionMap</code> in <code>scuttle</code>. However, we here prefer to use the <code>destiny</code> package to not only get the diffusion map components, but also the transition probability matrix that we will use later during the pseudo-time analysis.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># BiocManager::install("theislab/destiny")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://theislab.github.io/destiny/">"destiny"</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span>
<span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://theislab.github.io/destiny/reference/DiffusionMap-class.html">DiffusionMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We stored the diffusion components in the <code>reducedDim</code> slot so that we can plot the results using the <code>plotDimensionReduction</code> function.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"DiffusionMap"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://theislab.github.io/destiny/reference/destiny-generics.html">eigenvectors</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotDiffusionMap</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Population"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<p>The first diffusion component recapitulates the expected differentiation hierarchy from the AML model. One of the tips of the diffusion map mainly contains LSC cells. PROG cells then also appear as well as BLAST.</p>
</div>
</div>
<div id="pseudo-time-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#pseudo-time-analysis" class="anchor"></a>Pseudo-time analysis</h2>
<p>The objective of the pseudo-time analysis is to order cells along their differentiation tracks. To do so, we first need to define a root cell. Since the cells should theoretically derive from LSC cells, we define the root cell as the LSC cell at the tip of the diffusion map. We store this which cell is the root in the <code>colData</code>.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>DC1 <span class="op">=</span> <span class="va">dm</span><span class="op">$</span><span class="va">DC1</span>, Population <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">Population</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">$</span><span class="va">Population</span> <span class="op">==</span> <span class="st">"LSC"</span>, <span class="op">]</span>
<span class="va">root</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">DC1</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">DC1</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="va">sce</span><span class="op">$</span><span class="va">isRoot</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="va">root</span>, <span class="st">"isRoot"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></code></pre></div>
<p>We can now compute the diffusion pseudo-time using the <code>DPT</code> function. Again, we store the output in the <code>colData</code> after converting the pseudo-time to percentage.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://theislab.github.io/destiny/reference/DPT.html">DPT</a></span><span class="op">(</span><span class="va">dm</span>, tips <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">isRoot</span><span class="op">)</span><span class="op">)</span>
<span class="va">sce</span><span class="op">$</span><span class="va">dpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rank.html">rank</a></span><span class="op">(</span><span class="va">dpt</span><span class="op">$</span><span class="va">dpt</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<p>We map the pseudo-time on the diffusion map and on the UMAP plots.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotDiffusionMap</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"dpt"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"dpt"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-49-1.png" width="672"></p>
<p>We also map the pseudo-time on the FACS data to highlight the estimated differentation trajectory.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD34_APC.Cy7.A</span>, 
        y <span class="op">=</span> <span class="va">CD38_PE.A</span>,
        colour <span class="op">=</span> <span class="va">dpt</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">470</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">annotate</span><span class="op">(</span><span class="st">"label"</span>, 
             label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BLAST CD38+"</span>, <span class="st">"PROG"</span>, <span class="st">"BLAST CD38-"</span>, <span class="st">"LSC"</span><span class="op">)</span>,
             x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">300</span>, <span class="fl">700</span>, <span class="fl">300</span>, <span class="fl">700</span><span class="op">)</span>, 
             y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">750</span>, <span class="fl">750</span>, <span class="fl">200</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-50-1.png" width="700"></p>
<p>From this plot, the pseudo-time analysis follows the trajectory from the FACS data as descried in the literature, that is, LSC cells differentiate into PROG cells then into BLAST CD38+ cells and finish as BLAST CD38- cells. Interestingly, this plot also suggests an alternative trajectory going from LSC directly to BLAST CD38-. This was also observed in the <span class="citation">Schoof et al. (2021)</span> paper. This represents a good replication of their work, we were able to successfully replicate their work.</p>
</div>
<div id="cell-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-clustering" class="anchor"></a>Cell clustering</h2>
<p>Next to pseudo-time analysis, the authors also performed cell clustering. They applied the leiden algorithm to identify clusters from the data. Since this is only implemented in Python, we will again use the <code>reticulate</code> interface. First, we converted our data to an <code>AnnData</code> object.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/AnnData-Conversion.html">SCE2AnnData</a></span><span class="op">(</span><span class="va">sce</span>, X_name <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Next we load the <code>scanpy</code> module that contains the required function to perform PCA, that is used to build the nearest neighbour graph, that is used to identify clusters using the Leiden algorithm. The resulting clusters are stored in the <code>colData</code></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scanpy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/import.html">import</a></span><span class="op">(</span><span class="st">"scanpy"</span><span class="op">)</span>
<span class="va">scanpy</span><span class="op">$</span><span class="va">pp</span><span class="op">$</span><span class="fu">pca</span><span class="op">(</span><span class="va">adata</span>, random_state <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span>
<span class="va">scanpy</span><span class="op">$</span><span class="va">pp</span><span class="op">$</span><span class="fu">neighbors</span><span class="op">(</span><span class="va">adata</span>, n_neighbors <span class="op">=</span> <span class="fl">30L</span>, random_state <span class="op">=</span> <span class="fl">5L</span><span class="op">)</span>
<span class="va">scanpy</span><span class="op">$</span><span class="va">tl</span><span class="op">$</span><span class="fu">leiden</span><span class="op">(</span><span class="va">adata</span>, resolution <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span>
<span class="va">sce</span><span class="op">$</span><span class="va">leidenClust</span> <span class="op">&lt;-</span> <span class="va">adata</span><span class="op">$</span><span class="va">obs</span><span class="op">$</span><span class="va">leiden</span></code></pre></div>
<p>We map the clusters on the UMAP plot.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"leidenClust"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-53-1.png" width="700"></p>
<p>The clusters shown above differ from the clusters shown in the SCeptre notebook. As we will see in a later section the data obtained in this vignette differ slightly from the data obtained in the SCeptre notebook. The discrepancies between the clusters from two slightly different datasets puts into question the reliability of the identified clusters. This observation supports the observation we drew from the UMAP where no grouping structure is found. Furthermore, we know <em>a priori</em> the cell types thanks to the FACS staining and previous work from literature. We can plot the concordance between the identified cluster and the known cell types.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Count cells per cell type per cluster</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">df</span>, <span class="va">Population</span>, <span class="va">leidenClust</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">summarize</span><span class="op">(</span><span class="va">df</span>, n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Compute the relative frequency per cluster</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">df</span>, <span class="va">leidenClust</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">df</span>, rfreq <span class="op">=</span> <span class="va">n</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Plot bar chart</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">leidenClust</span>, 
        y <span class="op">=</span> <span class="va">rfreq</span>,
        fill <span class="op">=</span> <span class="va">Population</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span>
<span class="co">## Warning: Ignoring unknown parameters: binwidth, bins, pad</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-54-1.png" width="700"></p>
<p>The clustering does not recapitulate the known biological information. All the clusters contain several cell type (except for cluster 2), and all cell types are present in several clusters. So, this dataset, and maybe more generally this biological model, should not be described as a set of discrete clusters.</p>
</div>
<div id="find-marker-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#find-marker-genes" class="anchor"></a>Find marker genes</h2>
<p>In the SCeptre workflow, the authors searched for marker genes across the different clusters. As argued above, we don’t trust the clustering and will rather showcase how to identify marker gene between the different cell types from a <code>SingleCellExperiment</code> object. The functionality to perform this is provided in the <code>scran</code> package.</p>
<p>We use the <code>findMarkers function that generates a list of tables, where each table provide the result of the differential expression analysis  for one cell type against the others. We here use the t-test as the  statistical method and decide to report p-values based one the one  cluster vs all framework (see</code>?findMarkers` for more info). We also specify that we want up regulated genes.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"scran"</span><span class="op">)</span>
<span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers</a></span><span class="op">(</span><span class="va">sce</span>, 
                       assay.type <span class="op">=</span> <span class="fl">1</span>, 
                       groups <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">Population</span>, 
                       test.type <span class="op">=</span> <span class="st">"t"</span>,
                       direction <span class="op">=</span> <span class="st">"up"</span>,
                       pval.type <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>
<span class="va">markers</span>
<span class="co">## List of length 4</span>
<span class="co">## names(4): BLAST CD38- BLAST CD38+ LSC PROG</span></code></pre></div>
<p>We next combine the list of tables into a single table. We perform some extra data wrangling to also include the smallest fold change</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span>, 
                  <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">markers</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FDR"</span>, <span class="st">"summary.logFC"</span><span class="op">)</span><span class="op">]</span>,
                                    Population <span class="op">=</span> <span class="va">i</span><span class="op">)</span><span class="op">)</span>
<span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">markers</span><span class="op">)</span>
<span class="va">markers</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></code></pre></div>
<p>We keep only genes that are differentially expressed, that is for which the false discovery rate (FDR) is lower than 5 % and the log fold change between the given cluster and the others is larger than 1.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">markers</span> <span class="op">&lt;-</span> <span class="va">markers</span><span class="op">[</span><span class="va">markers</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">markers</span><span class="op">$</span><span class="va">summary.logFC</span> <span class="op">&gt;</span> <span class="fl">1</span>, <span class="op">]</span></code></pre></div>
<p>We plot the marker expression as heatmap using <code>ComplexHeatmaps</code>. First, we filter the data to include only marker genes.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="va">markers</span><span class="op">$</span><span class="va">gene</span>, <span class="op">]</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/jokergoo/ComplexHeatmap">"ComplexHeatmap"</a></span><span class="op">)</span></code></pre></div>
<p>We set a colouring scheme for the cell types using the <code>RColorBrewer</code> package.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"RColorBrewer"</span><span class="op">)</span>
<span class="va">popCols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"Set1"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">popCols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">Population</span><span class="op">)</span></code></pre></div>
<p>Next, we create column annotation to highlight the cell types on the columns of the heatmap.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">colAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/columnAnnotation.html">columnAnnotation</a></span><span class="op">(</span>Population <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">Population</span>, 
                             col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Population <span class="op">=</span> <span class="va">popCols</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Similarly, we create row annotation to highlight the cell types on the rows of the heatmap.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rowAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html">rowAnnotation</a></span><span class="op">(</span>Population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/anno_simple.html">anno_simple</a></span><span class="op">(</span><span class="va">markers</span><span class="op">$</span><span class="va">Population</span>, 
                                                   col <span class="op">=</span> <span class="va">popCols</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Finally, we define the colors of the heatmap using the <code>viridis</code> and the <code>circlize</code> color scheme. Note that we provide some trimming of the values for better rendering.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://sjmgarnier.github.io/viridis/">"viridis"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/jokergoo/circlize">"circlize"</a></span><span class="op">)</span>
<span class="va">vircols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html">viridis</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>
<span class="va">trimmin</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fl">2</span>
<span class="va">trimmax</span> <span class="op">&lt;-</span> <span class="fl">2</span>
<span class="va">vircols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="va">trimmin</span>, <span class="va">trimmax</span>, length.out <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>, <span class="va">vircols</span><span class="op">)</span></code></pre></div>
<p>We can now generate the heatmap.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, 
        name <span class="op">=</span> <span class="st">"logProteins"</span>,
        col <span class="op">=</span> <span class="va">vircols</span>,
        left_annotation <span class="op">=</span> <span class="va">rowAnnot</span>,
        top_annotation <span class="op">=</span> <span class="va">colAnnot</span>,
        cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,
        cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>,
        show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, 
        show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,
        column_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">Population</span><span class="op">)</span>,
        use_raster <span class="op">=</span> <span class="cn">TRUE</span>,
        heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">trimmin</span>, <span class="fl">0</span>, <span class="va">trimmax</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-63-1.png" width="700"></p>
<p>We can see that the more cells differentiate towards BLAST CD38- cells the more they expressed a unique set of genes. We could not find genes that are differentially over-expressed by PROG cells.</p>
<p>Let’s have a look at the top markers for the different cell populations.</p>
<div id="top-markers-of-the-blast-cd38--cells" class="section level3">
<h3 class="hasAnchor">
<a href="#top-markers-of-the-blast-cd38--cells" class="anchor"></a>Top markers of the BLAST CD38- cells</h3>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">markerBlastCD38neg</span> <span class="op">&lt;-</span> <span class="va">markers</span><span class="op">[</span><span class="va">markers</span><span class="op">$</span><span class="va">Population</span> <span class="op">==</span> <span class="st">"BLAST CD38-"</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">markerBlastCD38neg</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">markerBlastCD38neg</span><span class="op">$</span><span class="va">FDR</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span>
<span class="co">## DataFrame with 6 rows and 4 columns</span>
<span class="co">##                           FDR summary.logFC    Population          gene</span>
<span class="co">##                     &lt;numeric&gt;     &lt;numeric&gt;   &lt;character&gt;   &lt;character&gt;</span>
<span class="co">## ARL1              2.00556e-62       1.25825 BLAST CD38...          ARL1</span>
<span class="co">## RAB11FIP1         7.00637e-20       1.09573 BLAST CD38...     RAB11FIP1</span>
<span class="co">## MRPL20            1.24318e-18       1.06399 BLAST CD38...        MRPL20</span>
<span class="co">## CTAGE15; CTAGE15P 2.69564e-17       1.43480 BLAST CD38... CTAGE15; C...</span>
<span class="co">## NDUFA6            3.21412e-16       1.14144 BLAST CD38...        NDUFA6</span>
<span class="co">## FHL1              1.17468e-15       1.46280 BLAST CD38...          FHL1</span></code></pre></div>
<p>The most significant marker is ARL1 (ADP-ribosylation factor-like protein 1). It is annotated as involved in the protein localization and transport to the Golgi appartus.</p>
<p>We also plot the evolution of the ARL1 expression with respect to the pseudo-time.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="st">"ARL1"</span>, <span class="op">]</span>, 
                 <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">exprs</span>, 
        x <span class="op">=</span> <span class="va">dpt</span>,
        col <span class="op">=</span> <span class="va">Population</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"ARL1 abundance"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-65-1.png" width="700"></p>
<p>We can indeed observe a strong correlation between ARL1 expression and the pseudo time. However, the distribution of the point shows an undesired trend, that is that some observations are grouped along stripes. We already observed this for other SCP data and we showed that these stripes are artefacts of data processing (<span class="citation">Vanderaa and Gatto (2021)</span>).</p>
</div>
<div id="top-marker-of-the-blast-cd38-cells" class="section level3">
<h3 class="hasAnchor">
<a href="#top-marker-of-the-blast-cd38-cells" class="anchor"></a>Top marker of the BLAST CD38+ cells</h3>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">markerBlastCD38pos</span> <span class="op">&lt;-</span> <span class="va">markers</span><span class="op">[</span><span class="va">markers</span><span class="op">$</span><span class="va">Population</span> <span class="op">==</span> <span class="st">"BLAST CD38+"</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">markerBlastCD38pos</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">markerBlastCD38pos</span><span class="op">$</span><span class="va">FDR</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span>
<span class="co">## DataFrame with 6 rows and 4 columns</span>
<span class="co">##                FDR summary.logFC    Population        gene</span>
<span class="co">##          &lt;numeric&gt;     &lt;numeric&gt;   &lt;character&gt; &lt;character&gt;</span>
<span class="co">## P4HB   8.22144e-09       1.71286 BLAST CD38...        P4HB</span>
<span class="co">## CTSG   2.39614e-07       1.34137 BLAST CD38...        CTSG</span>
<span class="co">## ZNF586 1.81384e-06       1.25044 BLAST CD38...      ZNF586</span>
<span class="co">## GPI    7.04419e-06       1.16374 BLAST CD38...         GPI</span>
<span class="co">## MDC1   7.04419e-06       1.22853 BLAST CD38...        MDC1</span>
<span class="co">## ELANE  7.70134e-06       1.75155 BLAST CD38...       ELANE</span></code></pre></div>
<p>The most significant marker is P4HB (Protein disulfide-isomerase).</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="st">"P4HB"</span>, <span class="op">]</span>, 
                 <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">exprs</span>, 
        x <span class="op">=</span> <span class="va">dpt</span>,
        col <span class="op">=</span> <span class="va">Population</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"P4HB abundance"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-67-1.png" width="700"></p>
<p>For this marker, the trend between protein abundance and diffusion pseudo-time is bell-shaped. This is expected from the BLAST CD38+ cells since they represent the intermediate step of differentiation.</p>
<ul>
<li>Marker of the LSC cells</li>
</ul>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">markerBlastLSC</span> <span class="op">&lt;-</span> <span class="va">markers</span><span class="op">[</span><span class="va">markers</span><span class="op">$</span><span class="va">Population</span> <span class="op">==</span> <span class="st">"LSC"</span>, <span class="op">]</span>
<span class="va">markerBlastLSC</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">markerBlastLSC</span><span class="op">$</span><span class="va">FDR</span><span class="op">)</span>, <span class="op">]</span>
<span class="co">## DataFrame with 2 rows and 4 columns</span>
<span class="co">##             FDR summary.logFC  Population        gene</span>
<span class="co">##       &lt;numeric&gt;     &lt;numeric&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## PPIG  0.0301131       1.04888         LSC        PPIG</span>
<span class="co">## PRDX1 0.0369504       1.29379         LSC       PRDX1</span></code></pre></div>
<p>Only two markers are found significant for the LSC population. One of the two, <a href="https://www.uniprot.org/uniprot/Q06830">PRDX1</a> associated to cell proliferation and response to oxydative stress. These biological processes are expected to occur in LSC (<span class="citation">Trombetti et al. (2021)</span>).</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="st">"PRDX1"</span>, <span class="op">]</span>, 
                 <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">exprs</span>, 
        x <span class="op">=</span> <span class="va">dpt</span>,
        col <span class="op">=</span> <span class="va">Population</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"PRDX1 abundance"</span><span class="op">)</span></code></pre></div>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-69-1.png" width="700"></p>
<p>The relationship between PRDX1 expression and pseudo-time is anti-correlated with a strong expression by LSC at early estimated differentiation time and low expression at later time by BLAST cells.</p>
</div>
</div>
<div id="gene-set-enrichment" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-set-enrichment" class="anchor"></a>Gene set enrichment</h2>
<p><strong>UNDER PROGRESS</strong></p>
<p>Finally, we run a gene set enrichment analysis to provide a biologically meaningful interpretation of the markers we identified.</p>
<p>Let’s start with the BLAST CD38- cells.</p>
</div>
</div>
<div id="benchmarking" class="section level1">
<h1 class="hasAnchor">
<a href="#benchmarking" class="anchor"></a>Benchmarking</h1>
<p>Up to this point, we replicated the processing steps by SCeptre using our <code>scp</code> workflow and showed how to perform downstream analyses using R/Bioconductor software. We will now assess the quality of the replication and see how well we can recreate the original data. We retrieved the final data provided by the Schoof et al. </p>
<p>We will compare the dataset processed through SCeptre and the dataset processed through <code>scp</code> based on 3 metrics:</p>
<ul>
<li>The overlap of the selected cells</li>
<li>The overlap of the selected proteins</li>
<li>The difference of the processed quantitative data</li>
</ul>
<div id="compare-the-selected-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-the-selected-cells" class="anchor"></a>Compare the selected cells</h2>
<p>One of the above section filtered the single cells based on the summed abundance and the number of detected proteins. We will check that we could indeed reproduce the cell filter performed by SCeptre.</p>
<p><img src="schoof2021_files/figure-html/benchm_cell_filt-1.png" width="480"></p>
<p>The <code>scp</code> filtering step keeps 70 additional cells compared to SCeptre, and removed 1 cell that SCeptre kept, but the two workflows agreed on 2024 cells.</p>
</div>
<div id="compare-the-selected-proteins" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-the-selected-proteins" class="anchor"></a>Compare the selected proteins</h2>
<p>The SCeptre and the <code>scp</code> protein data have comparable dimensions, although they are not exactly the same. We plot the overlap of the selected proteins between the two workflows.</p>
<p><img src="schoof2021_files/figure-html/benchm_prot_filt-1.png" width="480"></p>
<p>The selected proteins highpy overlap between the two workflows. <code>scp</code> kept 4 proteins that were removed by SCeptre. 2723 proteins were selected by both workflows. Similarly to the filtered cells, the differences are negligible and demonstrate an almost perfect replication of the feature filtering steps.</p>
</div>
<div id="compare-the-processed-quantifications" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-the-processed-quantifications" class="anchor"></a>Compare the processed quantifications</h2>
<p>To assess the differences between the data matrices, we intersect the proteins and the cells to have comparable matrices.</p>
<pre><code>## Warning: Removed 8788 rows containing non-finite values (stat_bin).</code></pre>
<p><img src="schoof2021_files/figure-html/benchm_prot_diff-1.png" width="480"></p>
<p>95 % of the differences are contained within (-0.42, 0.53) with a sharp peak around 0. We compare this difference distribution to the distribution of the protein abundance in the SCeptre dataset.</p>
<pre><code>## Warning: Removed 246447 rows containing non-finite values (stat_bin).</code></pre>
<p><img src="schoof2021_files/figure-html/unnamed-chunk-71-1.png" width="480"></p>
<p>Compared to the shape of the quantitative data, the differences observed between the two workflows are negligeable. Therefore, the <code>scp</code> workflow accurately replicates the SCeptre workflow.</p>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>In this vignette, we have demonstrated that the <code>scp</code> package is able to accurately reproduce the analysis published in <span class="citation">Schoof et al. (2021)</span>. It also serves to showcase how to apply <code>scp</code> to real-life SCP data. Finally, we also show how to bridge and integrate software from two different programming languages. While this workflow is fully executed in R, we were able to include Python thanks to the <code>reticulate</code> interface.</p>
</div>
<div id="requirements" class="section level1">
<h1 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h1>
<div id="hardware-and-software" class="section level3">
<h3 class="hasAnchor">
<a href="#hardware-and-software" class="anchor"></a>Hardware and software</h3>
<p>The system details of the machine that built the vignette are:</p>
<pre><code>## Machine: Linux (4.15.0-158-generic)
## R version: R.4.1.0 (svn: 79334)
## RAM: 8.2 GB
## CPU: 4 core(s) - Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz</code></pre>
<p>Further information about the Python environment:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> IPython</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(IPython.sys_info())</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="co">## {'commit_hash': 'e76fa004a',</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="co">##  'commit_source': 'installation',</span></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a><span class="co">##  'default_encoding': 'utf-8',</span></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a><span class="co">##  'ipython_path': '/home/christophe/PhD/SCP.replication/inst/renvs/schoof2021/renv/python/virtualenvs/renv-python-3.8/lib/python3.8/site-packages/IPython',</span></span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a><span class="co">##  'ipython_version': '7.28.0',</span></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a><span class="co">##  'os_name': 'posix',</span></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a><span class="co">##  'platform': 'Linux-4.15.0-158-generic-x86_64-with-glibc2.10',</span></span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a><span class="co">##  'sys_executable': '/home/christophe/PhD/SCP.replication/inst/renvs/schoof2021/renv/python/virtualenvs/renv-python-3.8/bin/python',</span></span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a><span class="co">##  'sys_platform': 'linux',</span></span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a><span class="co">##  'sys_version': '3.8.5 (default, Sep  4 2020, 07:32:50) \n[GCC 7.3.0]'}</span></span></code></pre></div>
</div>
<div id="running-the-sceptre-module-in-r" class="section level3">
<h3 class="hasAnchor">
<a href="#running-the-sceptre-module-in-r" class="anchor"></a>Running the <code>sceptre</code> module in R</h3>
<p>In order to rerun this vignette, you will need to create a <code>conda</code> environment containing the required modules to run the SCeptre workflow.</p>
<p>First, you need to install Python (we used <a href="https://www.python.org/downloads/">Python3</a>) and <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> (or <a href="https://docs.conda.io/en/latest/miniconda.html">miniconda</a>).</p>
<p>Next, as mentioned in the <a href="https://github.com/bfurtwa/SCeptre#installation-from-github">author’s installation instructions</a>, create a new <code>conda</code> environment by running the following in a terminal.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create -n sceptre python=3.7</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate sceptre</span></code></pre></div>
<p>Then you need to install <code>sceptre</code> and its dependencies. To do so, download the <a href="https://github.com/bfurtwa/SCeptre">GitHub repository</a>, and <code>cd</code> to the downloaded <code>SCeptre/</code> folder. Then run in the terminal:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install .</span></code></pre></div>
<p>To make use of sceptre within an R session, simply call (from the R terminal):</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rstudio/reticulate">"reticulate"</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_condaenv</a></span><span class="op">(</span><span class="st">"sceptre"</span><span class="op">)</span></code></pre></div>
</div>
<div id="timing" class="section level3">
<h3 class="hasAnchor">
<a href="#timing" class="anchor"></a>Timing</h3>
<p>The total time required to compile this vignette is:</p>
<pre><code>## 30.02677 mins</code></pre>
</div>
<div id="memory" class="section level3">
<h3 class="hasAnchor">
<a href="#memory" class="anchor"></a>Memory</h3>
<p>The final <code>scp</code> memory size is:</p>
<pre><code>## [1] "0.3 Gb"</code></pre>
</div>
<div id="session-info" class="section level3">
<h3 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h3>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## R Under development (unstable) (2020-10-12 r79334)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 18.04.5 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] benchmarkme_1.0.7            circlize_0.4.13             </span>
<span class="co">##  [3] viridis_0.6.1                viridisLite_0.4.0           </span>
<span class="co">##  [5] RColorBrewer_1.1-2           ComplexHeatmap_2.9.4        </span>
<span class="co">##  [7] scran_1.21.3                 destiny_3.1.1               </span>
<span class="co">##  [9] scater_1.21.7                zellkonverter_1.3.2         </span>
<span class="co">## [11] reticulate_1.22              scuttle_1.3.1               </span>
<span class="co">## [13] patchwork_1.1.1              forcats_0.5.1               </span>
<span class="co">## [15] stringr_1.4.0                dplyr_1.0.7                 </span>
<span class="co">## [17] purrr_0.3.4                  readr_2.0.2                 </span>
<span class="co">## [19] tidyr_1.1.4                  tibble_3.1.5                </span>
<span class="co">## [21] ggplot2_3.3.5                tidyverse_1.3.1             </span>
<span class="co">## [23] scp_1.3.3                    scpdata_1.1.2               </span>
<span class="co">## [25] ExperimentHub_2.1.4          AnnotationHub_3.1.5         </span>
<span class="co">## [27] BiocFileCache_2.1.1          dbplyr_2.1.1                </span>
<span class="co">## [29] QFeatures_1.3.6              MultiAssayExperiment_1.19.12</span>
<span class="co">## [31] SingleCellExperiment_1.15.2  SummarizedExperiment_1.23.4 </span>
<span class="co">## [33] Biobase_2.53.0               GenomicRanges_1.45.0        </span>
<span class="co">## [35] GenomeInfoDb_1.29.8          IRanges_2.27.2              </span>
<span class="co">## [37] S4Vectors_0.31.4             BiocGenerics_0.39.2         </span>
<span class="co">## [39] MatrixGenerics_1.5.4         matrixStats_0.61.0          </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] rappdirs_0.3.3                ggthemes_4.2.4               </span>
<span class="co">##   [3] ragg_1.1.3                    bit64_4.0.5                  </span>
<span class="co">##   [5] knitr_1.36                    irlba_2.3.3                  </span>
<span class="co">##   [7] DelayedArray_0.19.4           data.table_1.14.2            </span>
<span class="co">##   [9] doParallel_1.0.16             KEGGREST_1.33.0              </span>
<span class="co">##  [11] RCurl_1.98-1.5                AnnotationFilter_1.17.1      </span>
<span class="co">##  [13] generics_0.1.0                ScaledMatrix_1.1.0           </span>
<span class="co">##  [15] RSQLite_2.2.8                 proxy_0.4-26                 </span>
<span class="co">##  [17] bit_4.0.4                     tzdb_0.1.2                   </span>
<span class="co">##  [19] xml2_1.3.2                    lubridate_1.7.10             </span>
<span class="co">##  [21] httpuv_1.6.3                  assertthat_0.2.1             </span>
<span class="co">##  [23] xfun_0.26                     hms_1.1.1                    </span>
<span class="co">##  [25] jquerylib_0.1.4               evaluate_0.14                </span>
<span class="co">##  [27] promises_1.2.0.1              DEoptimR_1.0-9               </span>
<span class="co">##  [29] fansi_0.5.0                   readxl_1.3.1                 </span>
<span class="co">##  [31] igraph_1.2.6                  DBI_1.1.1                    </span>
<span class="co">##  [33] benchmarkmeData_1.0.4         ellipsis_0.3.2               </span>
<span class="co">##  [35] RSpectra_0.16-0               backports_1.2.1              </span>
<span class="co">##  [37] bookdown_0.24                 sparseMatrixStats_1.5.3      </span>
<span class="co">##  [39] vctrs_0.3.8                   Cairo_1.5-12.2               </span>
<span class="co">##  [41] TTR_0.24.2                    abind_1.4-5                  </span>
<span class="co">##  [43] cachem_1.0.6                  RcppEigen_0.3.3.9.1          </span>
<span class="co">##  [45] withr_2.4.2                   robustbase_0.93-9            </span>
<span class="co">##  [47] vcd_1.4-8                     xts_0.12.1                   </span>
<span class="co">##  [49] cluster_2.1.2                 dir.expiry_1.1.0             </span>
<span class="co">##  [51] lazyeval_0.2.2                laeken_0.5.1                 </span>
<span class="co">##  [53] crayon_1.4.1                  basilisk.utils_1.5.0         </span>
<span class="co">##  [55] edgeR_3.35.1                  pkgconfig_2.0.3              </span>
<span class="co">##  [57] labeling_0.4.2                vipor_0.4.5                  </span>
<span class="co">##  [59] ProtGenerics_1.25.1           nnet_7.3-16                  </span>
<span class="co">##  [61] rlang_0.4.11                  lifecycle_1.0.1              </span>
<span class="co">##  [63] filelock_1.0.2                modelr_0.1.8                 </span>
<span class="co">##  [65] rsvd_1.0.5                    cellranger_1.1.0             </span>
<span class="co">##  [67] rprojroot_2.0.2               RcppHNSW_0.3.0               </span>
<span class="co">##  [69] lmtest_0.9-38                 Matrix_1.3-4                 </span>
<span class="co">##  [71] carData_3.0-4                 boot_1.3-28                  </span>
<span class="co">##  [73] zoo_1.8-9                     reprex_2.0.1                 </span>
<span class="co">##  [75] beeswarm_0.4.0                GlobalOptions_0.1.2          </span>
<span class="co">##  [77] rjson_0.2.20                  png_0.1-7                    </span>
<span class="co">##  [79] knn.covertree_1.0             bitops_1.0-7                 </span>
<span class="co">##  [81] Biostrings_2.61.2             blob_1.2.2                   </span>
<span class="co">##  [83] DelayedMatrixStats_1.15.4     shape_1.4.6                  </span>
<span class="co">##  [85] beachmat_2.9.1                scales_1.1.1                 </span>
<span class="co">##  [87] memoise_2.0.0                 magrittr_2.0.1               </span>
<span class="co">##  [89] hexbin_1.28.2                 zlibbioc_1.39.0              </span>
<span class="co">##  [91] compiler_4.1.0                dqrng_0.3.0                  </span>
<span class="co">##  [93] pcaMethods_1.85.0             clue_0.3-59                  </span>
<span class="co">##  [95] cli_3.0.1                     XVector_0.33.0               </span>
<span class="co">##  [97] ggplot.multistats_1.0.0       MASS_7.3-54                  </span>
<span class="co">##  [99] tidyselect_1.1.1              stringi_1.7.4                </span>
<span class="co">## [101] textshaping_0.3.5             highr_0.9                    </span>
<span class="co">## [103] yaml_2.2.1                    BiocSingular_1.9.1           </span>
<span class="co">## [105] locfit_1.5-9.4                ggrepel_0.9.1                </span>
<span class="co">## [107] sass_0.4.0                    tools_4.1.0                  </span>
<span class="co">## [109] parallel_4.1.0                rio_0.5.27                   </span>
<span class="co">## [111] rstudioapi_0.13               MsCoreUtils_1.5.0            </span>
<span class="co">## [113] foreach_1.5.1                 bluster_1.3.2                </span>
<span class="co">## [115] foreign_0.8-81                metapod_1.1.0                </span>
<span class="co">## [117] gridExtra_2.3                 smoother_1.1                 </span>
<span class="co">## [119] scatterplot3d_0.3-41          farver_2.1.0                 </span>
<span class="co">## [121] digest_0.6.28                 BiocManager_1.30.16          </span>
<span class="co">## [123] FNN_1.1.3                     shiny_1.7.1                  </span>
<span class="co">## [125] Rcpp_1.0.7                    car_3.0-11                   </span>
<span class="co">## [127] broom_0.7.9                   BiocVersion_3.14.0           </span>
<span class="co">## [129] later_1.3.0                   httr_1.4.2                   </span>
<span class="co">## [131] AnnotationDbi_1.55.1          colorspace_2.0-2             </span>
<span class="co">## [133] rvest_1.0.1                   fs_1.5.0                     </span>
<span class="co">## [135] ranger_0.13.1                 uwot_0.1.10                  </span>
<span class="co">## [137] statmod_1.4.36                pkgdown_1.6.1                </span>
<span class="co">## [139] sp_1.4-5                      basilisk_1.5.0               </span>
<span class="co">## [141] renv_0.14.0                   systemfonts_1.0.2            </span>
<span class="co">## [143] xtable_1.8-4                  jsonlite_1.7.2               </span>
<span class="co">## [145] R6_2.5.1                      pillar_1.6.3                 </span>
<span class="co">## [147] htmltools_0.5.2               mime_0.12                    </span>
<span class="co">## [149] glue_1.4.2                    fastmap_1.1.0                </span>
<span class="co">## [151] VIM_6.1.1                     BiocParallel_1.27.12         </span>
<span class="co">## [153] BiocNeighbors_1.11.0          class_7.3-19                 </span>
<span class="co">## [155] interactiveDisplayBase_1.31.2 codetools_0.2-18             </span>
<span class="co">## [157] utf8_1.2.2                    lattice_0.20-45              </span>
<span class="co">## [159] bslib_0.3.0                   curl_4.3.2                   </span>
<span class="co">## [161] ggbeeswarm_0.6.0              magick_2.7.3                 </span>
<span class="co">## [163] zip_2.2.0                     openxlsx_4.2.4               </span>
<span class="co">## [165] limma_3.49.4                  rmarkdown_2.11               </span>
<span class="co">## [167] desc_1.4.0                    munsell_0.5.0                </span>
<span class="co">## [169] GetoptLong_1.0.5              e1071_1.7-9                  </span>
<span class="co">## [171] GenomeInfoDbData_1.2.7        iterators_1.0.13             </span>
<span class="co">## [173] haven_2.4.3                   gtable_0.3.0</span></code></pre></div>
</div>
</div>
<div id="reference" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Schoof2021-pv" class="csl-entry">
Schoof, Erwin M, Benjamin Furtwängler, Nil Üresin, Nicolas Rapin, Simonas Savickas, Coline Gentil, Eric Lechman, Ulrich auf Dem Keller, John E Dick, and Bo T Porse. 2021. <span>“Quantitative Single-Cell Proteomics as a Tool to Characterize Cellular Hierarchies.”</span> <em>Nat. Commun.</em> 12 (1): 745679.
</div>
<div id="ref-ijms22052470" class="csl-entry">
Trombetti, Silvia, Elena Cesaro, Rosa Catapano, Raffaele Sessa, Alessandra Lo Bianco, Paola Izzo, and Michela Grosso. 2021. <span>“Oxidative Stress and ROS-Mediated Signaling in Leukemia: Novel Promising Perspectives to Eradicate Chemoresistant Cells in Myeloid Leukemia.”</span> <em>International Journal of Molecular Sciences</em> 22 (5). <a href="https://doi.org/10.3390/ijms22052470">https://doi.org/10.3390/ijms22052470</a>.
</div>
<div id="ref-reticulate" class="csl-entry">
Ushey, Kevin, JJ Allaire, and Yuan Tang. 2021. <em>Reticulate: Interface to ’Python’</em>. <a href="https://CRAN.R-project.org/package=reticulate">https://CRAN.R-project.org/package=reticulate</a>.
</div>
<div id="ref-Vanderaa2021-ue" class="csl-entry">
Vanderaa, Christophe, and Laurent Gatto. 2021. <span>“Replication of Single-Cell Proteomics Data Reveals Important Computational Challenges.”</span> <em>Expert Rev. Proteomics</em>, October.
</div>
<div id="ref-Wickham2019-fz" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the Tidyverse.”</span> <em>J. Open Source Softw.</em> 4 (43): 1686.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
