<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reproduction of the SCoPE2 analysis (Specht et al. 2021) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Reproduction of the SCoPE2 analysis (Specht et al. 2021)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/derks2022.html">Replication of the plexDIA analysis (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/leduc2022.html">Replication of the nPOP analysis (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Replication of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Reproduction of the SCoPE2 analysis (Specht et
al. 2021)</h1>
                        <h4 data-toc-skip class="author">Christophe
Vanderaa<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, Computational Biology, UCLouvain</h4>
                        <h4 data-toc-skip class="author">Laurent Gatto,
Computational Biology, UCLouvain</h4>
            
            <h4 data-toc-skip class="date">19 September 2022</h4>
      
      
      <div class="hidden name"><code>SCoPE2.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Recent advances in sample preparation, processing and mass
      spectrometry (MS) have allowed the emergence of MS-based
      single-cell proteomics (SCP). This vignette presents a robust and
      standardized workflow to reproduce the data analysis of SCoPE2,
      one of the pioneering works in the field developed by the Slavov
      Lab. The implementation uses well-defined Bioconductor classes
      that provide powerful tools for single-cell RNA sequencing and for
      shotgun proteomics. We demonstrate that our pipeline can reproduce
      the SCoPE2 analysis using only a few lines of code.
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SCoPE2 (<span class="citation">Specht et al. (2021)</span>) is the
first mass spectrometry (MS)-based single cell proteomics (SCP) protocol
that has been used to profile thousands of proteins in thousands of
single-cells. This is a technical milestone for SCP and it opens the
door for fine-grain understanding of biological processes within and
between cells at the protein level. The emergence of SCP data leads to
the need of new computational developments that can deal with the
specificities and challenges of this new type of data.</p>
<p>Although the authors provided all code and data required for fully
reproducing their analysis, the code is difficult to read, implements
many functions from scratch, and is based on generic table formats not
suited for easy SCP data manipulation. This makes their workflow
difficult to re-use for other analyses. Our goal is to bridge the gap
between data acquisition and data interpretation by providing a new
computational framework that is tailored for MS-SCP data. The SCoPE2
dataset is an ideal dataset to showcase the application of our
standardized software solution.</p>
<p>Let’s first load the replication package to make use of some helper
functions. Those functions are only meant for this reproduction vignette
and are not designed for general use.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SCP.replication"</span><span class="op">)</span></span>
<span><span class="co">## Warning: multiple methods tables found for 'aperm'</span></span>
<span><span class="co">## Warning: replacing previous import 'BiocGenerics::aperm' by</span></span>
<span><span class="co">## 'DelayedArray::aperm' when loading 'SummarizedExperiment'</span></span></code></pre></div>
<div class="section level4">
<h4 id="scp-and-the-scope2-workflow">
<code>scp</code> and the SCoPE2 workflow<a class="anchor" aria-label="anchor" href="#scp-and-the-scope2-workflow"></a>
</h4>
<p>The SCoPE2 code provided along with the article can be retrieved from
<a href="https://github.com/SlavovLab/SCoPE2/blob/master/code/SCoPE2_analysis.R" class="external-link">this
GitHub repository</a>. Note that we here used the part of the script
that performs the stringent feature selection (see the original paper
for more details). The objective of this vignette is to replicate the
SCoPE2 script while providing standardized, easy-to-read, and well
documented code. Therefore, our first contribution was to formalize the
SCoPE2 script into a conceptual flow chart.</p>
<div class="figure" style="text-align: center">
<img src="figs/SCoPE2-workflow.png" alt="Overview of the SCoPE2 workflow." width="100%"><p class="caption">
Overview of the SCoPE2 workflow.
</p>
</div>
<p>Blue boxes highlight steps that are already routinely performed in
bulk proteomics. The orange boxes depict the steps that are specific to
single-cell applications and require new implementations. We have
developed a new data framework dedicated to SCP data analysis that
combines two existing Bioconductor classes. The
<code>SingleCellExperiment</code> class provides an interface to many
cutting edge methods for single-cell analysis and the
<code>QFeatures</code> class facilitates manipulation and processing of
MS-based quantitative data (blue boxes in the flowchart). The <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/scp/inst/doc/scp.html" class="external-link"><code>scp</code>
vignette</a> provides detailed information about the data structure. The
<code>scp</code> package extends the functionality of
<code>QFeatures</code> for single-cell application. <code>scp</code>
offers a standardized implementation of the single-cell steps in the
flowchart above (orange boxes).</p>
<p>The required packages for running this workflow are listed below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Core packages of this workflow</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures" class="external-link">QFeatures</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scpdata</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">scp</a></span><span class="op">)</span></span>
<span><span class="co">## Utility packages for data manipulation and visualization</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="scpdata-and-the-scope2-dataset">
<code>scpdata</code> and the SCoPE2 dataset<a class="anchor" aria-label="anchor" href="#scpdata-and-the-scope2-dataset"></a>
</h4>
<p>We also implemented a data package called <code>scpdata</code>. It
distributes published MS-SCP datasets, such as the SCoPE2 dataset. The
datasets were downloaded from the data source provided in the
publication and formatted to a <code>QFeatures</code> object so that it
is compatible with our software. The underlying data storage is based on
the <code>ExperimentHub</code> package that provides a cloud-based
storage infrastructure.</p>
<p>The SCoPE2 dataset (<span class="citation">Specht et al.
(2021)</span>) is provided at different levels of processing:</p>
<ul>
<li>The <strong>raw data</strong> files that were generated by the
mass-spectrometer software.</li>
<li>A <strong>PSM data</strong> table obtained from the MaxQuant
software that performs spectrum identification and quantification. PSM
stands for <em>peptide to spectrum match</em> and refers to a MS
spectrum that could successfully be assigned to a peptide sequence. The
provided PSM table was further processed using DART-ID (<span class="citation">Chen, Franks, and Slavov (2019)</span>) to improve the
identification rate.</li>
<li>A <strong>peptide data</strong> table obtained from the SCoPE2
analysis script. This is the data after the running the last step of the
pipeline at peptide level.</li>
<li>A <strong>protein data</strong> table obtained as the final table
from the SCoPE2 analysis script. This is the data after the running the
last step of the pipeline at protein level.</li>
</ul>
<p>The workflow starts with the PSM table and will generate the peptide
and the protein data. The authors provided the PSM dataset as a data
table (<code>data.frame</code>) in the R data environment called <a href="https://drive.google.com/drive/folders/1Zhjik_JFjCQNIVjg63-fooJ4K0HZxWjV" class="external-link"><code>raw.RData</code></a>.
Peptide and protein data are shared as CSV files. We highly value the
effort the authors of SCoPE2 have made to publicly share all the data
generated in their project, from raw files to final expression tables
(see the Slavov Lab <a href="http://scope2.slavovlab.net/" class="external-link">website</a>).</p>
<p>We formatted the SCoPE2 dataset following our data framework. The
formatted data can be retrieved from the <code>scpdata</code> package
using the <code><a href="https://rdrr.io/pkg/scpdata/man/specht2019v3.html" class="external-link">specht2019v3()</a></code> function. All datasets in
<code>scpdata</code> are called after the first author and the date of
publication (the preprint and data were first published in 2019).
<code>v3</code> stands for the third version of the data that was
released in October 2020.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/specht2019v3.html" class="external-link">specht2019v3</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The data contain 179 different <code>SingleCellExperiment</code>
objects that we refer to as <strong>assays</strong>. Each assay contains
expression data along with feature metadata. Each row in an assay
represents a <strong>feature</strong> that can either be a PSM, a
peptide or a protein depending on the assay. Each column in an assay
represents a <strong>sample</strong>. In the SCoPE2 data, samples are
acquired as a TMT channel because several sample are multiplexed in a
single MS run, hence each assay contains multiple columns. Most samples
are single-cells, but some samples are blanks, references, carriers, …
(see later). Below, we show the overview of the <code>scp</code>
dataset.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span></span>
<span><span class="co">## An instance of class QFeatures containing 179 assays:</span></span>
<span><span class="co">##  [1] 190222S_LCA9_X_FP94AA: SingleCellExperiment with 2777 rows and 11 columns </span></span>
<span><span class="co">##  [2] 190222S_LCA9_X_FP94AB: SingleCellExperiment with 4348 rows and 11 columns </span></span>
<span><span class="co">##  [3] 190222S_LCA9_X_FP94AC: SingleCellExperiment with 4917 rows and 11 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [177] 191110S_LCB7_X_APNOV16plex2_Set_9: SingleCellExperiment with 4934 rows and 16 columns </span></span>
<span><span class="co">##  [178] peptides: SingleCellExperiment with 9354 rows and 1490 columns </span></span>
<span><span class="co">##  [179] proteins: SingleCellExperiment with 3042 rows and 1490 columns</span></span></code></pre></div>
<p>177 out of the 179 assays are PSM data, each assay corresponding to a
separate MS run. 63 contain 11 columns and 114 contain 16 columns. This
is because the SCoPE2 protocol first included TMT-11 multiplexing, but
soon the TMT-16 multiplexing was released and the authors decided to
switch to the latter to increase the throughput of the technique. Notice
that the assays were also acquired in 4 chromatographic batches. Here is
an overview of the distribution of the assays across the TMT and
chromatographic batches.</p>
<p><img src="SCoPE2_files/figure-html/plot_lcbatch-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The dataset also contains a <code>peptides</code> assay and a
<code>proteins</code> assay that hold peptide and protein level
information, respectively. Those were provided by the authors. The
objective of this vignette is to produce the <code>peptides</code> and
<code>proteins</code> assays from the 177 PSM assays following the same
procedure as the SCoPE2 script but using standardized
functionalities.</p>
<p>We extract the <code>peptides</code> and <code>proteins</code> assays
and keep them for later benchmarking. Using double brackets
<code>[[...]]</code> extracts the desired assay as a
<code>SingleCellExperiment</code> object. On the other hand, using
simple brackets <code>[row, col, assay]</code> subsets the desired
elements/assays but preserves the <code>QFeatures</code> data
structure.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peptides_SCoPE2</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">proteins_SCoPE2</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[[</span><span class="st">"proteins"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span>, , <span class="op">-</span><span class="op">(</span><span class="fl">178</span><span class="op">:</span><span class="fl">179</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span></code></pre></div>
<p>To avoid naming confusions during later benchmarking of the
replication, we name the peptide and protein data sets generated by the
SCoPE2 script <code>peptides_SCoPE2</code> and
<code>proteins_SCoPE2</code>, respectively.</p>
</div>
</div>
<div class="section level2">
<h2 id="filter-the-psm-data">Filter the PSM data<a class="anchor" aria-label="anchor" href="#filter-the-psm-data"></a>
</h2>
<p>After importing the data, the SCoPE2 analysis filters low-confidence
PSMs. Each PSM assay contains feature meta-information that are stored
in the assay <code>rowData</code>. The <code>QFeatures</code> package
allows to quickly filter the rows of an assay by using these
information. The available variables in the <code>rowData</code> are
listed below for each assay.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rowDataNames</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span></span>
<span><span class="co">## CharacterList of length 177</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AA"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AB"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AC"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AD"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AE"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AF"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AG"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AH"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AI"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## [["190222S_LCA9_X_FP94AJ"]] uid Sequence Length ... digest modseq peptide</span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## &lt;167 more elements&gt;</span></span></code></pre></div>
<div class="section level3">
<h3 id="filter-out-failed-runs-based-on-psm-content">Filter out failed runs based on PSM content<a class="anchor" aria-label="anchor" href="#filter-out-failed-runs-based-on-psm-content"></a>
</h3>
<p>First, only the assays that have sufficient PSMs are kept. The
authors keep an assay if it has over 500 PSMs. Before filtering, let’s
first look at the distribution of the number of PSMs per assay. Note
that we can easily extract the number of rows (here PSMs) and the number
of columns (TMT channels) of each assay using the <code>dims</code>
function implemented in <code>QFeatures</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nPSMs</span> <span class="op">&lt;-</span> <span class="fu">dims</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span></span></code></pre></div>
<p>Let’s have a look at the number of features that were identified in
the different runs.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">nPSMs</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nPSMs</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<p><img src="SCoPE2_files/figure-html/plot_dims-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>4 assays have very few number of PSMs, probably because those runs
failed. They are hence below the threshold of 500 and are removed from
the analysis. To perform this, we take advantage of the subsetting
method of a <code>QFeatures</code> object. It can be seen as a
three-order array: <span class="math inline">\(features \times samples
\times assay\)</span>. Hence, <code>QFeatures</code> supports
three-order subsetting <code>x[rows, columns, assays]</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span>, , <span class="va">nPSMs</span> <span class="op">&gt;</span> <span class="fl">500</span><span class="op">]</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span></code></pre></div>
<p>You may notice that a warning is thrown when performing the
subsetting step. This is because, as we expected, a few assays were
dropped. Removed assays are listed in the <code>metadata</code>
slot:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadata</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span></span>
<span><span class="co">## $drops.experiments</span></span>
<span><span class="co">## [1] "peptides"                          "proteins"                         </span></span>
<span><span class="co">## [3] "190222S_LCA9_X_FP94BD"             "191026S_LCB7_X_AP16plex_Set_15"   </span></span>
<span><span class="co">## [5] "191108S_LCB7_X_APNOV16plex_Set_24" "191108S_LCB7_X_APNOV16plex_Set_29"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-out-psms-with-high-false-discovery-rate">Filter out PSMs with high false discovery rate<a class="anchor" aria-label="anchor" href="#filter-out-psms-with-high-false-discovery-rate"></a>
</h3>
<p>Next, SCoPE2 filters PSMs based on the false discovery rate (FDR) for
identification. The PSM data were already processed with DART-ID (<span class="citation">Chen, Franks, and Slavov (2019)</span>), a python
software that updates the confidence in peptide identification using an
Bayesian inference approach. DART-ID outputs for every PSM the updated
posterior error probability (PEP). Filtering on the PEP is too
conservative and it is rather advised to filter based on FDR (<span class="citation">Käll et al. (2008)</span>). To control for the FDR, we
need to compute q-values, that correspond to the minimal FDR threshold
that would still select the associated feature.</p>
<p>We can use the <code>pep2qvalue</code> function to easily compute
q-values from the PEPs computed by MaxQuant or updated by DART-ID. In
the SCoPE2 workflow, the features are selected based on the FDR at PSM
level and at protein level. The DART-ID PEPs (<code>dart_PEP</code>) as
well as the protein information (<code>protein</code>) are automatically
retrieved from the <code>rowData</code> of each assay. The function will
store the computed q-values back in the <code>rowData</code>. It will be
stored under <code>qvalue_psm</code> and <code>qvalue_protein</code> for
the PSM and proteins q-values, respectively.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/pep2qvalue.html" class="external-link">pep2qvalue</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                  i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,</span>
<span>                  PEP <span class="op">=</span> <span class="st">"dart_PEP"</span>,</span>
<span>                  rowDataName <span class="op">=</span> <span class="st">"qvalue_psm"</span><span class="op">)</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/pep2qvalue.html" class="external-link">pep2qvalue</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                  i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,</span>
<span>                  groupBy <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>                  PEP <span class="op">=</span> <span class="st">"dart_PEP"</span>,</span>
<span>                  rowDataName <span class="op">=</span> <span class="st">"qvalue_protein"</span><span class="op">)</span></span></code></pre></div>
<p>We can extract the q-values from the <code>rowData</code> of several
assays using the <code>rbindRowData</code> function. It takes the
<code>rowData</code> of interest and returns a single
<code>DataFrame</code> table with variables of interest. We plot the
q-values along with the DART-ID PEPs using <code>ggplot2</code>
facets.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dart_PEP"</span>, <span class="st">"qvalue_psm"</span>, <span class="st">"qvalue_protein"</span><span class="op">)</span>,</span>
<span>                 names_to <span class="op">=</span> <span class="st">"measure"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span>rows <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">measure</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Warning: Transformation introduced infinite values in continuous x-axis</span></span>
<span><span class="co">## Warning: Removed 6690 rows containing non-finite values (stat_bin).</span></span></code></pre></div>
<p><img src="SCoPE2_files/figure-html/qvalue_plot-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>We filter the PSMs to control for a 1% PSM and protein FDR. We can
perform this on our <code>QFeatures</code> object by using the
<code>filterFeatures</code> function. The q-values are directly accessed
from the <code>rowData</code> of each assay.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                      <span class="op">~</span> <span class="va">qvalue_psm</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">qvalue_protein</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-out-contaminants">Filter out contaminants<a class="anchor" aria-label="anchor" href="#filter-out-contaminants"></a>
</h3>
<p>We will now remove PSMs that were matched to contaminant proteins
(the protein name starts with <code>CON</code>) or to the decoy database
(the protein name starts with <code>REV</code>). Again,
<code>filterFeatures</code> can directly access the protein names from
the <code>rowData</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                      <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"REV|CON"</span>, <span class="va">protein</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-out-noisy-spectra">Filter out noisy spectra<a class="anchor" aria-label="anchor" href="#filter-out-noisy-spectra"></a>
</h3>
<p>A PIF (parental ion fraction) smaller than 80 % indicates the
associated spectra is contaminated by co-isolated peptides and therefore
the quantification becomes unreliable. The PIF was computed by MaxQuant
and is readily available for filtering.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                      <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">PIF</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">PIF</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-out-psms-with-high-sample-to-carrier-ratio">Filter out PSMs with high sample to carrier ratio<a class="anchor" aria-label="anchor" href="#filter-out-psms-with-high-sample-to-carrier-ratio"></a>
</h3>
<p>The PSMs are next filtered based on the sample to carrier ratio
(SCR), that is the TMT ion intensity of a single-cell sample divided by
the TMT ion intensity of the carrier (200 cells) acquired during the
same run as the sample. It is expected that the carrier intensities are
much higher than the single-cell intensities. We implemented the
<code>computeSCR</code> function that computes the SCR for each PSM
averaged over all samples of interest in a given assay. A PSM is removed
when the mean SCR exceeds 10 %. To perform this, we need to tell the
function which columns are the samples of interest and which column is
the carrier. The <code>colData</code> of the <code>QFeatures</code>
object is used to define this.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">scp</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      Blank    Carrier Macrophage   Monocyte  Reference     Unused </span></span>
<span><span class="co">##        168        173       1198        462        173        284</span></span></code></pre></div>
<p>In this dataset, <code>SampleType</code> gives the type of sample
that is present in each TMT channel. The SCoPE2 protocol includes 5
types of samples:</p>
<ul>
<li>The carrier channels (<code>Carrier</code>) contain 200 cell
equivalents and are meant to boost the peptide identification rate.</li>
<li>The normalization channels (<code>Reference</code>) contain 5 cell
equivalents and are used to partially correct for between-run
variation.</li>
<li>The unused channels (<code>Unused</code>) are channels that are left
empty due to isotopic cross-contamination by the carrier channel.</li>
<li>The blank channels (<code>Blank</code>) contain samples that do not
contain any cell but are processed as single-cell samples.</li>
<li>The single-cell sample channels contain the single-cell samples of
interest (<code>Macrophage</code> or <code>Monocyte</code>).</li>
</ul>
<p>The <code>computeSCR</code> function expects the user to provide a
pattern (following regular expression syntax) that uniquely identifies a
carrier channel in each run and the samples or blanks. The function will
store the mean SCR of each feature in the <code>rowData</code> of each
assay.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/computeSCR.html" class="external-link">computeSCR</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                  i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,</span>
<span>                  colvar <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>                  carrierPattern <span class="op">=</span> <span class="st">"Carrier"</span>,</span>
<span>                  samplePattern <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fl">16</span>,</span>
<span>                  rowDataName <span class="op">=</span> <span class="st">"MeanSCR"</span><span class="op">)</span></span>
<span><span class="co">## Warning in computeSCR(scp, i = names(scp), colvar = "SampleType", carrierPattern</span></span>
<span><span class="co">## = "Carrier", : The pattern is numeric. This is only allowed for replicating the</span></span>
<span><span class="co">## SCoPE2 analysis and will later get defunct.</span></span></code></pre></div>
<p>In this step, we supplied numeric entries as
<code>samplePattern</code> instead of a regular expression pattern. This
is to replicate the SCoPE2 analysis. We refrain from using a numeric
pattern (as indicated by the warning above) because this can lead to
unnoticed artefacts. In fact, the empty channels of the TMT-16 runs but
not TMT-11 channels are here also used to compute the average SCR which
could have been avoided when using a character pattern.</p>
<p>Before applying the filter, we plot the distribution of the mean
SCR.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">MeanSCR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">200</span>, <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>               lty <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SCoPE2_files/figure-html/meanSCR_plot-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>A great majority of the PSMs have a mean SCR that is lower than 10%,
as expected. Interestingly, the mode of the distribution is located
close to 1%. This is expected since every sample channel contains a
single-cell and the carrier contains 200 cells leading to an expected
ratio of 0.5% (dashed line). We remove the PSMs for which the mean SCR
exceeds the 10% threshold.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                      <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">MeanSCR</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                          <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html" class="external-link">is.infinite</a></span><span class="op">(</span><span class="va">MeanSCR</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                          <span class="va">MeanSCR</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="normalize-to-reference">Normalize to reference<a class="anchor" aria-label="anchor" href="#normalize-to-reference"></a>
</h2>
<p>In order to partially correct for between-run variation, SCoPE2
computes relative reporter ion intensities. This means that intensities
measured for single-cells are divided by the reference channel (5-cell
equivalents). We use the <code>divideByReference</code> function that
divides channels of interest by the reference channel. Similarly to
<code>computeSCR</code>, we can point to the samples and the reference
columns in each assay using the annotation contained in the
<code>colData</code>. We will here divide all columns (using the regular
expression wildcard <code>.</code>) by the reference channel
(<code>Reference</code>). Notice that when taking all samples we also
include the reference channel itself. Hence, from now on, the reference
channels will contain only ones.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/divideByReference.html" class="external-link">divideByReference</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                         i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,</span>
<span>                         colvar <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>                         samplePattern <span class="op">=</span> <span class="st">"."</span>,</span>
<span>                         refPattern <span class="op">=</span> <span class="st">"Reference"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="aggregate-psm-data-to-peptide-data">Aggregate PSM data to peptide data<a class="anchor" aria-label="anchor" href="#aggregate-psm-data-to-peptide-data"></a>
</h2>
<p>Now that the PSM assays are processed, we can aggregate them to
peptides. This is performed using the
<code>aggregateFeaturesOverAssays</code> function. This is a wrapper
function in <code>scp</code> that sequentially calls the
<code>aggregateFeatures</code> from the <code>QFeatures</code> package
over the different assays. For each assay, the function aggregates
several PSMs into a unique peptide given an aggregating variable in the
<code>rowData</code> (peptide sequence) and a user-supplied aggregating
function (the median for instance). Regarding the aggregating function,
the SCoPE2 analysis removes duplicated peptide sequences per run by
taking the first non-missing value. While better alternatives are
documented in <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">QFeatures::aggregateFeatures</a></code>, we still use
this approach for the sake of replication and for illustrating that
custom functions can be applied.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">remove.duplicates</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">xx</span><span class="op">)</span> <span class="va">xx</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">xx</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<p>The aggregated peptide assays must be given a name. We here used the
original names with <code>peptides_</code> at the start.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peptideAssays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"peptides_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We now have all the required information to aggregate the PSMs in the
different batches to peptides.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/aggregateFeaturesOverAssays.html" class="external-link">aggregateFeaturesOverAssays</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                                   i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,</span>
<span>                                   fcol <span class="op">=</span> <span class="st">"peptide"</span>,</span>
<span>                                   name <span class="op">=</span> <span class="va">peptideAssays</span>,</span>
<span>                                   fun <span class="op">=</span> <span class="va">remove.duplicates</span><span class="op">)</span></span></code></pre></div>
<p>Under the hood, the <code>QFeatures</code> architecture preserves the
relationship between the aggregated assays. See <code><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">?AssayLinks</a></code>
for more information on relationships between assays. Notice that
<code>aggregateFeaturesOverAssays</code> created as many new assays as
the number of supplied assays.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span></span>
<span><span class="co">## An instance of class QFeatures containing 346 assays:</span></span>
<span><span class="co">##  [1] 190222S_LCA9_X_FP94AA: SingleCellExperiment with 1098 rows and 11 columns </span></span>
<span><span class="co">##  [2] 190222S_LCA9_X_FP94AB: SingleCellExperiment with 2420 rows and 11 columns </span></span>
<span><span class="co">##  [3] 190222S_LCA9_X_FP94AC: SingleCellExperiment with 2775 rows and 11 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [344] peptides_191110S_LCB7_X_APNOV16plex2_Set_7: SingleCellExperiment with 2379 rows and 16 columns </span></span>
<span><span class="co">##  [345] peptides_191110S_LCB7_X_APNOV16plex2_Set_8: SingleCellExperiment with 2480 rows and 16 columns </span></span>
<span><span class="co">##  [346] peptides_191110S_LCB7_X_APNOV16plex2_Set_9: SingleCellExperiment with 2308 rows and 16 columns</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="join-the-scope2-sets-in-one-assay">Join the SCoPE2 sets in one assay<a class="anchor" aria-label="anchor" href="#join-the-scope2-sets-in-one-assay"></a>
</h2>
<p>Up to now, we kept the data belonging to each MS run in separate
assays. We now combine all batches into a single assay. This can easily
be done using the <code>joinAssays</code> function from the
<code>QFeatures</code> package.</p>
<div class="section level3">
<h3 id="consensus-mapping-of-peptides-to-proteins">Consensus mapping of peptides to proteins<a class="anchor" aria-label="anchor" href="#consensus-mapping-of-peptides-to-proteins"></a>
</h3>
<p>We need to account for an issue in the data. <code>joinAssays</code>
will only keep the metadata variables that have the same value between
matching rows. However, some peptide sequences map to one protein in one
run and to another protein in another run. Hence, the protein sequence
is not constant for all peptides and is removed during joining. It is
important we keep the protein sequence in the <code>rowData</code> since
we will later need it to aggregate peptides to proteins. To avoid this
issue, we replace the problematic peptides to protein mappings through a
majority vote.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">173</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">peptide</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">## The majority vote happens here</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>protein <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">protein</span><span class="op">)</span>,</span>
<span>                                decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">peptide</span>, <span class="va">protein</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">peptide</span>, <span class="va">protein</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span></span>
<span>    <span class="va">ppMap</span></span>
<span><span class="va">consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">peptideAssays</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">peptide</span>, <span class="va">ppMap</span><span class="op">$</span><span class="va">peptide</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span>protein <span class="op">=</span> <span class="va">ppMap</span><span class="op">$</span><span class="va">protein</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">consensus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">peptideAssays</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">consensus</span></span></code></pre></div>
<p>This code chunk is hard to read, but this issue should represent a
special case and we therefore don’t consider this step worth
standardizing. Some peptides are mapped to different proteins because
the TMT-11 and TMT-16 data sets were analysed separately in MaxQuant.
Razor peptides, <em>i.e.</em> peptides that are found in several protein
groups, are assigned to the protein group with most peptides (<span class="citation">Tyanova, Temu, and Cox (2016)</span>,) and this can
vary from one analysis to another. Therefore, inconsistent peptide
mapping should not occur when all runs are preprocessed at once. We will
see later that this issue does not impair the replication of the SCoPE2
analysis, but it recalls that mapping of shared peptides to proteins is
not a trivial task.</p>
</div>
<div class="section level3">
<h3 id="cleaning-missing-data">Cleaning missing data<a class="anchor" aria-label="anchor" href="#cleaning-missing-data"></a>
</h3>
<p>Another important step before we join the assays is to replace zero
and infinite values by <code>NA</code>s. The zeros can be biological
zeros or technical zeros and differentiating between the two types is a
difficult task, they are therefore better considered as missing. The
infinite values arose during the normalization by the reference because
the channel values are divide by a zero from the reference channel. This
artefact could easily be avoided if we had replace the zeros by
<code>NA</code>s at the beginning of the workflow, what we strongly
recommend for future analyses.</p>
<p>The <code>infIsNA</code> and the <code>zeroIsNA</code> functions
automatically detect infinite and zero values, respectively, and replace
them with <code>NA</code>s. Those two functions are provided by the
<code>QFeatures</code> package.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">infIsNA</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="va">peptideAssays</span><span class="op">)</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="va">peptideAssays</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="join-assays">Join assays<a class="anchor" aria-label="anchor" href="#join-assays"></a>
</h3>
<p>Now that the peptides are correctly matched to proteins and missing
values are correctly formatted, we can join the assays.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                  i <span class="op">=</span> <span class="va">peptideAssays</span>,</span>
<span>                  name <span class="op">=</span> <span class="st">"peptides"</span><span class="op">)</span></span></code></pre></div>
<p><code>joinAssays</code> has created a new assay called
<code>peptides</code> that combines the previously aggregated peptide
assays.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span></span>
<span><span class="co">## An instance of class QFeatures containing 347 assays:</span></span>
<span><span class="co">##  [1] 190222S_LCA9_X_FP94AA: SingleCellExperiment with 1098 rows and 11 columns </span></span>
<span><span class="co">##  [2] 190222S_LCA9_X_FP94AB: SingleCellExperiment with 2420 rows and 11 columns </span></span>
<span><span class="co">##  [3] 190222S_LCA9_X_FP94AC: SingleCellExperiment with 2775 rows and 11 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [345] peptides_191110S_LCB7_X_APNOV16plex2_Set_8: SingleCellExperiment with 2480 rows and 16 columns </span></span>
<span><span class="co">##  [346] peptides_191110S_LCB7_X_APNOV16plex2_Set_9: SingleCellExperiment with 2308 rows and 16 columns </span></span>
<span><span class="co">##  [347] peptides: SingleCellExperiment with 15675 rows and 2458 columns</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="filter-single-cells-based-on-median-cv">Filter single-cells based on median CV<a class="anchor" aria-label="anchor" href="#filter-single-cells-based-on-median-cv"></a>
</h2>
<p>The SCoPE2 script proceeds with filtering the single-cells. The
filtering is mainly based on the median coefficient of variation (CV)
per cell. The median CV measures the consistency of quantification for a
group of peptides that belong to a protein. We remove cells that exhibit
high median CV over the different proteins. We compute the median CV per
cell using the <code>medianCVperCell</code> function from the
<code>scp</code> package. The function takes the protein information
from the <code>rowData</code> of the assays that will tell how to group
the features (peptides) when computing the CV. Note that we supply the
peptide assays before joining in a single assays
(<code>i = peptideAssays</code>). This is because SCoPE2 performs a
custom normalization (<code>norm = "SCoPE2"</code>). Each row in an
assay is normalized by a scaling factor. This scaling factor is the row
mean after dividing the columns by the median. The authors retained CVs
that are computed using at least 6 peptides (<code>nobs = 6</code>). See
the methods section in <span class="citation">Specht et al.
(2021)</span> for more information.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/medianCVperCell.html" class="external-link">medianCVperCell</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                       i <span class="op">=</span> <span class="va">peptideAssays</span>,</span>
<span>                       groupBy <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>                       nobs <span class="op">=</span> <span class="fl">6</span>,</span>
<span>                       na.rm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       colDataName <span class="op">=</span> <span class="st">"MedianCV"</span>,</span>
<span>                       norm <span class="op">=</span> <span class="st">"SCoPE2"</span><span class="op">)</span></span></code></pre></div>
<p>The computed CVs are stored in the <code>colData</code>. We can now
filter cells that have reliable quantifications. The blank samples are
not expected to have reliable quantifications and hence can be used to
estimate a null distribution of the CV. This distribution helps defining
a threshold that filters out single-cells that contain noisy
quantification.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">SampleType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Macrophage"</span>,  <span class="st">"Monocyte"</span>, <span class="st">"Blank"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">MedianCV</span>,</span>
<span>               fill <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.365</span><span class="op">)</span></span></code></pre></div>
<p><img src="SCoPE2_files/figure-html/medianCV_plot-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>We can see that the protein quantification for single-cells are much
more consistent within single-cell channels than within blank channels.
A threshold of 0.365 best separates single-cells from empty
channels.</p>
<p>We keep the cells that pass the median CV threshold. Furthermore, we
keep macrophages and monocytes as those represent the samples of
interest. Since this information is available from the
<code>colData</code>, we can apply this filter using
<code><a href="http://waldronlab.io/MultiAssayExperiment/reference/subsetBy.html" class="external-link">subsetByColData()</a></code> from the
<code>MultiAssayExperiment</code> package.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/subsetBy.html" class="external-link">subsetByColData</a></span><span class="op">(</span><span class="va">scp</span>, <span class="va">scp</span><span class="op">$</span><span class="va">MedianCV</span> <span class="op">&lt;</span> <span class="fl">0.365</span> <span class="op">&amp;</span> </span>
<span>                       <span class="va">scp</span><span class="op">$</span><span class="va">SampleType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Macrophage"</span>, <span class="st">"Monocyte"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">scp</span></span>
<span><span class="co">## An instance of class QFeatures containing 347 assays:</span></span>
<span><span class="co">##  [1] 190222S_LCA9_X_FP94AA: SingleCellExperiment with 1098 rows and 4 columns </span></span>
<span><span class="co">##  [2] 190222S_LCA9_X_FP94AB: SingleCellExperiment with 2420 rows and 5 columns </span></span>
<span><span class="co">##  [3] 190222S_LCA9_X_FP94AC: SingleCellExperiment with 2775 rows and 6 columns </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  [345] peptides_191110S_LCB7_X_APNOV16plex2_Set_8: SingleCellExperiment with 2480 rows and 11 columns </span></span>
<span><span class="co">##  [346] peptides_191110S_LCB7_X_APNOV16plex2_Set_9: SingleCellExperiment with 2308 rows and 11 columns </span></span>
<span><span class="co">##  [347] peptides: SingleCellExperiment with 15675 rows and 1527 columns</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="process-the-peptide-data">Process the peptide data<a class="anchor" aria-label="anchor" href="#process-the-peptide-data"></a>
</h2>
<p>In the SCoPE2 analysis, the peptide data is first transformed before
it is aggregated to proteins. The transformation steps are:
normalization, filter peptides based on missing data and
log-transformation.</p>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>The columns (samples) of the peptide data are first normalized by
dividing the relative intensities by the median relative intensities.
Then, the rows (peptides) are normalized by dividing the relative
intensities by the mean relative intensities. The first normalization is
available from the <code>normalize</code> function and accessible under
the <code>div.median</code> method. The second is not available from
<code>normalize</code>, but is easily performed using the
<code>sweep</code> function from the <code>QFeatures</code> package that
is inspired from the <code><a href="https://rdrr.io/r/base/sweep.html" class="external-link">base::sweep</a></code> function.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Scale column with median</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                 i <span class="op">=</span> <span class="st">"peptides"</span>,</span>
<span>                 method <span class="op">=</span> <span class="st">"div.median"</span>,</span>
<span>                 name <span class="op">=</span> <span class="st">"peptides_norm1"</span><span class="op">)</span></span>
<span><span class="co">## Scale rows with mean</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>             i <span class="op">=</span> <span class="st">"peptides_norm1"</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>             FUN <span class="op">=</span> <span class="st">"/"</span>,</span>
<span>             STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"peptides_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                              na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>             name <span class="op">=</span> <span class="st">"peptides_norm2"</span><span class="op">)</span></span></code></pre></div>
<p>Each normalization step is stored in a separate assay. An important
aspect to note here is that</p>
</div>
<div class="section level3">
<h3 id="remove-peptides-with-high-missing-rate">Remove peptides with high missing rate<a class="anchor" aria-label="anchor" href="#remove-peptides-with-high-missing-rate"></a>
</h3>
<p>Peptides that contain many missing values are not informative.
Therefore, we remove those with more than 99 % missing data. This is
done using the <code>filterNA</code> function from
<code>QFeatures</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterNA</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                i <span class="op">=</span> <span class="st">"peptides_norm2"</span>,</span>
<span>                pNA <span class="op">=</span> <span class="fl">0.99</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="log-transformation">Log-transformation<a class="anchor" aria-label="anchor" href="#log-transformation"></a>
</h3>
<p>The last processing step of the peptide data before aggregating to
proteins is to log-transform the data. SCoPE2 uses a base 2
log-transformation and this is implemented in <code>logTransform</code>
in <code>QFeatures</code>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                    base <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                    i <span class="op">=</span> <span class="st">"peptides_norm2"</span>,</span>
<span>                    name <span class="op">=</span> <span class="st">"peptides_log"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-scope2-peptide-data">The SCoPE2 peptide data<a class="anchor" aria-label="anchor" href="#the-scope2-peptide-data"></a>
</h3>
<p>Before exporting the data as <code>Peptides-raw.csv</code>, the
authors performed an additional normalization step. This step is however
not considered in the remainder of the workflow.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Center columns with median</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                 i <span class="op">=</span> <span class="st">"peptides_log"</span>,</span>
<span>                 method <span class="op">=</span> <span class="st">"center.median"</span>,</span>
<span>                 name <span class="op">=</span> <span class="st">"peptides_log_norm1"</span><span class="op">)</span></span>
<span><span class="co">## Center rows with mean</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>             i <span class="op">=</span> <span class="st">"peptides_log_norm1"</span>,</span>
<span>             MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>             FUN <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>             STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"peptides_log_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                              na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>             name <span class="op">=</span> <span class="st">"peptides_scp"</span><span class="op">)</span></span></code></pre></div>
<p>We name the last assay <code>peptides_scp</code>. We will contrast
this data matrix to <code>peptides_SCoPE2</code> later in this vignette
to assess the accuracy of the replication.</p>
</div>
</div>
<div class="section level2">
<h2 id="aggregate-peptide-data-to-protein-data">Aggregate peptide data to protein data<a class="anchor" aria-label="anchor" href="#aggregate-peptide-data-to-protein-data"></a>
</h2>
<p>Similarly to aggregating PSM data to peptide data, we can aggregate
peptide data to protein data using the <code>aggregateFeatures</code>
function. Note that we here use the median as a summarizing
function.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                         i <span class="op">=</span> <span class="st">"peptides_log"</span>,</span>
<span>                         name <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                         fcol <span class="op">=</span> <span class="st">"protein"</span>,</span>
<span>                         fun <span class="op">=</span> <span class="fu">matrixStats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html" class="external-link">colMedians</a></span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="process-the-protein-data">Process the protein data<a class="anchor" aria-label="anchor" href="#process-the-protein-data"></a>
</h2>
<p>The protein data is processed in three steps: normalization,
imputation and batch correction.</p>
<div class="section level3">
<h3 id="normalization-1">Normalization<a class="anchor" aria-label="anchor" href="#normalization-1"></a>
</h3>
<p>Normalization is performed similarly to peptide normalization. We use
the same functions, but since the data were log-transformed at the
peptide level, we subtract/center by the statistic (median or mean)
instead of dividing.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Center columns with median</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                 i <span class="op">=</span> <span class="st">"proteins"</span>,</span>
<span>                 method <span class="op">=</span> <span class="st">"center.median"</span>,</span>
<span>                 name <span class="op">=</span> <span class="st">"proteins_norm1"</span><span class="op">)</span></span>
<span><span class="co">## Center rows with mean</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>             i <span class="op">=</span> <span class="st">"proteins_norm1"</span>,</span>
<span>             MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>             FUN <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>             STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                              na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>             name <span class="op">=</span> <span class="st">"proteins_norm2"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="imputation">Imputation<a class="anchor" aria-label="anchor" href="#imputation"></a>
</h3>
<p>The protein data contains a lot of missing values. The graph below
shows the distribution of the proportion missingness in cells. Cells
contain on average over 75 % missing values!</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">longFormat</span><span class="op">(</span><span class="va">scp</span><span class="op">[</span>, , <span class="st">"proteins_norm2"</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">colname</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>missingness <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">missingness</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span></code></pre></div>
<p><img src="SCoPE2_files/figure-html/missing_plot-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The missing data is imputed using K nearest neighbors. The SCoPE2
script runs KNN with k = 3. We made a wrapper around the author’s code
to apply imputation to our <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/imputeKnnSCoPE2.html">imputeKnnSCoPE2</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                       i <span class="op">=</span> <span class="st">"proteins_norm2"</span>,</span>
<span>                       name <span class="op">=</span> <span class="st">"proteins_impd"</span>,</span>
<span>                       k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><code>QFeatures</code> provides the <code>impute</code> function that
serves as an interface to different imputation algorithms among which
the KNN algorithm from <code>impute::impute.knn</code>. However, the KNN
implementation in SCoPE2 and in <code>impute.knn</code> are different.
SCoPE2 performs KNN imputation in the sample space, meaning that data
from neighbouring cells are used to impute the central cell, whereas
<code>impute::impute.knn</code> performs KNN imputation in the feature
space, meaning that data from neighbouring features are used to impute
the missing values from the central features. We provide the code for
KNN imputation with <code>QFeatures</code> but do not run in order to
reproduce the SCoPE2 analysis.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>              i <span class="op">=</span> <span class="st">"proteins_norm2"</span>,</span>
<span>              method <span class="op">=</span> <span class="st">"knn"</span>,</span>
<span>              k <span class="op">=</span> <span class="fl">3</span>, rowmax <span class="op">=</span> <span class="fl">1</span>, colmax<span class="op">=</span> <span class="fl">1</span>,</span>
<span>              maxp <span class="op">=</span> <span class="cn">Inf</span>, rng.seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h3>
<p>The final step is to model the remaining batch effects and correct
for it. The data were acquired as a series of MS runs. Recall we had 177
assays at the beginning of the workflow. Each MS run can be subjected to
technical perturbations that lead to differences in the data. This must
be accounted for to avoid attributing biological effects to technical
effects. The <code>ComBat</code> algorithm (<span class="citation">Johnson, Li, and Rabinovic (2007)</span>) is used in
the SCoPE2 script to correct for those batch effects.
<code>ComBat</code> is part of the <code>sva</code> package. It requires
a batch variable, in this case the LC-MS/MS run, and adjusts for batch
effects, while protecting variables of interest, the sample type in this
case. All the information is contained in the <code>colData</code> of
the <code>QFeatures</code> object. We first extract the assays with the
associated <code>colData</code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-helpers.html" class="external-link">getWithColData</a></span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_impd"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span></code></pre></div>
<p>We next store the batch variable and create the design matrix. We
then perform the batch correction and overwrite the data matrix. Recall
the data matrix can be accessed using the <code>assay</code> function.
Note also that we here use the <code>ComBatv3.34</code> function that is
provided in this package. This function was taken from an older version
of the <code>sva</code> package (version <code>3.34.0</code>). This is
to replicate the SCoPE2 results. More recent versions will avoid batch
correction for proteins that show no variance within at least one batch
and this occurs for a significant proportion of the proteins (artefact
of imputation).</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Set</span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">SampleType</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ComBatv3.34.html">ComBatv3.34</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span>                          batch <span class="op">=</span> <span class="va">batch</span>,</span>
<span>                          mod <span class="op">=</span> <span class="va">model</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we add the batch corrected assay to the
<code>QFeatures</code> object and create the feature links.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sce</span>,</span>
<span>             name <span class="op">=</span> <span class="st">"proteins_batchC"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLinkOneToOne</a></span><span class="op">(</span>from <span class="op">=</span> <span class="st">"proteins_impd"</span>,</span>
<span>                         to <span class="op">=</span> <span class="st">"proteins_batchC"</span><span class="op">)</span> <span class="op">-&gt;</span></span>
<span>    <span class="va">scp</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-scope2-protein-data">The SCoPE2 protein data<a class="anchor" aria-label="anchor" href="#the-scope2-protein-data"></a>
</h3>
<p>Before exporting the data as <code>Proteins-processed.csv</code>, the
authors performed an additional normalization step.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Center columns with median</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>                 i <span class="op">=</span> <span class="st">"proteins_batchC"</span>,</span>
<span>                 method <span class="op">=</span> <span class="st">"center.median"</span>,</span>
<span>                 name <span class="op">=</span> <span class="st">"proteins_batchC_norm1"</span><span class="op">)</span></span>
<span><span class="co">## Center rows with mean</span></span>
<span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">scp</span>,</span>
<span>             i <span class="op">=</span> <span class="st">"proteins_batchC_norm1"</span>,</span>
<span>             MARGIN <span class="op">=</span> <span class="fl">1</span>,</span>
<span>             FUN <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>             STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_batchC_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                              na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>             name <span class="op">=</span> <span class="st">"proteins_scp"</span><span class="op">)</span></span></code></pre></div>
<p>We named the last assay <code>proteins_scp</code>. We will contrast
this data matrix to <code>proteins_SCoPE2</code> later in the next
section to assess the accuracy of the replication.</p>
</div>
</div>
<div class="section level2">
<h2 id="benchmarking-the-replication">Benchmarking the replication<a class="anchor" aria-label="anchor" href="#benchmarking-the-replication"></a>
</h2>
<p>We will now compare the data that were provided by Specht and
colleagues with the data generated in this vignette. We will compare the
peptide and the protein data.</p>
<div class="section level3">
<h3 id="compare-the-selected-cells">Compare the selected cells<a class="anchor" aria-label="anchor" href="#compare-the-selected-cells"></a>
</h3>
<p>One of the above section filtered the single cells based on the
median coefficient of variation per cell. We will check that we could
indeed reproduce the SCoPE2 cell filter. Note that the peptide and
protein data contain the same set of cells.</p>
<p><img src="SCoPE2_files/figure-html/benchm_cell_filt-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The <code>scp</code> filtering step keeps 37 additional cells
compared to SCoPE2.</p>
</div>
<div class="section level3">
<h3 id="compare-peptide-data">Compare peptide data<a class="anchor" aria-label="anchor" href="#compare-peptide-data"></a>
</h3>
<p>The SCoPE2 and the <code>scp</code> peptide data have comparable
dimensions, although they are not exactly the same. The <code>scp</code>
peptide data contains 17 peptides that were not selected by SCoPE2 and
SCoPE2 selected 38 peptides that were not selected in this replication.
The two analyses agree on 9316 peptides. Similarly to the filtered
cells, the differences are negligible and demonstrate an almost perfect
replication of the feature filtering steps.</p>
<p><img src="SCoPE2_files/figure-html/benchm_pep_filt-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>To assess the differences between the data matrices, we intersect the
peptides and the cells to have comparable matrices. The range of the
differences between the 2 datasets is contained in -0.2 and 0.2, with a
very sharp peak around 0. Therefore, we can say that the
<code>scp</code> workflow could accurately replicate the SCoPE2 peptide
data.</p>
<p><img src="SCoPE2_files/figure-html/benchm_pep_diff-1.png" width="70%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="compare-protein-data">Compare protein data<a class="anchor" aria-label="anchor" href="#compare-protein-data"></a>
</h3>
<p>Similarly to the peptide data, the SCoPE2 and the <code>scp</code>
protein data contain different but highly overlapping (&gt;99 %) sets of
proteins, indicating a successful replication of the aggregation of the
protein data.</p>
<p><img src="SCoPE2_files/figure-html/benchm_prot_filt-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>The range of differences between SCoPE2 and <code>scp</code> is much
wider for the protein data than for the peptide data. Still, there is a
sharp peak around 0 indicating no dramatic bias was introduced. The
small differences observed at peptide level may have propagated during
protein imputation, batch correction and normalization.</p>
<p><img src="SCoPE2_files/figure-html/benchm_prot_diff-1.png" width="70%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>To assess the separability between macrophages and monocytes, Specht
and colleagues perform weighted principal component analysis (PCA) on
the protein data. The weights are applied to the protein space and the
weight of a protein is proportional to the amount of correlation between
that protein and the other proteins. The implementation of the weighted
PCA was taken from the SCoPE2 script and wrapped in the
<code>pcaSCoPE2</code> function that is accessible from within this
replication package. We compute the PCA on the SCopE2 protein data and
plot the results. This is figure 4a in <span class="citation">Specht et
al. (2021)</span>.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pcaSCoPE2.html">pcaSCoPE2</a></span><span class="op">(</span><span class="va">proteins_SCoPE2</span><span class="op">)</span></span>
<span><span class="co">## Compute percent explained variance</span></span>
<span><span class="va">pcaPercentVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">pcaRes</span><span class="op">$</span><span class="va">values</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pcaRes</span><span class="op">$</span><span class="va">values</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">## Plot PCA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>PC <span class="op">=</span> <span class="va">pcaRes</span><span class="op">$</span><span class="va">vectors</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>           <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">proteins_SCoPE2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC.1</span>, </span>
<span>        y <span class="op">=</span> <span class="va">PC.2</span>,</span>
<span>        col <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC1 ("</span>, <span class="va">pcaPercentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"%)"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC2 ("</span>, <span class="va">pcaPercentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"%)"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"SCoPE2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SCoPE2_files/figure-html/PCA_SCoPE2-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>Below is the resulting PCA when applied on the data processed using
the <code>scp</code> package.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pcaSCoPE2.html">pcaSCoPE2</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_scp"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## Compute percent explained variance</span></span>
<span><span class="va">pcaPercentVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">pcaRes</span><span class="op">$</span><span class="va">values</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pcaRes</span><span class="op">$</span><span class="va">values</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">## Plot PCA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>PC <span class="op">=</span> <span class="va">pcaRes</span><span class="op">$</span><span class="va">vectors</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>           <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC.1</span>, </span>
<span>        y <span class="op">=</span> <span class="va">PC.2</span>,</span>
<span>        col <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC1 ("</span>, <span class="va">pcaPercentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"%)"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC2 ("</span>, <span class="va">pcaPercentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"%)"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"SCoPE2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SCoPE2_files/figure-html/PCA_scp-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>These two plots are highly similar and further demonstrate the
ability to accurately reproduce the SCoPE2 analysis using
<code>scp</code>.</p>
<p>As a side note, the <code>scater</code> Bioconductor packages
provides a suite of functions to perform dimension reduction for
<code>SingleCellExperiment</code> objects. Since our assays are all
<code>SingleCellExperiment</code> objects, we can perform and store
standard PCA on the protein data in just a few commands.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-helpers.html" class="external-link">getWithColData</a></span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_scp"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">## Perform PCA, see ?runPCA for more info about arguments</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span>ncomponents <span class="op">=</span> <span class="fl">50</span>,</span>
<span>           ntop <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>           scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           exprs_values <span class="op">=</span> <span class="fl">1</span>,</span>
<span>           name <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">## Plotting is performed in a single line of code</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span>colour_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></span>
<span><span class="co">## Warning: Ignoring redundant column names in 'colData(x)':</span></span></code></pre></div>
<p><img src="SCoPE2_files/figure-html/PCA_scater-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>This plot is different from the previous plot because this PCA is not
weighted.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this vignette, we have demonstrated that the <code>scp</code>
package is able to accurately reproduce the analysis published in
SCoPE2. We not only support the reliability of the published work, but
we also offer a formalization and standardization of the pipeline by
means of easy-to-read and highly documented code. This workflow can
serve as a starting ground to improve upon the current methods and to
design new modelling tools dedicated to single-cell proteomics.</p>
</div>
<div class="section level2">
<h2 id="reproduce-this-vignette">Reproduce this vignette<a class="anchor" aria-label="anchor" href="#reproduce-this-vignette"></a>
</h2>
<p>You can reproduce this vignette using <code>Docker</code>:</p>
<pre><code>docker pull cvanderaa/scp_replication_docker:v1
docker run \
    -e PASSWORD=bioc \
    -p 8787:8787 \
    cvanderaa/scp_replication_docker:v1</code></pre>
<p>Open your browser and go to <a href="http://localhost:8787" class="external-link uri">http://localhost:8787</a>. The USER is <code>rstudio</code>
and the password is <code>bioc</code>. You can find the vignette in the
<code>vignettes</code> folder.</p>
<p>See the <a href="https://uclouvain-cbio.github.io/SCP.replication/index.html" class="external-link">website
home page</a> for more information.</p>
</div>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<div class="section level4">
<h4 id="hardware-and-software">Hardware and software<a class="anchor" aria-label="anchor" href="#hardware-and-software"></a>
</h4>
<p>The system details of the machine that built the vignette are:</p>
<pre><code><span><span class="co">## Machine: Linux (5.15.0-46-generic)</span></span>
<span><span class="co">## R version: R.4.2.1 (svn: 82513)</span></span>
<span><span class="co">## RAM: 16.5 GB</span></span>
<span><span class="co">## CPU: 16 core(s) - 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="timing">Timing<a class="anchor" aria-label="anchor" href="#timing"></a>
</h4>
<p>The total time required to compile this vignette is:</p>
<pre><code><span><span class="co">## 30.41182 mins</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="memory">Memory<a class="anchor" aria-label="anchor" href="#memory"></a>
</h4>
<p>The final <code>scp</code> memory size is:</p>
<pre><code><span><span class="co">## [1] "2.4 Gb"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h4>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] benchmarkme_1.0.8           scater_1.25.6              </span></span>
<span><span class="co">##  [3] scuttle_1.7.4               patchwork_1.1.2            </span></span>
<span><span class="co">##  [5] forcats_0.5.2               stringr_1.4.1              </span></span>
<span><span class="co">##  [7] dplyr_1.0.10                purrr_0.3.4                </span></span>
<span><span class="co">##  [9] readr_2.1.2                 tidyr_1.2.1                </span></span>
<span><span class="co">## [11] tibble_3.1.8                ggplot2_3.3.6              </span></span>
<span><span class="co">## [13] tidyverse_1.3.2             SCP.replication_0.2.1      </span></span>
<span><span class="co">## [15] SingleCellExperiment_1.19.0 scpdata_1.5.0              </span></span>
<span><span class="co">## [17] ExperimentHub_2.5.0         AnnotationHub_3.5.1        </span></span>
<span><span class="co">## [19] BiocFileCache_2.5.0         dbplyr_2.2.1               </span></span>
<span><span class="co">## [21] scp_1.7.0                   QFeatures_1.7.1            </span></span>
<span><span class="co">## [23] MultiAssayExperiment_1.23.8 SummarizedExperiment_1.27.3</span></span>
<span><span class="co">## [25] Biobase_2.57.1              GenomicRanges_1.49.1       </span></span>
<span><span class="co">## [27] GenomeInfoDb_1.33.7         IRanges_2.31.2             </span></span>
<span><span class="co">## [29] S4Vectors_0.35.3            BiocGenerics_0.43.4        </span></span>
<span><span class="co">## [31] MatrixGenerics_1.9.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## [33] BiocStyle_2.25.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] utf8_1.2.2                    reticulate_1.26              </span></span>
<span><span class="co">##   [3] tidyselect_1.1.2              RSQLite_2.2.17               </span></span>
<span><span class="co">##   [5] AnnotationDbi_1.59.1          grid_4.2.1                   </span></span>
<span><span class="co">##   [7] BiocParallel_1.31.12          munsell_0.5.0                </span></span>
<span><span class="co">##   [9] ScaledMatrix_1.5.1            codetools_0.2-18             </span></span>
<span><span class="co">##  [11] ragg_1.2.2                    withr_2.5.0                  </span></span>
<span><span class="co">##  [13] colorspace_2.0-3              filelock_1.0.2               </span></span>
<span><span class="co">##  [15] highr_0.9                     knitr_1.40                   </span></span>
<span><span class="co">##  [17] labeling_0.4.2                GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">##  [19] bit64_4.0.5                   farver_2.1.1                 </span></span>
<span><span class="co">##  [21] rprojroot_2.0.3               vctrs_0.4.1                  </span></span>
<span><span class="co">##  [23] generics_0.1.3                xfun_0.33                    </span></span>
<span><span class="co">##  [25] doParallel_1.0.17             R6_2.5.1                     </span></span>
<span><span class="co">##  [27] ggbeeswarm_0.6.0              clue_0.3-61                  </span></span>
<span><span class="co">##  [29] rsvd_1.0.5                    locfit_1.5-9.6               </span></span>
<span><span class="co">##  [31] MsCoreUtils_1.9.1             AnnotationFilter_1.21.0      </span></span>
<span><span class="co">##  [33] bitops_1.0-7                  cachem_1.0.6                 </span></span>
<span><span class="co">##  [35] DelayedArray_0.23.1           assertthat_0.2.1             </span></span>
<span><span class="co">##  [37] promises_1.2.0.1              scales_1.2.1                 </span></span>
<span><span class="co">##  [39] googlesheets4_1.0.1           beeswarm_0.4.0               </span></span>
<span><span class="co">##  [41] gtable_0.3.1                  beachmat_2.13.4              </span></span>
<span><span class="co">##  [43] OrgMassSpecR_0.5-3            benchmarkmeData_1.0.4        </span></span>
<span><span class="co">##  [45] sva_3.45.0                    rlang_1.0.5                  </span></span>
<span><span class="co">##  [47] genefilter_1.79.0             systemfonts_1.0.4            </span></span>
<span><span class="co">##  [49] splines_4.2.1                 lazyeval_0.2.2               </span></span>
<span><span class="co">##  [51] gargle_1.2.1                  broom_1.0.1                  </span></span>
<span><span class="co">##  [53] BiocManager_1.30.18           yaml_2.3.5                   </span></span>
<span><span class="co">##  [55] modelr_0.1.9                  backports_1.4.1              </span></span>
<span><span class="co">##  [57] httpuv_1.6.6                  tools_4.2.1                  </span></span>
<span><span class="co">##  [59] bookdown_0.29                 ellipsis_0.3.2               </span></span>
<span><span class="co">##  [61] jquerylib_0.1.4               Rcpp_1.0.9                   </span></span>
<span><span class="co">##  [63] sparseMatrixStats_1.9.0       zlibbioc_1.43.0              </span></span>
<span><span class="co">##  [65] RCurl_1.98-1.8                viridis_0.6.2                </span></span>
<span><span class="co">##  [67] haven_2.5.1                   ggrepel_0.9.1                </span></span>
<span><span class="co">##  [69] cluster_2.1.4                 fs_1.5.2                     </span></span>
<span><span class="co">##  [71] magrittr_2.0.3                reprex_2.0.2                 </span></span>
<span><span class="co">##  [73] googledrive_2.0.0             ProtGenerics_1.29.0          </span></span>
<span><span class="co">##  [75] hms_1.1.2                     mime_0.12                    </span></span>
<span><span class="co">##  [77] evaluate_0.16                 xtable_1.8-4                 </span></span>
<span><span class="co">##  [79] XML_3.99-0.10                 readxl_1.4.1                 </span></span>
<span><span class="co">##  [81] gridExtra_2.3                 compiler_4.2.1               </span></span>
<span><span class="co">##  [83] crayon_1.5.1                  htmltools_0.5.3              </span></span>
<span><span class="co">##  [85] mgcv_1.8-40                   later_1.3.0                  </span></span>
<span><span class="co">##  [87] tzdb_0.3.0                    lubridate_1.8.0              </span></span>
<span><span class="co">##  [89] DBI_1.1.3                     MASS_7.3-58.1                </span></span>
<span><span class="co">##  [91] rappdirs_0.3.3                Matrix_1.5-1                 </span></span>
<span><span class="co">##  [93] cli_3.4.0                     parallel_4.2.1               </span></span>
<span><span class="co">##  [95] igraph_1.3.4                  pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [97] pkgdown_2.0.6                 foreach_1.5.2                </span></span>
<span><span class="co">##  [99] xml2_1.3.3                    annotate_1.75.0              </span></span>
<span><span class="co">## [101] vipor_0.4.5                   bslib_0.4.0                  </span></span>
<span><span class="co">## [103] XVector_0.37.1                rvest_1.0.3                  </span></span>
<span><span class="co">## [105] digest_0.6.29                 Biostrings_2.65.6            </span></span>
<span><span class="co">## [107] rmarkdown_2.16                cellranger_1.1.0             </span></span>
<span><span class="co">## [109] edgeR_3.39.6                  DelayedMatrixStats_1.19.0    </span></span>
<span><span class="co">## [111] curl_4.3.2                    shiny_1.7.2                  </span></span>
<span><span class="co">## [113] lifecycle_1.0.2               nlme_3.1-159                 </span></span>
<span><span class="co">## [115] jsonlite_1.8.0                BiocNeighbors_1.15.1         </span></span>
<span><span class="co">## [117] desc_1.4.2                    viridisLite_0.4.1            </span></span>
<span><span class="co">## [119] limma_3.53.7                  fansi_1.0.3                  </span></span>
<span><span class="co">## [121] pillar_1.8.1                  lattice_0.20-45              </span></span>
<span><span class="co">## [123] KEGGREST_1.37.3               fastmap_1.1.0                </span></span>
<span><span class="co">## [125] httr_1.4.4                    survival_3.4-0               </span></span>
<span><span class="co">## [127] interactiveDisplayBase_1.35.0 glue_1.6.2                   </span></span>
<span><span class="co">## [129] iterators_1.0.14              png_0.1-7                    </span></span>
<span><span class="co">## [131] BiocVersion_3.16.0            bit_4.0.4                    </span></span>
<span><span class="co">## [133] stringi_1.7.8                 sass_0.4.2                   </span></span>
<span><span class="co">## [135] BiocBaseUtils_0.99.12         blob_1.2.3                   </span></span>
<span><span class="co">## [137] textshaping_0.3.6             BiocSingular_1.13.1          </span></span>
<span><span class="co">## [139] memoise_2.0.1                 irlba_2.3.5</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="licence">Licence<a class="anchor" aria-label="anchor" href="#licence"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
licence</a> licence.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Chen2019-uc" class="csl-entry">
Chen, Albert Tian, Alexander Franks, and Nikolai Slavov. 2019.
<span>“<span>DART-ID</span> Increases Single-Cell Proteome
Coverage.”</span> <em>PLoS Comput. Biol.</em> 15 (7): e1007082.
</div>
<div id="ref-Johnson2007-nc" class="csl-entry">
Johnson, W Evan, Cheng Li, and Ariel Rabinovic. 2007. <span>“Adjusting
Batch Effects in Microarray Expression Data Using Empirical Bayes
Methods.”</span> <em>Biostatistics</em> 8 (1): 118–27.
</div>
<div id="ref-Kall2008-hb" class="csl-entry">
Käll, Lukas, John D Storey, Michael J MacCoss, and William Stafford
Noble. 2008. <span>“Posterior Error Probabilities and False Discovery
Rates: Two Sides of the Same Coin.”</span> <em>J. Proteome Res.</em> 7
(1): 40–44.
</div>
<div id="ref-Specht2021-jm" class="csl-entry">
Specht, Harrison, Edward Emmott, Aleksandra A Petelski, R Gray Huffman,
David H Perlman, Marco Serra, Peter Kharchenko, Antonius Koller, and
Nikolai Slavov. 2021. <span>“Single-Cell Proteomic and Transcriptomic
Analysis of Macrophage Heterogeneity Using <span>SCoPE2</span>.”</span>
<em>Genome Biol.</em> 22 (1): 50.
</div>
<div id="ref-Tyanova2016-yj" class="csl-entry">
Tyanova, Stefka, Tikira Temu, and Juergen Cox. 2016. <span>“The
<span>MaxQuant</span> Computational Platform for Mass Spectrometry-Based
Shotgun Proteomics.”</span> <em>Nat. Protoc.</em> 11 (12): 2301–19.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:christophe.vanderaa@uclouvain.be" class="email">christophe.vanderaa@uclouvain.be</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
