<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scplainer: reanalysis of the plexDIA dataset (Derks et al. 2022) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scplainer: reanalysis of the plexDIA dataset (Derks et al. 2022)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/brunner2022.html">Replication of the cell cycle state study (Brunner et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/derks2022.html">Replication of the plexDIA analysis (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/leduc2022.html">Replication of the nPOP analysis (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Replication of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/scplainer_derks2022.html">scplainer: reanalysis of the plexDIA dataset (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/scplainer_leduc2022.html">scplainer: reanalysis of the nPOP dataset (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/scplainer_woo2022.html">scplainer: reanalysis of the macrophage activation dataset (Woo et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scplainer: reanalysis of the plexDIA dataset
(Derks et al. 2022)</h1>
                        <h4 data-toc-skip class="author">Christophe
Vanderaa, Computational Biology, UCLouvain</h4>
                        <h4 data-toc-skip class="author">Laurent Gatto,
Computational Biology, UCLouvain</h4>
            
            <h4 data-toc-skip class="date">6 December 2023</h4>
      
      
      <div class="hidden name"><code>scplainer_derks2022.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette, we will analyse the <code>derks2022</code> dataset.
The data were acquired using the plexDIA protocol (<a href="http://dx.doi.org/10.1038/s41587-022-01389-w" class="external-link">Derks et
al. 2022</a>). The authors performed mTRAQ-based labelling in DIA mode,
and showed that plexDIA could increase the cell throughput while
benefitting from the advantages of DIA.</p>
<p>Before delving into the scplainer workflow, we recommend first
familiarising with the <code>scp</code> and <code>QFeatures</code>
packages (see <a href="https://uclouvain-cbio.github.io/scp/index.html" class="external-link">tutorials</a>).</p>
</div>
<div class="section level2">
<h2 id="packages-and-data">Packages and data<a class="anchor" aria-label="anchor" href="#packages-and-data"></a>
</h2>
<p>We rely on several packages to compile this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Core packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">"scp"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"scpdata"</span><span class="op">)</span></span>
<span><span class="co">## Utility packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/scater/" class="external-link">"scater"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"camprotR"</span><span class="op">)</span></span></code></pre></div>
<p>The data set is available from the <code>scpdata</code> package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/derks2022.html" class="external-link">derks2022</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The data set contains 186 samples, including 12 negative control
samples, 12 bulk samples, 52 monocytes (U-937), 55 melanoma cells, and
55 pancreatic ductal adenocarcinoma cells (PDAC). The samples were
labelled with mTRAQ-3 acquired either with a Q-Exactive (n = 153) or a
timsTOF-SCP instrument (n = 33).</p>
</div>
<div class="section level2">
<h2 id="minimal-data-processing">Minimal data processing<a class="anchor" aria-label="anchor" href="#minimal-data-processing"></a>
</h2>
<p>The minimal data processing workflow consists of 5 main steps:</p>
<ul>
<li>Data cleaning</li>
<li>Feature quality control</li>
<li>Sample quality control</li>
<li>Peptide data assembly</li>
<li>Log2-transformation</li>
</ul>
<div class="section level3">
<h3 id="cleaning-data">Cleaning data<a class="anchor" aria-label="anchor" href="#cleaning-data"></a>
</h3>
<p>The data available in <code>scpdata</code> were provided by the
authors and were analysed with DIA-NN. We will start with the precursor
data and take advantage of DIA-NN’s capability to transfer precursor
identities across mTRAQ labels within the same runs. These data are
available in the <code>qe_prec_extracted</code> and
<code>tims_prec_extracted</code> assays.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="va">derks</span><span class="op">[</span>, , <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"qe_prec_extracted"</span>, <span class="st">"tims_prec_extracted"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Next, we clean the cell type annotation and remove the
<code>_t</code> tags contained in cell type annotations that refer to
samples acquired with timsTOF-SCP. Note that samples annotated with a
<code>_DB</code> are bulk samples.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span><span class="op">$</span><span class="va">Celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_t"</span>, <span class="st">""</span>, <span class="va">derks</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">derks</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    Melanoma Melanoma_DB         Neg        PDAC     PDAC_DB       U-937 </span></span>
<span><span class="co">##          55           1          12          55           1          52 </span></span>
<span><span class="co">##    U-937_DB </span></span>
<span><span class="co">##           1</span></span></code></pre></div>
<p>The instrument information is available in a dedicated
descriptor.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">derks</span><span class="op">$</span><span class="va">Instrument</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Q-Exactive timsTOFSCP </span></span>
<span><span class="co">##        144         33</span></span></code></pre></div>
<p>The data were acquired using mTRAQ-3 labelling ($$0, $$4 and
$$8).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">derks</span><span class="op">$</span><span class="va">Label</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  0  4  8 </span></span>
<span><span class="co">## 59 59 59</span></span></code></pre></div>
<p>The data were acquired as part of 59 MS acquisition batches.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">derks</span><span class="op">$</span><span class="va">Set</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] 59</span></span></code></pre></div>
<p>Next, we remove feature annotations that won’t be used in the
remainder of the vignette. This is to avoid overcrowding of the
annotation tables later in the vignette.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">requiredRowData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Protein.Group"</span>, <span class="st">"Protein.Ids"</span>, <span class="st">"Protein.Names"</span>, <span class="st">"Genes"</span>, </span>
<span>    <span class="st">"First.Protein.Description"</span>, <span class="st">"Proteotypic"</span>, <span class="st">"Stripped.Sequence"</span>,</span>
<span>    <span class="st">"Modified.Sequence"</span>, <span class="st">"Precursor.Charge"</span>, <span class="st">"Precursor.Id"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">selectRowData</a></span><span class="op">(</span><span class="va">derks</span>, <span class="va">requiredRowData</span><span class="op">)</span></span></code></pre></div>
<p>We replace zeros with missing values. A zero may be a true zero (the
feature is not present in the sample) or because of technical
limitations (due to the technology or the computational pre-processing).
Because we are not able to distinguish between the two, zeros should be
replaced with <code>NA</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Below is an overview of the cleaned <code>Qfeatures</code> object
used as input data for the data analysis.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span></span>
<span><span class="co">## An instance of class QFeatures containing 2 assays:</span></span>
<span><span class="co">##  [1] qe_prec_extracted: SingleCellExperiment with 8590 rows and 144 columns </span></span>
<span><span class="co">##  [2] tims_prec_extracted: SingleCellExperiment with 8722 rows and 33 columns</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-quality-control">Feature quality control<a class="anchor" aria-label="anchor" href="#feature-quality-control"></a>
</h3>
<p>We remove contaminants peptides that may propagate technical
artefacts and bias data modelling. To identify contaminants we retrieve
the UniProt sequences in cRAP through the <code>camprotR</code> package.
We also remove any peptide that maps to a keratin sequence.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contaminants</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/camprotR/man/get_ccp_crap.html" class="external-link">get_ccp_crap</a></span><span class="op">(</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span></span>
<span><span class="va">keratins</span> <span class="op">&lt;-</span> <span class="st">"KRT|^K\\d.\\d"</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">derks</span>, <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="va">contaminants</span>, <span class="va">Protein.Ids</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                          <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="va">keratins</span>, <span class="va">Protein.Names</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sample-quality-control">Sample quality control<a class="anchor" aria-label="anchor" href="#sample-quality-control"></a>
</h3>
<p>We next remove low-quality cells. The quality control criteria
are:</p>
<ul>
<li>We remove samples with low number of detected peptides. The
criterion is computed as follows:</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/countUniqueFeatures.html" class="external-link">countUniqueFeatures</a></span><span class="op">(</span></span>
<span>    <span class="va">derks</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span>,</span>
<span>    groupBy <span class="op">=</span> <span class="st">"Stripped.Sequence"</span>, colDataName <span class="op">=</span> <span class="st">"NumberPeptides"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>We remove samples with low median intensity. The metric (note we
will later use it for normalisation) is computed as follows:</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MedianIntensity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu">experiments</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">colMedians</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">out</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">MedianIntensity</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">MedianIntensity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">MedianIntensity</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">MedianIntensity</span><span class="op">)</span>, <span class="st">"MedianIntensity"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">MedianIntensity</span></span></code></pre></div>
<ul>
<li>We remove the samples that have a high median coefficient of
variation (CV). The CV is computed within each sample, by grouping the
peptides that belong to the same protein or protein group. Computing the
median CV requires peptide normalisation (<code>norm = "SCoPE2"</code>)
to avoid that differences in ionisation obscure the quantification
accuracy. This is computed as follows:</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/medianCVperCell.html" class="external-link">medianCVperCell</a></span><span class="op">(</span></span>
<span>    <span class="va">derks</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span>,</span>
<span>    groupBy <span class="op">=</span> <span class="st">"Protein.Ids"</span>, nobs <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    norm <span class="op">=</span> <span class="st">"SCoPE2"</span>,</span>
<span>    colDataName <span class="op">=</span> <span class="st">"MedianCV"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We plot the metrics used to perform sample quality control. Since two
different instruments were used, we expect different metric
distributions for each instrument and we plot the metrics
separately.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">derks</span><span class="op">$</span><span class="va">Instrument</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">pl</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Instrument</span> <span class="op">==</span> <span class="va">i</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">MedianIntensity</span>,</span>
<span>            x <span class="op">=</span> <span class="va">NumberPeptides</span>,</span>
<span>            color <span class="op">=</span> <span class="va">MedianCV</span>,</span>
<span>            shape <span class="op">=</span> <span class="va">Celltype</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_color_continuous</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">pl</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/qc_plot-1.png" width="768" style="display: block; margin: auto;"></p>
<p>There are a few suspicious cells with a low number of detected
peptides, a low median intensity and a high median CV. For the
Q-Exactive dataset, we remove the group of cells that cluster with the
negative control samples, that are cells with less than 750 peptides and
a median intensity lower than 9.4. For the timsTOF-SCP dataset, we
remove the group cells characterised by a median intensity lower than
6.5. We apply the QC filter but also remove any sample that is not a
single cell.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span><span class="op">$</span><span class="va">passQC</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">derks</span><span class="op">$</span><span class="va">Instrument</span> <span class="op">==</span> <span class="st">"Q-Exactive"</span> <span class="op">&amp;</span></span>
<span>                    <span class="va">derks</span><span class="op">$</span><span class="va">MedianIntensity</span> <span class="op">&gt;</span> <span class="fl">9.4</span> <span class="op">&amp;</span></span>
<span>                    <span class="va">derks</span><span class="op">$</span><span class="va">NumberPeptides</span> <span class="op">&gt;</span> <span class="fl">750</span><span class="op">)</span> <span class="op">|</span></span>
<span>                <span class="op">(</span><span class="va">derks</span><span class="op">$</span><span class="va">Instrument</span> <span class="op">==</span> <span class="st">"timsTOFSCP"</span> <span class="op">&amp;</span></span>
<span>                    <span class="va">derks</span><span class="op">$</span><span class="va">MedianIntensity</span> <span class="op">&gt;</span> <span class="fl">6.5</span><span class="op">)</span></span>
<span><span class="va">derks</span><span class="op">$</span><span class="va">isSingleCell</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Neg|DB$"</span>, <span class="va">derks</span><span class="op">$</span><span class="va">Celltype</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">derks</span>, <span class="va">derks</span><span class="op">$</span><span class="va">passQC</span> <span class="op">&amp;</span> <span class="va">derks</span><span class="op">$</span><span class="va">isSingleCell</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="peptide-data-assembly">Peptide data assembly<a class="anchor" aria-label="anchor" href="#peptide-data-assembly"></a>
</h3>
<p>For now, we worked with precursor-level data. We here aggregate
precursors into peptides. However, we need to solve a minor issue in the
feature annotations. When aggregating peptides, we will only keep the
feature annotations that have the common data. However, the protein
groups differ among a few precursor from the same peptide. Hence, the
protein groups and the associated gene names are not constant for some
peptides that are removed during aggregation. To avoid this, we ensure
consistant protein groups across precursor from the same peptide using
majority vote.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ppMap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Stripped.Sequence</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">## The majority vote happens here</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span></span>
<span>        Protein.Group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">Protein.Group</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>        Genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">Genes</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Stripped.Sequence</span>, <span class="va">Protein.Group</span>, <span class="va">Genes</span><span class="op">)</span></span>
<span><span class="va">consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Stripped.Sequence</span>, <span class="va">ppMap</span><span class="op">$</span><span class="va">Stripped.Sequence</span><span class="op">)</span></span>
<span>    <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>        Protein.Group <span class="op">=</span> <span class="va">ppMap</span><span class="op">$</span><span class="va">Protein.Group</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span>,</span>
<span>        Genes <span class="op">=</span> <span class="va">ppMap</span><span class="op">$</span><span class="va">Genes</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">consensus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">derks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">consensus</span></span></code></pre></div>
<p>We then can safely perform peptide aggregation.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peptideAssays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"peptides_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span></span>
<span>    <span class="va">derks</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span>,</span>
<span>    fcol <span class="op">=</span> <span class="st">"Stripped.Sequence"</span>,</span>
<span>    name <span class="op">=</span> <span class="va">peptideAssays</span>,</span>
<span>    fun <span class="op">=</span> <span class="va">colMedians</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Moreover, the data are stored in separate assays for each instrument.
We here combine the two assays into one.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="va">peptideAssays</span>, name <span class="op">=</span> <span class="st">"peptides"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="log-transformation">Log-transformation<a class="anchor" aria-label="anchor" href="#log-transformation"></a>
</h3>
<p>We log2-transform the quantification data.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">derks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">derks</span>, i <span class="op">=</span> <span class="st">"peptides"</span>, name <span class="op">=</span> <span class="st">"peptides_log"</span><span class="op">)</span></span></code></pre></div>
<p>Here is an overview of the data processing:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">derks</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="data-modelling">Data modelling<a class="anchor" aria-label="anchor" href="#data-modelling"></a>
</h2>
<p>Model the data using scplainer, a linear regression model implemented
in <code>scp</code>. The model is applied on a
<code>SingleCellExperiment</code> so we extract it from the processed
data set along with the <code>colData</code></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">derks</span>, <span class="st">"peptides_log"</span><span class="op">)</span></span></code></pre></div>
<p>First, we must specify which variables to include in the model. We
here include 4 variables:</p>
<ul>
<li>
<code>MedianIntensity</code>: this is the normalisation factor used
to correct for cell-specific technical differences.</li>
<li>
<code>Set</code>: this is used to perform batch correction. We
consider each acquisition run to be a batch.</li>
<li>
<code>Label</code>: this is used to correct for mTRAQ effects.</li>
<li>
<code>Celltype</code>: this is the biological variable of interest.
It captures the difference between macrophages and monocytes</li>
</ul>
<p><code><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelWorkflow()</a></code> fits linear regression models to the
data, where the model is adapted for each peptide depending on its
pattern of missing values.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelWorkflow</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>,</span>
<span>    formula <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="co">## intercept</span></span>
<span>        <span class="co">## normalisation</span></span>
<span>        <span class="va">MedianIntensity</span> <span class="op">+</span></span>
<span>        <span class="co">## batch effects</span></span>
<span>        <span class="va">Set</span> <span class="op">+</span> <span class="va">Label</span> <span class="op">+</span></span>
<span>        <span class="co">## biological variability</span></span>
<span>        <span class="va">Celltype</span><span class="op">)</span></span></code></pre></div>
<p>Once the model is prepared, we can explore the distribution of the
n/p ratios.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelFilterPlot</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;"></p>
<p>About 40 % of the peptides do not have sufficient observations to
estimate the model. You could consider <span class="math inline">\(n/p\)</span> a rough average of the number of
replicates per parameter to fit (for categorical variables, the number
of replicates per group). We recommend moving the threshold away from 1
to increase statistical power and remove noisy peptides. This comes of
course at the cost of less peptides included in the analysis.</p>
</div>
<div class="section level2">
<h2 id="model-analysis">Model analysis<a class="anchor" aria-label="anchor" href="#model-analysis"></a>
</h2>
<p>The model analysis consists of three steps:</p>
<ul>
<li>Analysis of variance</li>
<li>Differential abundance analysis</li>
<li>Component analysis</li>
</ul>
<div class="section level3">
<h3 id="analysis-of-variance">Analysis of variance<a class="anchor" aria-label="anchor" href="#analysis-of-variance"></a>
</h3>
<p>The analysis of variance explores the proportion of data captures by
each variable in the model.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">vaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVarianceAnalysis</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## DataFrameList of length 5</span></span>
<span><span class="co">## names(5): Residuals MedianIntensity Set Label Celltype</span></span>
<span><span class="va">vaRes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## DataFrame with 4037 rows and 4 columns</span></span>
<span><span class="co">##            feature         SS        df percentExplainedVar</span></span>
<span><span class="co">##        &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt;           &lt;numeric&gt;</span></span>
<span><span class="co">## 1    AAAAAAALQA...  7.4410176        42           1.3163177</span></span>
<span><span class="co">## 2    AAAEDVNVTF...  5.7902940         4           2.1208123</span></span>
<span><span class="co">## 3    AAAEVNQDYG...  7.6434457        11           8.6315618</span></span>
<span><span class="co">## 4        AAALEFLNR  0.0125425         1           0.0134479</span></span>
<span><span class="co">## 5        AAANEQLTR 13.7693132        22           4.1179696</span></span>
<span><span class="co">## ...            ...        ...       ...                 ...</span></span>
<span><span class="co">## 4033    QTDVGITHFR  10.021577         8           11.149523</span></span>
<span><span class="co">## 4034    VAIYSPDGVR   0.538425         3            0.633974</span></span>
<span><span class="co">## 4035       YGDAFIR   1.301915         3            3.389192</span></span>
<span><span class="co">## 4036      SDTFINLR   3.207778         7            5.320569</span></span>
<span><span class="co">## 4037      VFIEDVSR   6.065779         5            4.867759</span></span></code></pre></div>
<p>The results are a list of tables, one table for each variable. Each
table reports for each peptide the variance captures (<code>SS</code>),
the residual degrees of freedom for estimating the variance
(<code>df</code>) and the percentage of total variance explained
(<code>percentExplainedVar</code>). To better explore the results, we
add the annotations available in the <code>rowData</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Stripped.Sequence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>By default, we explore the variance for all peptides combined:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span><span class="va">vaRes</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-23-1.png" width="30%" style="display: block; margin: auto;"></p>
<p>We explore the top 20 peptides that are have the highest percentage
of variance explained by the biological variable (top) or by the batch
variable (bottom).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Celltype"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Set"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-24-1.png" width="960" style="display: block; margin: auto;"></p>
<p>We can also group these peptide according to their protein. We simply
need to provide the <code>fcol</code> argument.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Celltype"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span>, fcol <span class="op">=</span> <span class="st">"Genes"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Set"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span>, fcol <span class="op">=</span> <span class="st">"Genes"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-25-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a>
</h3>
<p>Next, we explore the model output to understand the differences
between the cell types. The difference of interest is specified using
the <code>contrast</code> argument. The first element points to the
variable to test and the two following element are the groups of
interest to compare. You can provide multiple contrast in a list.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpDifferentialAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Celltype"</span>, <span class="st">"Melanoma"</span>, <span class="st">"PDAC"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Celltype"</span>, <span class="st">"Melanoma"</span>, <span class="st">"U-937"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Celltype"</span>, <span class="st">"PDAC"</span>, <span class="st">"U-937"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## List of length 3</span></span>
<span><span class="co">## names(3): Celltype_Melanoma_vs_PDAC Celltype_Melanoma_vs_U.937 Celltype_PDAC_vs_U.937</span></span>
<span><span class="va">daRes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## DataFrame with 4037 rows and 7 columns</span></span>
<span><span class="co">##            feature  Estimate        SE        Df tstatistic      pvalue</span></span>
<span><span class="co">##        &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1    AAAAAAALQA...  0.363610 0.0961336        42   3.782347 0.000485767</span></span>
<span><span class="co">## 2    AAAEDVNVTF...  0.422119 1.0414839         4   0.405305 0.705994818</span></span>
<span><span class="co">## 3    AAAEVNQDYG...  0.883843 0.3117647        11   2.834969 0.016227253</span></span>
<span><span class="co">## 4        AAALEFLNR -1.817356 0.1650310         1 -11.012211 0.057652227</span></span>
<span><span class="co">## 5        AAANEQLTR -0.110730 0.2340733        22  -0.473055 0.640833487</span></span>
<span><span class="co">## ...            ...       ...       ...       ...        ...         ...</span></span>
<span><span class="co">## 4033    QTDVGITHFR  0.389972  1.033239         8   0.377427  0.71567241</span></span>
<span><span class="co">## 4034    VAIYSPDGVR -4.447359  0.467208         3  -9.519019  0.00245869</span></span>
<span><span class="co">## 4035       YGDAFIR  0.691946  0.701938         3   0.985766  0.39692989</span></span>
<span><span class="co">## 4036      SDTFINLR  0.137518  0.609705         7   0.225548  0.82799636</span></span>
<span><span class="co">## 4037      VFIEDVSR -5.064273  0.984345         5  -5.144816  0.00363063</span></span>
<span><span class="co">##            padj</span></span>
<span><span class="co">##       &lt;numeric&gt;</span></span>
<span><span class="co">## 1    0.00341817</span></span>
<span><span class="co">## 2    0.84068845</span></span>
<span><span class="co">## 3    0.06279106</span></span>
<span><span class="co">## 4    0.16187589</span></span>
<span><span class="co">## 5    0.79421530</span></span>
<span><span class="co">## ...         ...</span></span>
<span><span class="co">## 4033  0.8459663</span></span>
<span><span class="co">## 4034  0.0131652</span></span>
<span><span class="co">## 4035  0.5967268</span></span>
<span><span class="co">## 4036  0.9084303</span></span>
<span><span class="co">## 4037  0.0181854</span></span></code></pre></div>
<p>Similarly to analysis of variance, the results are a list of tables,
one table for each contrast. Each table reports for each peptide the
estimated difference between the two groups, the standard error
associated to the estimation, the degrees of freedom, the t-statistics,
the associated p-value and the p-value FDR-adjusted for multiple testing
across all peptides. Again, to better explore the results, we add the
annotations available in the <code>rowData</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, </span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Stripped.Sequence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We then visualize the results using a volcano plot. The function
below return a volcano plot for each contrast.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span><span class="va">daRes</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-28-1.png" width="960" style="display: block; margin: auto;"></p>
<p>To help interpretation of the results, we will label the peptides
with their protein name. Also we increase the number of labels shown on
the plot. Finally, we can add colors to the plot. For instance, let’s
explore the impact of the number of observations using the <span class="math inline">\(n/p\)</span> ratio. We create a new annotation
table, add it to the results and redraw the plot. The <span class="math inline">\(n/p\)</span> ratio is retrieved using
scpModelFilterNPRatio</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">np</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-class.html" class="external-link">scpModelFilterNPRatio</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">np</span><span class="op">)</span>, npRatio <span class="op">=</span> <span class="va">np</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, top <span class="op">=</span> <span class="fl">30</span>, textBy <span class="op">=</span> <span class="st">"Genes"</span>, </span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">npRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-29-1.png" width="960" style="display: block; margin: auto;"></p>
<p>As expected, higher number of observations (higher <span class="math inline">\(n/p\)</span>) lead to increased statistical power
and hence to more significant results.</p>
<p>Finally, we offer functionality to report results at the protein
level.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpDifferentialAggregate</a></span><span class="op">(</span><span class="va">daRes</span>, fcol <span class="op">=</span> <span class="st">"Genes"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span>top <span class="op">=</span> <span class="fl">30</span>, textBy <span class="op">=</span> <span class="st">"Genes"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-30-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="component-analysis">Component analysis<a class="anchor" aria-label="anchor" href="#component-analysis"></a>
</h3>
<p>Finally, we perform component analysis to link the modelled effects
to the cellular heterogeneity. We here run an APCA+ (extended
ANOVA-simultaneous principal component analysis) for the sample type
effect. In other words, we perform a PCA on the data that is capture by
the sample type variable along with the residuals (unmodelled data).</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">caRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, ncomp <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">"APCA"</span>, effect <span class="op">=</span> <span class="st">"Celltype"</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] "APCA"</span></span>
<span><span class="co">## [1] "Celltype"</span></span>
<span><span class="co">## List of length 2</span></span>
<span><span class="co">## names(2): bySample byFeature</span></span></code></pre></div>
<p>The results are contained in a list with 2 elements.
<code>bySample</code> contains the PC scores, that is the component
results in sample space. <code>byFeature</code> contains the
eigenvectors, that is the component results in feature space.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span></span>
<span><span class="co">## List of length 3</span></span>
<span><span class="co">## names(3): unmodelled residuals APCA_Celltype</span></span></code></pre></div>
<p>Each of the two elements contains components results for the data
before modelling (<code>unmodelled</code>), for the residuals or for the
APCA on the sample type variable (<code>APCA_Celltype</code>).</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## DataFrame with 122 rows and 3 columns</span></span>
<span><span class="co">##                                                                                      PC1</span></span>
<span><span class="co">##                                                                                &lt;numeric&gt;</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1146.raw.4                                                 55.8240</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1146.raw.8                                                 39.1033</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1147.raw.0                                                 18.2279</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1147.raw.4                                                 83.2346</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1148.raw.0                                                 40.2493</span></span>
<span><span class="co">## ...                                                                                  ...</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_8_Slot2.8_1_222.d.4  -76.8229</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_8_Slot2.8_1_222.d.8  -94.2885</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.0  -80.4816</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.4 -105.6186</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.8  -79.6811</span></span>
<span><span class="co">##                                                                                       PC2</span></span>
<span><span class="co">##                                                                                 &lt;numeric&gt;</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1146.raw.4                                                  34.5480</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1146.raw.8                                                 -39.6912</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1147.raw.0                                                 -31.4929</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1147.raw.4                                                 -67.5675</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1148.raw.0                                                 -46.8058</span></span>
<span><span class="co">## ...                                                                                   ...</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_8_Slot2.8_1_222.d.4   1.455085</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_8_Slot2.8_1_222.d.8  12.151391</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.0  -0.922176</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.4  10.664048</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.8 -59.582158</span></span>
<span><span class="co">##                                                                                         cell</span></span>
<span><span class="co">##                                                                                  &lt;character&gt;</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1146.raw.4                                               F..JD.plex...</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1146.raw.8                                               F..JD.plex...</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1147.raw.0                                               F..JD.plex...</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1147.raw.4                                               F..JD.plex...</span></span>
<span><span class="co">## F..JD.plexDIA.nPOP.wJD1148.raw.0                                               F..JD.plex...</span></span>
<span><span class="co">## ...                                                                                      ...</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_8_Slot2.8_1_222.d.4 F..JD.plex...</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_8_Slot2.8_1_222.d.8 F..JD.plex...</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.0 F..JD.plex...</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.4 F..JD.plex...</span></span>
<span><span class="co">## F..JD.plexDIA.Bruker.OneDrive_1_3.9.2022.20220308_npop_SCP_9_Slot2.9_1_223.d.8 F..JD.plex...</span></span></code></pre></div>
<p>Each elements is a table with the computed componoents. Let’s explore
the component analysis results in cell space. Similarly to the previous
explorations, we annotate the results.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResCells</span> <span class="op">&lt;-</span> <span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">caResCells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span><span class="va">caResCells</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span></span></code></pre></div>
<p>We then generate the component plot, colouring by
<code>Celltype</code>. To assess the impact of batch effects, we shape
the points according to the plate batch (cf intro) as well.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>,</span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">Celltype</span>, shape <span class="op">=</span> <span class="va">Instrument</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-34-1.png" width="864" style="display: block; margin: auto;"></p>
<p>While the data before modelling is mainly driven by batch effects
(instruments are separated on the first PC), the APCA clearly separates
the three cell populations. The PCA on the residuals suggests that no
additional unmodelled effects influence the data globally, indicating
that our model accurately captures the variation in the data.</p>
<p>We use the same approach to explore the component results in feature
space.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResPeps</span> <span class="op">&lt;-</span> <span class="va">caRes</span><span class="op">$</span><span class="va">byFeature</span></span>
<span><span class="va">caResPeps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">caResPeps</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Stripped.Sequence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We plot the compenents in peptide-space.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plCApeps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResPeps</span>, pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.8</span>, alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plCApeps</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-36-1.png" width="864" style="display: block; margin: auto;"></p>
<p>We can also combine the exploration of the components in cell and
peptide space. This is possible thanks to biplots.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biplots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentBiplot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>, <span class="va">caResPeps</span>, </span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">Celltype</span>, shape <span class="op">=</span> <span class="va">Instrument</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    labelParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span>, max.overlaps <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>    textBy <span class="op">=</span> <span class="st">"Genes"</span>, top <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">biplots</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-37-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Finally, we offer functionality to aggregate the results at the
protein level instead of the peptide level.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResProts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentAggregate</a></span><span class="op">(</span><span class="va">caResPeps</span>, fcol <span class="op">=</span> <span class="st">"Genes"</span><span class="op">)</span></span>
<span><span class="va">biplots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentBiplot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>, <span class="va">caResProts</span>, </span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">Celltype</span>, shape <span class="op">=</span> <span class="va">Instrument</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    labelParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span>, max.overlaps <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>    textBy <span class="op">=</span> <span class="st">"Genes"</span>, top <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">biplots</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_derks2022_files/figure-html/unnamed-chunk-38-1.png" width="864" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="interactive-visualisation-with-isee">Interactive visualisation with iSEE<a class="anchor" aria-label="anchor" href="#interactive-visualisation-with-isee"></a>
</h2>
<p>You can manually explore the data through an interactive interface
thanks to using <code>iSEE</code>:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/iSEE/iSEE" class="external-link">"iSEE"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code>R Under development (unstable) (2023-07-27 r84768)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 23.04

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.11.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.11.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Prague
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] camprotR_0.0.0.9000         scater_1.31.0              
 [3] scuttle_1.13.0              SingleCellExperiment_1.25.0
 [5] dplyr_1.1.3                 patchwork_1.1.3            
 [7] ggplot2_3.4.4               scpdata_1.9.2              
 [9] ExperimentHub_2.11.0        AnnotationHub_3.11.0       
[11] BiocFileCache_2.11.1        dbplyr_2.4.0               
[13] scp_1.11.3                  QFeatures_1.13.0           
[15] MultiAssayExperiment_1.29.0 SummarizedExperiment_1.33.0
[17] Biobase_2.63.0              GenomicRanges_1.55.1       
[19] GenomeInfoDb_1.39.0         IRanges_2.37.0             
[21] S4Vectors_0.41.1            BiocGenerics_0.49.1        
[23] MatrixGenerics_1.15.0       matrixStats_1.1.0          
[25] BiocStyle_2.31.0           

loaded via a namespace (and not attached):
  [1] later_1.3.1                   bitops_1.0-7                 
  [3] filelock_1.0.2                tibble_3.2.1                 
  [5] lpsymphony_1.31.0             preprocessCore_1.65.0        
  [7] XML_3.99-0.15                 lifecycle_1.0.4              
  [9] doParallel_1.0.17             rprojroot_2.0.4              
 [11] lattice_0.22-5                MASS_7.3-60.1                
 [13] magrittr_2.0.3                limma_3.59.1                 
 [15] sass_0.4.7                    rmarkdown_2.25               
 [17] jquerylib_0.1.4               yaml_2.3.7                   
 [19] metapod_1.11.0                httpuv_1.6.12                
 [21] MsCoreUtils_1.15.1            DBI_1.1.3                    
 [23] RColorBrewer_1.1-3            abind_1.4-5                  
 [25] zlibbioc_1.49.0               purrr_1.0.2                  
 [27] AnnotationFilter_1.27.0       RCurl_1.98-1.13              
 [29] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      
 [31] ggrepel_0.9.4                 irlba_2.3.5.1                
 [33] MSnbase_2.29.0                pkgdown_2.0.7                
 [35] DelayedMatrixStats_1.25.0     ncdf4_1.21                   
 [37] codetools_0.2-19              DelayedArray_0.29.0          
 [39] tidyselect_1.2.0              farver_2.1.1                 
 [41] ScaledMatrix_1.11.0           viridis_0.6.4                
 [43] nipals_0.8                    jsonlite_1.8.7               
 [45] BiocNeighbors_1.21.0          ellipsis_0.3.2               
 [47] iterators_1.0.14              systemfonts_1.0.5            
 [49] foreach_1.5.2                 tools_4.4.0                  
 [51] ragg_1.2.6                    Rcpp_1.0.11                  
 [53] glue_1.6.2                    BiocBaseUtils_1.5.0          
 [55] gridExtra_2.3                 SparseArray_1.3.0            
 [57] xfun_0.41                     withr_2.5.2                  
 [59] BiocManager_1.30.22           fastmap_1.1.1                
 [61] fansi_1.0.5                   digest_0.6.33                
 [63] rsvd_1.0.5                    R6_2.5.1                     
 [65] mime_0.12                     textshaping_0.3.7            
 [67] colorspace_2.1-0              RSQLite_2.3.3                
 [69] utf8_1.2.4                    tidyr_1.3.0                  
 [71] generics_0.1.3                robustbase_0.99-0            
 [73] httr_1.4.7                    S4Arrays_1.3.0               
 [75] pkgconfig_2.0.3               gtable_0.3.4                 
 [77] blob_1.2.4                    impute_1.77.0                
 [79] XVector_0.43.0                htmltools_0.5.7              
 [81] bookdown_0.36                 MALDIquant_1.22.1            
 [83] ProtGenerics_1.35.0           clue_0.3-65                  
 [85] scales_1.2.1                  png_0.1-8                    
 [87] knitr_1.45                    rstudioapi_0.15.0            
 [89] curl_5.1.0                    cachem_1.0.8                 
 [91] stringr_1.5.0                 BiocVersion_3.19.1           
 [93] parallel_4.4.0                vipor_0.4.5                  
 [95] AnnotationDbi_1.65.2          mzID_1.41.0                  
 [97] vsn_3.71.0                    desc_1.4.2                   
 [99] pillar_1.9.0                  grid_4.4.0                   
[101] vctrs_0.6.4                   pcaMethods_1.95.0            
[103] slam_0.1-50                   promises_1.2.1               
[105] BiocSingular_1.19.0           IHW_1.31.0                   
[107] beachmat_2.19.0               xtable_1.8-4                 
[109] cluster_2.1.4                 beeswarm_0.4.0               
[111] evaluate_0.23                 cli_3.6.1                    
[113] compiler_4.4.0                rlang_1.1.2                  
[115] crayon_1.5.2                  labeling_0.4.3               
[117] fdrtool_1.2.17                affy_1.81.0                  
[119] plyr_1.8.9                    fs_1.6.3                     
[121] ggbeeswarm_0.7.2              stringi_1.7.12               
[123] viridisLite_0.4.2             BiocParallel_1.37.0          
[125] munsell_0.5.0                 Biostrings_2.71.1            
[127] lazyeval_0.2.2                Matrix_1.6-1.1               
[129] sparseMatrixStats_1.15.0      bit64_4.0.5                  
[131] KEGGREST_1.43.0               statmod_1.5.0                
[133] shiny_1.7.5.1                 highr_0.10                   
[135] interactiveDisplayBase_1.41.0 mzR_2.37.0                   
[137] igraph_1.5.1                  memoise_2.0.1                
[139] affyio_1.73.0                 bslib_0.5.1                  
[141] DEoptimR_1.1-3                bit_4.0.5                    </code></pre>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="fu">citation</span>(<span class="st">"scp"</span>)</span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>To cite the scp package <span class="cf">in</span> publications use<span class="sc">:</span></span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a>  Vanderaa, Christophe, and Laurent Gatto. <span class="fl">2023.</span>  Revisiting the Thorny</span>
<span id="cb46-5"><a href="#cb46-5" tabindex="-1"></a>  Issue of Missing Values <span class="cf">in</span> Single<span class="sc">-</span>Cell Proteomics. Journal of</span>
<span id="cb46-6"><a href="#cb46-6" tabindex="-1"></a>  Proteome Research <span class="dv">22</span> (<span class="dv">9</span>)<span class="sc">:</span> 2775–<span class="fl">84.</span></span>
<span id="cb46-7"><a href="#cb46-7" tabindex="-1"></a></span>
<span id="cb46-8"><a href="#cb46-8" tabindex="-1"></a>  Vanderaa Christophe and Laurent Gatto. The current state of</span>
<span id="cb46-9"><a href="#cb46-9" tabindex="-1"></a>  single<span class="sc">-</span>cell proteomics data analysis. Current Protocols <span class="dv">3</span> (<span class="dv">1</span>)<span class="sc">:</span> e658.;</span>
<span id="cb46-10"><a href="#cb46-10" tabindex="-1"></a>  doi<span class="sc">:</span> https<span class="sc">:</span><span class="er">//</span>doi.org<span class="sc">/</span><span class="fl">10.1002</span><span class="sc">/</span><span class="fu">cpz1.658</span> (<span class="dv">2023</span>).</span>
<span id="cb46-11"><a href="#cb46-11" tabindex="-1"></a></span>
<span id="cb46-12"><a href="#cb46-12" tabindex="-1"></a>  Vanderaa Christophe and Laurent Gatto. Replication of Single<span class="sc">-</span>Cell</span>
<span id="cb46-13"><a href="#cb46-13" tabindex="-1"></a>  Proteomics Data Reveals Important Computational Challenges. Expert</span>
<span id="cb46-14"><a href="#cb46-14" tabindex="-1"></a>  Review of Proteomics, 1–<span class="dv">9</span> (<span class="dv">2021</span>).</span>
<span id="cb46-15"><a href="#cb46-15" tabindex="-1"></a></span>
<span id="cb46-16"><a href="#cb46-16" tabindex="-1"></a>To see these entries <span class="cf">in</span> BibTeX format, use <span class="st">'print(&lt;citation&gt;,</span></span>
<span id="cb46-17"><a href="#cb46-17" tabindex="-1"></a><span class="st">bibtex=TRUE)'</span>, <span class="st">'toBibtex(.)'</span>, or set</span>
<span id="cb46-18"><a href="#cb46-18" tabindex="-1"></a><span class="st">'options(citation.bibtex.max=999)'</span>.</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
license</a> license.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
