<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SCP data analysis of the macrophage activation dataset (Woo et al. 2022) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SCP data analysis of the macrophage activation dataset (Woo et al. 2022)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/brunner2022.html">Replication of the cell cycle state study (Brunner et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/derks2022.html">Replication of the plexDIA analysis (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/leduc2022.html">Replication of the nPOP analysis (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Replication of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/scplainer_leduc2022.html">scplainer: reanalysis of the nPOP dataset (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/scplainer_woo2022.html">SCP data analysis of the macrophage activation dataset (Woo et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SCP data analysis of the macrophage activation
dataset (Woo et al. 2022)</h1>
                        <h4 data-toc-skip class="author">Christophe
Vanderaa, Computational Biology, UCLouvain</h4>
                        <h4 data-toc-skip class="author">Laurent Gatto,
Computational Biology, UCLouvain</h4>
            
            <h4 data-toc-skip class="date">10 November 2023</h4>
      
      
      <div class="hidden name"><code>scplainer_woo2022.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette, we will analyse the <code>woo2022_macophage</code>
dataset. The data were acquired using the TIFF acquisition protocol (<a href="http://dx.doi.org/10.1016/j.cels.2022.02.003" class="external-link">Woo et
al. 2022</a>). The authors performed a label-free experiment in DDA
mode, and showed that TIFF could improve the sensitivity and accuracy of
label-free single-cell proteomics (SCP).</p>
</div>
<div class="section level2">
<h2 id="packages-and-data">Packages and data<a class="anchor" aria-label="anchor" href="#packages-and-data"></a>
</h2>
<p>We rely on several packages to compile this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Core packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">"scp"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"scpdata"</span><span class="op">)</span></span>
<span><span class="co">## Utility packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/scater/" class="external-link">"scater"</a></span><span class="op">)</span></span></code></pre></div>
<p>The data set is available from the <code>scpdata</code> package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/woo2022_macrophage.html" class="external-link">woo2022_macrophage</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## see ?scpdata and browseVignettes('scpdata') for documentation</span></span>
<span><span class="co">## loading from cache</span></span></code></pre></div>
<p>The data set contains 155 RAW264.7 cells: 54 are unstimulated cells,
52 are LPS-stimulated cells at 24 h, and 49 LPS-stimulated cells at 48
h.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">woo</span><span class="op">$</span><span class="va">Treatment</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   CON LPS24 LPS48 </span></span>
<span><span class="co">##    54    52    49</span></span></code></pre></div>
<p>The data were acquired as part of 4 sample preparation chips that we
will consider as a potential source for batch effects.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">woo</span><span class="op">$</span><span class="va">Chip</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  1  2  3  4 </span></span>
<span><span class="co">## 43 41 27 44</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="minimal-data-processing">Minimal data processing<a class="anchor" aria-label="anchor" href="#minimal-data-processing"></a>
</h2>
<p>The minimal data processing workflow consist of 5 main steps:</p>
<ul>
<li>Data cleaning</li>
<li>Feature quality control</li>
<li>Sample quality control</li>
<li>Log2-transformation</li>
</ul>
<div class="section level3">
<h3 id="cleaning-data">Cleaning data<a class="anchor" aria-label="anchor" href="#cleaning-data"></a>
</h3>
<p>The data available in <code>scpdata</code> were provided authors and
were analysed with MaxQuant. We here start with the non-normalised
peptide data. We recommend starting with data with least prior
processing. We therefore remove the maxLFQ normalised data and the
protein data generated by MaxQuant.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">woo</span><span class="op">)</span></span>
<span><span class="co">## [1] "peptides_intensity" "peptides_LFQ"       "proteins_intensity"</span></span>
<span><span class="co">## [4] "proteins_iBAQ"      "proteins_LFQ"</span></span>
<span><span class="va">assaysToRemove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"peptides_LFQ"</span>, <span class="st">"proteins_intensity"</span>, <span class="st">"proteins_iBAQ"</span>, <span class="st">"proteins_LFQ"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">removeAssay</a></span><span class="op">(</span><span class="va">woo</span>, <span class="va">assaysToRemove</span><span class="op">)</span></span>
<span><span class="co">## harmonizing input:</span></span>
<span><span class="co">##   removing 620 sampleMap rows not in names(experiments)</span></span></code></pre></div>
<p>Below is an overview of the <code>Qfeatures</code> object used as
input data for this data analysis.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span></span>
<span><span class="co">## An instance of class QFeatures containing 1 assays:</span></span>
<span><span class="co">##  [1] peptides_intensity: SingleCellExperiment with 10469 rows and 155 columns</span></span></code></pre></div>
<p>We remove feature annotations that won’t be used in the remainder of
the vignette. This is to avoid overcrowding of the annotation tables
later in the vignette.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">requiredRowData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Sequence"</span>, <span class="st">"Proteins"</span>, <span class="st">"Leading.razor.protein"</span>, <span class="st">"Gene.names"</span>,</span>
<span>    <span class="st">"Protein.names"</span>, <span class="st">"Potential.contaminant"</span>, <span class="st">"Reverse"</span>, <span class="st">"PEP"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">selectRowData</a></span><span class="op">(</span><span class="va">woo</span>, <span class="va">requiredRowData</span><span class="op">)</span></span></code></pre></div>
<p>We replace zeros by missing values. A zero may be a true (the feature
is not present in the sample) or because of technical limitations (due
to the technology or the computational pre-processing). Because we are
not able to distinguish between the two, zeros should be replaced with
<code>NA</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">woo</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">woo</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-quality-control">Feature quality control<a class="anchor" aria-label="anchor" href="#feature-quality-control"></a>
</h3>
<p>We remove low-quality PSMs that may propagate technical artefacts and
bias data modelling. The quality control criteria are:</p>
<ul>
<li>We remove contaminants and decoy peptides.</li>
<li>We remove low-confidence peptides with less than 1% FDR.</li>
</ul>
<p>First, we need to convert PEP into q-value to control for FDR.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/pep2qvalue.html" class="external-link">pep2qvalue</a></span><span class="op">(</span></span>
<span>    <span class="va">woo</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">woo</span><span class="op">)</span>, PEP <span class="op">=</span> <span class="st">"PEP"</span>, rowDataName <span class="op">=</span> <span class="st">"qvalue"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>All the criteria to perform feature filtering are stored in the
<code>rowData</code>. We remove any peptide matched to a contaminant
protein or to a decoy peptide and control for 1% FDR. Here is an
overview of the distributions of each criteria.</p>
<pre><code><span><span class="co">## Warning: Removed 427 rows containing non-finite values (`stat_bin()`).</span></span>
<span><span class="co">## Warning: Removed 2 rows containing missing values (`geom_bar()`).</span></span></code></pre>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Peptide identification are already controlled for 1% FDR. We here
remove contaminant and decoy peptides.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span></span>
<span>    <span class="va">woo</span>, <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span> </span>
<span>        <span class="va">Potential.contaminant</span> <span class="op">!=</span> <span class="st">"+"</span><span class="op">)</span></span>
<span><span class="co">## 'Reverse' found in 1 out of 1 assay(s)</span></span>
<span><span class="co">## 'Potential.contaminant' found in 1 out of 1 assay(s)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cell-quality-control">Cell quality control<a class="anchor" aria-label="anchor" href="#cell-quality-control"></a>
</h3>
<p>Similarly to the features, we also remove low-quality cells. The
quality control criteria are:</p>
<ul>
<li>We remove samples with low number of detected peptides. The
criterion is computed as follows:</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/countUniqueFeatures.html" class="external-link">countUniqueFeatures</a></span><span class="op">(</span></span>
<span>    <span class="va">woo</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">woo</span><span class="op">)</span>, colDataName <span class="op">=</span> <span class="st">"NumberPeptides"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>We remove samples with low median intensity. The metric (note we
will later use it for normalization) is computed as follows:</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span><span class="op">$</span><span class="va">MedianIntensity</span> <span class="op">&lt;-</span> <span class="fu">colMedians</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">woo</span><span class="op">[[</span><span class="st">"peptides_intensity"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>We remove the samples that have a high median coefficients of
variation (CV). The CV is computed within each sample, by grouping the
peptides that belong to the same protein or protein group. This is
computed as follows:</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/medianCVperCell.html" class="external-link">medianCVperCell</a></span><span class="op">(</span></span>
<span>    <span class="va">woo</span>, i <span class="op">=</span> <span class="st">"peptides_intensity"</span>, groupBy <span class="op">=</span> <span class="st">"Leading.razor.protein"</span>,</span>
<span>    nobs <span class="op">=</span> <span class="fl">5</span>, norm <span class="op">=</span> <span class="st">"SCoPE2"</span>, colDataName <span class="op">=</span> <span class="st">"MedianCV"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We plot the metrics used to perform sample quality control.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">woo</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>        y <span class="op">=</span> <span class="va">MedianIntensity</span>,</span>
<span>        x <span class="op">=</span> <span class="va">NumberPeptides</span>,</span>
<span>        color <span class="op">=</span> <span class="va">MedianCV</span>,</span>
<span>        shape <span class="op">=</span> <span class="va">Treatment</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_color_continuous</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/plot_cell_qc-1.png" width="672" style="display: block; margin: auto;"></p>
<p>There are a few suspicious cells with low number of detected peptides
and high median CV. We will remove cells with less than 400 peptides and
with a median CV higher than 0.5. We will not use the median intensity
as a QC metric. We apply the filter and keep only single cells that pass
the quality control.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">passQC</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">woo</span><span class="op">$</span><span class="va">MedianCV</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">woo</span><span class="op">$</span><span class="va">MedianCV</span> <span class="op">&lt;</span> <span class="fl">0.5</span> <span class="op">&amp;</span></span>
<span>    <span class="va">woo</span><span class="op">$</span><span class="va">NumberPeptides</span> <span class="op">&gt;</span> <span class="fl">400</span></span>
<span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">woo</span>, <span class="va">passQC</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="log-transformation">Log-transformation<a class="anchor" aria-label="anchor" href="#log-transformation"></a>
</h3>
<p>We log2-transform the quantification data.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">woo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">woo</span>, i <span class="op">=</span> <span class="st">"peptides_intensity"</span>, name <span class="op">=</span> <span class="st">"peptides_log"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="data-modelling">Data modelling<a class="anchor" aria-label="anchor" href="#data-modelling"></a>
</h2>
<p>Model the data using the linear regression model implemented in
<code>scp</code>. The model is applied on a
<code>SingleCellExperiment</code> so we extract it from the processed
data set along with the <code>colData</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">woo</span>, <span class="st">"peptides_log"</span><span class="op">)</span></span></code></pre></div>
<p>First, we must specify which variables to include in the model. We
here include 3 variables:</p>
<ul>
<li>
<code>MedianIntensity</code>: this is the normalization factor used
to correct for cell-specific technical differences.</li>
<li>
<code>Chip</code>: the sample preparation chip is a potential source
of batch effect.</li>
<li>
<code>Treatment</code>: this is the biological variable of
interest.</li>
</ul>
<p><code><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelWorkflow()</a></code> fits linear regression models to the
data, where the model is adapted for each peptide depending on its
pattern of missing values.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelWorkflow</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, </span>
<span>    formula <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="co">## intercept</span></span>
<span>        <span class="co">## normalization</span></span>
<span>        <span class="va">MedianIntensity</span> <span class="op">+</span></span>
<span>        <span class="co">## batch effects</span></span>
<span>        <span class="va">Chip</span> <span class="op">+</span> </span>
<span>        <span class="co">## biological variability</span></span>
<span>        <span class="va">Treatment</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Once the model is prepared, we can explore the distribution of the
n/p ratios.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-class.html" class="external-link">scpModelFilterThreshold</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelFilterPlot</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Many peptides do not have sufficient observations to estimate the
model. We have chosen to continue the analysis with peptides that have
<span class="math inline">\(n/p &gt;= 2\)</span>. You could consider
<span class="math inline">\(n/p\)</span> a rough average of the number
of replicates per parameter to fit (for categorical variables, the
number of replicates per group). We recommend moving the threshold away
from 1 to increase statistical power and remove noisy peptides. This
comes of course at the cost of less peptides included in the
analysis.</p>
</div>
<div class="section level2">
<h2 id="model-analysis">Model analysis<a class="anchor" aria-label="anchor" href="#model-analysis"></a>
</h2>
<p>The model analysis consists of three steps:</p>
<ul>
<li>Variance analysis</li>
<li>Differential abundance analysis</li>
<li>Component analysis</li>
</ul>
<div class="section level3">
<h3 id="variance-analysis">Variance analysis<a class="anchor" aria-label="anchor" href="#variance-analysis"></a>
</h3>
<p>The variance analysis explores the proportion of data captures by
each variable in the model.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">vaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVarianceAnalysis</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## DataFrameList of length 4</span></span>
<span><span class="co">## names(4): Residuals MedianIntensity Chip Treatment</span></span>
<span><span class="va">vaRes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## DataFrame with 4042 rows and 4 columns</span></span>
<span><span class="co">##            feature        SS        df percentExplainedVar</span></span>
<span><span class="co">##        &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;           &lt;numeric&gt;</span></span>
<span><span class="co">## 1       AAAAATAATK  0.877832        12             9.23296</span></span>
<span><span class="co">## 2    AAAEVNQEYG...  2.143368        19            27.01360</span></span>
<span><span class="co">## 3        AAAFEQLQK 32.771158        36            69.11668</span></span>
<span><span class="co">## 4       AAAMANNLQK  3.751524         7            18.80773</span></span>
<span><span class="co">## 5        AAANEQLTR 15.121272        70            42.58844</span></span>
<span><span class="co">## ...            ...       ...       ...                 ...</span></span>
<span><span class="co">## 4038     YYPTEDVPR   1.96949         5            53.07102</span></span>
<span><span class="co">## 4039     YYSIASSSK   2.59158         7             9.97151</span></span>
<span><span class="co">## 4040     YYTLEEIQK  13.98114        55            43.16651</span></span>
<span><span class="co">## 4041      YYTPTISR   5.94088        33            45.93421</span></span>
<span><span class="co">## 4042 YYVTIIDAPG...  36.10122        39            56.63360</span></span></code></pre></div>
<p>The results are a list of tables, one table for each variable. Each
table reports for each peptide the variance captures (<code>SS</code>),
the residual degrees of freedom for estimating the variance
(<code>df</code>) and the percentage of total variance explained
(<code>percentExplainedVar</code>). To better explore the results, we
add the annotations available in the <code>rowData</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>By default, we explore the variance for all peptides combined:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span><span class="va">vaRes</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-13-1.png" width="240" style="display: block; margin: auto;"></p>
<p>We explore the top 20 peptides that are have the highest percentage
of variance explained by the biological variable (top) or by the batch
variable (bottom).</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Treatment"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Chip"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-14-1.png" width="960" style="display: block; margin: auto;"></p>
<p>We can also group these peptide according to their protein. We simply
need to provide the <code>fcol</code> argument.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Treatment"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span>, fcol <span class="op">=</span> <span class="st">"Gene.names"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Chip"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span>, fcol <span class="op">=</span> <span class="st">"Gene.names"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-15-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a>
</h3>
<p>Next, we explore the model output to understand the differences
between treatment conditions. The difference of interest is specified
using the <code>contrast</code> argument. The first element points to
the variable to test and the two following element are the groups of
interest to compare. Since we have 3 groups, we explore the three
possible contrasts, provided as a list.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpDifferentialAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Treatment"</span>, <span class="st">"CON"</span>, <span class="st">"LPS24"</span><span class="op">)</span>,</span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Treatment"</span>, <span class="st">"CON"</span>, <span class="st">"LPS48"</span><span class="op">)</span>,</span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Treatment"</span>, <span class="st">"LPS24"</span>, <span class="st">"LPS48"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## List of length 3</span></span>
<span><span class="co">## names(3): Treatment_CON_vs_LPS24 Treatment_CON_vs_LPS48 Treatment_LPS24_vs_LPS48</span></span></code></pre></div>
<p>Similarly to variance analysis, the results are a list of tables, one
table for each contrast.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">daRes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## DataFrame with 4042 rows and 7 columns</span></span>
<span><span class="co">##            feature  Estimate        SE        Df tstatistic      pvalue</span></span>
<span><span class="co">##        &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1       AAAAATAATK        NA        NA        12         NA          NA</span></span>
<span><span class="co">## 2    AAAEVNQEYG... -0.132380 0.0915688        19   -1.44569   0.1645585</span></span>
<span><span class="co">## 3        AAAFEQLQK  0.425190 0.2284671        36    1.86106   0.0709183</span></span>
<span><span class="co">## 4       AAAMANNLQK        NA        NA         7         NA          NA</span></span>
<span><span class="co">## 5        AAANEQLTR -0.145354 0.0747546        70   -1.94442   0.0558642</span></span>
<span><span class="co">## ...            ...       ...       ...       ...        ...         ...</span></span>
<span><span class="co">## 4038     YYPTEDVPR        NA        NA         5         NA          NA</span></span>
<span><span class="co">## 4039     YYSIASSSK  0.116941 0.4432525         7   0.263825 0.799513070</span></span>
<span><span class="co">## 4040     YYTLEEIQK -0.374134 0.0916765        55  -4.081020 0.000146227</span></span>
<span><span class="co">## 4041      YYTPTISR        NA        NA        33         NA          NA</span></span>
<span><span class="co">## 4042 YYVTIIDAPG...  0.787044 0.2081162        39   3.781753 0.000522754</span></span>
<span><span class="co">##             padj</span></span>
<span><span class="co">##        &lt;numeric&gt;</span></span>
<span><span class="co">## 1             NA</span></span>
<span><span class="co">## 2       0.256461</span></span>
<span><span class="co">## 3       0.129164</span></span>
<span><span class="co">## 4             NA</span></span>
<span><span class="co">## 5       0.106305</span></span>
<span><span class="co">## ...          ...</span></span>
<span><span class="co">## 4038          NA</span></span>
<span><span class="co">## 4039 0.859425862</span></span>
<span><span class="co">## 4040 0.000642391</span></span>
<span><span class="co">## 4041          NA</span></span>
<span><span class="co">## 4042 0.001944492</span></span></code></pre></div>
<p>Each table reports for each peptide the estimated difference between
the two groups, the standard error associated to the estimation, the
degrees of freedom, the t-statistics, the associated p-value and the
p-value FDR-adjusted for multiple testing across all peptides. Again, to
better explore the results, we add the annotations available in the
<code>rowData</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, </span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We then visualize the results using a volcano plot. The function
below return a volcano plot for each contrast. We here will show the
results for the first contrast.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span><span class="va">daRes</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;"></p>
<p>To help interpretation of the results, we will label the peptides
with their protein name. Also we increase the number of labels shown on
the plot. Finally, we can add colors to the plot. For instance, let’s
explore the impact of the number of observations using the <span class="math inline">\(n/p\)</span> ratio. We create a new annotation
table, add it to the results and redraw the plot. The <span class="math inline">\(n/p\)</span> ratio is retrieved using
scpModelFilterNPRatio. We show the improved plots for all contrasts.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">np</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-class.html" class="external-link">scpModelFilterNPRatio</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">np</span><span class="op">)</span>, npRatio <span class="op">=</span> <span class="va">np</span><span class="op">)</span>, </span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, top <span class="op">=</span> <span class="fl">30</span>, textBy <span class="op">=</span> <span class="st">"Gene.names"</span>, </span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">npRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-20-1.png" width="960" style="display: block; margin: auto;"></p>
<p>As expected, higher number of observations (higher <span class="math inline">\(n/p\)</span>) lead to increased statistical power
and hence to more significant results. We can already see that some
proteins are upregulated upon treatment such as heat shock proteins,
Npm1 and Mdh2.</p>
<p>Finally, we offer functionality to report results at the protein
level.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpDifferentialAggregate</a></span><span class="op">(</span><span class="va">daRes</span>, fcol <span class="op">=</span> <span class="st">"Gene.names"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span>top <span class="op">=</span> <span class="fl">30</span>, textBy <span class="op">=</span> <span class="st">"Gene.names"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-21-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="component-analysis">Component analysis<a class="anchor" aria-label="anchor" href="#component-analysis"></a>
</h3>
<p>Finally, we perform component analysis to link the modelled effects
to the cellular heterogeneity. We here run an APCA+ (extended
ANOVA-simultaneous principal component analysis) for the treatment
effect. In other words, we perform a PCA on the data that is capture by
the treatment variable along with the residuals (unmodelled data).</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">caRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, ncomp <span class="op">=</span> <span class="fl">20</span>, method <span class="op">=</span> <span class="st">"APCA"</span>, effect <span class="op">=</span> <span class="st">"Treatment"</span>, </span>
<span>    maxiter <span class="op">=</span> <span class="fl">200</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] "APCA"</span></span>
<span><span class="co">## [1] "Treatment"</span></span>
<span><span class="co">## List of length 2</span></span>
<span><span class="co">## names(2): bySample byFeature</span></span></code></pre></div>
<p>The results are contained in a list with 2 elements.
<code>bySample</code> contains the PC scores, that is the component
results in sample space. <code>byFeature</code> contains the
eigenvectors, that is the component results in feature space.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span></span>
<span><span class="co">## List of length 3</span></span>
<span><span class="co">## names(3): unmodelled residuals APCA_Treatment</span></span></code></pre></div>
<p>Each of the two elements contains components results for the data
before modelling (<code>unmodelled</code>), for the residuals or for the
APCA on the sample type variable (<code>APCA_Treatment</code>). Each
elements is a table with the computed components. Let’s explore the
component analysis results in cell space. Similarly to the previous
explorations, we annotate the results.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResCells</span> <span class="op">&lt;-</span> <span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">caResCells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span><span class="va">caResCells</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span></span></code></pre></div>
<p>We then generate the component plot, colouring by
<code>Treatment</code>. To assess the impact of batch effects, we shape
the points according to the sample preparation batch (cf intro) as
well.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>,</span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">Treatment</span>, shape <span class="op">=</span> <span class="va">Chip</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-24-1.png" width="864" style="display: block; margin: auto;"></p>
<p>The data before modelling already shows some data pattern driven by
the treatment as control samples are mostly characterized by low PC1
values, while treated cells show on average higher PC1 scores. However,
the data also shows batch effects as cells from the same batch tend to
group together. Batch effects are removed upon modelling and treatment
conditions are better separated although they still overlap. There is
not apparent structure in the residuals, indicating that the remaining
variation may be attributed to noise.</p>
<p>We explore further the variance that is hidden in later components.
In the chunk above, we computed 20 principal components. To explore the
patterns they capture, we reduce these 20 PCs to 2 dimensions using
t-SNE. We rely on the <code>scater</code> package to compute the
t-SNE.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">caResCells</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ca</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">pcs</span> <span class="op">&lt;-</span> <span class="va">ca</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^PC"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ca</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="va">tsne</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">calculateTSNE</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pcs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">ca</span>, <span class="va">tsne</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TSNE1</span>,</span>
<span>            y <span class="op">=</span> <span class="va">TSNE2</span>,</span>
<span>            colour <span class="op">=</span> <span class="va">Treatment</span>,</span>
<span>            shape <span class="op">=</span> <span class="va">Chip</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_woo2022_files/figure-html/unnamed-chunk-25-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Taking the 20 principal components into account, we can see good
separation between treatment after modelling while correctly mixing the
batch effects. We can also clearly see a clear clustering of the sample
preparation batches when data is not modelled. Again, the residuals show
no interesting clustering.</p>
</div>
</div>
<div class="section level2">
<h2 id="interactive-visualisation-with-isee">Interactive visualisation with iSEE<a class="anchor" aria-label="anchor" href="#interactive-visualisation-with-isee"></a>
</h2>
<p>You can manually explore the data through an interactive interface
thanks to using the iSEE:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/iSEE/iSEE" class="external-link">"iSEE"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code>R Under development (unstable) (2023-07-27 r84768)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 23.04

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.11.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.11.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Prague
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.31.0               scuttle_1.13.0             
 [3] SingleCellExperiment_1.25.0 dplyr_1.1.3                
 [5] patchwork_1.1.3             ggplot2_3.4.4              
 [7] scpdata_1.9.2               ExperimentHub_2.11.0       
 [9] AnnotationHub_3.11.0        BiocFileCache_2.11.1       
[11] dbplyr_2.4.0                scp_1.11.3                 
[13] QFeatures_1.13.0            MultiAssayExperiment_1.29.0
[15] SummarizedExperiment_1.33.0 Biobase_2.63.0             
[17] GenomicRanges_1.55.1        GenomeInfoDb_1.39.0        
[19] IRanges_2.37.0              S4Vectors_0.41.1           
[21] BiocGenerics_0.49.1         MatrixGenerics_1.15.0      
[23] matrixStats_1.1.0           BiocStyle_2.31.0           

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3            jsonlite_1.8.7               
  [3] magrittr_2.0.3                ggbeeswarm_0.7.2             
  [5] farver_2.1.1                  rmarkdown_2.25               
  [7] fs_1.6.3                      zlibbioc_1.49.0              
  [9] ragg_1.2.6                    vctrs_0.6.4                  
 [11] memoise_2.0.1                 DelayedMatrixStats_1.25.0    
 [13] RCurl_1.98-1.13               BiocBaseUtils_1.5.0          
 [15] htmltools_0.5.7               S4Arrays_1.3.0               
 [17] curl_5.1.0                    BiocNeighbors_1.21.0         
 [19] SparseArray_1.3.0             sass_0.4.7                   
 [21] bslib_0.5.1                   desc_1.4.2                   
 [23] cachem_1.0.8                  igraph_1.5.1                 
 [25] mime_0.12                     lifecycle_1.0.4              
 [27] pkgconfig_2.0.3               rsvd_1.0.5                   
 [29] Matrix_1.6-1.1                R6_2.5.1                     
 [31] fastmap_1.1.1                 GenomeInfoDbData_1.2.11      
 [33] shiny_1.7.5.1                 clue_0.3-65                  
 [35] digest_0.6.33                 fdrtool_1.2.17               
 [37] colorspace_2.1-0              AnnotationDbi_1.65.2         
 [39] rprojroot_2.0.4               irlba_2.3.5.1                
 [41] textshaping_0.3.7             lpsymphony_1.31.0            
 [43] RSQLite_2.3.3                 beachmat_2.19.0              
 [45] labeling_0.4.3                filelock_1.0.2               
 [47] fansi_1.0.5                   httr_1.4.7                   
 [49] abind_1.4-5                   compiler_4.4.0               
 [51] bit64_4.0.5                   withr_2.5.2                  
 [53] BiocParallel_1.37.0           viridis_0.6.4                
 [55] DBI_1.1.3                     highr_0.10                   
 [57] MASS_7.3-60.1                 rappdirs_0.3.3               
 [59] DelayedArray_0.29.0           tools_4.4.0                  
 [61] vipor_0.4.5                   beeswarm_0.4.0               
 [63] interactiveDisplayBase_1.41.0 httpuv_1.6.12                
 [65] glue_1.6.2                    promises_1.2.1               
 [67] grid_4.4.0                    Rtsne_0.16                   
 [69] cluster_2.1.4                 generics_0.1.3               
 [71] gtable_0.3.4                  tidyr_1.3.0                  
 [73] ScaledMatrix_1.11.0           BiocSingular_1.19.0          
 [75] metapod_1.11.0                utf8_1.2.4                   
 [77] XVector_0.43.0                ggrepel_0.9.4                
 [79] BiocVersion_3.19.1            pillar_1.9.0                 
 [81] stringr_1.5.0                 later_1.3.1                  
 [83] lattice_0.22-5                bit_4.0.5                    
 [85] tidyselect_1.2.0              Biostrings_2.71.1            
 [87] knitr_1.45                    gridExtra_2.3                
 [89] bookdown_0.36                 ProtGenerics_1.35.0          
 [91] IHW_1.31.0                    xfun_0.41                    
 [93] stringi_1.7.12                lazyeval_0.2.2               
 [95] yaml_2.3.7                    evaluate_0.23                
 [97] codetools_0.2-19              nipals_0.8                   
 [99] MsCoreUtils_1.15.1            tibble_3.2.1                 
[101] BiocManager_1.30.22           cli_3.6.1                    
[103] xtable_1.8-4                  systemfonts_1.0.5            
[105] munsell_0.5.0                 jquerylib_0.1.4              
[107] Rcpp_1.0.11                   png_0.1-8                    
[109] parallel_4.4.0                ellipsis_0.3.2               
[111] pkgdown_2.0.7                 blob_1.2.4                   
[113] AnnotationFilter_1.27.0       sparseMatrixStats_1.15.0     
[115] bitops_1.0-7                  viridisLite_0.4.2            
[117] slam_0.1-50                   scales_1.2.1                 
[119] purrr_1.0.2                   crayon_1.5.2                 
[121] rlang_1.1.2                   KEGGREST_1.43.0              </code></pre>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">citation</span>(<span class="st">"scp"</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>To cite the scp package <span class="cf">in</span> publications use<span class="sc">:</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  Vanderaa, Christophe, and Laurent Gatto. <span class="fl">2023.</span>  Revisiting the Thorny</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  Issue of Missing Values <span class="cf">in</span> Single<span class="sc">-</span>Cell Proteomics. Journal of</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  Proteome Research <span class="dv">22</span> (<span class="dv">9</span>)<span class="sc">:</span> 2775–<span class="fl">84.</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>  Vanderaa Christophe and Laurent Gatto. The current state of</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  single<span class="sc">-</span>cell proteomics data analysis. Current Protocols <span class="dv">3</span> (<span class="dv">1</span>)<span class="sc">:</span> e658.;</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>  doi<span class="sc">:</span> https<span class="sc">:</span><span class="er">//</span>doi.org<span class="sc">/</span><span class="fl">10.1002</span><span class="sc">/</span><span class="fu">cpz1.658</span> (<span class="dv">2023</span>).</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>  Vanderaa Christophe and Laurent Gatto. Replication of Single<span class="sc">-</span>Cell</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>  Proteomics Data Reveals Important Computational Challenges. Expert</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>  Review of Proteomics, 1–<span class="dv">9</span> (<span class="dv">2021</span>).</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>To see these entries <span class="cf">in</span> BibTeX format, use <span class="st">'print(&lt;citation&gt;,</span></span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a><span class="st">bibtex=TRUE)'</span>, <span class="st">'toBibtex(.)'</span>, or set</span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a><span class="st">'options(citation.bibtex.max=999)'</span>.</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
license</a> license.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
