<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scplainer: reanalysis of the nPOP dataset (Leduc et al. 2022) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scplainer: reanalysis of the nPOP dataset (Leduc et al. 2022)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/brunner2022.html">Replication of the cell cycle state study (Brunner et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/derks2022.html">Replication of the plexDIA analysis (Derks et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/leduc2022.html">Replication of the nPOP analysis (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Replication of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/scplainer_leduc2022.html">scplainer: reanalysis of the nPOP dataset (Leduc et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/scplainer_woo2022.html">SCP data analysis of the macrophage activation dataset (Woo et al. 2022)</a>
    </li>
    <li>
      <a href="../articles/williams2020_tmt.html">Reproducing the multiplexed SCP analysis by Williams et al. 2020</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scplainer: reanalysis of the nPOP dataset (Leduc
et al. 2022)</h1>
                        <h4 data-toc-skip class="author">Christophe
Vanderaa<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, Computational Biology, UCLouvain</h4>
                        <h4 data-toc-skip class="author">Laurent Gatto,
Computational Biology, UCLouvain</h4>
            
            <h4 data-toc-skip class="date">10 November 2023</h4>
      
      
      <div class="hidden name"><code>scplainer_leduc2022.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette, we will analyse the <code>leduc2022_pSCoPE</code>
dataset. The data were acquired using the nPOP acquisition protocole.
nPOP (<a href="http://dx.doi.org/10.1101/2021.04.24.441211" class="external-link">Leduc et
al. 2022</a>) is an upgrade of the SCoPE2 protocol (<a href="http://dx.doi.org/10.1186/s13059-021-02267-5" class="external-link">Specht et
al. 2021</a> and <a href="http://dx.doi.org/10.1038/s41596-021-00616-z" class="external-link">Petelski et
al. 2021</a>), where the mPOP sample preparation method is replaced by
the nPOP method. nPOP processes samples using the Cellenion dispensing
device and uses DMSO as lysis reagent instead of a freeze-thaw
procedure. They also include the prioritised data acquisition mode as
described by <a href="http://dx.doi.org/10.1101/2022.03.16.484655" class="external-link">Huffman et
al. 2022</a>. Single cells were labelled with TMT-18 and MS data were
acquired in prioritised DDA mode.</p>
</div>
<div class="section level2">
<h2 id="packages-and-data">Packages and data<a class="anchor" aria-label="anchor" href="#packages-and-data"></a>
</h2>
<p>We rely on several packages to compile this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Core packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">"scp"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"scpdata"</span><span class="op">)</span></span>
<span><span class="co">## Utility packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/jorainer/ensembldb" class="external-link">"ensembldb"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"EnsDb.Hsapiens.v86"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"bluster"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/scater/" class="external-link">"scater"</a></span><span class="op">)</span></span></code></pre></div>
<p>The data set is available from the <code>scpdata</code> package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/leduc2022_pSCoPE.html" class="external-link">leduc2022_pSCoPE</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## see ?scpdata and browseVignettes('scpdata') for documentation</span></span>
<span><span class="co">## loading from cache</span></span></code></pre></div>
<p>The data set primarily consists of melanoma cells and monocytes. The
data set also contains carrier samples, negative control samples,
reference samples and empty wells (<code>Unused</code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Carrier  Melanoma  Monocyte  Negative Reference    Unused </span></span>
<span><span class="co">##       134       878       877       120       134       269</span></span></code></pre></div>
<p>The data were acquired using TMT-18 labelling.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">Channel</span><span class="op">)</span></span>
<span><span class="co">##  [1] "TMT126"  "TMT127N" "TMT127C" "TMT128N" "TMT128C" "TMT129N" "TMT129C"</span></span>
<span><span class="co">##  [8] "TMT130N" "TMT130C" "TMT131N" "TMT131C" "TMT132N" "TMT132C" "TMT133N"</span></span>
<span><span class="co">## [15] "TMT133C" "TMT134N" "TMT134C" "TMT135N"</span></span></code></pre></div>
<p>The data were acquired as part of 134 MS acquisition batches.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">Set</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] 134</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">Set</span><span class="op">)</span>, <span class="st">"..."</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">Set</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## eAL00219 eAL00219 eAL00219 eAL00219 eAL00219 eAL00219 ... wAL00286 wAL00286 wAL00286 wAL00286 wAL00286 wAL00286</span></span></code></pre></div>
<p>Finally, samples were prepared with the nPOP protocol using 2 glass
slides. The information is stored under <code>lcbatch</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">lcbatch</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    A    C </span></span>
<span><span class="co">## 1548  864</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="minimal-data-processing">Minimal data processing<a class="anchor" aria-label="anchor" href="#minimal-data-processing"></a>
</h2>
<p>The minimal data processing workflow consist of 5 main steps:</p>
<ul>
<li>Data cleaning</li>
<li>Feature quality control</li>
<li>Sample quality control</li>
<li>Peptide data assembly</li>
<li>Log2-transformation</li>
</ul>
<div class="section level3">
<h3 id="cleaning-data">Cleaning data<a class="anchor" aria-label="anchor" href="#cleaning-data"></a>
</h3>
<p>We remove the assays that were processed by the authors. The vignette
only uses the data generated by Maxquant.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assaysToRemove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"peptides"</span>, <span class="st">"peptides_log"</span>, <span class="st">"proteins_norm2"</span>, <span class="st">"proteins_processed"</span><span class="op">)</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">removeAssay</a></span><span class="op">(</span><span class="va">leduc</span>, <span class="va">assaysToRemove</span><span class="op">)</span></span>
<span><span class="co">## Warning: 'experiments' dropped; see 'drops()'</span></span>
<span><span class="co">## harmonizing input:</span></span>
<span><span class="co">##   removing 6185 sampleMap rows not in names(experiments)</span></span></code></pre></div>
<p>We remove feature annotations that won’t be used in the remainder of
the vignette. This is to avoid overcrowding of the annotation tables
later in the vignette.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">requiredRowData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Sequence"</span>, <span class="st">"Leading.razor.protein.symbol"</span>, </span>
<span>    <span class="st">"Leading.razor.protein.id"</span>, <span class="st">"Reverse"</span>, <span class="st">"Potential.contaminant"</span>,</span>
<span>    <span class="st">"Leading.razor.protein"</span>, <span class="st">"PIF"</span>, <span class="st">"dart_qval"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">selectRowData</a></span><span class="op">(</span><span class="va">leduc</span>, <span class="va">requiredRowData</span><span class="op">)</span></span></code></pre></div>
<p>We replace zeros by missing values. A zero may be a true (the feature
is not present in the sample) or because of technical limitations (due
to the technology or the computational pre-processing). Because we are
not able to distinguish between the two, zeros should be replaced with
<code>NA</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-quality-control">Feature quality control<a class="anchor" aria-label="anchor" href="#feature-quality-control"></a>
</h3>
<p>We remove low-quality PSMs that may propagate technical artefacts and
bias data modelling. The quality control criteria are:</p>
<ul>
<li>We remove contaminants and decoy peptides.</li>
<li>We remove PSMs with low spectral purity.</li>
<li>We remove low-confidence peptides, as defined by the false discovery
rate (FDR) computed by the DART-ID algorithm.</li>
<li>We remove PSMs for which the signal in single-cell samples exceeds
5% of the signal in 100-cell samples, i.e. the carrier.</li>
</ul>
<p>All the criteria were readily computed except for the single cell to
carrier ratio. We compute this using <code><a href="https://UCLouvain-CBIO.github.io/scp/reference/computeSCR.html" class="external-link">computeSCR()</a></code>. The
results are stored in the <code>rowData</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/computeSCR.html" class="external-link">computeSCR</a></span><span class="op">(</span></span>
<span>    <span class="va">leduc</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>, colvar <span class="op">=</span> <span class="st">"SampleType"</span>, </span>
<span>    samplePattern <span class="op">=</span> <span class="st">"Mel|Macro"</span>, carrierPattern <span class="op">=</span> <span class="st">"Carrier"</span>,</span>
<span>    sampleFUN <span class="op">=</span> <span class="st">"mean"</span>, rowDataName <span class="op">=</span> <span class="st">"MeanSCR"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here is an overview of the distributions of each criteria</p>
<pre><code><span><span class="co">## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span>
<span><span class="co">## Warning: Removed 181437 rows containing non-finite values</span></span>
<span><span class="co">## (`stat_bin()`).</span></span>
<span><span class="co">## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span>
<span><span class="co">## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span>
<span><span class="co">## Warning: Removed 205172 rows containing non-finite values</span></span>
<span><span class="co">## (`stat_bin()`).</span></span></code></pre>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We filter and keep the features that pass the quality control
criteria.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span></span>
<span>    <span class="va">leduc</span>, <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">Potential.contaminant</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span></span>
<span>        <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"REV|CON"</span>, <span class="va">Leading.razor.protein</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">PIF</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">PIF</span> <span class="op">&gt;</span> <span class="fl">0.6</span> <span class="op">&amp;</span></span>
<span>        <span class="va">dart_qval</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span></span>
<span>        <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">MeanSCR</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">MeanSCR</span> <span class="op">&lt;</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## 'Reverse' found in 134 out of 134 assay(s)</span></span>
<span><span class="co">## 'Potential.contaminant' found in 134 out of 134 assay(s)</span></span>
<span><span class="co">## 'Leading.razor.protein' found in 134 out of 134 assay(s)</span></span>
<span><span class="co">## 'PIF' found in 134 out of 134 assay(s)</span></span>
<span><span class="co">## 'dart_qval' found in 134 out of 134 assay(s)</span></span>
<span><span class="co">## 'MeanSCR' found in 134 out of 134 assay(s)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sample-quality-control">Sample quality control<a class="anchor" aria-label="anchor" href="#sample-quality-control"></a>
</h3>
<p>Similarly to the features, we also remove low-quality cells. The
quality control criteria are:</p>
<ul>
<li>We remove samples with low number of detected peptides. The
criterion is computed as follows:</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/countUniqueFeatures.html" class="external-link">countUniqueFeatures</a></span><span class="op">(</span></span>
<span>    <span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>, groupBy <span class="op">=</span> <span class="st">"Sequence"</span>,</span>
<span>    colDataName <span class="op">=</span> <span class="st">"NumberPeptides"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>We remove samples with low median intensity. The metric (note we
will later use it for normalization) is computed as follows:</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MedianIntensity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu">experiments</span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">colMedians</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">out</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">MedianIntensity</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">MedianIntensity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">MedianIntensity</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">MedianIntensity</span><span class="op">)</span>, <span class="st">"MedianIntensity"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">MedianIntensity</span></span></code></pre></div>
<ul>
<li>We remove the samples that have a high median coefficients of
variation (CV). The CV is computed within each sample, by grouping the
peptides that belong to the same protein or protein group. This is
computed as follows:</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/medianCVperCell.html" class="external-link">medianCVperCell</a></span><span class="op">(</span></span>
<span>    <span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>, groupBy <span class="op">=</span> <span class="st">"Leading.razor.protein.symbol"</span>,</span>
<span>    nobs <span class="op">=</span> <span class="fl">3</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span>, colDataName <span class="op">=</span> <span class="st">"MedianCV"</span>, norm <span class="op">=</span> <span class="st">"SCoPE2"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Warning in medianCVperCell(leduc, i = names(leduc), groupBy = "Leading.razor.protein.symbol", : The median CV could not be computed for one or more samples. You may want to try a smaller value for 'nobs'.</span></span></code></pre></div>
<ul>
<li>We also remove the samples that are not single cells because we will
no longer need them.</li>
</ul>
<p>We plot the metrics used to perform sample quality control.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>        y <span class="op">=</span> <span class="va">MedianIntensity</span>,</span>
<span>        x <span class="op">=</span> <span class="va">NumberPeptides</span>,</span>
<span>        color <span class="op">=</span> <span class="va">MedianCV</span>,</span>
<span>        shape <span class="op">=</span> <span class="va">SampleType</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_color_continuous</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span></span>
<span><span class="co">## Warning: Removed 2 rows containing missing values (`geom_point()`).</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We apply the filter and keep only single cells that pass the quality
control.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">passQC</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">$</span><span class="va">MedianCV</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">leduc</span><span class="op">$</span><span class="va">MedianCV</span> <span class="op">&lt;</span> <span class="fl">0.6</span> <span class="op">&amp;</span></span>
<span>    <span class="va">leduc</span><span class="op">$</span><span class="va">MedianIntensity</span> <span class="op">&gt;</span> <span class="fl">6</span> <span class="op">&amp;</span> <span class="va">leduc</span><span class="op">$</span><span class="va">MedianIntensity</span> <span class="op">&lt;</span> <span class="fl">8</span> <span class="op">&amp;</span></span>
<span>    <span class="va">leduc</span><span class="op">$</span><span class="va">NumberPeptides</span> <span class="op">&gt;</span> <span class="fl">750</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Mono|Mel"</span>, <span class="va">leduc</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu">subsetByColData</span><span class="op">(</span><span class="va">leduc</span>, <span class="va">passQC</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="peptide-data-assembly">Peptide data assembly<a class="anchor" aria-label="anchor" href="#peptide-data-assembly"></a>
</h3>
<p>For now each MS acquisition run is stored separately in an assay. We
here combine these assays in one. The issue is that PSMs are specific to
each run. We therefore aggregate the PSMs to peptides.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peptideAssays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"peptides_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">leduc</span>,</span>
<span>                           i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span>,</span>
<span>                           fcol <span class="op">=</span> <span class="st">"Sequence"</span>,</span>
<span>                           name <span class="op">=</span> <span class="va">peptideAssays</span>,</span>
<span>                           fun <span class="op">=</span> <span class="va">colMedians</span>,</span>
<span>                           na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Next to that, we must adapt the peptide to protein mapping. When
joining all assay, we will only keep the feature annotations that have
the common data. However, some peptide sequences map to one protein in
one run and to another protein in another run. Hence, the protein name
is not constant for all peptides and is removed during joining. It is
important we keep the protein sequence in the <code>rowData</code>, in
case we later want to aggregate, model or infer protein level
quantification.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ppMap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^pep"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Sequence</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">## The majority vote happens here</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Leading.razor.protein.symbol <span class="op">=</span></span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">Leading.razor.protein.symbol</span><span class="op">)</span>,</span>
<span>                          decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>           Leading.razor.protein.id <span class="op">=</span></span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">Leading.razor.protein.id</span><span class="op">)</span>,</span>
<span>                          decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Sequence</span>, <span class="va">Leading.razor.protein.symbol</span>, <span class="va">Leading.razor.protein.id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">Sequence</span>, <span class="va">Leading.razor.protein.symbol</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">peptideAssays</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Sequence</span>, <span class="va">ppMap</span><span class="op">$</span><span class="va">Sequence</span><span class="op">)</span></span>
<span>    <span class="fu">DataFrame</span><span class="op">(</span>Leading.razor.protein.symbol <span class="op">=</span></span>
<span>                  <span class="va">ppMap</span><span class="op">$</span><span class="va">Leading.razor.protein.symbol</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span>,</span>
<span>              Leading.razor.protein.id <span class="op">=</span> </span>
<span>                  <span class="va">ppMap</span><span class="op">$</span><span class="va">Leading.razor.protein.id</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">consensus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">peptideAssays</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">consensus</span></span></code></pre></div>
<p>The data can now be joined.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html" class="external-link">joinAssays</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="va">peptideAssays</span>, </span>
<span>                    name <span class="op">=</span> <span class="st">"peptides"</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we also convert uniprot protein identifier to gene
symbols.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proteinIds</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Leading.razor.protein.id</span></span>
<span><span class="va">proteinConversionDf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">transcripts</a></span><span class="op">(</span></span>
<span>    <span class="va">EnsDb.Hsapiens.v86</span>, </span>
<span>    columns <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>    return.type <span class="op">=</span> <span class="st">"data.frame"</span>,</span>
<span>    filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">UniprotFilter</a></span><span class="op">(</span><span class="va">proteinIds</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">matchedIndex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">proteinIds</span>, <span class="va">proteinConversionDf</span><span class="op">$</span><span class="va">uniprot_id</span><span class="op">)</span></span>
<span><span class="va">geneName</span> <span class="op">&lt;-</span> <span class="va">proteinConversionDf</span><span class="op">$</span><span class="va">gene_name</span><span class="op">[</span><span class="va">matchedIndex</span><span class="op">]</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span><span class="op">[[</span><span class="st">"peptides"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="va">geneName</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="log-transformation">Log-transformation<a class="anchor" aria-label="anchor" href="#log-transformation"></a>
</h3>
<p>We log2-transform the quantification data.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">leduc</span>, i <span class="op">=</span> <span class="st">"peptides"</span>, name <span class="op">=</span> <span class="st">"peptides_log"</span><span class="op">)</span></span></code></pre></div>
<p>Here is an overview of the data processing:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">leduc</span><span class="op">)</span></span>
<span><span class="co">## Warning in plot.QFeatures(leduc): The QFeatures object contains many assays.</span></span>
<span><span class="co">## You may want to consider creating an interactive plot (set 'interactive =</span></span>
<span><span class="co">## TRUE')</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="data-modelling">Data modelling<a class="anchor" aria-label="anchor" href="#data-modelling"></a>
</h2>
<p>Model the data using the linear regression model implemented in
<code>scp</code>. The model is applied on a
<code>SingleCellExperiment</code> so we extract it from the processed
data set along with the <code>colData</code></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">leduc</span>, <span class="st">"peptides_log"</span><span class="op">)</span></span></code></pre></div>
<p>First, we must specify which variables to include in the model. We
here include 4 variables:</p>
<ul>
<li>
<code>MedianIntensity</code>: this is the normalization factor used
to correct for cell-specific technical differences.</li>
<li>
<code>Channel</code>: this is used to correct for TMT effects.</li>
<li>
<code>Set</code>: this is used to perform batch correction. We
consider each acquisition run to be a batch.</li>
<li>
<code>SampleType</code>: this is the biological variable of
interest. It capture the difference between macrophages and
monocytes</li>
</ul>
<p><code><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelWorkflow()</a></code> fits linear regression models to the
data, where the model is adapted for each peptide depending on its
pattern of missing values.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelWorkflow</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>,</span>
<span>    formula <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="co">## intercept</span></span>
<span>        <span class="co">## normalization</span></span>
<span>        <span class="va">MedianIntensity</span> <span class="op">+</span></span>
<span>        <span class="co">## batch effects</span></span>
<span>        <span class="va">Channel</span> <span class="op">+</span> <span class="va">Set</span> <span class="op">+</span> </span>
<span>        <span class="co">## biological variability</span></span>
<span>        <span class="va">SampleType</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Once the model is prepared, we can explore the distribution of the
n/p ratios.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-class.html" class="external-link">scpModelFilterThreshold</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelFilterPlot</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">## To change the threshold, use:</span></span>
<span><span class="co">## scpModelFilterThreshold(object, name) &lt;- threshold</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Many peptides do not have sufficient observations to estimate the
model. We have chosen to continue the analysis with peptides that have
<span class="math inline">\(n/p &gt;= 3\)</span>. You could consider
<span class="math inline">\(n/p\)</span> a rough average of the number
of replicates per parameter to fit (for categorical variables, the
number of replicates per group). We recommend moving the threshold away
from 1 to increase statistical power and remove noisy peptides. This
comes of course at the cost of less peptides included in the
analysis.</p>
</div>
<div class="section level2">
<h2 id="model-analysis">Model analysis<a class="anchor" aria-label="anchor" href="#model-analysis"></a>
</h2>
<p>The model analysis consists of three steps:</p>
<ul>
<li>Variance analysis</li>
<li>Differential abundance analysis</li>
<li>Component analysis</li>
</ul>
<div class="section level3">
<h3 id="variance-analysis">Variance analysis<a class="anchor" aria-label="anchor" href="#variance-analysis"></a>
</h3>
<p>The variance analysis explores the proportion of data captures by
each variable in the model.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">vaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVarianceAnalysis</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## DataFrameList of length 5</span></span>
<span><span class="co">## names(5): Residuals MedianIntensity Channel Set SampleType</span></span>
<span><span class="va">vaRes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## DataFrame with 6854 rows and 4 columns</span></span>
<span><span class="co">##            feature        SS        df percentExplainedVar</span></span>
<span><span class="co">##        &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;           &lt;numeric&gt;</span></span>
<span><span class="co">## 1          GTLVQTK   428.956      1502            15.58356</span></span>
<span><span class="co">## 2    NSNPALNDNL...   350.165      1357            25.62569</span></span>
<span><span class="co">## 3    ITVTSEVPFS...   273.791      1365            22.35432</span></span>
<span><span class="co">## 4    LVLVGDGGTG...   242.713      1471             9.67771</span></span>
<span><span class="co">## 5          LVAQLYK   361.345      1175            36.60871</span></span>
<span><span class="co">## ...            ...       ...       ...                 ...</span></span>
<span><span class="co">## 6850 LIHPDEDISL...   75.4117       290             22.2277</span></span>
<span><span class="co">## 6851 MQEHSDQVPV...  125.2367       494             45.1561</span></span>
<span><span class="co">## 6852 HGAEVIDTPV...   38.6879       126             58.5003</span></span>
<span><span class="co">## 6853 TETQEKNPLP...  404.3668       790             14.5120</span></span>
<span><span class="co">## 6854 TLTTVQGIAD...   77.5062       253             66.6051</span></span></code></pre></div>
<p>The results are a list of tables, one table for each variable. Each
table reports for each peptide the variance captures (<code>SS</code>),
the residual degrees of freedom for estimating the variance
(<code>df</code>) and the percentage of total variance explained
(<code>percentExplainedVar</code>). To better explore the results, we
add the annotations available in the <code>rowData</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>By default, we explore the variance for all peptides combined:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span><span class="va">vaRes</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-24-1.png" width="240" style="display: block; margin: auto;"></p>
<p>We explore the top 20 peptides that are have the highest percentage
of variance explained by the biological variable (top) or by the batch
variable (bottom).</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Set"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-25-1.png" width="960" style="display: block; margin: auto;"></p>
<p>We can also group these peptide according to their protein. We simply
need to provide the <code>fcol</code> argument.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span>, fcol <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-VarianceAnalysis.html" class="external-link">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Set"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span>, fcol <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-26-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a>
</h3>
<p>Next, we explore the model output to understand the differences
between melanoma cells and monocytes. The difference of interest is
specified using the <code>contrast</code> argument. The first element
points to the variable to test and the two following element are the
groups of interest to compare. You can provide multiple contrast in a
list.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpDifferentialAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleType"</span>, <span class="st">"Melanoma"</span>, <span class="st">"Monocyte"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## List of length 1</span></span>
<span><span class="co">## names(1): SampleType_Melanoma_vs_Monocyte</span></span>
<span><span class="va">daRes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## DataFrame with 6854 rows and 7 columns</span></span>
<span><span class="co">##            feature   Estimate        SE        Df tstatistic       pvalue</span></span>
<span><span class="co">##        &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## 1          GTLVQTK -0.2348267 0.0461710      1502   -5.08602  4.11671e-07</span></span>
<span><span class="co">## 2    NSNPALNDNL... -0.2964675 0.0476322      1357   -6.22410  6.44196e-10</span></span>
<span><span class="co">## 3    ITVTSEVPFS...  0.1699631 0.0422754      1365    4.02038  6.12828e-05</span></span>
<span><span class="co">## 4    LVLVGDGGTG... -0.0575517 0.0359411      1471   -1.60128  1.09529e-01</span></span>
<span><span class="co">## 5          LVAQLYK  0.2392563 0.0586715      1175    4.07790  4.85057e-05</span></span>
<span><span class="co">## ...            ...        ...       ...       ...        ...          ...</span></span>
<span><span class="co">## 6850 LIHPDEDISL...  0.1568293 0.1001117       290   1.566543  1.18312e-01</span></span>
<span><span class="co">## 6851 MQEHSDQVPV... -0.0352213 0.0803697       494  -0.438241  6.61403e-01</span></span>
<span><span class="co">## 6852 HGAEVIDTPV...  0.0925308 0.1753991       126   0.527545  5.98743e-01</span></span>
<span><span class="co">## 6853 TETQEKNPLP... -2.8525348 0.1002023       790 -28.467765 3.09874e-123</span></span>
<span><span class="co">## 6854 TLTTVQGIAD...  0.1569952 0.1299189       253   1.208409  2.28018e-01</span></span>
<span><span class="co">##              padj</span></span>
<span><span class="co">##         &lt;numeric&gt;</span></span>
<span><span class="co">## 1     3.30397e-06</span></span>
<span><span class="co">## 2     6.93142e-09</span></span>
<span><span class="co">## 3     3.63979e-04</span></span>
<span><span class="co">## 4     2.69653e-01</span></span>
<span><span class="co">## 5     2.91886e-04</span></span>
<span><span class="co">## ...           ...</span></span>
<span><span class="co">## 6850  2.84930e-01</span></span>
<span><span class="co">## 6851  8.40398e-01</span></span>
<span><span class="co">## 6852  8.01521e-01</span></span>
<span><span class="co">## 6853 2.35986e-120</span></span>
<span><span class="co">## 6854  4.57104e-01</span></span></code></pre></div>
<p>Similarly to variance analysis, the results are a list of tables, one
table for each contrast. Each table reports for each peptide the
estimated difference between the two groups, the standard error
associated to the estimation, the degrees of freedom, the t-statistics,
the associated p-value and the p-value FDR-adjusted for multiple testing
across all peptides. Again, to better explore the results, we add the
annotations available in the <code>rowData</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, </span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We then visualize the results using a volcano plot. The function
below return a volcano plot for each contrast.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span><span class="va">daRes</span><span class="op">)</span></span>
<span><span class="co">## $SampleType_Melanoma_vs_Monocyte</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;"></p>
<p>To help interpretation of the results, we will label the peptides
with their protein name. Also we increase the number of labels shown on
the plot. Finally, we can add colors to the plot. For instance, let’s
explore the impact of the number of observations using the <span class="math inline">\(n/p\)</span> ratio. We create a new annotation
table, add it to the results and redraw the plot. The <span class="math inline">\(n/p\)</span> ratio is retrieved using
scpModelFilterNPRatio</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">np</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-class.html" class="external-link">scpModelFilterNPRatio</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">np</span><span class="op">)</span>, npRatio <span class="op">=</span> <span class="va">np</span><span class="op">)</span>, </span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, top <span class="op">=</span> <span class="fl">30</span>, textBy <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">npRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## $SampleType_Melanoma_vs_Monocyte</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-30-1.png" width="672" style="display: block; margin: auto;"></p>
<p>As expected, higher number of observations (higher <span class="math inline">\(n/p\)</span>) lead to increased statistical power
and hence to more significant results.</p>
<p>Proteins such as VIM, LGALS3, CALU, LMNA, CTTN, are more abundant in
melanoma cells compared to monocytes. On the other hand, proteins such
as LCP1, CORO1A, ARHGDIB, TMSB4X are more abundant in monocytes compared
to melanoma cells.</p>
<p>Finally, we offer functionality to report results at the protein
level.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpDifferentialAggregate</a></span><span class="op">(</span><span class="va">daRes</span>, fcol <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span>top <span class="op">=</span> <span class="fl">30</span>, textBy <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="co">## $SampleType_Melanoma_vs_Monocyte</span></span>
<span><span class="co">## Warning: ggrepel: 3 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="component-analysis">Component analysis<a class="anchor" aria-label="anchor" href="#component-analysis"></a>
</h3>
<p>Finally, we perform component analysis to link the modelled effects
to the cellular heterogeneity. We here run an APCA+ (extended
ANOVA-simultaneous principal component analysis) for the sample type
effect. In other words, we perform a PCA on the data that is capture by
the sample type variable along with the residuals (unmodelled data).</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">caRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, ncomp <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">"APCA"</span>, effect <span class="op">=</span> <span class="st">"SampleType"</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] "APCA"</span></span>
<span><span class="co">## [1] "SampleType"</span></span>
<span><span class="co">## List of length 2</span></span>
<span><span class="co">## names(2): bySample byFeature</span></span></code></pre></div>
<p>The results are contained in a list with 2 elements.
<code>bySample</code> contains the PC scores, that is the component
results in sample space. <code>byFeature</code> contains the
eigenvectors, that is the component results in feature space.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span></span>
<span><span class="co">## List of length 3</span></span>
<span><span class="co">## names(3): unmodelled residuals APCA_SampleType</span></span></code></pre></div>
<p>Each of the two elements contains components results for the data
before modelling (<code>unmodelled</code>), for the residuals or for the
APCA on the sample type variable (<code>APCA_SampleType</code>).</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## DataFrame with 1656 rows and 3 columns</span></span>
<span><span class="co">##                    PC1        PC2          cell</span></span>
<span><span class="co">##              &lt;numeric&gt;  &lt;numeric&gt;   &lt;character&gt;</span></span>
<span><span class="co">## eAL00219RI5   34.82634    25.4084 eAL00219RI...</span></span>
<span><span class="co">## eAL00219RI6   27.30071    28.3922 eAL00219RI...</span></span>
<span><span class="co">## eAL00219RI7   10.85533    29.7354 eAL00219RI...</span></span>
<span><span class="co">## eAL00219RI8    2.81552    28.9838 eAL00219RI...</span></span>
<span><span class="co">## eAL00219RI9   47.84411    29.4004 eAL00219RI...</span></span>
<span><span class="co">## ...                ...        ...           ...</span></span>
<span><span class="co">## wAL00286RI12 -32.95100  -3.641834 wAL00286RI...</span></span>
<span><span class="co">## wAL00286RI14 -36.19707  -0.283055 wAL00286RI...</span></span>
<span><span class="co">## wAL00286RI16   9.79076 -13.134253 wAL00286RI...</span></span>
<span><span class="co">## wAL00286RI17   2.63224 -16.199820 wAL00286RI...</span></span>
<span><span class="co">## wAL00286RI18  -1.84038 -18.643383 wAL00286RI...</span></span></code></pre></div>
<p>Each elements is a table with the computed componoents. Let’s explore
the component analysis results in cell space. Similarly to the previous
explorations, we annotate the results.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResCells</span> <span class="op">&lt;-</span> <span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">caResCells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span><span class="va">caResCells</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span></span></code></pre></div>
<p>We then generate the component plot, colouring by
<code>SampleType</code>. To assess the impact of batch effects, we shape
the points according to the plate batch (cf intro) as well.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>,</span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">SampleType</span>, shape <span class="op">=</span> <span class="va">lcbatch</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-35-1.png" width="864" style="display: block; margin: auto;"></p>
<p>While the data before modelling is mainly driven by batch effects,
the APCA clearly separates the two cell populations. Interestingly, the
PCA on the residuals suggests that there is a small subpopulation that
we did not model. We will explore this later during downstream
analysis.</p>
<p>We use the same approach to explore the component results in feature
space.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResPeps</span> <span class="op">&lt;-</span> <span class="va">caRes</span><span class="op">$</span><span class="va">byFeature</span></span>
<span><span class="va">caResPeps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">caResPeps</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We plot the compenents in peptide-space.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plCApeps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResPeps</span>, pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.8</span>, alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plCApeps</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-37-1.png" width="864" style="display: block; margin: auto;"></p>
<p>We can also combine the exploration of the components in cell and
peptide space. This is possible thanks to biplots.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biplots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentBiplot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>, <span class="va">caResPeps</span>, </span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">SampleType</span>, shape <span class="op">=</span> <span class="va">lcbatch</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    labelParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span>, max.overlaps <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>    textBy <span class="op">=</span> <span class="st">"gene"</span>, top <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">biplots</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span>
<span><span class="co">## Warning: Removed 4 rows containing missing values</span></span>
<span><span class="co">## (`geom_label_repel()`).</span></span>
<span><span class="co">## Warning: ggrepel: 16 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-38-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Finally, we offer functionality to aggregate the results at the
protein level instead of the peptide level.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResProts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentAggregate</a></span><span class="op">(</span><span class="va">caResPeps</span>, fcol <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="co">## Components may no longer be orthogonal after aggregation.</span></span>
<span><span class="va">biplots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentBiplot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>, <span class="va">caResProts</span>, </span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">SampleType</span>, shape <span class="op">=</span> <span class="va">lcbatch</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    labelParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span>, max.overlaps <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>    textBy <span class="op">=</span> <span class="st">"gene"</span>, top <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">biplots</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-39-1.png" width="864" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="interactive-visualisation-with-isee">Interactive visualisation with iSEE<a class="anchor" aria-label="anchor" href="#interactive-visualisation-with-isee"></a>
</h2>
<p>You can manually explore the data through an interactive interface
thanks to using the iSEE:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/iSEE/iSEE" class="external-link">"iSEE"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h2>
<p>The final step in this analyses in the identification of the small
melanoma subpopulation. To achieve this, we perform an APCA+ with more
components to capture most of the variability. Then, we cluster the
cells to highlight the subpopulations. We then retrieve the batch
corrected data. Finally, we run a new model on the melanoma data to find
marker proteins that are strongly differentially abundant between the
two cell subpopulations.</p>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>Before applying the cluster algorithm, we compute the 20 first APCA+
components on the sample type. We limit the PCA algorithm to 50
iterations to avoid an overly long run and will not compute the PCA for
the unmodelled data and the residuals because we won’t use it during the
downstream analysis.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">apcaSampleType</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-ComponentAnalysis.html" class="external-link">scpComponentAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, ncomp <span class="op">=</span> <span class="fl">20</span>, method <span class="op">=</span> <span class="st">"APCA"</span>, effect <span class="op">=</span> <span class="st">"SampleType"</span>, </span>
<span>    residuals <span class="op">=</span> <span class="cn">FALSE</span>, unmodelled <span class="op">=</span> <span class="cn">FALSE</span>, maxiter <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">bySample</span><span class="op">$</span><span class="va">APCA_SampleType</span></span>
<span><span class="co">## [1] "APCA"</span></span>
<span><span class="co">## [1] "SampleType"</span></span></code></pre></div>
<p>The resulting components are stored in the
<code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcCols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^PC"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">apcaSampleType</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"APCA"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">apcaSampleType</span><span class="op">[</span>, <span class="va">pcCols</span><span class="op">]</span></span></code></pre></div>
<p>Next, we visualize these 20 components by further decreasing the
dimensions using t-SNE.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"APCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-42-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Finally, we cluster cells based on a shared nearest neighbour
embedding followed by community detection using the Louvain
algorithm.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">algo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">30</span>, type <span class="op">=</span> <span class="st">"rank"</span>, cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">Cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"APCA"</span><span class="op">)</span>, BLUSPARAM <span class="op">=</span> <span class="va">algo</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Cluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-43-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The clustering retrieved 3 cell population, more or less splitting
the melanoma cells in 2 subpopulation. Note that <span class="math inline">\(k\)</span> was arbitrarily set to 30. A thorough
analysis would optimize K based on an objective criterion, such as
maximizing the average silhouette width.</p>
<p>Below is an overview of the unsupervised clusters against the known
annotations provided by the authors after thorough investigation.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span></span>
<span>    Cluster <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">Cluster</span>, </span>
<span>    Population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">SampleType</span>, <span class="va">sce</span><span class="op">$</span><span class="va">MelanomaSubCluster</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##        Population</span></span>
<span><span class="co">## Cluster Melanoma A Melanoma B Melanoma NA Monocyte NA</span></span>
<span><span class="co">##       1          2          0           0         769</span></span>
<span><span class="co">##       2        691          4          70          55</span></span>
<span><span class="co">##       3          1         53          11           0</span></span></code></pre></div>
<p>We annotate the the unsupervised clusters using the available
annotations.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">Cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">recode</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">Cluster</span>, <span class="st">"1"</span> <span class="op">=</span> <span class="st">"Monocyte"</span>, <span class="st">"2"</span> <span class="op">=</span> <span class="st">"Melanoma A"</span>, <span class="st">"3"</span> <span class="op">=</span> <span class="st">"Melanoma B"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h3>
<p>Based on the model fitted in the previous section, we generate
batch-corrected data, that is data with only the effect of cell type and
the residual data. We also remove the intercept.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">scebr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DataCorrection.html" class="external-link">scpRemoveBatchEffect</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, effects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Set"</span>, <span class="st">"Channel"</span>, <span class="st">"MedianIntensity"</span><span class="op">)</span>,</span>
<span>    intercept <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 6854 1656 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): ''</span></span>
<span><span class="co">## rownames(6854): GTLVQTK NSNPALNDNLEK ... TETQEKNPLPSK TLTTVQGIADDYDK</span></span>
<span><span class="co">## rowData names(6): Sequence Reverse ... Leading.razor.protein.symbol</span></span>
<span><span class="co">##   gene</span></span>
<span><span class="co">## colnames(1656): eAL00219RI5 eAL00219RI6 ... wAL00286RI17 wAL00286RI18</span></span>
<span><span class="co">## colData names(13): Set Channel ... cell Cluster</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre></div>
<p>Note that the batch-corrected data still contain missing values.</p>
</div>
<div class="section level3">
<h3 id="marker-proteins">Marker proteins<a class="anchor" aria-label="anchor" href="#marker-proteins"></a>
</h3>
<p>To identify markers proteins, we run a differential abundance
analysis as presented in the previous section. We create a new model,
focusing only on melanoma data as we don’t want to compare
subpopulations with monocytes.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceMel</span> <span class="op">&lt;-</span> <span class="va">scebr</span><span class="op">[</span>, <span class="va">scebr</span><span class="op">$</span><span class="va">SampleType</span> <span class="op">==</span> <span class="st">"Melanoma"</span><span class="op">]</span></span>
<span><span class="va">sceMel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-Workflow.html" class="external-link">scpModelWorkflow</a></span><span class="op">(</span><span class="va">sceMel</span>, formula <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">Cluster</span><span class="op">)</span></span></code></pre></div>
<p>We then apply the same procedure as described above.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpDifferentialAnalysis</a></span><span class="op">(</span><span class="va">sceMel</span>, contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Cluster"</span>, <span class="st">"Melanoma A"</span>, <span class="st">"Melanoma B"</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/scpAnnotateResults.html" class="external-link">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sceMel</span><span class="op">)</span>, </span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/ScpModel-DifferentialAnalysis.html" class="external-link">scpVolcanoPlot</a></span><span class="op">(</span><span class="va">daRes</span>, top <span class="op">=</span> <span class="fl">100</span>, textBy <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="co">## $Cluster_Melanoma.A_vs_Melanoma.B</span></span>
<span><span class="co">## Warning: Removed 234 rows containing missing values (`geom_point()`).</span></span>
<span><span class="co">## Warning: Removed 2 rows containing missing values (`geom_text_repel()`).</span></span>
<span><span class="co">## Warning: ggrepel: 56 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre></div>
<p><img src="scplainer_leduc2022_files/figure-html/unnamed-chunk-46-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Interestingly, subpopulation be is characterized by higher abundances
of MATR3 and SFPQ (involved in response to DNA double-strand break),
P4HB and PDIA3 (involved in disulfide bond rearrangement). On the other
hand, melanoma subpopulation A shows more elongation factors (EEF1B3,
EEF2, ), proteins associated with tubulin (TBCA, TUBB4B, TUBA1C, …) or
proliferation control proteins (PA2G4, ENO1). This is in line with
previous findings from the authors.</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code>R Under development (unstable) (2023-07-27 r84768)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 23.04

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.11.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.11.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Prague
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.31.0               scuttle_1.13.0             
 [3] SingleCellExperiment_1.25.0 bluster_1.13.0             
 [5] dplyr_1.1.3                 EnsDb.Hsapiens.v86_2.99.0  
 [7] ensembldb_2.27.0            AnnotationFilter_1.27.0    
 [9] GenomicFeatures_1.55.1      AnnotationDbi_1.65.2       
[11] patchwork_1.1.3             ggplot2_3.4.4              
[13] scpdata_1.9.2               ExperimentHub_2.11.0       
[15] AnnotationHub_3.11.0        BiocFileCache_2.11.1       
[17] dbplyr_2.4.0                scp_1.11.3                 
[19] QFeatures_1.13.0            MultiAssayExperiment_1.29.0
[21] SummarizedExperiment_1.33.0 Biobase_2.63.0             
[23] GenomicRanges_1.55.1        GenomeInfoDb_1.39.0        
[25] IRanges_2.37.0              S4Vectors_0.41.1           
[27] BiocGenerics_0.49.1         MatrixGenerics_1.15.0      
[29] matrixStats_1.1.0           BiocStyle_2.31.0           

loaded via a namespace (and not attached):
  [1] later_1.3.1                   BiocIO_1.13.0                
  [3] bitops_1.0-7                  filelock_1.0.2               
  [5] tibble_3.2.1                  lpsymphony_1.31.0            
  [7] XML_3.99-0.15                 lifecycle_1.0.4              
  [9] rprojroot_2.0.4               lattice_0.22-5               
 [11] MASS_7.3-60.1                 magrittr_2.0.3               
 [13] sass_0.4.7                    rmarkdown_2.25               
 [15] jquerylib_0.1.4               yaml_2.3.7                   
 [17] metapod_1.11.0                httpuv_1.6.12                
 [19] cowplot_1.1.1                 MsCoreUtils_1.15.1           
 [21] DBI_1.1.3                     RColorBrewer_1.1-3           
 [23] abind_1.4-5                   zlibbioc_1.49.0              
 [25] Rtsne_0.16                    purrr_1.0.2                  
 [27] RCurl_1.98-1.13               rappdirs_0.3.3               
 [29] GenomeInfoDbData_1.2.11       ggrepel_0.9.4                
 [31] irlba_2.3.5.1                 pkgdown_2.0.7                
 [33] DelayedMatrixStats_1.25.0     codetools_0.2-19             
 [35] DelayedArray_0.29.0           xml2_1.3.5                   
 [37] tidyselect_1.2.0              farver_2.1.1                 
 [39] ScaledMatrix_1.11.0           viridis_0.6.4                
 [41] nipals_0.8                    GenomicAlignments_1.39.0     
 [43] jsonlite_1.8.7                BiocNeighbors_1.21.0         
 [45] ellipsis_0.3.2                systemfonts_1.0.5            
 [47] tools_4.4.0                   progress_1.2.2               
 [49] ragg_1.2.6                    Rcpp_1.0.11                  
 [51] glue_1.6.2                    BiocBaseUtils_1.5.0          
 [53] gridExtra_2.3                 SparseArray_1.3.0            
 [55] xfun_0.41                     withr_2.5.2                  
 [57] BiocManager_1.30.22           fastmap_1.1.1                
 [59] fansi_1.0.5                   digest_0.6.33                
 [61] rsvd_1.0.5                    R6_2.5.1                     
 [63] mime_0.12                     textshaping_0.3.7            
 [65] colorspace_2.1-0              biomaRt_2.59.0               
 [67] RSQLite_2.3.3                 utf8_1.2.4                   
 [69] tidyr_1.3.0                   generics_0.1.3               
 [71] rtracklayer_1.63.0            prettyunits_1.2.0            
 [73] httr_1.4.7                    S4Arrays_1.3.0               
 [75] pkgconfig_2.0.3               gtable_0.3.4                 
 [77] blob_1.2.4                    XVector_0.43.0               
 [79] htmltools_0.5.7               bookdown_0.36                
 [81] ProtGenerics_1.35.0           clue_0.3-65                  
 [83] scales_1.2.1                  png_0.1-8                    
 [85] knitr_1.45                    rjson_0.2.21                 
 [87] curl_5.1.0                    cachem_1.0.8                 
 [89] stringr_1.5.0                 BiocVersion_3.19.1           
 [91] parallel_4.4.0                vipor_0.4.5                  
 [93] restfulr_0.0.15               desc_1.4.2                   
 [95] pillar_1.9.0                  grid_4.4.0                   
 [97] vctrs_0.6.4                   slam_0.1-50                  
 [99] promises_1.2.1                BiocSingular_1.19.0          
[101] IHW_1.31.0                    beachmat_2.19.0              
[103] xtable_1.8-4                  cluster_2.1.4                
[105] beeswarm_0.4.0                evaluate_0.23                
[107] cli_3.6.1                     compiler_4.4.0               
[109] Rsamtools_2.19.2              rlang_1.1.2                  
[111] crayon_1.5.2                  labeling_0.4.3               
[113] fdrtool_1.2.17                fs_1.6.3                     
[115] ggbeeswarm_0.7.2              stringi_1.7.12               
[117] viridisLite_0.4.2             BiocParallel_1.37.0          
[119] munsell_0.5.0                 Biostrings_2.71.1            
[121] lazyeval_0.2.2                Matrix_1.6-1.1               
[123] hms_1.1.3                     sparseMatrixStats_1.15.0     
[125] bit64_4.0.5                   KEGGREST_1.43.0              
[127] shiny_1.7.5.1                 highr_0.10                   
[129] interactiveDisplayBase_1.41.0 igraph_1.5.1                 
[131] memoise_2.0.1                 bslib_0.5.1                  
[133] bit_4.0.5                    </code></pre>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">citation</span>(<span class="st">"scp"</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>To cite the scp package <span class="cf">in</span> publications use<span class="sc">:</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>  Vanderaa, Christophe, and Laurent Gatto. <span class="fl">2023.</span>  Revisiting the Thorny</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  Issue of Missing Values <span class="cf">in</span> Single<span class="sc">-</span>Cell Proteomics. Journal of</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>  Proteome Research <span class="dv">22</span> (<span class="dv">9</span>)<span class="sc">:</span> 2775–<span class="fl">84.</span></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>  Vanderaa Christophe and Laurent Gatto. The current state of</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>  single<span class="sc">-</span>cell proteomics data analysis. Current Protocols <span class="dv">3</span> (<span class="dv">1</span>)<span class="sc">:</span> e658.;</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>  doi<span class="sc">:</span> https<span class="sc">:</span><span class="er">//</span>doi.org<span class="sc">/</span><span class="fl">10.1002</span><span class="sc">/</span><span class="fu">cpz1.658</span> (<span class="dv">2023</span>).</span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>  Vanderaa Christophe and Laurent Gatto. Replication of Single<span class="sc">-</span>Cell</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>  Proteomics Data Reveals Important Computational Challenges. Expert</span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>  Review of Proteomics, 1–<span class="dv">9</span> (<span class="dv">2021</span>).</span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-16"><a href="#cb57-16" aria-hidden="true" tabindex="-1"></a>To see these entries <span class="cf">in</span> BibTeX format, use <span class="st">'print(&lt;citation&gt;,</span></span>
<span id="cb57-17"><a href="#cb57-17" aria-hidden="true" tabindex="-1"></a><span class="st">bibtex=TRUE)'</span>, <span class="st">'toBibtex(.)'</span>, or set</span>
<span id="cb57-18"><a href="#cb57-18" aria-hidden="true" tabindex="-1"></a><span class="st">'options(citation.bibtex.max=999)'</span>.</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
license</a> license.</p>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:christophe.vanderaa@uclouvain.be" class="email">christophe.vanderaa@uclouvain.be</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
