<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife) • SCP.replication</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)">
<meta property="og:description" content="SCP.replication">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCP.replication</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCoPE2.html">Reproduction of the SCoPE2 analysis (Specht et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/liang2020.html">Exploring the autoPOTS data (Liang et al. 2020)</a>
    </li>
    <li>
      <a href="../articles/schoof2021.html">Reproduction of the AML model analysis (Schoof et al. 2021)</a>
    </li>
    <li>
      <a href="../articles/zhu2019EL.html">Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="zhu2019EL_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Reproduction of the hair-cell development analysis (Zhu et al. 2019, eLife)</h1>
                        <h4 class="author">Christophe Vanderaa<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, Computational Biology, UCLouvain</h4>
                        <h4 class="author">Laurent Gatto, Computational Biology, UCLouvain</h4>
            
            <h4 class="date">5 Oktober 2021</h4>
      
      
      <div class="hidden name"><code>zhu2019EL.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This vignette presents a robust and standardized workflow to reproduce the data analysis published in Zhu et al. (2019, eLife), the first article to have published a trajectory analysis on mass spectrometry (MS)-based single-cell proteomics (SCP) data. The replication of the analysis is performed using the scp package. This package is based on well-defined Bioconductor classes that provide powerful tools for single-cell RNA sequencing and for MS-based proteomics. We demonstrate that the scp package can be applied to label-free SCP data and can be used to highlight some issues in the data.
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The hair-cell development study (<span class="citation">Zhu et al. (2019)</span>) is the first study to apply developmental trajectory analysis to mass spectrometry (MS)-based single-cell proteomics (SCP) data. It is also the first paper to apply SCP to a biological research question as a tool rather than as a proof of concept. This makes reproducing the analysis very appealing as it allows to demonstrate our newly developed software is suited for answering biologically meaningful questions and provides a robust tool for standardized and reproducible research.</p>
<p>The research question here is to better characterize the differentiation of supporting cells into hair cells in the utricle of E15 chick embryos. MS-SCP was conducted on samples containing between 1 and 20 cells. The 20-cell samples were used to perform differential expression analysis between hair cells from supporting cells. The single-cell samples were used to conduct developmental trajectory analysis. This later analysis will be the focus of this vignette.</p>
<p>No code is provided to reproduce the original results. The data was analysed to some extend in Excel and differential expression analysis and trajectory analysis were perform in R using the <code>limma</code> and <code>CellTrails</code> packages, respectively. Our attempt of reproduction is solely based on the methods section.</p>
<p>Before starting, we load the replication package to make use of some helper functions. Those functions are only meant for this reproduction vignette and are not designed for general use.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"SCP.replication"</span><span class="op">)</span></code></pre></div>
<div id="scp-workflow-and-data-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#scp-workflow-and-data-structure" class="anchor"></a><code>scp</code> workflow and data structure</h3>
<p>The data processing workflow we carry out in this vignette is inferred from the methods section of the paper. Since no reproducible data analysis script was provided, we are not able to accurately evaluate the accuracy of the replication of the results and will rely on the information in the article to assess whether we can come up with the same conclusions.</p>
<p>To perform SCP data analysis, we have developed a new data framework that combines two existing Bioconductor classes. The <code>SingleCellExperiment</code> class provides an interface to many cutting edge methods for single-cell analysis and the <code>QFeatures</code> class facilitates manipulation and processing of MS-based quantitative data. The <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/scp/inst/doc/scp.html"><code>scp</code> vignette</a> provides detailed information about the data structure. The <code>scp</code> package extends the functionality of <code>QFeatures</code> for single-cell application.</p>
<p>The required packages for running this workflow are listed below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Core packages of this workflow</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/QFeatures">QFeatures</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scpdata</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp">scp</a></span><span class="op">)</span>
<span class="co">## Utility packages for data manipulation and visualization</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="co">## Other packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">CellTrails</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">sva</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">splines</span><span class="op">)</span></code></pre></div>
</div>
<div id="scpdata-and-the-hair-cell-development-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#scpdata-and-the-hair-cell-development-dataset" class="anchor"></a><code>scpdata</code> and the hair-cell development dataset</h3>
<p>We also implemented a data package called <code>scpdata</code>. It distributes published MS-SCP datasets, such as the dataset that will be analysed in this vignette. The data were downloaded from the data source provided in the publication and formatted to a <code>QFeatures</code> object so that it is compatible with our software. The underlying data storage is based on the <code>ExperimentHub</code> package that provides a cloud-based storage infrastructure.</p>
<p>The dataset was retrieved from the PRIDE repository (accession ID: PXD014256). The annotation file that is provided in this repository contains mismatched information. We therefore received a corrected version from the authors. The dataset contains three types of quantitative data: quantified peptide to spectrum match (PSM) data, peptide data and protein data. They were generated after running MaxQuant on the raw MS files.</p>
<p>The formatted data can be retrieved from the <code>scpdata</code> package using the <code><a href="https://rdrr.io/pkg/scpdata/man/zhu2019EL.html">zhu2019EL()</a></code> function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/zhu2019EL.html">zhu2019EL</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The data contain 62 different <code>SingleCellExperiment</code> objects that we refer to as <strong>assays</strong>. Each assay contains expression data along with feature metadata. Each row in an assay represents a <strong>feature</strong> that can either be a PSM, a peptide or a protein depending on the assay. Each column in an assay represents a <strong>sample</strong>. Below, we show the overview of the <code>scp</code> dataset.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 62 assays:</span>
<span class="co">##  [1] 1H1a: SingleCellExperiment with 152 rows and 1 columns </span>
<span class="co">##  [2] 1H1b: SingleCellExperiment with 267 rows and 1 columns </span>
<span class="co">##  [3] 1H1c: SingleCellExperiment with 128 rows and 1 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [60] 2N0c: SingleCellExperiment with 61 rows and 1 columns </span>
<span class="co">##  [61] peptides: SingleCellExperiment with 3444 rows and 60 columns </span>
<span class="co">##  [62] proteins: SingleCellExperiment with 840 rows and 60 columns</span></code></pre></div>
<p>The 60 first are the peptide to spectrum match (PSM) for each run. PSMs were quantified and identified by MaxQuant (<span class="citation">Tyanova, Temu, and Cox (2016)</span>). The samples were acquired using a label-free protocol meaning that every run contains an individual sample and therefore every assay contains a single column. The before last assay contains the peptide quantifications and the last assay contains the protein quantifications, both also generated by MaxQuant.</p>
<p>The sample annotation can be found in the <code>colData</code> of the dataset. There three important annotation fields:</p>
<ul>
<li>
<code>Cells.per.well</code>: the samples contains either 0 (blank), 1 (single-cell), 3, 5, or 20 cells</li>
<li>
<code>FM1.43.signal</code>: single-cells from chicken utricles were first separated using FACS based on FM1-43 labelling, which was reported to label more strongly hair cells than supporting cells (<span class="citation">Zhu et al. (2019)</span>).</li>
<li>
<code>Experiment</code>: the samples were acquired in two separate experiments (batches). From the MS file names (see <code>Raw.file</code> in the <code>colData</code>), the two experiments seemed to be acquired with one year interval.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>
<span class="co">## DataFrame with 60 rows and 8 columns</span>
<span class="co">##       Sample.name FM1.43.signal Cells.per.well Biological.replicate</span>
<span class="co">##       &lt;character&gt;   &lt;character&gt;      &lt;numeric&gt;            &lt;numeric&gt;</span>
<span class="co">## 1H1a         1H1a          High              1                    1</span>
<span class="co">## 1H1b         1H1b          High              1                    2</span>
<span class="co">## 1H1c         1H1c          High              1                    3</span>
<span class="co">## 1H1d         1H1d          High              1                    4</span>
<span class="co">## 1H1e         1H1e          High              1                    5</span>
<span class="co">## ...           ...           ...            ...                  ...</span>
<span class="co">## 2L20b       2L20b           Low             20                    2</span>
<span class="co">## 2L20c       2L20c           Low             20                    3</span>
<span class="co">## 2N0a         2N0a           Low              0                    1</span>
<span class="co">## 2N0b         2N0b           Low              0                    2</span>
<span class="co">## 2N0c         2N0c           Low              0                    3</span>
<span class="co">##       Technical.replicate      Raw.file    QuantCol  Experiment</span>
<span class="co">##                 &lt;numeric&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 1H1a                    1 701155_Sin...   Intensity           1</span>
<span class="co">## 1H1b                    1 701157_Sin...   Intensity           1</span>
<span class="co">## 1H1c                    1 701166_Sin...   Intensity           1</span>
<span class="co">## 1H1d                    1 701176_Sin...   Intensity           1</span>
<span class="co">## 1H1e                    1 701190_Sin...   Intensity           1</span>
<span class="co">## ...                   ...           ...         ...         ...</span>
<span class="co">## 2L20b                   1 Single_Hai...   Intensity           2</span>
<span class="co">## 2L20c                   1 Single_Hai...   Intensity           2</span>
<span class="co">## 2N0a                    1 Single_Hai...   Intensity           2</span>
<span class="co">## 2N0b                    1 Single_Hai...   Intensity           2</span>
<span class="co">## 2N0c                    1 Single_Hai...   Intensity           2</span></code></pre></div>
<p>Throughout the vignette, we will <code>ggplot2</code> from the <code>tidyverse</code> project (<span class="citation">Wickham et al. (2019)</span>) to visualize the data. For example, we can show the distribution of the number of cells per sample according to the experiment and the FM1-43 signal.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">Cells.per.well</span><span class="op">)</span>,
        fill <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"count"</span>, position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span> <span class="va">Experiment</span>, labeller <span class="op">=</span> <span class="va">label_both</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-3-1.png" width="672"></p>
</div>
</div>
<div id="data-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#data-processing" class="anchor"></a>Data processing</h1>
<div id="add-ibaq-assay" class="section level2">
<h2 class="hasAnchor">
<a href="#add-ibaq-assay" class="anchor"></a>Add iBAQ assay</h2>
<p>The author ran MaxQuant with iBAQ computation enabled. Those values are currently stored in the <code>rowData</code> of the <code>proteins</code> assays. We will add an additional assay containing the iBAQ values. We can easily create a new assay of class <code>SingleCellExperiment</code> from a data table using the <code>readSingleCellExperiment</code> function. We first extract the <code>rowData</code> and we retrieve the columns that contains the iBAQ values (those column names start with <code>iBAQ</code>).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibaqData</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">[[</span><span class="st">"proteins"</span><span class="op">]</span><span class="op">]</span>
<span class="va">ibaqCols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"^iBAQ."</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ibaqData</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ibaqCols</span><span class="op">)</span>
<span class="co">## [1] "iBAQ.1H1a"  "iBAQ.1H1b"  "iBAQ.1H1c"  "iBAQ.1H1d"  "iBAQ.1H1e" </span>
<span class="co">## [6] "iBAQ.1H20a"</span></code></pre></div>
<p>Next, we can supply the data table and the iBAQ quantification columns to <code>readSingleCellExperiment</code>. We use the <code>Protein</code> information in the table to set the row names of the assay. Finally, we adapt to column names to match the column names in the <code>scp</code> data set.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibaqSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/readSingleCellExperiment.html">readSingleCellExperiment</a></span><span class="op">(</span>table <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">ibaqData</span><span class="op">)</span>,
                                    ecol <span class="op">=</span> <span class="va">ibaqCols</span>,
                                    fnames <span class="op">=</span> <span class="st">"Protein"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ibaqSCE</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"iBAQ."</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ibaqSCE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We add the iBAQ assay in the <code>scp</code> data set using the <code>addAssay</code> function. We also preserve the one-to-one relationship between protein quantifications and the iBAQ quantifications using the <code>addAssayLinkOneToOne</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">addAssay</a></span><span class="op">(</span><span class="va">scp</span>, <span class="va">ibaqSCE</span>, name <span class="op">=</span> <span class="st">"proteins_iBAQ"</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html">addAssayLinkOneToOne</a></span><span class="op">(</span><span class="va">scp</span>, from <span class="op">=</span> <span class="st">"proteins"</span>, to <span class="op">=</span> <span class="st">"proteins_iBAQ"</span><span class="op">)</span></code></pre></div>
<p>In the article, the authors claim that:</p>
<blockquote>
<p>(…) the total iBAQ attributed to keratins (e.g., human skin contamination) was only 0.1% of the total (…)</p>
</blockquote>
<p>Let’s check this by first computing the total iBAQ across all samples. We can access the quantification data using the <code>assay</code> method.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">iBAQ</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_iBAQ"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="va">iBAQtotal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">iBAQ</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>To compute the total iBAQ for contaminating proteins, we need to select only the contaminating proteins. This is performed using the <code>filterFeatures</code> function which will keep features that pass a filter applied on the <code>rowData</code> variables. In this case, the <code>Potential.contaminant</code> field generated by MaxQuant allows to easily filter out contaminating proteins. Note that we ignore <code>P07477</code> corresponding to trypsin because the author’s claim is about keratin contamination.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scpfilt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>, <span class="op">~</span> <span class="va">Potential.contaminant</span> <span class="op">==</span> <span class="st">"+"</span> <span class="op">&amp;</span>
                              <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"P07477"</span>, <span class="va">Protein</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We can now compute the total iBAQ for the contaminating proteins.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">iBAQ</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">scpfilt</span><span class="op">[[</span><span class="st">"proteins_iBAQ"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="va">iBAQContam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">iBAQ</span>,  na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"Proportion of contaminating iBAQ signal:"</span>,
    <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">iBAQContam</span> <span class="op">/</span> <span class="va">iBAQtotal</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"%\n"</span><span class="op">)</span>
<span class="co">## Proportion of contaminating iBAQ signal: 29.23 %</span></code></pre></div>
<p>Surprisingly, the proportion of contaminating iBAQ signal that we compute here is about 30 % which is much higher than the reported 0.1 %.</p>
</div>
<div id="remove-decoys-and-contaminating-proteins" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-decoys-and-contaminating-proteins" class="anchor"></a>Remove decoys and contaminating proteins</h2>
<p>As shown above, a significant proportion of the quantification is attributed to contaminating proteins. We will therefore remove those along with decoy peptides. This step is not mentioned in the article, but this is commonly applied and will avoid artefacts.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,
                      <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span> <span class="op">&amp;</span>
                          <span class="va">Potential.contaminant</span> <span class="op">!=</span> <span class="st">"+"</span><span class="op">)</span></code></pre></div>
</div>
<div id="format-missing-data" class="section level2">
<h2 class="hasAnchor">
<a href="#format-missing-data" class="anchor"></a>Format missing data</h2>
<p>In MS-SCP, zeros can be biological zeros or technical zeros and distinguishing the two types is a difficult task, they are therefore better considered as missing to avoid artefacts in downstream analyses. The <code>zeroIsNA</code> function automatically detects zero values and replace them with <code>NA</code>s. Those two functions are provided by the <code>QFeatures</code> package.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">zeroIsNA</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="technical-performance" class="section level1">
<h1 class="hasAnchor">
<a href="#technical-performance" class="anchor"></a>Technical performance</h1>
<p>In this section, we reproduce some of the plots in figure 1. This figure depicts different technical aspects of the data such as number of detected unique peptides and protein groups, or the distribution of iBAQ values. We also propose additional plots to further assess the quality of the acquisition.</p>
<div class="figure" style="text-align: center">
<img src="figs/zhu2019EL-figure1.jpg" alt="Figures 1 from Zhu et al. 2019." width="100%"><p class="caption">
Figures 1 from Zhu et al. 2019.
</p>
</div>
<div id="number-of-proteins-vs-number-of-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#number-of-proteins-vs-number-of-cells" class="anchor"></a>Number of proteins vs number of cells</h2>
<p>Let’s assess the number of protein groups identified in each sample with respect to the number of cells in the sample. The authors were also interested whether the cell type would influence the number of identified peptides. Only the data from experiment 1 is used to generate the plot (see legend of Figure 1 in article).</p>
<p>Counting the number of proteins per sample is straightforward thanks to the <code>countUniqueFeatures</code> function from <code>QFeatures</code>. We select the protein assay, count the number of unique protein groups (the function will automatically ignore missing data) and store the counts in a new variable in the <code>colData</code> (we call it here <code>Protein_counts</code>).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/countUniqueFeatures.html">countUniqueFeatures</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="st">"proteins"</span>,
                           colDataName <span class="op">=</span> <span class="st">"Protein_counts"</span><span class="op">)</span></code></pre></div>
<p>We can plot the counts from the colData, but we first filter the data coming from experiment 1 and average the counts to replicate the figure in the article. This is performed using functionality from the <code>dplyr</code> package.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="co">## Take only Experiment 1</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Experiment</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Compute mean and standard error per cell number</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">FM1.43.signal</span>, <span class="va">Cells.per.well</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">Protein_counts</span><span class="op">)</span>,
              stderr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">Protein_counts</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">Protein_counts</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">df</span></code></pre></div>
<p>We then plot the data using <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Cells.per.well</span>,
        y <span class="op">=</span> <span class="va">mean</span>,
        group <span class="op">=</span> <span class="va">FM1.43.signal</span>,
        color <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_errorbar</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">mean</span> <span class="op">-</span> <span class="va">stderr</span>,
                      ymax <span class="op">=</span> <span class="va">mean</span> <span class="op">+</span> <span class="va">stderr</span><span class="op">)</span>,
                  color <span class="op">=</span> <span class="st">"grey"</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">'lm'</span>,
                formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/poly.html">poly</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span><span class="op">)</span>,
                color <span class="op">=</span> <span class="st">"grey"</span>, lwd <span class="op">=</span><span class="fl">0.5</span>, se <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect to match the figure design</span>
    <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ylab</span><span class="op">(</span><span class="st">"Proteins or groups"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"# cells"</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/Protein_count_plot-1.png" width="672"></p>
<p>This plot replicates the figure 1G in the article. More features are found in the FM1-43 high cells compared to the FM1-43 low cells.</p>
</div>
<div id="number-of-peptides-vs-number-of-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#number-of-peptides-vs-number-of-cells" class="anchor"></a>Number of peptides vs number of cells</h2>
<p>The authors also showed the number of unique peptides identified in each sample with respect to the number of cells in the sample. They further split this information based on the FM1-43 signal level and on the type of identification (direct MS/MS matching or match between run) when applicable. Only the data from experiment 1 is used to generate the plot (see legend of Figure 1 in article).</p>
<p>Counting the peptides could be again performed using <code>countUniqueFeatures</code>, but since we want to count with respect to the type of identification method, we will need some more sophisticated code. First, we collect the PSM meta-information contained in the <code>rowData</code> of the assays. This is where the type of identification method (<code>Type</code>) is stored.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">rbindRowData</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">60</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample.name"</span>, <span class="st">"Type"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>Next, we combine the rowData with the sample annotation contained in the <code>colData</code>. This is performed using the <code>left_join</code> function from <code>dplyr</code>. The sample name is used as common variable to control the data joining. Joining the two tables allow to include information about the FM1-43 signal and the number of cells per sample.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">rd</span><span class="op">)</span>,
                <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span>,
                by <span class="op">=</span> <span class="st">"Sample.name"</span><span class="op">)</span></code></pre></div>
<p>We next constrain the data to only samples from experiment 1. We also encode the type of PSM identification method to be either direct matching (MS/MS) or match between run (MBR).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">Experiment</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">df</span>, Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">Type</span> <span class="op">==</span> <span class="st">"MULTI-MATCH"</span>, <span class="st">"MBR"</span>, <span class="st">"MS/MS"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We then compute the number of feature per sample and per identification type. The counts are also averaged to plot the mean counts and associated standard error.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">%&gt;%</span>
    <span class="co">## Count the number of peptides in each sample per identification type</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">FM1.43.signal</span>, <span class="va">Cells.per.well</span>, <span class="va">Type</span>, <span class="va">Sample.name</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Compute the average count</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">FM1.43.signal</span>, <span class="va">Cells.per.well</span>, <span class="va">Type</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span>,
              stderr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">df</span></code></pre></div>
<p>Finally, we plot the data using <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Cells.per.well</span>,
        y <span class="op">=</span> <span class="va">mean</span>,
        shape <span class="op">=</span> <span class="va">Type</span>,
        fill <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_errorbar</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">mean</span> <span class="op">-</span> <span class="va">stderr</span>,
                      ymax <span class="op">=</span> <span class="va">mean</span> <span class="op">+</span> <span class="va">stderr</span><span class="op">)</span>,
                  color <span class="op">=</span> <span class="st">"grey"</span>,
                  width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">stat_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">'lm'</span>,
                formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/poly.html">poly</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span><span class="op">)</span>,
                color <span class="op">=</span> <span class="st">"grey"</span>,
                lwd <span class="op">=</span><span class="fl">0.5</span>, se <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span> <span class="op">~</span> <span class="va">FM1.43.signal</span>,
                labeller <span class="op">=</span> <span class="fu">labeller</span><span class="op">(</span>.cols <span class="op">=</span> <span class="va">label_both</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect to match the figure design</span>
    <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"black"</span><span class="op">)</span>, guide <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_shape_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">22</span>, <span class="fl">21</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ylab</span><span class="op">(</span><span class="st">"Unique peptides"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"# cells"</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-8-1.png" width="576"></p>
<p>This graph replicates figure 1 D and E in the article. There is an almost linear trend between number of peptides and the number of cells per sample when considering direct MS/MS identification. This plot also shows that more features are found in the FM1-43 high cells compared to the FM1-43 low cells.</p>
</div>
<div id="total-ibaq-vs-number-of-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#total-ibaq-vs-number-of-cells" class="anchor"></a>Total iBAQ vs number of cells</h2>
<p>Figure 1F in the article shows the distribution of the total iBAQ with respect to the number of cells per samples. It is expected that samples with more cells exhibit more signal and this plot assess whether the quantification is linear with respect to the number of cells present in the sample.</p>
<p>Again, the data is taken from experiment 1. We first extract the quantitative data along with sample annotation. We use the <code>longFormat</code> function to collect the data as a long table format that is particularly useful to work with <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">longdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[</span>, , <span class="st">"proteins_iBAQ"</span><span class="op">]</span>,
                      colvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample.name"</span>, <span class="st">"Experiment"</span>,
                                      <span class="st">"Cells.per.well"</span>, <span class="st">"FM1.43.signal"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>As for the number of proteins and peptides, we compute the average and standard error of the iBAQ values over the number of cells per sample and the FM1-43 signal. We also again constrain the data only to data generated by experiment 1.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">longdat</span><span class="op">)</span>, <span class="va">Experiment</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">df</span> <span class="op">%&gt;%</span>
    <span class="co">## Compute the total iBAQ</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">Sample.name</span>, <span class="va">FM1.43.signal</span>, <span class="va">Cells.per.well</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">summarise</span><span class="op">(</span>total <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">value</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Compute mean and standard error per cell number</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">FM1.43.signal</span>, <span class="va">Cells.per.well</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span>,
              stderr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">df</span></code></pre></div>
<p>Finally, we plot the data.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Cells.per.well</span>,
        y <span class="op">=</span> <span class="va">mean</span>,
        group <span class="op">=</span> <span class="va">FM1.43.signal</span>,
        color <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_errorbar</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">mean</span> <span class="op">-</span> <span class="va">stderr</span>,
                      ymax <span class="op">=</span> <span class="va">mean</span> <span class="op">+</span> <span class="va">stderr</span><span class="op">)</span>,
                  color <span class="op">=</span> <span class="st">"grey"</span>,
                  width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">'lm'</span>,
                formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/poly.html">poly</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span><span class="op">)</span>,
                color <span class="op">=</span> <span class="st">"grey"</span>,
                lwd <span class="op">=</span><span class="fl">0.5</span>, se <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect</span>
    <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"black"</span><span class="op">)</span>, guide <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ylab</span><span class="op">(</span><span class="st">"total iBAQ"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"# cells"</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-11-1.png" width="288"></p>
<p>Again, the above figure accurately reproduces figure 1F from the article.</p>
</div>
<div id="total-ibaq-in-single-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#total-ibaq-in-single-cells" class="anchor"></a>Total iBAQ in single-cells</h2>
<p>Another quality check is to plot the iBAQ distribution for single-cells and blank samples. This distribution is shown for samples coming from experiment 2 that contains more single-cell samples. The authors showed that the total iBAQ is not a good criterion to distinguish blank and single-cell samples.</p>
<p>Using the same table generated by <code>longFormat</code>, we filter the data generated by experiment 2 and only containing blank or single-cells samples. FM1-43 signal coming from blanks is considered as noise, and we sume the total iBAQ for each sample.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">longdat</span><span class="op">)</span>,
       <span class="va">Experiment</span> <span class="op">==</span> <span class="fl">2</span> <span class="op">&amp;</span> <span class="va">Cells.per.well</span> <span class="op">&lt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## M1-43 signal from blanks is noise</span>
    <span class="fu">mutate</span><span class="op">(</span>FM1.43.signal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">Cells.per.well</span> <span class="op">==</span> <span class="fl">0</span>,
                                  <span class="st">"noise"</span>, <span class="va">FM1.43.signal</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Compute total iBAQ</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">Sample.name</span>, <span class="va">FM1.43.signal</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">summarise</span><span class="op">(</span>tiBAQ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">value</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">df</span></code></pre></div>
<p>We plot the distribution of the total iBAQ.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">tiBAQ</span>,
        fill <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">5E5</span>,
                   position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect</span>
    <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>, <span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_x_continuous</span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/label_scientific.html">scientific</a></span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"Total iBAQ"</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-13-1.png" width="480"></p>
<p>Surprisingly, this plot does not reporduce figure 1H. We do not observe the same values for the total iBAQ and the number of samples is different (we have 21, while the article shows 23 samples). The value ranges are also different: <span class="math inline">\((0, 2.10^6)\)</span> against <span class="math inline">\((0, 4.10^6)\)</span>.</p>
</div>
<div id="number-of-proteins-in-single-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#number-of-proteins-in-single-cells" class="anchor"></a>Number of proteins in single-cells</h2>
<p>While the total iBAQ cannot distinguish blanks from single cells, the distribution of the number of proteins identified can separate single cells from blank samples. Recall that we already stored the protein counts in the <code>colData</code>. We can filter those counts to contain data only for experiment 2 and blank and single-cell samples.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="co">## Keep only single cells or blanks coming from experiment 2</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Experiment</span> <span class="op">==</span> <span class="fl">2</span> <span class="op">&amp;</span>
               <span class="va">Cells.per.well</span> <span class="op">&lt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>FM1.43.signal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">Cells.per.well</span> <span class="op">==</span> <span class="fl">0</span>,
                                  <span class="st">"noise"</span>, <span class="va">FM1.43.signal</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Plot</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Protein_counts</span>,
        fill <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">20</span>,
                   position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect</span>
    <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>, <span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"Proteins or groups"</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-14-1.png" width="480"></p>
<p>This plot should replicate figure 1I in the article. Surprisingly, we observe the same values for single-cell samples (except there are 2 missing samples), but the blank samples are are not all contained in the 0-20 protein bin! Furthermore, binning the data is misleading. A better plot is shown below. More than half of the single-cell samples have <strong>less or equal</strong> identified proteins compared to blank samples.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="co">## Keep only single cells or blanks coming from experiment 2</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Experiment</span> <span class="op">==</span> <span class="fl">2</span> <span class="op">&amp;</span>
               <span class="va">Cells.per.well</span> <span class="op">&lt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>FM1.43.signal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">Cells.per.well</span> <span class="op">==</span> <span class="fl">0</span>,
                                  <span class="st">"noise"</span>, <span class="va">FM1.43.signal</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Plot</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">Protein_counts</span>,
        x <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_jitter</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>col <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span>, height <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect</span>
    <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>, <span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-15-1.png" width="384"></p>
</div>
<div id="intensity-distribution-within-single-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#intensity-distribution-within-single-cells" class="anchor"></a>Intensity distribution within single-cells</h2>
<p>An aspect of the data that has been overlooked in the original article is the distribution of the PSM intensities in individual cells. Since the number of single-cells is rather low, we can easily show all distributions on a single figure. Furthermore, we can also include the type of identification as additional information. This is stored in the PSM assays.</p>
<p>We collect the data using the same functionality as presented before.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[</span>, , <span class="fl">1</span><span class="op">:</span><span class="fl">60</span><span class="op">]</span>,
           colvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Cells.per.well"</span>, <span class="st">"FM1.43.signal"</span><span class="op">)</span>,
           rowvars <span class="op">=</span> <span class="st">"Type"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="co">## Keep only single cells and blanks</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Cells.per.well</span> <span class="op">&lt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Clean data</span>
    <span class="fu">mutate</span><span class="op">(</span>Cells.per.well <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">Cells.per.well</span><span class="op">)</span>,
           FM1.43.signal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">Cells.per.well</span> <span class="op">==</span> <span class="fl">0</span>,
                                  <span class="st">"noise"</span>, <span class="va">FM1.43.signal</span><span class="op">)</span>,
           Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">Type</span> <span class="op">==</span> <span class="st">"MULTI-MATCH"</span>, <span class="st">"MBR"</span>, <span class="st">"MS/MS"</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">df</span></code></pre></div>
<p>We plot the PSM intensities for each sample separately. The distribution within each sample is best summarized using boxplots. We colour the PSMs based on the peptide identification method.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">assay</span>,
        y <span class="op">=</span> <span class="va">value</span>,
        colour <span class="op">=</span> <span class="va">Type</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_boxplot</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"grey40"</span>, outlier.colour <span class="op">=</span> <span class="st">"transparent"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_jitter</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span> <span class="va">FM1.43.signal</span>, scales <span class="op">=</span> <span class="st">"free_x"</span>,
               space <span class="op">=</span> <span class="st">"free_x"</span>, labeller <span class="op">=</span> <span class="va">label_both</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect</span>
    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey40"</span>, <span class="st">"orange2"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span>,
          legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ylab</span><span class="op">(</span><span class="st">"PSM intensity"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"Sample"</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>Two observations can be drawn from this plot:</p>
<ol style="list-style-type: decimal">
<li>We can clearly notice that MBR allows to increase the sensitivity of the identification as low intensity PSMs are mostly identified by MBR. This is expected as low intensity PSMs have an associated spectrum that is harder to identify. However, it is surprising to see that the PSMs with highest intensities are also identified using MBR, those PSMs could be artefacts from the matching.</li>
<li>Many samples (n = 11) have low number of PSMs. Those samples cannot be differentiated from blanks (three last samples) and this indicates that the run failed for those samples. We would advice to remove those samples from the analysis.</li>
</ol>
</div>
</div>
<div id="developmental-trajectory-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#developmental-trajectory-analysis" class="anchor"></a>Developmental trajectory analysis</h1>
<p>This part of the analysis is truly single-cell analysis. The developmental trajectory analysis infers the developmental states and paths that the acquired are following or have achieved and hence elucidating the mechanisms of differentiation of supporting cells towards specialized hair cells in the utricle.</p>
<div id="the-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#the-workflow" class="anchor"></a>The workflow</h2>
<p>The method section for this analysis mentions:</p>
<blockquote>
<p>For single-cell analysis, we filtered all identifications that were robustly detected in at least five cells, leaving a total of 75 proteins (or protein groups). The iBAQ values for proteins in the 30 single cells were normalized to the median of the cells’ mean expression and log2-transformed; non detected values were kept as zeros to avoid imputation artifacts. The empirical Bayes framework available in the sva R package was used to remove the batch covariate, while accounting for the FM1-43 levels as covariate of interest. The same framework has been used previously to correct for batch effects in protein mass spectrometry data. The resulting values are referred to as log2-normalized iBAQ (niBAQ) units. The relationship between protein expression variance and its average expression was fitted using a log-log cubic smoothing spline with four degrees of freedom; 37 proteins with a higher average expression than a log2niBAQ value of 1.0 and a higher variance than the fit (Figure 4A) were kept for dimensionality reduction. The lower-dimensional manifold(four dimensions) was computed and a trajectory was inferred using the CellTrails R package. Protein expression dynamics were calculated by fitting the rolling mean with a window size of five of the log2 niBAQ values and the pseudotime axis using a cubic smoothing spline function with 4-degrees of freedom.</p>
</blockquote>
<p>The results of the trajectory analysis on the SCP data is shown in figure 4 in the paper.</p>
<div class="figure" style="text-align: center">
<img src="figs/zhu2019EL-figure4.jpg" alt="Figure 4 from Zhu et al. 2019." width="100%"><p class="caption">
Figure 4 from Zhu et al. 2019.
</p>
</div>
<p>First, we subset the data to keep only single-cell data. Note that we use <code>drop = TRUE</code> to drop any MS run that does not contain a single cell.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span>, <span class="va">scp</span><span class="op">$</span><span class="va">Cells.per.well</span> <span class="op">==</span> <span class="fl">1</span>, drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span>
<span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 63 assays:</span>
<span class="co">##  [1] 1H1a: SingleCellExperiment with 129 rows and 1 columns </span>
<span class="co">##  [2] 1H1b: SingleCellExperiment with 242 rows and 1 columns </span>
<span class="co">##  [3] 1H1c: SingleCellExperiment with 104 rows and 1 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [61] peptides: SingleCellExperiment with 3133 rows and 28 columns </span>
<span class="co">##  [62] proteins: SingleCellExperiment with 808 rows and 28 columns </span>
<span class="co">##  [63] proteins_iBAQ: SingleCellExperiment with 808 rows and 28 columns</span></code></pre></div>
<p>We can already see there is an issue with respect to the method section. They document an analysis on 30 single-cells, but we here see that only 28 cells are present in this dataset. Furthermore, as mentioned in the previous section, some single-cell runs have failed and should be removed. This seriously limits the trust we can have in the results.</p>
</div>
<div id="feature-selection-based-on-protein-detection-rate" class="section level2">
<h2 class="hasAnchor">
<a href="#feature-selection-based-on-protein-detection-rate" class="anchor"></a>Feature selection based on protein detection rate</h2>
<p>The first selection filter applied by the authors is to keep only proteins that are identified in at least 5 cells.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_iBAQ"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span>
<span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">sel</span><span class="op">)</span><span class="op">[</span><span class="va">sel</span><span class="op">]</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span><span class="va">sel</span>, , <span class="op">]</span>
<span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_iBAQ"</span><span class="op">]</span><span class="op">]</span>
<span class="co">## class: SingleCellExperiment </span>
<span class="co">## dim: 81 28 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(1): ''</span>
<span class="co">## rownames(81): ENSGALP00000000021 ENSGALP00000000304 ...</span>
<span class="co">##   ENSGALP00000073233 ENSGALP00000073404</span>
<span class="co">## rowData names(336): Protein.IDs Majority.protein.IDs ...</span>
<span class="co">##   Oxidation..M..site.positions Protein</span>
<span class="co">## colnames(28): 1H1a 1H1b ... 2L1i 2L1j</span>
<span class="co">## colData names(0):</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span></code></pre></div>
<p>We are left with 81 selected proteins against 75 proteins kept by the authors. Before proceeding with the workflow, we will assess the overlap between the selected proteins in this vignette and the selected proteins in the article. Figure 4E provides a heatmap with the names of the 75 selected proteins.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">selpaper</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"LRRC59"</span>, <span class="st">"PGM1"</span>, <span class="st">"HBBA"</span>, <span class="st">"ERP44"</span>, <span class="st">"KRT7"</span>, <span class="st">"SRSF6"</span>,
              <span class="st">"TPM3"</span>, <span class="st">"RAN"</span>, <span class="st">"HSP90B1"</span>, <span class="st">"AGR3"</span>, <span class="st">"TMSB4X"</span>, <span class="st">"HIST1H101"</span>,
              <span class="st">"EEF1A2"</span>, <span class="st">"PDIA3"</span>, <span class="st">"EEF2"</span>, <span class="st">"CAPRIN1"</span>, <span class="st">"ARID4A"</span>, <span class="st">"RPN1"</span>,
              <span class="st">"CCT7"</span>, <span class="st">"CNN2"</span>, <span class="st">"HIST2H3"</span>, <span class="st">"PPIA"</span>, <span class="st">"ARF3"</span>, <span class="st">"HSPA5"</span>,
              <span class="st">"GST2L"</span>, <span class="st">"SYN3"</span>, <span class="st">"DPYSL2"</span>, <span class="st">"CLTC"</span>, <span class="st">"ANXA2"</span>, <span class="st">"VDA2"</span>,
              <span class="st">"PKM"</span>, <span class="st">"VIM"</span>, <span class="st">"RAB15"</span>, <span class="st">"ACTG1"</span>, <span class="st">"CIRBP"</span>, <span class="st">"YWHAE"</span>,
              <span class="st">"HNRNPA2B1"</span>, <span class="st">"H2AFV"</span>, <span class="st">"HNRNPM"</span>, <span class="st">"RPS27A"</span>, <span class="st">"GOT2"</span>,
              <span class="st">"HNRNPH2"</span>, <span class="st">"HNRNPH3"</span>, <span class="st">"LDHB"</span>, <span class="st">"HSPA8"</span>, <span class="st">"TUBB4B"</span>,
              <span class="st">"ZC3H14"</span>, <span class="st">"MYO6"</span>, <span class="st">"ATP5A1W"</span>, <span class="st">"PRDX1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"CALM2"</span>,
              <span class="st">"TUBA1A1"</span>, <span class="st">"ENO1"</span>, <span class="st">"OCM"</span>, <span class="st">"GAPDH"</span>, <span class="st">"LOC107051177"</span>,
              <span class="st">"CALB2"</span>, <span class="st">"CKB"</span>, <span class="st">"SLC25A6"</span>, <span class="st">"OTOF"</span>, <span class="st">"MDH2"</span>, <span class="st">"HSP90AA1"</span>,
              <span class="st">"ATP5F1B"</span>, <span class="st">"CALR"</span>, <span class="st">"MDH1"</span>, <span class="st">"CRABP1"</span>, <span class="st">"RPS28"</span>, <span class="st">"PGK1"</span>,
              <span class="st">"HIST1H2B5L"</span>, <span class="st">"HIST1H4D"</span>, <span class="st">"COX4I1"</span>, <span class="st">"TPI1"</span>, <span class="st">"RTN1"</span>,
              <span class="st">"TXN"</span><span class="op">)</span></code></pre></div>
<p>The proteins names in the <code>sce</code> data are encoded as Ensembl ID. We convert them to protein symbols using the <code>biomaRt</code> package.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gallus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"ensembl"</span>, dataset <span class="op">=</span> <span class="st">"ggallus_gene_ensembl"</span><span class="op">)</span>
<span class="va">conversionDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">sel</span>,
                      attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_peptide_id"</span>, <span class="st">"uniprot_gn_symbol"</span><span class="op">)</span>,
                      filters <span class="op">=</span> <span class="st">"ensembl_peptide_id"</span>,
                      mart <span class="op">=</span> <span class="va">gallus</span><span class="op">)</span>
<span class="co">## Remove duplicate gene symbol</span>
<span class="va">conversionDF</span> <span class="op">&lt;-</span> <span class="va">conversionDF</span><span class="op">[</span><span class="va">conversionDF</span><span class="op">$</span><span class="va">uniprot_gn_symbol</span> <span class="op">!=</span> <span class="st">"CRABP-I"</span>, <span class="op">]</span>
<span class="va">conversionDF</span> <span class="op">&lt;-</span> <span class="va">conversionDF</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">conversionDF</span><span class="op">$</span><span class="va">ensembl_peptide_id</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">selscp</span> <span class="op">&lt;-</span> <span class="va">conversionDF</span><span class="op">$</span><span class="va">uniprot_gn_symbol</span></code></pre></div>
<p>We next plot the number of proteins found only in this replication, only in the paper, or in both the replication and paper.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">allprots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">selscp</span>, <span class="va">selpaper</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>paper <span class="op">=</span> <span class="va">allprots</span> <span class="op">%in%</span> <span class="va">selpaper</span>,
           scp <span class="op">=</span> <span class="va">allprots</span> <span class="op">%in%</span> <span class="va">selscp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate_all</span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"present"</span>, <span class="st">"absent"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">paper</span>, <span class="va">scp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">summarise</span><span class="op">(</span>count <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">paper</span>,
        y <span class="op">=</span> <span class="va">scp</span>,
        label <span class="op">=</span> <span class="va">count</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>size <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_text</span><span class="op">(</span>nudge_y <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect</span>
    <span class="fu">scale_size_continuous</span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>,
          axis.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-22-1.png" width="288"></p>
<p>There is only about 60 % overlap between the proteins selected here and the proteins selected in the paper. This already means it is impossible to accurately replicate the trajectory analysis results since the data sets do not agree from the start.</p>
</div>
<div id="normalization" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization" class="anchor"></a>Normalization</h2>
<p>The single-cells were normalized to the median of the cells’ mean expression.</p>
<p>We first compute the normalization factor. The normalization factor for each cell is the mean iBAQ of the cell. We then scale the factor by the median of the factor.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Compute sample mean</span>
<span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_iBAQ"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span>
    <span class="va">assay</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fl">2</span>, <span class="va">mean</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">nf</span>
<span class="co">## Scale by the median</span>
<span class="va">nf</span> <span class="op">&lt;-</span> <span class="va">nf</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">nf</span><span class="op">)</span></code></pre></div>
<p>Then, we apply the resulting normalization factor to the data and create a new assay. This can be done using the <code>sweep</code> method that behaves like the base <code>sweep</code> function for matrices, but takes a <code>QFeatures</code> object and an index to the assay to modify. A new assay we call <code>proteins_niBAQ</code> is added to the dataset.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">sweep</a></span><span class="op">(</span><span class="va">scp</span>,
             i <span class="op">=</span> <span class="st">"proteins_iBAQ"</span>,
             MARGIN <span class="op">=</span> <span class="fl">2</span>, <span class="co">## (by column)</span>
             FUN <span class="op">=</span> <span class="st">"/"</span>,
             STATS <span class="op">=</span> <span class="va">nf</span>,
             name <span class="op">=</span> <span class="st">"proteins_niBAQ"</span><span class="op">)</span></code></pre></div>
</div>
<div id="log-transformation" class="section level2">
<h2 class="hasAnchor">
<a href="#log-transformation" class="anchor"></a>Log-transformation</h2>
<p>The normalized values are then log-transformed. This is implemented in the <code>logTransform</code> function in <code>QFeatures</code>. We use a base 2 log-transformation as indicated in the methods section. The function will add a new assay that we call <code>proteins_log2niBAQ</code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">logTransform</a></span><span class="op">(</span><span class="va">scp</span>,
                    base <span class="op">=</span> <span class="fl">2</span>,
                    i <span class="op">=</span> <span class="st">"proteins_niBAQ"</span>,
                    name <span class="op">=</span> <span class="st">"proteins_log2niBAQ"</span><span class="op">)</span></code></pre></div>
</div>
<div id="imputation" class="section level2">
<h2 class="hasAnchor">
<a href="#imputation" class="anchor"></a>Imputation</h2>
<p>Missing data in the log-transformed assay are imputed with zeros. We replicate this using the <code>impute</code> function from <code>QFeatures</code>, specifying the <code>zero</code> method.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/impute.html">impute</a></span><span class="op">(</span><span class="va">scp</span>,
              i <span class="op">=</span> <span class="st">"proteins_log2niBAQ"</span>,
              method <span class="op">=</span> <span class="st">"zero"</span><span class="op">)</span></code></pre></div>
</div>
<div id="batch-correction" class="section level2">
<h2 class="hasAnchor">
<a href="#batch-correction" class="anchor"></a>Batch correction</h2>
<p>Next, the data is batch corrected using the <code>ComBat</code> algorithm. The algorithm requires a design matrix to protect the covariate against the batch correction and therefore include <code>FM1.43.signal</code> in the model. We also need to supply the covariate that indicates the batch structure we want to correct against, <code>Experiment</code> in this case.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get the colData of the assay of interest</span>
<span class="va">cd</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_log2niBAQ"</span><span class="op">)</span><span class="op">)</span>
<span class="co">## Build the design matrix for batch correction</span>
<span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">FM1.43.signal</span>, data <span class="op">=</span> <span class="va">cd</span><span class="op">)</span>
<span class="co">## Get the batch covariate</span>
<span class="va">batch</span> <span class="op">&lt;-</span> <span class="va">cd</span><span class="op">$</span><span class="va">Experiment</span></code></pre></div>
<p>We then extract the data matrix containing the quantitative values and run <code>ComBat</code>. Note that we here use the <code>ComBatv3.34</code> function that is provided in the <code>SCP.replication</code> package. This function was taken from an older version of the <code>sva</code> package (version <code>3.34.0</code>). This is to use the same software as in the article. More recent versions will avoid batch correction for proteins that show no variance within at least one batch and this occurs for 16 out of 81 proteins.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_log2niBAQ"</span><span class="op">]</span><span class="op">]</span>
<span class="co">## Perform batch correction</span>
<span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ComBatv3.34.html">ComBatv3.34</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,
                          batch <span class="op">=</span> <span class="va">batch</span>,
                          mod <span class="op">=</span> <span class="va">model</span><span class="op">)</span>
<span class="co">## Standardizing Data across genes</span>
<span class="co">## Add assay</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">addAssay</a></span><span class="op">(</span><span class="va">scp</span>,
                <span class="va">sce</span>,
                name <span class="op">=</span> <span class="st">"proteins_log2niBAQ_bc"</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html">addAssayLinkOneToOne</a></span><span class="op">(</span><span class="va">scp</span>,
                            from <span class="op">=</span> <span class="st">"proteins_log2niBAQ"</span>,
                            to <span class="op">=</span> <span class="st">"proteins_log2niBAQ_bc"</span><span class="op">)</span></code></pre></div>
</div>
<div id="fitting-the-trajectory" class="section level2">
<h2 class="hasAnchor">
<a href="#fitting-the-trajectory" class="anchor"></a>Fitting the trajectory</h2>
<p>The trajectory fitting is performed using <code>CellTrails</code>. <code>CellTrails</code> is a Bioconductor package initially developed to perform trajectory analysis on single-cell RNA sequencing (scRNA-Seq) data (<span class="citation">Ellwanger et al. (2018)</span>). The software is provided with high quality documentation, such as a user <a href="https://dcellwanger.github.io/CellTrails-handbook/">guide</a>.</p>
<p><code>CellTrails</code> relies on the <code>SingleCellExperiment</code> data class. Since our framework also relies on <code>SingleCellExperiment</code>, our dataset can be readily supplied to <code>CellTrails</code>. We only need to extract the assay of interest, that is the log-normalized and batch correct protein iBAQ values.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_log2niBAQ_bc"</span><span class="op">)</span>
<span class="va">sce</span>
<span class="co">## class: SingleCellExperiment </span>
<span class="co">## dim: 81 28 </span>
<span class="co">## metadata(0):</span>
<span class="co">## assays(1): ''</span>
<span class="co">## rownames(81): ENSGALP00000000021 ENSGALP00000000304 ...</span>
<span class="co">##   ENSGALP00000073233 ENSGALP00000073404</span>
<span class="co">## rowData names(336): Protein.IDs Majority.protein.IDs ...</span>
<span class="co">##   Oxidation..M..site.positions Protein</span>
<span class="co">## colnames(28): 1H1a 1H1b ... 2L1i 2L1j</span>
<span class="co">## colData names(9): Sample.name FM1.43.signal ... Experiment</span>
<span class="co">##   Protein_counts</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span></code></pre></div>
<p><code>CellTrails</code> assumes that the data was processed using a classical scRNA-Seq pipeline and expect the assay to be called <code>logcounts</code>. Hence, we adapt the name of the data matrix although quantitative data <strong>are not</strong> logcounts.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assayNames</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"logcounts"</span></code></pre></div>
<p>There are several steps to run <code>CellTrails</code>, each step containing several parameters. Only little information is available from the methods section and will therefore heavily rely on the <code>CellTrails</code> user guide and assume that default parameters were applied.</p>
<div id="select-highly-variable-proteins" class="section level3">
<h3 class="hasAnchor">
<a href="#select-highly-variable-proteins" class="anchor"></a>Select highly variable proteins</h3>
<p>Highly variable proteins are identified based on the relationship between the mean of log expression and the variance of log expression. This is commonly performed in scRNA-Seq and a general trend is fit to estimate the technical variability. Genes that are significantly higher than this trend might contain interesting biological variability and are therefore selected. In this case a smoothing spline is fit and proteins above the trend are kept.</p>
<p>We compute the mean and variance of the protein log expressions and store them in a table. Using the functions <code>rowMeans</code> (from <code>base</code>) and <code>rowVars</code> (from <code>matrixStats</code>), we can directly compute those metrics from the quantification matrix.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rvar</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="va">rmean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,
                 Mean <span class="op">=</span> <span class="va">rmean</span>,
                 Variance <span class="op">=</span> <span class="va">rvar</span><span class="op">)</span></code></pre></div>
<p>We next model the mean-variance relationship using b-spline regression. Proteins are considered to contain interesting variability if the variance is higher than the fitted trend.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span><span class="op">$</span><span class="va">Fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">Variance</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/splines/bs.html">bs</a></span><span class="op">(</span><span class="va">Mean</span>, df <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,
             data <span class="op">=</span> <span class="va">df</span><span class="op">)</span><span class="op">$</span><span class="va">fitted.values</span>
<span class="va">df</span><span class="op">$</span><span class="va">VarianceLevel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">Variance</span> <span class="op">&gt;</span> <span class="va">df</span><span class="op">$</span><span class="va">Fit</span>, <span class="st">"High"</span>, <span class="st">"Low"</span><span class="op">)</span></code></pre></div>
<p>The plot below attempts to reproduce figure 4A. While the fitted trend is similar than in the original figure, the ranges for both mean and variance are different between the two graph. This implies that the methods section did not provide sufficient information to accurately reproduce the workflow.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Mean</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>col <span class="op">=</span> <span class="va">VarianceLevel</span>,
                   y <span class="op">=</span> <span class="va">Variance</span><span class="op">)</span>,
               fill <span class="op">=</span> <span class="st">"transparent"</span>, shape <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">Fit</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-33-1.png" width="384"></p>
<p>The variance level is stored in the <code>rowData</code> so that it can be accessed during the data embedding.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">VarianceLevel</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">$</span><span class="va">VarianceLevel</span></code></pre></div>
</div>
<div id="spectral-embedding" class="section level3">
<h3 class="hasAnchor">
<a href="#spectral-embedding" class="anchor"></a>Spectral embedding</h3>
<p>The spectral embedding compresses the protein quantification data in a few components so to reveal trajectory information between cells. The embedding is computed directly from the <code>SingleCellExperiment</code> object. The optimal number of latent components are automatically identified from the scree plot. The embedding is computed only using genes with high variance.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">embedding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/embedSamples.html">embedSamples</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">VarianceLevel</span> <span class="op">==</span> <span class="st">"High"</span>, <span class="op">]</span><span class="op">)</span>
<span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/findSpectrum.html">findSpectrum</a></span><span class="op">(</span><span class="va">embedding</span><span class="op">$</span><span class="va">eigenvalues</span>, frac <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-35-1.png" width="384"></p>
<p>In this vignette, we find that 3 components are sufficient to represent the latent trajectory, but the authors used 4 latent dimensions. To stick to the replication, we will use 4 components as well. The embedding is stored in the object as latent space information.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/latentSpace.html">latentSpace</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">embedding</span><span class="op">$</span><span class="va">components</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></code></pre></div>
<p>We plot the computed embedding and colour the single-cell depending on the FM1-43 signal.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/plotManifold.html">plotManifold</a></span><span class="op">(</span><span class="va">sce</span>,
             color_by <span class="op">=</span> <span class="st">"phenoName"</span>,
             name <span class="op">=</span> <span class="st">"FM1.43.signal"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"grey40"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-37-1.png" width="480"></p>
<p>This plot is similar to figure 4B as we can see that the embedding almost perfectly separates the two cell types. Two FM1-43 low cells are found in the high group and three FM1-43 high cells are found in the low group. This is also found in the article.</p>
<p>Another interesting dimension reduction technique is PCA. PCA performs dimension reduction but keeping information about the amount of variance that is extracted by each component. We compute the PCA using the <code>runPCA</code> from the <code>scater</code> package.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>,
        shape_by <span class="op">=</span> <span class="st">"Experiment"</span>,
        colour_by <span class="op">=</span> <span class="st">"FM1.43.signal"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"grey40"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-38-1.png" width="480"></p>
<p>The PCA plot also shows good separation between the two cell types. Moreover, the batch correction allowed a good mixing of the two experimental batches. Almost half of the variance is explained by the two components, indicating a string biological effect.</p>
</div>
<div id="clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering</h3>
<p>The clustering allows to identify trajectory states from the embedding. The <code>findState</code> requires a few parameters. Since no description of the parameters is provided in the article, we set all parameters to default, except for the <code>max_pval</code>. We increased the threshold from <span class="math inline">\(10^{-4}\)</span> to <span class="math inline">\(0.05\)</span> so that at least 2 states are found.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/findStates.html">findStates</a></span><span class="op">(</span><span class="va">sce</span>,
                 min_size <span class="op">=</span> <span class="fl">0.01</span>,
                 min_feat <span class="op">=</span> <span class="fl">5</span>,
                 max_pval <span class="op">=</span> <span class="fl">0.05</span>,
                 min_fc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/states.html">states</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cl</span></code></pre></div>
<p>The states correlate well with the FM1-43 staining.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">CellTrails.state</span>,
               fill <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"count"</span>,
             position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"grey40"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-40-1.png" width="384"></p>
</div>
<div id="trajectory-fit" class="section level3">
<h3 class="hasAnchor">
<a href="#trajectory-fit" class="anchor"></a>Trajectory fit</h3>
<p>Next, we fit the trajectory. The states are connected. Since two states were found, the connection is trivial. The state connections are projected onto the latent space. Samples are projected using this backbone trajectory.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/connectStates.html">connectStates</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/fitTrajectory.html">fitTrajectory</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<p>The trajectory layout must then be computed. This is done using the <a href="https://www.yworks.com/products/yed"><em>yED</em></a> software that will build a tree graph layout from the data. This is performed manually (see the <code>CellTrails</code> <a href="https://dcellwanger.github.io/CellTrails-handbook/celltrails-maps.html#graph-layout">vignette</a> for more information). The graph layout is stored in the <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/write.ygraphml.html">write.ygraphml</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce</span>,
               file <span class="op">=</span> <span class="st">'zhu2019EL_CellTrails.graphml'</span>,
               color_by <span class="op">=</span> <span class="st">'phenoName'</span>,
               name <span class="op">=</span> <span class="st">'state'</span>,
               node_label <span class="op">=</span> <span class="st">'state'</span><span class="op">)</span></code></pre></div>
<p>The <code>graphml</code> file was further processed manually using the yED sofwtare as suggested in the user guide. This allows to arrange the fitted trajectory. This is saved back to file, reloaded and stored in the data as <code>trajLayout</code>.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/read.ygraphml.html">read.ygraphml</a></span><span class="op">(</span><span class="st">"zhu2019EL_CellTrails.graphml"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/trajLayout.html">trajLayout</a></span><span class="op">(</span><span class="va">sce</span>, adjust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tl</span></code></pre></div>
<p>We plot the output trajectory.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/plotMap.html">plotMap</a></span><span class="op">(</span><span class="va">sce</span>,
        color_by <span class="op">=</span> <span class="st">"phenoName"</span>,
        name <span class="op">=</span> <span class="st">"FM1.43.signal"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"grey40"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-44-1.png" width="384"></p>
<p>This trajectory is trivial because only a simple line can be drawn between two states.</p>
</div>
<div id="pseudotime" class="section level3">
<h3 class="hasAnchor">
<a href="#pseudotime" class="anchor"></a>Pseudotime</h3>
<p>The trajectory inference also yields pseudotime inference. We define a trail, since only two states were found (labeled <code>H1</code> and <code>H2</code>) only one trail is possible. We add the new trail and stored the pseudotime in the <code>colData</code>.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/addTrail.html">addTrail</a></span><span class="op">(</span><span class="va">sce</span>, from <span class="op">=</span> <span class="st">"H1"</span>, to <span class="op">=</span> <span class="st">"H2"</span>, name <span class="op">=</span> <span class="st">"trail"</span><span class="op">)</span>
<span class="va">sce</span><span class="op">$</span><span class="va">Pseudotime</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/trails.html">trails</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/CellTrails/man/trails.html">trails</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></code></pre></div>
<p>We plot the pseudotime across the different samples to replicate figure 4D.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>,
           sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">sample</span>, levels <span class="op">=</span> <span class="va">sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">Pseudotime</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>,
               y <span class="op">=</span> <span class="va">Pseudotime</span>,
               color <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect to match the figure design</span>
    <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"grey40"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="zhu2019EL_files/figure-html/unnamed-chunk-46-1.png" width="576"></p>
<p>The pseudotime distribution across cells is rather similar to figure 4D, although the order of the cells is much different. Note that the pseudo-time scale is inverted compared to the original figure, but setting the root to the other state would have solved this difference without changing the interpretation.</p>
<p>We try to replicate figure 4F by plotting the expression against the pseudotime. In the article the rolling means are used and add a layer smoke on the data. We therefore will ignore the rolling mean step and plot the log2-niBAQ directly. We however were not able to reproduce the plots.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>Pseudotime <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="va">colname</span>, <span class="st">"Pseudotime"</span><span class="op">]</span>,
           FM1.43.signal <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="va">colname</span>, <span class="st">"FM1.43.signal"</span><span class="op">]</span>,
           Symbol <span class="op">=</span> <span class="va">conversionDF</span><span class="op">$</span><span class="va">uniprot_gn_symbol</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">rowname</span>, <span class="va">conversionDF</span><span class="op">$</span><span class="va">ensembl_peptide_id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Use protein symbols instead of Enseml ids</span>
    <span class="fu">mutate</span><span class="op">(</span>Symbol <span class="op">=</span> <span class="va">conversionDF</span><span class="op">$</span><span class="va">uniprot_gn_symbol</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">rowname</span>, <span class="va">conversionDF</span><span class="op">$</span><span class="va">ensembl_peptide_id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## Keep only proteins shown in figure 4F</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Symbol</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AGR3"</span>, <span class="st">"TMSB4X"</span>, <span class="st">"OCM"</span>,  <span class="st">"GAPDH"</span>, <span class="st">"CALB2"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Pseudotime</span>,
        y <span class="op">=</span> <span class="va">value</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">stat_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span>,
                formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/splines/bs.html">bs</a></span><span class="op">(</span><span class="va">x</span>, df <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,
                se <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">FM1.43.signal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="co">## Customize the plot aspect to match the figure design</span>
    <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">Symbol</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"green3"</span>,  <span class="st">"grey40"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>The nanoPOTs technology holds promising perspectives in the field of single-cell proteomics. The <span class="citation">Zhu et al. (2019)</span> study demonstrated that the nanoPOTS technology can be applied to answer biologically relevant questions.</p>
<p>The replication of the data analysis presented in this vignette confirmed that the technology is able to distinguish between utricle hair cells (FM1-43 high) and supporting cells (FM1-43 low) at single-cell resolution. We were however surprised that we could not confirm the technical performance of the technology for single-cells. Unfortunately, we noticed that about 30 % of single-cell sample acquisitions failed and showed identification rates comparable to blank samples.</p>
<p>Replication of the trajectory analysis was hindered by the lack of comprehensive description of the workflow or the availability of analysis scripts. This led to discrepancies between the datasets used in this vignette and the original work. Feature selection only yielded about 60 % overlap (47 proteins were found in both datasets out 74 proteins found in this vignette). We argue that performing a trajectory analysis on this limited amount of single-cells and limited amount of proteins is rather optimistic. Further improvement of the nanoPOTS technology are required to improve throughput and sensitivity of the method allowing for a principled analysis. This is also pointed out in the article where scRNA-Seq could provide a more detailed map of the differentiation of supporting cells to hair cells.</p>
</div>
<div id="requirements" class="section level1">
<h1 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h1>
<div id="hardware-and-software" class="section level3">
<h3 class="hasAnchor">
<a href="#hardware-and-software" class="anchor"></a>Hardware and software</h3>
<p>The system details of the machine that built the vignette are:</p>
<pre><code>## Machine: Linux (4.15.0-158-generic)
## R version: R.4.1.0 (svn: 79334)
## RAM: 8.2 GB
## CPU: 4 core(s) - Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz</code></pre>
</div>
<div id="timing" class="section level3">
<h3 class="hasAnchor">
<a href="#timing" class="anchor"></a>Timing</h3>
<p>The total time required to compile this vignette is:</p>
<pre><code>## 5.166169 mins</code></pre>
</div>
<div id="memory" class="section level3">
<h3 class="hasAnchor">
<a href="#memory" class="anchor"></a>Memory</h3>
<p>The final <code>scp</code> memory size is:</p>
<pre><code>## [1] "8.2 Mb"</code></pre>
</div>
<div id="session-info" class="section level3">
<h3 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h3>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## R Under development (unstable) (2020-10-12 r79334)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 18.04.5 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=de_BE.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=de_BE.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=de_BE.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=de_BE.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] splines   stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] benchmarkme_1.0.7            sva_3.41.0                  </span>
<span class="co">##  [3] BiocParallel_1.27.12         genefilter_1.75.1           </span>
<span class="co">##  [5] mgcv_1.8-37                  nlme_3.1-153                </span>
<span class="co">##  [7] scater_1.21.7                scuttle_1.3.1               </span>
<span class="co">##  [9] CellTrails_1.11.0            biomaRt_2.49.4              </span>
<span class="co">## [11] patchwork_1.1.1              forcats_0.5.1               </span>
<span class="co">## [13] stringr_1.4.0                dplyr_1.0.7                 </span>
<span class="co">## [15] purrr_0.3.4                  readr_2.0.2                 </span>
<span class="co">## [17] tidyr_1.1.4                  tibble_3.1.5                </span>
<span class="co">## [19] ggplot2_3.3.5                tidyverse_1.3.1             </span>
<span class="co">## [21] SCP.replication_0.2.1        scpdata_1.1.0               </span>
<span class="co">## [23] ExperimentHub_2.1.4          AnnotationHub_3.1.5         </span>
<span class="co">## [25] BiocFileCache_2.1.1          dbplyr_2.1.1                </span>
<span class="co">## [27] scp_1.3.3                    QFeatures_1.3.6             </span>
<span class="co">## [29] MultiAssayExperiment_1.19.12 SingleCellExperiment_1.15.2 </span>
<span class="co">## [31] SummarizedExperiment_1.23.4  Biobase_2.53.0              </span>
<span class="co">## [33] GenomicRanges_1.45.0         GenomeInfoDb_1.29.8         </span>
<span class="co">## [35] IRanges_2.27.2               S4Vectors_0.31.5            </span>
<span class="co">## [37] BiocGenerics_0.39.2          MatrixGenerics_1.5.4        </span>
<span class="co">## [39] matrixStats_0.61.0          </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] utf8_1.2.2                    tidyselect_1.1.1             </span>
<span class="co">##   [3] RSQLite_2.2.8                 AnnotationDbi_1.55.1         </span>
<span class="co">##   [5] grid_4.1.0                    Rtsne_0.15                   </span>
<span class="co">##   [7] munsell_0.5.0                 ScaledMatrix_1.1.0           </span>
<span class="co">##   [9] codetools_0.2-18              ragg_1.1.3                   </span>
<span class="co">##  [11] withr_2.4.2                   colorspace_2.0-2             </span>
<span class="co">##  [13] filelock_1.0.2                highr_0.9                    </span>
<span class="co">##  [15] knitr_1.36                    rstudioapi_0.13              </span>
<span class="co">##  [17] dtw_1.22-3                    labeling_0.4.2               </span>
<span class="co">##  [19] GenomeInfoDbData_1.2.7        farver_2.1.0                 </span>
<span class="co">##  [21] bit64_4.0.5                   rprojroot_2.0.2              </span>
<span class="co">##  [23] vctrs_0.3.8                   generics_0.1.0               </span>
<span class="co">##  [25] xfun_0.26                     doParallel_1.0.16            </span>
<span class="co">##  [27] R6_2.5.1                      ggbeeswarm_0.6.0             </span>
<span class="co">##  [29] clue_0.3-59                   rsvd_1.0.5                   </span>
<span class="co">##  [31] locfit_1.5-9.4                MsCoreUtils_1.5.0            </span>
<span class="co">##  [33] AnnotationFilter_1.17.1       bitops_1.0-7                 </span>
<span class="co">##  [35] cachem_1.0.6                  DelayedArray_0.19.4          </span>
<span class="co">##  [37] assertthat_0.2.1              promises_1.2.0.1             </span>
<span class="co">##  [39] scales_1.1.1                  beeswarm_0.4.0               </span>
<span class="co">##  [41] gtable_0.3.0                  beachmat_2.9.1               </span>
<span class="co">##  [43] benchmarkmeData_1.0.4         rlang_0.4.11                 </span>
<span class="co">##  [45] systemfonts_1.0.2             lazyeval_0.2.2               </span>
<span class="co">##  [47] broom_0.7.9                   BiocManager_1.30.16          </span>
<span class="co">##  [49] yaml_2.2.1                    reshape2_1.4.4               </span>
<span class="co">##  [51] modelr_0.1.8                  backports_1.2.1              </span>
<span class="co">##  [53] httpuv_1.6.3                  tools_4.1.0                  </span>
<span class="co">##  [55] bookdown_0.24                 ellipsis_0.3.2               </span>
<span class="co">##  [57] jquerylib_0.1.4               proxy_0.4-26                 </span>
<span class="co">##  [59] Rcpp_1.0.7                    plyr_1.8.6                   </span>
<span class="co">##  [61] sparseMatrixStats_1.5.3       progress_1.2.2               </span>
<span class="co">##  [63] zlibbioc_1.39.0               RCurl_1.98-1.5               </span>
<span class="co">##  [65] prettyunits_1.1.1             rpart_4.1-15                 </span>
<span class="co">##  [67] viridis_0.6.1                 haven_2.4.3                  </span>
<span class="co">##  [69] ggrepel_0.9.1                 cluster_2.1.2                </span>
<span class="co">##  [71] fs_1.5.0                      magrittr_2.0.1               </span>
<span class="co">##  [73] reprex_2.0.1                  ProtGenerics_1.25.1          </span>
<span class="co">##  [75] hms_1.1.1                     mime_0.12                    </span>
<span class="co">##  [77] evaluate_0.14                 xtable_1.8-4                 </span>
<span class="co">##  [79] XML_3.99-0.8                  readxl_1.3.1                 </span>
<span class="co">##  [81] gridExtra_2.3                 compiler_4.1.0               </span>
<span class="co">##  [83] crayon_1.4.1                  htmltools_0.5.2              </span>
<span class="co">##  [85] later_1.3.0                   tzdb_0.1.2                   </span>
<span class="co">##  [87] lubridate_1.7.10              DBI_1.1.1                    </span>
<span class="co">##  [89] EnvStats_2.4.0                MASS_7.3-54                  </span>
<span class="co">##  [91] rappdirs_0.3.3                BiocStyle_2.21.3             </span>
<span class="co">##  [93] cba_0.2-21                    Matrix_1.3-4                 </span>
<span class="co">##  [95] cli_3.0.1                     parallel_4.1.0               </span>
<span class="co">##  [97] igraph_1.2.6                  pkgconfig_2.0.3              </span>
<span class="co">##  [99] pkgdown_1.6.1                 foreach_1.5.1                </span>
<span class="co">## [101] xml2_1.3.2                    annotate_1.71.0              </span>
<span class="co">## [103] vipor_0.4.5                   bslib_0.3.0                  </span>
<span class="co">## [105] XVector_0.33.0                rvest_1.0.1                  </span>
<span class="co">## [107] digest_0.6.28                 Biostrings_2.61.2            </span>
<span class="co">## [109] rmarkdown_2.11                cellranger_1.1.0             </span>
<span class="co">## [111] dendextend_1.15.1             edgeR_3.35.1                 </span>
<span class="co">## [113] DelayedMatrixStats_1.15.4     curl_4.3.2                   </span>
<span class="co">## [115] shiny_1.7.1                   lifecycle_1.0.1              </span>
<span class="co">## [117] jsonlite_1.7.2                BiocNeighbors_1.11.0         </span>
<span class="co">## [119] desc_1.4.0                    viridisLite_0.4.0            </span>
<span class="co">## [121] limma_3.49.4                  fansi_0.5.0                  </span>
<span class="co">## [123] pillar_1.6.3                  lattice_0.20-45              </span>
<span class="co">## [125] KEGGREST_1.33.0               fastmap_1.1.0                </span>
<span class="co">## [127] httr_1.4.2                    survival_3.2-13              </span>
<span class="co">## [129] interactiveDisplayBase_1.31.2 glue_1.4.2                   </span>
<span class="co">## [131] iterators_1.0.13              png_0.1-7                    </span>
<span class="co">## [133] BiocVersion_3.14.0            bit_4.0.4                    </span>
<span class="co">## [135] stringi_1.7.4                 sass_0.4.0                   </span>
<span class="co">## [137] blob_1.2.2                    textshaping_0.3.5            </span>
<span class="co">## [139] BiocSingular_1.9.1            memoise_2.0.0                </span>
<span class="co">## [141] renv_0.14.0                   irlba_2.3.3                  </span>
<span class="co">## [143] maptree_1.4-7</span></code></pre></div>
</div>
</div>
<div id="reference" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Ellwanger2018-vn" class="csl-entry">
Ellwanger, Daniel C, Mirko Scheibinger, Rachel A Dumont, Peter G Barr-Gillespie, and Stefan Heller. 2018. <span>“Transcriptional Dynamics of <span>Hair-Bundle</span> Morphogenesis Revealed with <span>CellTrails</span>.”</span> <em>Cell Rep.</em> 23 (10): 2901–2914.e13.
</div>
<div id="ref-Tyanova2016-yj" class="csl-entry">
Tyanova, Stefka, Tikira Temu, and Juergen Cox. 2016. <span>“The <span>MaxQuant</span> Computational Platform for Mass Spectrometry-Based Shotgun Proteomics.”</span> <em>Nat. Protoc.</em> 11 (12): 2301–19.
</div>
<div id="ref-Wickham2019-fz" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the Tidyverse.”</span> <em>J. Open Source Softw.</em> 4 (43): 1686.
</div>
<div id="ref-Zhu2019-ja" class="csl-entry">
Zhu, Ying, Mirko Scheibinger, Daniel Christian Ellwanger, Jocelyn F Krey, Dongseok Choi, Ryan T Kelly, Stefan Heller, and Peter G Barr-Gillespie. 2019. <span>“Single-Cell Proteomics Reveals Changes in Expression During Hair-Cell Development.”</span> <em>Elife</em> 8 (November).
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:christophe.vanderaa@uclouvain.be" class="email">christophe.vanderaa@uclouvain.be</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
